---
title: "PhILR PCoA and PERMANOVA - Zero Replacement Comparison"
---

# Preamble

PhILR instructions are following the introducory vignette: http://bioconductor.org/packages/release/bioc/vignettes/philr/inst/doc/philr-intro.html

Purpose of this notebook is to show effects of different zero replacement
methods on the PhILR + PCoA ordination of the data. Note this is a 
close reconstruction of main Figure 1 of the paper (database: nt, genus level 
and the 0.7 min support threshold, withinvariation bad samples removal method, 
with sources but without controls)

# Preparation

## For script conversion

As this notebook became more and more parameterised, I decided to make it
easy to convert into a script. But for the script, we also need to define
input arguments for the options in chunk 1.

```{r purl = FALSE}
## Name options CHANGE HERE
db <- "nt" ## nt or refseq
tax_level <- "genus" ## genus or species
sources <- "withSources" ## "withSources" "noSources" 
controls <- "noControls" ## "withControls" "noControls"
bad_samples <- "out" ## in or out
sample_filter <- "withinvariation" # onepcburnin, twocburnin, fivepcburnin, tenpcburnin, withinvariation or none
script <- F
minsupp_multiplier <- 7
minsupp_threshold <- 0.01 * minsupp_multiplier
###########################################
```

## Data Loading and Cleaning

Load libraries

```{r  fig.height=3.5, fig.width=7}
library(tidyverse) ## for general data cleaning
library(taxize) ## for NCBI taxonomic info collection
library(ape) ## for tree manipulation
library(phyloseq) ## for data format as input for PhILR
library(philr) ## for data transform
library(vegan) ## for statistical testing
library(ggtree) ## for tree visualisation
library(patchwork) ## for further visualisation assistance
library(broom) ## for saving stats summary tables
library(plotly) ## for interactive plots
library(zCompositions) ## for alternate zero-replacement
```

Load already generated data from MEGAN and metadata. We also need to export the 
same data with as a tree from MEGAN with the option: file > export > tree.

```{r  fig.height=3.5, fig.width=7}

## The tree related to the OTU table
if (tax_level == "genus" & db == "nt") {
 otu_tree <- read.tree("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/megan.backup/Evolution-Comparison_20190401_nt_prokaryotes_genus.nwk")
} else if (tax_level == "species" & db == "nt") {
 otu_tree <- read.tree("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/megan.backup/Evolution-Comparison_20190401_nt_prokaryotes_species.nwk")
} else if (tax_level == "genus" & db == "refseq") {
 otu_tree <- read.tree("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/megan.backup/Evolution-Comparison_20190410_refseq_prokaryotes_genus.nwk")
} else if (tax_level == "species" & db == "refseq") {
  otu_tree <- read.tree("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/megan.backup/Evolution-Comparison_20190410_refseq_prokaryotes_species.nwk")
}


## OTU tables
if (tax_level == "genus" & db == "nt") {
 otu_table <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/megan.backup/Evolution-Comparison_MEGAN_20190401-ex_absolute_genus_prokaryotes_summarised_nt.txt")
} else if (tax_level == "species" & db == "nt") {
 otu_table <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/megan.backup/Evolution-Comparison_MEGAN_20190401-ex_absolute_species_prokaryotes_summarised_nt.txt")
} else if (tax_level == "genus" & db == "refseq") {
 otu_table <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/megan.backup/Evolution-Comparison_MEGAN_20190410-ex_absolute_genus_prokaryotes_summarised_refseq.txt")
} else if (tax_level == "species" & db == "refseq") {
 otu_table <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/megan.backup/Evolution-Comparison_MEGAN_20190410-ex_absolute_species_all_summarised_refseq.txt")
}


## Predicted contaminant taxa to remove
if (tax_level == "genus" & db == "nt") {
 taxa_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/decontam.backup/decontam_taxa_to_remove_megan_nt_genus_combined_0.99_190411.tsv")
} else if (tax_level == "species" & db == "nt") {
 taxa_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/decontam.backup/decontam_taxa_to_remove_megan_nt_species_combined_0.99_190411.tsv")
} else if (tax_level == "genus" & db == "refseq") {
 taxa_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/decontam.backup/decontam_taxa_to_remove_megan_refseq_genus_combined_0.99_190411.tsv")
} else if (tax_level == "species" & db == "refseq") {
 taxa_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/decontam.backup/decontam_taxa_to_remove_megan_refseq_species_combined_0.99_190411.tsv")
}

## Metadata
raw_metadata <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/00-documentation.backup/02-calculus_microbiome-deep_evolution-individualscontrolssources_metadata_20190509.tsv")

## Bad samples to remove

if (sample_filter == "sourcetracker") {
 samples_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/sourcetracker.backup/sourcetracker_filtering_results_190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "onepcburnin") {
 samples_to_remove <- read_tsv("home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter1pc_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "twopcburnin") {
 samples_to_remove <- read_tsv("home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter2pc_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "fivepcburnin") {
 samples_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter5pc_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "tenpcburnin") {
 samples_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter10pc_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "withinvariation") {
 samples_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninwithinfluctuationSDvariation_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "onepcburnin") {
 samples_to_remove <- read_tsv("home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter1pc_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "twopcburnin") {
 samples_to_remove <- read_tsv("home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter2pc_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "fivepcburnin") {
 samples_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter5pc_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "tenpcburnin") {
 samples_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter10pc_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "withinvariation") {
 samples_to_remove <- read_tsv("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninwithinfluctuationSDvariation_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
}



```

Clean up to remove samples not required and then remove any OTUs that
now have no counts. Also remove OTUs that are likely lab contaminants

```{r  fig.height=3.5, fig.width=7}
data_cleaner <- function(x) {
 colnames(x) <- gsub("_S.*_L.*_R1_.*.fastq.combined.fq.prefixed.extractunmapped.bam","", colnames(x))
 colnames(x) <- gsub("_S.*_L00.*_R1_.*.fastq.merged.prefixed.hg19unmapped", "", colnames(x))
 colnames(x) <- gsub("_S.*_L00.*_R1_.*.fastq.extractunmapped.bam", "", colnames(x))
 colnames(x) <- gsub("_S.*_L.*_R1_.*.fastq.merged", "", colnames(x))
 colnames(x) <- gsub(".prefixed.hg19unmapped", "", colnames(x))
 colnames(x) <- gsub("_S0_L003_R1_001.sorted.bam.unmapped", "", colnames(x))
 colnames(x)[1] <- "Taxon"
 return(x)
}

## Remove col cruft of Metadata
raw_metadata <- rename(raw_metadata, Individual = `#SampleID`)

## Remove bad sources from OTU table
if (bad_samples == "in") {
 otu_table <- otu_table %>% 
 data_cleaner
} else if (bad_samples == "out") {
 otu_table <- otu_table %>% 
  data_cleaner %>% 
  dplyr::select(-one_of(samples_to_remove %>% 
                         left_join(raw_metadata, 
                                   by = c("sample" = "Individual")) %>%
                         dplyr::select(sample, SourceSink, Sample_or_Control) %>% 
                         filter(SourceSink == "sink", 
                                Sample_or_Control == "Sample") %>% 
                          pull(sample)) 
 )
}
 


## Conditional filtering out of sources and/or Controls
if (sources == "withSources") {
 NA
} else if (sources == "noSources") {
 otu_table <- otu_table %>% 
 dplyr::select(Taxon, one_of(filter(raw_metadata, SourceSink == "sink") %>% 
             pull(Individual)))
}

if (controls == "withControls") {
 NA
} else if (controls == "noControls") {
 otu_table <- otu_table %>% 
 dplyr::select(Taxon, one_of(filter(raw_metadata, 
                                    Sample_or_Control == "Sample") %>% 
             pull(Individual)), 
             contains("ARS"))
}


## Filter taxa not passing min support threshold 
if (db == "nt") {
 otu_table <- otu_table %>% 
  gather(Individual, Value, 2:ncol(.)) %>% 
  left_join(dplyr::select(raw_metadata, Individual, Min_Support_Reads_Threshold_MALT)) %>%
  mutate(Threshold = Min_Support_Reads_Threshold_MALT * minsupp_multiplier) %>%
  mutate(Threshold = as.numeric(Threshold)) %>%
  mutate(Filter_Passed = if_else(Value >= Threshold, 1, 0)) %>% 
  filter(Filter_Passed == 1) %>%
  dplyr::select(Taxon, Individual, Value) %>%
  spread(Individual, Value, fill = 0)
} else if (db == "refseq") {
 otu_table <- otu_table %>% 
  gather(Individual, Value, 2:ncol(.)) %>% 
  left_join(dplyr::select(raw_metadata, Individual, Min_Support_Reads_Threshold_MALT_refseq)) %>%
  mutate(Threshold = Min_Support_Reads_Threshold_MALT_refseq * minsupp_multiplier) %>%
  mutate(Threshold = as.numeric(Threshold)) %>%
  mutate(Filter_Passed = if_else(Value >= Threshold, 1, 0)) %>% 
  filter(Filter_Passed == 1) %>%
  dplyr::select(Taxon, Individual, Value) %>%
  spread(Individual, Value, fill = 0)
}
 
## Convert to matrix
otu_matrix <- as.matrix(dplyr::select(otu_table, -Taxon))
rownames(otu_matrix) <- otu_table$Taxon

## Remove any taxa that were unique to the bad samples
pos_otus <- rowSums(otu_matrix)
pos_otus <- pos_otus[pos_otus != 0] 

## Remove lab contaminants
otu_matrix <- subset(otu_matrix, !rownames(otu_matrix) %in% (taxa_to_remove %>% pull))
otu_matrix_final <- subset(otu_matrix, rownames(otu_matrix) %in% names(pos_otus))

rownames(otu_matrix_final) <- gsub(" ", "_", rownames(otu_matrix_final))
```

## Make Taxonomy Table

We also need to make our taxonomy table at species level. This takes a long 
to run, so I've generated ones in the past and re-load them when needed.

```{r eval = FALSE}
taxonomy_to_tibble <- function(x) {
 y <- gsub("\\[", "", x)
 y <- gsub("\\]", "", y)
 y <- gsub("\\(.*\\)", "", y)
 
 ttt_uid <- taxize::get_uid(y, db = "ncbi") ## inspired by myTAI::taxonomy()
 ttt_result <- taxize::classification(ttt_uid)[[1]] ## inspired by myTAI::taxonomy()
 
 if ( is.na(ttt_result)[1] ) {
  ttt_out <- x  
 } else {
  ttt_out <- as_tibble(ttt_result %>% 
   mutate(taxon = paste(x)) %>%
   filter(rank %in% c("superkingdom", "kingdom", "phylum", "class", "order", "family", "genus", "species")) %>%
   dplyr::select(name, rank, taxon) %>%
   spread(rank, name))
 }
 return(ttt_out)
  Sys.sleep(2)
}

# taxonomy_summary <- rownames(otu_matrix_final) %>% map(., .f = taxonomy_to_tibble)

taxonomy_summary <- tibble(taxon = character(), 
              superkingdom = character(),
              kingdom = character(),
              phylum = character(),
              class = character(), 
              order = character(),
              family = character(), 
              genus = character(),
              species = character()
              )

n <- 0
tot <- nrow(otu_matrix_final)
fail_taxa <- c()

## Temp due to curl timeout, change number beginning øf range depending on
## where n got to at time out
otu_matrix_final_sub <- otu_matrix_final[903:nrow(otu_matrix_final),]

## If time out, replace otu_matrix_final with otu_matrix_final_sub and re-run
for(i in rownames(otu_matrix_final_sub)) {
 print(paste(i," - ", format(round((n / tot) * 100), nsmall = 2), "%", sep = ""))
 n <- n + 1
 taxonomy_temp <-taxonomy_to_tibble(i)
 
 if (length(taxonomy_temp) = 1) {
  fail_taxa <-append(fail_taxa, i)
 } else {
   taxonomy_summary <- bind_rows(taxonomy_summary, taxonomy_temp)
 }
}

taxonomy_summary <- taxonomy_summary %>% distinct()

```

And save

```{r eval = FALSE}
save(taxonomy_summary, file = paste("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/philr.backup/0-evolution-philr_taxonomytable", "_", db,"_", tax_level, "_", sources, "_", controls, "_badsamples", bad_samples, "_", sample_filter, "_", minsupp_threshold, "_" , format(Sys.Date(), "%Y%m%d"),".robj", sep = ""))
```

For the failed taxa, manually look up paths and add taxonomic paths and add them
and save.

```{r eval = FALSE}
## To check for the ones still missing
missing <- setdiff(rownames(otu_matrix_final), rownames(taxonomy_table_final))

out <- tibble(taxon = character(), 
              superkingdom = character(),
              kingdom = character(),
              phylum = character(),
              class = character(), 
              order = character(),
              family = character(), 
              genus = character(),
              species = character()
              )

n <- 0
tot <- length(missing)
fail_taxa <- c()

for(i in missing[1:length(missing)]) {
 print(paste(i," - ", format(round((n / tot) * 100), nsmall = 2), "%", sep = ""))
 n <- n + 1
 taxonomy_temp <- taxonomy_to_tibble(i)
 
 if (length(taxonomy_temp) == 1) {
  fail_taxa <- append(fail_taxa, i)
 } else {
   out <- bind_rows(out, taxonomy_temp)
 }
}

taxonomy_summary <- bind_rows(taxonomy_summary, out) %>% distinct()


## Some manual additions
temp <- read_tsv("~/Downloads/Genus_extra.csv")
taxonomy_summary <- bind_rows(taxonomy_summary, temp) %>% distinct()

save(taxonomy_summary, file = "/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/philr.backup/0-evolution-philr_taxonomytable_withsources_withControls_all_20190404.robj")
```

Load a previously made taxonomy table

```{r  fig.height=3.5, fig.width=7}
if (db == "nt") {
 load("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/philr.backup/0-evolution-philr_taxonomytable_withsources_withblanks_all_20190404.robj")
} else if (db == "refseq") {
 load("/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/philr.backup/0-evolution-philr_taxonomytable_refseq_species_withSources_withControls_20190115.robj")
}

taxonomy_table <- as.data.frame(taxonomy_summary[2:ncol(taxonomy_summary)])
rownames(taxonomy_table) <- taxonomy_summary$taxon 
taxonomy_table <- as.matrix.data.frame(taxonomy_table)


## To make sure OTU names match tree, which in import replaced spaces with 
## undercores
#rownames(taxonomy_table) <- gsub(" ", "_", rownames(taxonomy_table))

## Filter for genus level
if (tax_level == "genus") {
 taxonomy_table_final <- taxonomy_table[,c("superkingdom", 
                                           "kingdom", 
                                           "phylum", 
                                           "class", 
                                           "order", 
                                           "family", 
                                           "genus")] %>% 
  as.data.frame.matrix
 rownames(taxonomy_table_final) <- c()
 taxonomy_table_final <- subset(taxonomy_table_final, 
                                genus %in% gsub("_", 
                                                " ", 
                                                rownames(otu_matrix_final))) %>% 
   unique()
 taxonomy_table_final <- unique(taxonomy_table_final)
 row.names(taxonomy_table_final) <- taxonomy_table_final[,"genus"]
 taxonomy_table_final <- as.matrix(taxonomy_table_final)
 taxonomy_table_final <- unique(taxonomy_table_final)

 } else if (tax_level == "species") {
 taxonomy_table_final <- taxonomy_table
}
```


## Check Tree Format Valid

As need a purely bifurcating tree (thus do not allow polytomy), and need a
rooted tree we should check this with:

```{r  fig.height=3.5, fig.width=7}
is.rooted(otu_tree)
is.binary.tree(otu_tree)
```

If we don't have a binary tree, solve polytomies with 

```{r  fig.height=3.5, fig.width=7}
otu_tree <- multi2di(otu_tree)
is.binary.tree(otu_tree)
```

Or if this doesn't work we can remove singletons (as in single-intermediate 
nodes between a leaf and a parent node) with

```{r  fig.height=3.5, fig.width=7}
otu_tree <- collapse.singles(otu_tree) # from http://blog.phytools.org/2018/05/when-phylogeny-fails-isbinary-but-is.html
is.binary.tree(otu_tree)

```

## Prepare Metdata

Prepare sample metadata

```{r  fig.height=3.5, fig.width=7}
meta_data <- as.data.frame(raw_metadata[2:ncol(raw_metadata)], 
              row.names = raw_metadata$Individual, 
              colnames = col.names(raw_metadata[2:col(raw_metadata),]))
rownames(meta_data) <- raw_metadata$Individual

```


# Clustering Analysis

## PhILR Transform

### Make PhyloSeq object

Make our Phyloseq object.

```{r  fig.height=3.5, fig.width=7}
ps_data <- phyloseq(otu_table(otu_matrix_final, taxa_are_rows = TRUE), tax_table(taxonomy_table_final), phy_tree(otu_tree), sample_data(meta_data))
```

### Zero Replacement

We also need to a run zero removal procedure as the *LR methods require
positive numbers. Here we will use both a pseudocount (note might be flawed - 
see Fodor paper), and also the 'multiplicative simple replacement' method
from the zCompositions package as suggested in Palarea-Alaladejo and 
Martín-Fernández (2015) _Chemometrics Intellig Lab Syst_ (and also used
in https://github.com/ggloor/CoDa_microbiome_tutorial/wiki/Part-1:-Exploratory-Compositional-PCA-biplot)

For psuedocount we can just do this with phyloseq.

```{r  fig.height=3.5, fig.width=7}
ps_data <- transform_sample_counts(ps_data, function(x) x + 1)
```

For the CZM

```{r, fig.height=3.5, fig.width=7}
otu_matrix_czm <- cmultRepl(t(otu_matrix_final),  label = 0, method = "CZM")

ps_data_czm <- phyloseq(otu_table(t(otu_matrix_czm), taxa_are_rows = T), tax_table(taxonomy_table_final), phy_tree(otu_tree), sample_data(meta_data))
```


### Final Checks

We fix the tree nodes with and something else from the PhILR tutorial I don't 
fully understand with name balancing and transpose to sample row/taxa column
composition convention.

```{r  fig.height=3.5, fig.width=7}
phy_tree(ps_data) <- makeNodeLabel(phy_tree(ps_data), method = "number", prefix = 'n')
name.balance(phy_tree(ps_data), tax_table(ps_data), 'n1')
otu.table <- t(otu_table(ps_data))
tree <- phy_tree(ps_data)

phy_tree(ps_data_czm) <- makeNodeLabel(phy_tree(ps_data_czm), method = "number", prefix = 'n')
name.balance(phy_tree(ps_data_czm), tax_table(ps_data_czm), 'n1')
otu.table <- t(otu_table(ps_data_czm))
tree <- phy_tree(ps_data_czm)

```

Some last checks

```{r  fig.height=3.5, fig.width=7}
otu.table <- t(otu_table(ps_data))
tree <- phy_tree(ps_data)
metadata <- sample_data(ps_data)
tax <- tax_table(ps_data)

otu.table_czm <- t(otu_table(ps_data_czm))
tree_czm <- phy_tree(ps_data_czm)
metadata_czm <- sample_data(ps_data_czm)
tax_czm <- tax_table(ps_data_czm)

otu.table[1:2,1:2] # OTU Table
tree
head(metadata,2)
```

### Run PhILR

Now run PhILR (hopefully)... here goes...

```{r  fig.height=3.5, fig.width=7}
ps_data.philr <- philr(otu.table, 
            tree,
            part.weights = "enorm.x.gm.counts",
            ilr.weights = "uniform")

ps_data.philr[1:5,1:5]

ps_data.philr_czm <- philr(otu.table_czm, 
            tree_czm,
            part.weights = "enorm.x.gm.counts",
            ilr.weights = "uniform")

ps_data.philr_czm[1:5,1:5]
```

And we can now do ordination within the PhILR space with 
Euclidean distances

```{r  fig.height=3.5, fig.width=7}
ps_data.dist <- dist(ps_data.philr, method = "euclidean")
ps_data.pcoa <- ordinate(ps_data, 'PCoA', distance = ps_data.dist)

ps_data.dist_czm <- dist(ps_data.philr_czm, method = "euclidean")
ps_data.pcoa_czm <- ordinate(ps_data_czm, 'PCoA', distance = ps_data.dist_czm)
```

### Plot PhILR

Pre-plotting formatting

```{r  fig.height=3.5, fig.width=7}
env_colours <- c("#1f78b4", 
                 "#6a3d9a", 
                 "#6a3d9a", 
                 "#6a3d9a", 
                 "#33a02c", 
                 "#33a02c", 
                 "#33a02c", 
                 "#33a02c", 
                 "#e31a1c", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#d9d9d9", 
                 "#d9d9d9",
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9",
                 "#d9d9d9")

names(env_colours) <- c("Howler_Monkey", 
           "Gorilla_1", 
           "Gorilla_2", 
           "Gorilla_3", 
           "Chimp_1", 
           "Chimp_2", 
           "Chimp_3", 
           "Chimp_4",
           "Neanderthal", 
           "PreagriculturalHuman_1", 
           "PreagriculturalHuman_2", 
           "PreantibioticHuman_1", 
           "PreantibioticHuman_2", 
           "ModernDayHuman_1", 
           "ModernDayHuman_2", 
           "ExtractionControl", 
           "LibraryControl", 
           "ruralGut", 
           "urbanGut",
           "sediment", 
           "skin", 
           "subPlaque",
           "supPlaque",
           "EnvironmentalControl"
           )

 env_shapes <- c(8,
                0,
                1,
                2,
                0,
                1,
                2,
                5,
                11,
                0,
                12,
                1,
                10,
                2,
                6,
                10,
                13,
                7,
                8,
                14,
                3,
                4,
                12,
                0)

 
 names(env_shapes) <- c("Howler_Monkey", 
           "Gorilla_1", 
           "Gorilla_2", 
           "Gorilla_3", 
           "Chimp_1", 
           "Chimp_2", 
           "Chimp_3", 
           "Chimp_4",
           "Neanderthal", 
           "PreagriculturalHuman_1", 
           "PreagriculturalHuman_2", 
           "PreantibioticHuman_1", 
           "PreantibioticHuman_2", 
           "ModernDayHuman_1", 
           "ModernDayHuman_2", 
           "ExtractionControl", 
           "LibraryControl", 
           "ruralGut", 
           "sediment", 
           "skin", 
           "subPlaque",
           "supPlaque", 
           "urbanGut",
           "EnvironmentalControl"
          )
 
common_colours <- c(Alouatta = "#1f78b4", Gorilla = "#6a3d9a", Pan = "#33a02c", 
          `Homo (Neanderthal)` = "#ff7f00", 
          `Homo (Modern Human)` = "#ff7f00", ExtractionControl = "#d9d9d9", 
          LibraryControl = "#d9d9d9", Plaque = "#d9d9d9", Gut = "#d9d9d9", 
          Skin = "#d9d9d9", Sediment = "#d9d9d9", EnvironmentalControl = "#d9d9d9")

common_shapes <- c(Alouatta = 8, Gorilla = 0, Pan = 1, 
          `Homo (Neanderthal)` = 2, `Homo (Modern Human)` = 6,
          ExtractionControl = 10, LibraryControl = 13, Plaque = 9, 
          Gut = 4, Skin = 14, Sediment = 7, EnvironmentalControl = 12)


 
 
```


And plot with host populations

```{r  fig.height=3.5, fig.width=7}

## Convert to tidy data and add metadata
ps_data.pcoa_tib <- as_tibble(ps_data.pcoa$vectors, rownames = "Individual") %>%
 left_join(as_tibble(metadata, rownames = "Individual"))

ps_data.pcoa_tib$Env <- factor(ps_data.pcoa_tib$Env, levels = c("Howler_Monkey", 
           "Gorilla_1", 
           "Gorilla_2", 
           "Gorilla_3", 
           "Chimp_1", 
           "Chimp_2", 
           "Chimp_3", 
           "Chimp_4",
           "Neanderthal", 
           "PreagriculturalHuman_1", 
           "PreagriculturalHuman_2", 
           "PreantibioticHuman_1", 
           "PreantibioticHuman_2", 
           "ModernDayHuman_1", 
           "ModernDayHuman_2", 
           "ExtractionControl", 
           "LibraryControl", 
           "ruralGut", 
           "sediment",
           "skin", 
           "subPlaque",
           "supPlaque", 
           "urbanGut",
           "EnvironmentalControl")
           )

## Extract percentage variation
percentage_data <- round(100 * ps_data.pcoa$values[1] / sum(ps_data.pcoa$values[1]), 2)
rownames(percentage_data) <- rownames(ps_data.pcoa$value)

## Colour/Shapes by sample type and source/sink
pcoa_plot_hostpops1 <- ggplot(ps_data.pcoa_tib, aes(Axis.1, Axis.2, colour = Env, shape = Env, text = Individual)) +
 geom_point(size = 2) +
 scale_colour_manual(values = env_colours) +
 scale_shape_manual(values = env_shapes) +
 xlab(paste("Axis 1 (", percentage_data[1,], "%)", sep = "")) +
 ylab(paste("Axis 2 (", percentage_data[2,], "%)", sep = "")) +
 theme_minimal(base_family = "Roboto", base_size = 7) +
 labs(shape = "Host Population", colour = "Host Population") +
 guides(shape = guide_legend(ncol = 2), colour = guide_legend(ncol = 2))


if (script == F) {
 pcoa_plot_hostpops1
}

# ggsave(paste("01a-philr_pcoa_malt_euclidean_axis1axis2_hostpopulations_", db , "_", tax_level, "_", sources, "_", controls, "_badsamples", bad_samples, "_", sample_filter, "_", minsupp_threshold, "_czm_" , format(Sys.Date(), "%Y%m%d"),".pdf", sep = ""), plot = pcoa_plot_hostpops, "/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/philr.backup/", device = cairo_pdf, width = 7, height = 3.5, units = "in", dpi = 600)


## Axis 3 and 2
pcoa_plot_hostpops2 <- ggplot(ps_data.pcoa_tib, aes(Axis.3, Axis.2, colour = Env, shape = Env)) +
 geom_point(size = 2) +
 scale_colour_manual(values = env_colours) +
 scale_shape_manual(values = env_shapes) +
 xlab(paste("Axis 3 (", percentage_data[3,], "%)", sep = "")) +
 ylab(paste("Axis 2 (", percentage_data[2,], "%)", sep = "")) +
 theme_minimal(base_family = "Roboto", base_size = 7) +
 labs(shape = "Host Population", colour = "Host Population") + 
 guides(shape = guide_legend(ncol = 2), colour = guide_legend(ncol = 2))



if (script == F) {
pcoa_plot_hostpops2
}

# ggsave(paste("01b-philr_pcoa_malt_euclidean_axis3axis2_hostpopulations_", db , "_", tax_level, "_", sources, "_", controls, "_badsamples", bad_samples, "_", sample_filter, "_", minsupp_threshold, "_" , format(Sys.Date(), "%Y%m%d"),".pdf", sep = ""), plot = pcoa_plot_hostpops, "/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/philr.backup/", device = cairo_pdf, width = 7, height = 3.5, units = "in", dpi = 600)



```

And for czm

And plot with host populations

```{r  fig.height=3.5, fig.width=7}

## Convert to tidy data and add metadata
ps_data.pcoa_tib_czm <- as_tibble(ps_data.pcoa_czm$vectors, rownames = "Individual") %>%
 left_join(as_tibble(metadata, rownames = "Individual"))

ps_data.pcoa_tib_czm$Env <- factor(ps_data.pcoa_tib_czm$Env, levels = c("Howler_Monkey", 
           "Gorilla_1", 
           "Gorilla_2", 
           "Gorilla_3", 
           "Chimp_1", 
           "Chimp_2", 
           "Chimp_3", 
           "Chimp_4",
           "Neanderthal", 
           "PreagriculturalHuman_1", 
           "PreagriculturalHuman_2", 
           "PreantibioticHuman_1", 
           "PreantibioticHuman_2", 
           "ModernDayHuman_1", 
           "ModernDayHuman_2", 
           "ExtractionControl", 
           "LibraryControl", 
           "ruralGut", 
           "sediment",
           "skin", 
           "subPlaque",
           "supPlaque", 
           "urbanGut",
           "EnvironmentalControl")
           )

## Extract percentage variation
percentage_data_czm <- round(100 * ps_data.pcoa_czm$values[1] / sum(ps_data.pcoa_czm$values[1]), 2)
rownames(percentage_data_czm) <- rownames(ps_data.pcoa_czm$value)

## Colour/Shapes by sample type and source/sink
pcoa_plot_hostpops1_czm <- ggplot(ps_data.pcoa_tib_czm, aes(Axis.1, Axis.2, colour = Env, shape = Env, text = Individual)) +
 geom_point(size = 2) +
 scale_colour_manual(values = env_colours) +
 scale_shape_manual(values = env_shapes) +
 xlab(paste("Axis 1 (", percentage_data_czm[1,], "%)", sep = "")) +
 ylab(paste("Axis 2 (", percentage_data_czm[2,], "%)", sep = "")) +
 theme_minimal(base_family = "Roboto", base_size = 7) +
 labs(shape = "Host Population", colour = "Host Population") +
 guides(shape = guide_legend(ncol = 2), colour = guide_legend(ncol = 2))


if (script == F) {
 pcoa_plot_hostpops1_czm
}

# ggsave(paste("01a-philr_pcoa_malt_euclidean_axis1axis2_hostpopulations_", db , "_", tax_level, "_", sources, "_", controls, "_badsamples", bad_samples, "_", sample_filter, "_", minsupp_threshold, "_czm_" , format(Sys.Date(), "%Y%m%d"),".pdf", sep = ""), plot = pcoa_plot_hostpops, "/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/philr.backup/", device = cairo_pdf, width = 7, height = 3.5, units = "in", dpi = 600)


## Axis 3 and 2
pcoa_plot_hostpops2_czm <- ggplot(ps_data.pcoa_tib_czm, aes(Axis.3, Axis.2, colour = Env, shape = Env)) +
 geom_point(size = 2) +
 scale_colour_manual(values = env_colours) +
 scale_shape_manual(values = env_shapes) +
 xlab(paste("Axis 3 (", percentage_data_czm[3,], "%)", sep = "")) +
 ylab(paste("Axis 2 (", percentage_data_czm[2,], "%)", sep = "")) +
 theme_minimal(base_family = "Roboto", base_size = 7) +
 labs(shape = "Host Population", colour = "Host Population") + 
 guides(shape = guide_legend(ncol = 2), colour = guide_legend(ncol = 2))



if (script == F) {
pcoa_plot_hostpops2_czm
}

# ggsave(paste("01b-philr_pcoa_malt_euclidean_axis3axis2_hostpopulations_", db , "_", tax_level, "_", sources, "_", controls, "_badsamples", bad_samples, "_", sample_filter, "_", minsupp_threshold, "_czm_" , format(Sys.Date(), "%Y%m%d"),".pdf", sep = ""), plot = pcoa_plot_hostpops, "/home/fellows/projects1/microbiome_calculus/evolution/04-analysis/screening/philr.backup/", device = cairo_pdf, width = 7, height = 3.5, units = "in", dpi = 600)



```

Compare

```{r}
final_plot <- pcoa_plot_hostpops1 + 
  theme(legend.position = "none") + 
  labs(title = "Pseudocount") + 
  pcoa_plot_hostpops1_czm + 
  theme(legend.position = "none") + 
  labs(title = "Count Zero Multiplicative")

ggplot2::ggsave("/home/fellows/Dropbox (MPI SHH)/Projects/0-Calculus_Oral_Evolution/1-Oral_Deep_Evolution/06-manuscripts/SupplementaryFigures/PCoA_ZeroReplacementComparison/SuppFigSXX_PhiLRPCoA_ZeroReplacementComparison.pdf", 
                plot = final_plot, 
                device = cairo_pdf, 
                width = 7, 
                height = 3.5, 
                units = "in")

ggplot2::ggsave("/home/fellows/Dropbox (MPI SHH)/Projects/0-Calculus_Oral_Evolution/1-Oral_Deep_Evolution/06-manuscripts/SupplementaryFigures/PCoA_ZeroReplacementComparison/SuppFigSXX_PhiLRPCoA_ZeroReplacementComparison.png", 
                plot = final_plot, 
                device = "png", 
                width = 7, 
                height = 3.5, 
                units = "in")

```


