---
title: "R Notebook"
output: html_notebook
---

# Preamble

Here we are going to follow the [tutorial](https://cran.r-project.org/web/packages/indicspecies/vignettes/indicspeciesTutorial.pdf) 
in the `indicspecies` package to
run 'indicator analysis' to pick taxa that are 'indicators' for a particular
group. The nice thing about this particular package is it has implemented
a 'combination' function, allowing detection of taxa that is also present 
across multiple groups.

# Prepraration

## Data Loading

First we load the libraries

```{r}
library(dplyr)
library(stringr)
library(tidyr)
library(readr)
library(indicspecies)
library(tibble)
```

The required data is the OTU table and a vector of groupings, based on the 
Individual order in the OTU table. The OTU table needs to be in the form of a 
matrix, with sites (or Individuals) in rows, and species in columns.

We can set out options here

```{r}
## Name options CHANGE HERE
db <- "nt" ## nt or refseq or mp2
tax_level <- "species" ## genus or species
sources <- "noSources" ## "withSources" "noSources" 
controls <- "noControls" ## "withControls" "noControls"
bad_samples <- "out" ## in or out
sample_filter <- "withinvariation" # onepcburnin, twocburnin, fivepcburnin, tenpcburnin, withinvariation or none
view_taxon_only <- "none" #"Actinomyces|Campylobacter|Capnocytophaga|Corynebacterium|Desulfomicrobium|Fusobacterium|Fretibacterium|Mogibacterium|Mycobacterium|Olsenella|Ottowia|Parvimonas|Prevotella|Porphyromonas|Pseudopropionibacterium|Selenomonas|Streptococcus|Treponema|Tannerella" #(will be printed as Coregenus); any valid genus name in filtered OTU table or 'none'
minsupp_multiplier <- 4
prevalence_filter <- 5 # interger of minimum number of individuals taxon required or 0 for no filtering

minsupp_threshold <- 0.01 * minsupp_multiplier
```

Load already generated data from MEGAN and metadata. We also need to export the 
same data with as a tree from MEGAN with the option: file > export > tree.

```{r  fig.height=3.5, fig.width=7}
## OTU tables
if (tax_level == "genus" & db == "nt") {
 otu_table <- read_tsv("../04-analysis/screening/megan.backup/Evolution-Comparison_MEGAN_20190401-ex_absolute_genus_prokaryotes_summarised_nt.txt")
} else if (tax_level == "species" & db == "nt") {
 otu_table <- read_tsv("../04-analysis/screening/megan.backup/Evolution-Comparison_MEGAN_20190401-ex_absolute_species_prokaryotes_summarised_nt.txt")
} else if (tax_level == "genus" & db == "refseq") {
 otu_table <- read_tsv("../04-analysis/screening/megan.backup/Evolution-Comparison_MEGAN_20190410-ex_absolute_genus_prokaryotes_summarised_refseq.txt")
} else if (tax_level == "species" & db == "refseq") {
 otu_table <- read_tsv("../04-analysis/screening/megan.backup/Evolution-Comparison_MEGAN_20190410-ex_absolute_species_prokaryotes_summarised_refseq.txt")
} else if (tax_level == "genus" & db == "mp2") {
 otu_table <- read_tsv("../04-analysis/screening/metaphlan2/output/mp2_merged_abundance_table_all_20190401.txt", comment = '#') %>%
    rename(Clade = ID) %>%
    filter(!grepl("t__[[:alnum:]]", Clade)) %>%
    filter(!grepl("s__[[:alnum:]]", Clade)) %>%
    filter( grepl("g__[[:alnum:]]", Clade)) %>%
    mutate(Clade = purrr::map(Clade, 
                       function(x) str_split_fixed(x, 
                                                   "\\|g__", 
                                                   n = 2)[2]) %>% unlist) %>%
    mutate(Clade = gsub("_", " ", Clade)) %>%
    rename(`#Datasets` = Clade)
 
} else if (tax_level == "species" & db == "mp2") {
 otu_table <- read_tsv("../04-analysis/screening/metaphlan2/output/mp2_merged_abundance_table_all_20190401.txt", comment = '#') %>%
    rename(Clade = ID) %>%
    filter(!grepl("t__[[:alnum:]]", Clade)) %>%
    filter(grepl("s__[[:alnum:]]", Clade)) %>%
    mutate(Clade = purrr::map(Clade, 
                       function(x) str_split_fixed(x, 
                                                   "\\|s__", 
                                                   n = 2)[2]) %>% unlist) %>%
    mutate(Clade = gsub("_", " ", Clade)) %>%
    rename(`#Datasets` = Clade)
}


## Predicted contaminant taxa to remove
if (tax_level == "genus" & db == "nt") {
 taxa_to_remove <- read_tsv("../04-analysis/screening/decontam.backup/decontam_taxa_to_remove_megan_nt_genus_combined_0.99_190411.tsv")
} else if (tax_level == "species" & db == "nt") {
 taxa_to_remove <- read_tsv("../04-analysis/screening/decontam.backup/decontam_taxa_to_remove_megan_nt_species_combined_0.99_190411.tsv")
} else if (tax_level == "genus" & db == "refseq") {
 taxa_to_remove <- read_tsv("../04-analysis/screening/decontam.backup/decontam_taxa_to_remove_megan_refseq_genus_combined_0.99_190411.tsv")
} else if (tax_level == "species" & db == "refseq") {
 taxa_to_remove <- read_tsv("../04-analysis/screening/decontam.backup/decontam_taxa_to_remove_megan_refseq_species_combined_0.99_190411.tsv")
} else if (tax_level == "genus" & db == "mp2") {
 taxa_to_remove <- read_tsv("../04-analysis/screening/decontam.backup/decontam_taxa_to_remove_mp2_mp2_genus_combined_0.99_190411.tsv")
} else if (tax_level == "species" & db == "mp2") {
 taxa_to_remove <- read_tsv("../04-analysis/screening/decontam.backup/decontam_taxa_to_remove_mp2_mp2_species_combined_0.9_190411.tsv")
}

## Metadata
raw_metadata <- read_tsv("../00-documentation.backup/02-calculus_microbiome-deep_evolution-individualscontrolssources_metadata_20190509.tsv")

## Bad samples to remove

if (sample_filter == "sourcetracker") {
 samples_to_remove <- read_tsv("../04-analysis/screening/sourcetracker.backup/sourcetracker_filtering_results_190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "onepcburnin") {
 samples_to_remove <- read_tsv("home/fellows../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter1pc_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "twopcburnin") {
 samples_to_remove <- read_tsv("home/fellows../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter2pc_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "fivepcburnin") {
 samples_to_remove <- read_tsv("../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter5pc_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "tenpcburnin") {
 samples_to_remove <- read_tsv("../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter10pc_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "nt" && sample_filter == "withinvariation") {
 samples_to_remove <- read_tsv("../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninwithinfluctuationSDvariation_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "onepcburnin") {
 samples_to_remove <- read_tsv("home/fellows../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter1pc_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "twopcburnin") {
 samples_to_remove <- read_tsv("home/fellows../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter2pc_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "fivepcburnin") {
 samples_to_remove <- read_tsv("../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter5pc_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "tenpcburnin") {
 samples_to_remove <- read_tsv("../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninfilter10pc_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "refseq" && sample_filter == "withinvariation") {
 samples_to_remove <- read_tsv("../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninwithinfluctuationSDvariation_refseq_fractionOralThreshold_65_20190509.tsv") %>% 
  filter(more_env == T)
} else if (db == "mp2" && sample_filter == "withinvariation") {
 samples_to_remove <- read_tsv("../04-analysis/screening/cumulative_decay.backup/cumulativeproportiondecay_burninwithinfluctuationSDvariation_nt_fractionOralThreshold_50_20190509.tsv") %>% 
  filter(more_env == T)
}

```

## Data Cleaning

Clean up to remove samples not required and then remove any OTUs that
now have no counts. Also remove OTUs that are likely lab contaminants

```{r fig.height=3.5, fig.width=7}

data_cleaner <- function(x) {
 colnames(x) <- gsub("_S.*_L.*_R1_.*.fastq.combined.fq.prefixed.extractunmapped.bam","", colnames(x))
 colnames(x) <- gsub("_S.*_L00.*_R1_.*.fastq.merged.prefixed.hg19unmapped", "", colnames(x))
 colnames(x) <- gsub("_S.*_L00.*_R1_.*.fastq.extractunmapped.bam", "", colnames(x))
 colnames(x) <- gsub("_S.*_L.*_R1_.*.fastq.merged", "", colnames(x))
 colnames(x) <- gsub(".prefixed.hg19unmapped", "", colnames(x))
 colnames(x) <- gsub("_S0_L003_R1_001.sorted.bam.unmapped", "", colnames(x))
colnames(x) <- gsub(".mp2profile", "", colnames(x))
 colnames(x)[1] <- "Taxon"
 return(x)
}

## Remove col cruft of Metadata
raw_metadata <- rename(raw_metadata, Individual = `#SampleID`)

## Remove bad sources from OTU table
if (bad_samples == "in") {
 otu_table <- otu_table %>% data_cleaner
} else if (bad_samples == "out") {
 otu_table <- otu_table %>% 
  data_cleaner %>% 
  dplyr::select(-one_of(samples_to_remove %>% 
                         left_join(raw_metadata, 
                                   by = c("sample" = "Individual")) %>%
                         dplyr::select(sample, SourceSink, Sample_or_Control) %>% 
                         filter(SourceSink == "sink", 
                                Sample_or_Control == "Sample") %>% 
                          pull(sample)) 
 )
}
 


## Conditional filtering out of sources and/or Controls
if (sources == "withSources") {
 NA
} else if (sources == "noSources") {
 otu_table <- otu_table %>% 
 dplyr::select(Taxon, one_of(filter(raw_metadata, SourceSink == "sink") %>% 
             pull(Individual)))
}

if (controls == "withControls") {
 NA
} else if (controls == "noControls") {
 otu_table <- otu_table %>% 
 dplyr::select(Taxon, one_of(filter(raw_metadata, 
                                    Sample_or_Control == "Sample") %>% 
             pull(Individual)), 
             contains("ARS"))
}


## Filter taxa not passing min support threshold 
if (db == "nt") {
 otu_table <- otu_table %>% 
  gather(Individual, Value, 2:ncol(.)) %>% 
  left_join(dplyr::select(raw_metadata, Individual, Min_Support_Reads_Threshold_MALT)) %>%
  mutate(Threshold = Min_Support_Reads_Threshold_MALT * minsupp_multiplier) %>%
  mutate(Threshold = as.numeric(Threshold)) %>%
  mutate(Filter_Passed = if_else(Value >= Threshold, 1, 0)) %>% 
  filter(Filter_Passed == 1) %>%
  dplyr::select(Taxon, Individual, Value) %>%
  spread(Individual, Value, fill = 0)
} else if (db == "refseq") {
 otu_table <- otu_table %>% 
  gather(Individual, Value, 2:ncol(.)) %>% 
  left_join(dplyr::select(raw_metadata, Individual, Min_Support_Reads_Threshold_MALT_refseq)) %>%
  mutate(Threshold = Min_Support_Reads_Threshold_MALT_refseq * minsupp_multiplier) %>%
  mutate(Threshold = as.numeric(Threshold)) %>%
  mutate(Filter_Passed = if_else(Value >= Threshold, 1, 0)) %>% 
  filter(Filter_Passed == 1) %>%
  dplyr::select(Taxon, Individual, Value) %>%
  spread(Individual, Value, fill = 0)
} else if (db == "mp2") {
  otu_table <- otu_table %>% 
    gather(Individual, Value, 2:ncol(.)) %>%
    mutate(Threshold = 0 * minsupp_multiplier) %>%
    mutate(Value = as.numeric(Value)) %>%
    mutate(Filter_Passed = if_else(Value > Threshold, 1, 0)) %>%
    filter(Filter_Passed == 1) %>%
    dplyr::select(Taxon, Individual, Value) %>%
    spread(Individual, Value, fill = 0)
}

 
## Convert to matrix
otu_matrix <- as.matrix(dplyr::select(otu_table, -Taxon))
rownames(otu_matrix) <- otu_table$Taxon

## Remove any taxa that were unique to the bad samples
pos_otus <- rowSums(otu_matrix)
pos_otus <- pos_otus[pos_otus != 0] 

## Remove lab contaminants
otu_matrix <- subset(otu_matrix, !rownames(otu_matrix) %in% (taxa_to_remove %>% pull))
otu_matrix_final <- subset(otu_matrix, rownames(otu_matrix) %in% names(pos_otus))

## Select for only genus of interested
if (view_taxon_only == "none") {
  NULL
} else {
  if ( length(grep(view_taxon_only, rownames(otu_matrix_final))) == 0 ) {
    stop("The searched for taxa does not exist: is it in your OTU table, otu comtaminants or falls below min-support threshold?")
  } else {
    otu_matrix_final <- subset(otu_matrix_final, 
                             grepl(view_taxon_only, rownames(otu_matrix_final)))
  }
}

rownames(otu_matrix_final) <- gsub(" ", "_", rownames(otu_matrix_final))
```

Apply prevalence filter

```{r}
if (prevalence_filter == 0) {
  NULL
} else {
  prevalence_filter <- as.numeric(prevalence_filter)
  passed_prevalence_list <- otu_matrix_final %>% 
    as_tibble(rownames = "Taxon") %>% 
    gather(Sample, Count, 2:ncol(.)) %>% 
    group_by(Taxon) %>% 
    summarise(Passes_Prevalence = sum(Count > 0)) %>% 
    filter(Passes_Prevalence >= prevalence_filter) %>% 
    pull(Taxon)
  otu_matrix_final <- subset(otu_matrix_final, 
                             grepl(paste(passed_prevalence_list, collapse = "|"), 
                                   rownames(otu_matrix_final)))
}

otu_table_clean <- otu_matrix_final %>% as_tibble(rownames = "Taxon")
```


## Indicspecies formatting

Convert the OTU table to a 'matrix', with Individuals order via groups. 
*IMPORTANT*: Then convert this matrix into a **dataframe**, as the multipatt 
function doesn't actually take a matrix (as the tutorial seesm to suggest!)! If 
you give a matrix it gives a 'can't find dimmnames' error.

Convert to long to order the individuals by host genus

```{r}
## For OTU table
otu_table_long <- gather(otu_table_clean, 
                         Individual, 
                         Count, 
                         2:ncol(otu_table_clean))

otu_table_long <- left_join(otu_table_long, raw_metadata) %>%
  arrange(Individual) %>%
  arrange(Host_Genus)


otu_table_long$Host_Genus <- factor(otu_table_long$Host_Genus,
                                  levels = c(unique(sort(otu_table_long$Host_Genus))))
```


Convert Individual into a factor, based on a unique list of Individual levels 
(which is already in host order because of the ordering of the Host_Common factors)

```{r}
otu_table_long <- otu_table_long %>% 
  mutate(Individual = factor(Individual, unique(Individual)))
```

Covert OTU data into numeric matrix then back ot a dataframe

```{r}
data_dataframe <- select(otu_table_long, Taxon, Individual, Count) %>% 
  filter(!is.na(Taxon)) %>%
  spread(Taxon, Count)

data_matrix <- data.matrix(data_dataframe[1:nrow(data_dataframe), 
                                                2:ncol(data_dataframe)])
rownames(data_matrix) <- data_dataframe$Individual
  
data_final <- data.frame(data_matrix[1:nrow(data_matrix), 1:ncol(data_matrix)])
```

To generate the vector at host genus level, but corresponding to the order
of the individuals themselves

```{r}
group_vector <- otu_table_long %>% 
  select(Individual, Host_Genus) %>% 
  distinct() %>% 
  pull(Host_Genus)
```


Then run indicator analysis!

```{r}
indval <- multipatt(data_final, 
                    group_vector,
                    control = how(nperm = 999))
```

To visualise the results

```{r}
summary(indval)
```

Clean up results to tidy format, fixing names and indicating if host
specific (core) or shared across multiple host.

```{r}
indval_filtered <- as.tibble(indval$sign, rownames = "Taxon") %>% 
  filter(p.value < 0.05) 

indval_filtered <- indval_filtered %>% 
  gather(Group, Group_Y_N, 2:(ncol(.) - 3)) %>% 
  group_by(Taxon) %>% 
  summarise(No_Groups = sum(Group_Y_N)) %>%
  mutate(Group_Type = if_else(No_Groups == 1, "Single", "Multiple")) %>%
  select(Taxon, Group_Type) %>%
  right_join(indval_filtered) %>%
  mutate(Taxon = gsub("\\.", " ", Taxon), 
         Taxon = gsub("sp  ", "sp. ", Taxon))

indval_filtered <- indval_filtered %>% 
  mutate(Combination = colnames(indval$comb)[index]) %>%
  mutate(Combination = gsub("\\+", "\\:", Combination))

```

And optionally save

```{r eval = T}
write_tsv(indval_filtered, 
          paste0("../04-analysis/screening/indicspecies.backup/",
             "IndicatorAnalysis_results_" ,
             db , "_", 
             tax_level, "_", 
             sources, "_", 
             controls, 
             "_badsamples", bad_samples, "_", 
             sample_filter, "_", 
             "additionaltaxafilteringto_", gsub("\\|", "", view_taxon_only), "_",
             minsupp_threshold, "_" , 
             "minprevalencethreshold_", prevalence_filter, "_",
             format(Sys.Date(), "%Y%m%d"),".tsv"))
```
