---
title: "Amylase Binding Protein Reads from Superreference"
output: html_notebook
---

## Set up

Set up

```{r}
library(tidyverse)
library(ggbeeswarm)
```

Load genus/stats superrefernece stats (previously generated in notebook 
`031-supperferencemapped_stats`)

```{r}
superreference_data <- read_tsv("../04-analysis/deep/competitive_mapping.backup/superreference_mapping_statistics_raw_20190522.tsv.gz")

raw_metadata_deep <- read_tsv("../00-documentation.backup/17-Deep_Sequencing_HumanFiltering_Results_20200220.tsv")

abp_data <- Sys.glob("../04-analysis/deep/eager/superreference_mapping/output/Streptococcus/*/5-DeDup/") %>%
  list.files("*reads.tsv.gz", full.names = T) %>%
  enframe(name = NULL, value = "file_path") %>%
  mutate(file_contents = map(file_path, ~read_tsv(.x, 
                                                  col_names = c("chrom", "start_coordinate", "end_coordinate", "name", "reads"), 
                                                  col_types = cols("c", "d", "d", "c", "d"))),
         individual = map(file_path, ~str_split(.x, "/") %>% unlist() %>% tail(n = 4) %>% head(n = 1)) %>% unlist,
         abp_type = map(file_path, ~ basename(.x) %>% 
                          str_split("_") %>% 
                          pluck(1,6) %>% 
                          str_split("\\.") %>% 
                          unlist %>% 
                          pluck(4) %>%
                        str_replace("reads", "-like")) %>% unlist)
```

Set some ordering and aesthetics

```{r}
host_general_order <- c("ExtractionBlank", "LibraryBlank", "Howler", "Gorilla", "Chimp", "Neanderthal", "PreagriculturalHuman", "PreantibioticHuman", "ModernDayHuman")
host_genus_order <- c("Control", "Alouatta", "Gorilla", "Pan", "Homo")

common_colours <- c(`0` = "black", Alouatta = "#1f78b4", Gorilla = "#6a3d9a", Pan = "#33a02c", 
          `Homo (Neanderthal)` = "#fdbf6f", 
          `Homo (Modern Human)` = "#ff7f00", ExtractionControl = "#d9d9d9", 
          LibraryControl = "#d9d9d9", Plaque = "#d9d9d9", Gut = "#d9d9d9", 
          Skin = "#d9d9d9", Sediment = "#d9d9d9", EnvironmentalControl = "#d9d9d9")

common_shapes <- c(Alouatta = 8, Gorilla = 0, Pan = 1, 
          `Homo (Neanderthal)` = 2, `Homo (Modern Human)` = 6,
          ExtractionControl = 3, LibraryControl = 7, Plaque = 9, 
          Gut = 7, Skin = 14, Sediment = 10, EnvironmentalControl = 12)

 env_shapes <- c(8,
                0,
                1,
                2,
                0,
                1,
                2,
                5,
                11,
                0,
                12,
                1,
                10,
                2,
                6,
                10,
                13,
                7,
                8,
                14,
                3,
                4,
                12,
                0)

names(env_shapes) <- c("Howler_Monkey", 
           "Gorilla_1", 
           "Gorilla_2", 
           "Gorilla_3", 
           "Chimp_1", 
           "Chimp_2", 
           "Chimp_3", 
           "Chimp_4",
           "Neanderthal", 
           "PreagriculturalHuman_1", 
           "PreagriculturalHuman_2", 
           "PreantibioticHuman_1", 
           "PreantibioticHuman_2", 
           "ModernDayHuman_1", 
           "ModernDayHuman_2", 
           "ExtractionControl", 
           "LibraryControl", 
           "ruralGut", 
           "sediment", 
           "skin", 
           "subPlaque",
           "supPlaque", 
           "urbanGut",
           "EnvironmentalControl")
```

## Calculations

Get total Streptococcus super-reference reads.

```{r}
## Collapse the contigs
superreference_data_meta <- superreference_data %>% 
  group_by(genus, sample, species) %>%
  summarise(total_no_reads = sum(no_reads),
             total_no_bases_covered = sum(no_bases_covered),
             total_species_length = sum(feature_length),
             mean_mean_depth_coverage = mean(mean_depth_coverage)
             ) %>%
    mutate(percent_covered = (total_no_bases_covered / total_species_length) * 100) %>%
  group_by(sample) %>%
  mutate(total_dataset_reads = sum(total_no_reads)) %>%
  left_join(raw_metadata_deep %>% 
              select(sample_name, Env, Description, Host_General, 
                     Host_Genus, Host_Common, Age, total_unmapped_reads), 
            by = c("sample" = "sample_name")
            ) %>%
    filter(Host_Genus != "Control")

superreference_data_total <- superreference_data_meta %>%
  filter(genus == "Streptococcus") %>%
  group_by(sample) %>%
  summarise(total_genus_reads = sum(total_no_reads)) %>%
  rename(individual = sample)
```


Now clean up abp reads and calculate sums per sample and gene

```{r}
abp_data_clean <- abp_data %>% 
  select(individual, abp_type, file_contents) %>%
  unnest() %>%
  group_by(individual, abp_type) %>%
  summarise(total_abp_reads = sum(reads))
```

Join together

```{r}
data_combined <- left_join(superreference_data_total, abp_data_clean) %>% 
  mutate(`Ratio of abp-associated reads to total Streptococcus reads` = total_abp_reads / total_genus_reads) %>%
  rename(Individual = individual,
         )
```

Set ordering

```{r}
data_combined <- data_combined %>% 
  mutate(Group = if_else(grepl("DJA|EBO|KNP|MTM|OME", Individual), "Primate", "Non-Primate"),
         Group = factor(Group, levels = c("Primate", "Non-Primate"))) %>%
  left_join(raw_metadata_deep, by = c("Individual" = "sample_name")) %>%
  mutate(Env = factor(Env, levels = names(env_shapes)),
         Host_Common = factor(Host_Common, levels = names(common_shapes))) %>%
  arrange(Env) %>%
  mutate(Individual = as_factor(Individual))
```

To save

```{r eval = F}
## And save
write_tsv(data_combined, "../04-analysis/deep/streptococcus_investigation.backup/Streptococcus_superreference_amylaseReadsOverAllReads_rawdata.tsv")
```

## Plotting

And plot

```{r}
final_plot <- ggplot(data_combined, aes(Individual, `Ratio of abp-associated reads to total Streptococcus reads`, fill = Host_Common)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = common_colours) +
  facet_grid(abp_type ~ Group, scales = "free_x") +
  theme_minimal(base_size = 7, base_family = "Roboto") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        strip.background = element_rect(colour = "lightgrey"),
        legend.position = "none")

final_plot
```

to save

```{r eval = F}
ggsave(plot = final_plot, 
       "../04-analysis/deep/streptococcus_investigation.backup/Streptococcus_superreference_amylaseReadsOverAllReads_plot.pdf",
      device = cairo_pdf,
      width = 3.5,
      height = 3.5,
      units = "in",
      dpi = 600)
```

Or

```{r}
ggplot(data_combined, aes(Group, `Ratio of abp-associated reads to total Streptococcus reads`)) +
  geom_boxplot() +
  geom_beeswarm(aes(fill = Host_Common, colour = Host_Common, shape = Host_Common)) +
  scale_fill_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  scale_colour_manual(values = common_colours) +
  facet_grid(~ abp_type, scales = "free_x") +
  theme_minimal(base_size = 7, base_family = "Roboto") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

