---
title: "Oral Evolution Functional Changes with HUMAnN2"
output: html_notebook
---

This looks at functional differences in the calculus samples of the 
oral deep evolution project

## Notebook Preparation

Load the packages for data import and processing.
```{r load_libraries}
library("data.table")
library(psych)
library(mixOmics)
library(vegan)
library(reshape2)
library(janitor)
library(usedist)
library(tidyverse)
library(gplots)
library(ggrepel)
library(viridis)
library(cowplot)
```

Set the working directory of the notebook to the scripts folder. 

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath("."))
```

 Read in the data HUMAnN2 path abundance data file 
```{r data_file}
# read in the full humann2 table that has spaces in names removed
fullset.humann2 <- fread("../04-analysis/screening/humann2/tables/humann2_pathabundance_20190408_nospace.tsv")


# Fix the column headers
fullset.humann2 <- dplyr::rename(fullset.humann2, Pathway = `#_Pathway`)
colnames(fullset.humann2) <- gsub("_S.*_L.*_R1_.*.fastq.merged.prefixed.hg19unmapped_Abundance-CPM","", colnames(fullset.humann2))
colnames(fullset.humann2) <- gsub("_S.*_L.*_R1_.*.fastq.merged_Abundance-CPM","", colnames(fullset.humann2))
colnames(fullset.humann2) <- gsub("_S.*_L.*_R1_.*.fastq.combined.fq.prefixed.extractunmapped.bam_Abundance-CPM","", colnames(fullset.humann2))

# remove problematic sample LIB017.A0101
fullset.humann2 <- fullset.humann2 %>%
  dplyr::select(-LIB017.A0101)

# Finish fixing the column headers
colnames(fullset.humann2) <- gsub(".A0101.SG1.2","", colnames(fullset.humann2))
colnames(fullset.humann2) <- gsub(".A0101","", colnames(fullset.humann2))
colnames(fullset.humann2) <- gsub(".A0102","", colnames(fullset.humann2))
colnames(fullset.humann2) <- gsub(".B0101","", colnames(fullset.humann2))

```

Prepare the metadata
```{r metadata_file}
# read in the metadata
fullmetadata <- fread("../00-documentation.backup/02-calculus_microbiome-deep_evolution-individualscontrolssources_metadata_20190523.tsv")
fullmetadata <- fullmetadata %>% rename(SampleID = `#SampleID`)

selectmetadata <- fullmetadata %>%
  rename(Sample = SampleID) %>%
dplyr::select(one_of(c("individual_id","Sample","Host_General","Host_Genus","Host_Common")))

fullset.humann2_metadata <- fullset.humann2 %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathway,CPM) %>% 
                                         inner_join(., selectmetadata, by = "individual_id") %>%
                                         select(-one_of("Sample","Host_General","Host_Genus","Host_Common")) %>%
                                         gather("Pathway","CPM",2:ncol(.)) %>%
                                         spread(individual_id,CPM)

selectmetadata <- inner_join(selectmetadata, fullset.humann2_metadata %>%
                               gather("individual_id","CPM",2:ncol(.)) %>%
                               spread(Pathway,CPM) %>%
                               select(individual_id))

# remove all the species-specific rows and the metadata rows
community_level <- fullset.humann2_metadata %>%
  filter(!str_detect(Pathway, "\\||UNMAPPED|UNINTEGRATED"))

row.names(community_level) <- community_level$Pathway
community_level <- community_level %>% select(-Pathway)

# remove the columns that sum to 0
fullset_noNA0.humann2 <- as.data.frame(sapply(community_level, as.numeric))
row.names(fullset_noNA0.humann2) <- row.names(community_level)

fullset_noNA0.humann2_filtered <- fullset_noNA0.humann2 %>%
  rownames_to_column("category") %>%
  adorn_totals(., where = "row", name = "Sum") %>%
  gather("individual_id","CPM", 2:ncol(.)) %>%
  spread(category,CPM, fill = 0) %>%
  filter(Sum > 0) %>%
  select(-Sum) %>%
  gather("category","CPM", 2:ncol(.)) %>%
  spread(individual_id,CPM)
  
```

Set the colors and shapes for plots
```{r colors_shapes}
env_colours <- c("#1f78b4", 
                 "#6a3d9a", 
                 "#6a3d9a", 
                 "#6a3d9a", 
                 "#33a02c", 
                 "#33a02c", 
                 "#33a02c", 
                 "#33a02c", 
                 "#e31a1c", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#d9d9d9", 
                 "#d9d9d9",
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9",
                 "#d9d9d9")

names(env_colours) <- c("Howler_Monkey", 
                        "Gorilla_1", 
                        "Gorilla_2", 
                        "Gorilla_3", 
                        "Chimp_1", 
                        "Chimp_2", 
                        "Chimp_3", 
                        "Chimp_4",
                        "Neanderthal", 
                        "PreagriculturalHuman_1", 
                        "PreagriculturalHuman_2", 
                        "PreantibioticHuman_1", 
                        "PreantibioticHuman_2", 
                        "ModernDayHuman_1", 
                        "ModernDayHuman_2", 
                        "ExtractionControl", 
                        "LibraryControl", 
                        "ruralGut", 
                        "urbanGut",
                        "sediment", 
                        "skin", 
                        "subPlaque",
                        "supPlaque",
                        "EnvironmentalControl")

env_shapes <- c(8,
                0,
                1,
                2,
                0,
                1,
                2,
                5,
                11,
                0,
                12,
                1,
                10,
                2,
                6,
                10,
                13,
                7,
                8,
                14,
                3,
                4,
                12,
                0)


names(env_shapes) <- c("Howler_Monkey", 
                       "Gorilla_1", 
                       "Gorilla_2", 
                       "Gorilla_3", 
                       "Chimp_1", 
                       "Chimp_2", 
                       "Chimp_3", 
                       "Chimp_4",
                       "Neanderthal", 
                       "PreagriculturalHuman_1", 
                       "PreagriculturalHuman_2", 
                       "PreantibioticHuman_1", 
                       "PreantibioticHuman_2", 
                       "ModernDayHuman_1", 
                       "ModernDayHuman_2", 
                       "ExtractionControl", 
                       "LibraryControl", 
                       "ruralGut", 
                       "sediment", 
                       "skin", 
                       "subPlaque",
                       "supPlaque", 
                       "urbanGut",
                       "EnvironmentalControl")

common_colours <- c(Alouatta = "#1f78b4", Gorilla = "#6a3d9a", Pan = "#33a02c", 
                    `Homo (Neanderthal)` = "#fdbf6f", 
                    `Homo (Modern Human)` = "#ff7f00", ExtractionControl = "#d9d9d9", 
                    LibraryControl = "#d9d9d9", Plaque = "#d9d9d9", Gut = "#d9d9d9", 
                    Skin = "#d9d9d9", Sediment = "#d9d9d9", EnvironmentalControl = "#d9d9d9")

common_shapes <- c(Alouatta = 8, Gorilla = 0, Pan = 1, 
                   `Homo (Neanderthal)` = 2, `Homo (Modern Human)` = 6,
                   ExtractionControl = 3, LibraryControl = 13, Plaque = 14, 
                   Gut = 7, Skin = 9, Sediment = 10, EnvironmentalControl = 12)


genus_colours <- c(Alouatta = "#1f78b4", Gorilla = "#6a3d9a", Pan = "#33a02c", 
                   Homo = "#ff7f00",  ExtractionBlank = "#d9d9d9", 
                   LibraryBlank ="#d9d9d9", Plaque = "#d9d9d9", Gut = "#d9d9d9", 
                   Skin = "#d9d9d9", Sediment = "#d9d9d9", Control = "#d9d9d9")

general_colors <- c(Howler = "#1f78b4", Gorilla = "#6a3d9a", Chimp = "#33a02c", 
                    Neanderthal = "#ff7f00", PreagriculturalHuman = "#ff7f00",
                    PreantibioticHuman = "#ff7f00", ModernDayHuman = "#ff7f00", ExtractionControl = "#d9d9d9", 
                    LibraryControl = "#d9d9d9", plaque = "#d9d9d9", gut = "#d9d9d9", 
                    skin = "#d9d9d9", sediment = "#d9d9d9", EnvironmentalControl = "#d9d9d9")

general_shapes <- c(Howler = 8, Gorilla = 0, Chimp = 1, 
                   Neanderthal = 2, PreagriculturalHuman = 0,
                   PreantibioticHuman = 1, ModernDayHuman = 6,
                   ExtractionControl = 3, LibraryControl = 13, plaque = 14, 
                   gut = 7, skin = 9, sediment = 10, EnvironmentalControl = 12)

```

Perform a PCA to look for outliers in the dataset
```{r pathway_pca}

# transpose the tables for input to mixOmics
fullset_noNA0.humann2T <- fullset_noNA0.humann2_filtered %>%
  gather("individual_id","CPM",2:ncol(.)) %>%
  spread(category,CPM, fill = 0)

fullset_noNA0.humann2T.rownames <- fullset_noNA0.humann2T$individual_id
fullset_noNA0.humann2T <- fullset_noNA0.humann2T %>% select(-individual_id)
fullset_noNA0.humann2T <- as.data.frame(sapply(fullset_noNA0.humann2T, as.numeric))
rownames(fullset_noNA0.humann2T) <- fullset_noNA0.humann2T.rownames

fullset_noNA0.humann2T = fullset_noNA0.humann2T+1

# run a PCA to look at the plot for outliers/"bad" samples
evolutionh2.pca <- mixOmics::pca(fullset_noNA0.humann2T, ncomp = 3, logratio = 'CLR')
plot(evolutionh2.pca)
evolutionh2.pca

evolutionh2.variates <- as.data.frame(evolutionh2.pca$variates$X)
evolutionh2.variates <- evolutionh2.variates %>%
  rownames_to_column("individual_id")
  
evolutionh2.pca.varmet <- inner_join(evolutionh2.variates, fullmetadata, by = "individual_id")

evolutionh2.pca.varmet$Host_Common <- factor(evolutionh2.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl","LibraryControl"))

pca_oral_h2_pc1pc2_allsamples <- ggplot(as.data.frame(evolutionh2.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(evolutionh2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(evolutionh2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7)

pca_oral_h2_pc1pc2_allsamples

# ggsave("../pca_oral_h2_pc1pc2_allsamples.pdf", plot = pca_oral_h2_pc1pc2_allsamples, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```

List the outliers from the PCA plots and remove them along with the sources and
blanks from the data table and metadata table. At this point we also split the
data analysis into 2 lines - oral and calculus. Oral includes the Human
Microbiome Project plaque samples, while calculus does not.
```{r paths_outliers_pca}
# List the outliers from the PCA plots
outliersH2_paths <- c("BAN001","BIT001","BSQ001","CDC007","ECO006",
              "ECO010","FDM001","FUM003","GDC002","GDC004",
              "GDC005","GUE001","IBA001","IBA002","LOB001",
              "MTM002","MTM003","MTM006","MTM011","MTM012",
              "MTM013","MTS001","MTS002","MTS003","OAK001",
              "SPM001","SpyNew","SpyOld","WAL001","ZAF001")


fullset_noNA0.humann2_outrem <- fullset_noNA0.humann2_filtered %>%
  select(-one_of(outliersH2_paths)) %>%
  gather("individual_id","CPM", 2:ncol(.)) %>%
  spread(category,CPM)

fullset_noNA0.humann2_outrem_rownames <- fullset_noNA0.humann2_outrem$individual_id
fullset_noNA0.humann2_outrem <- fullset_noNA0.humann2_outrem %>%
                                            select(-individual_id)

fullset_noNA0.humann2_outrem <- as.data.frame(sapply(fullset_noNA0.humann2_outrem, as.numeric))
rownames(fullset_noNA0.humann2_outrem) <- fullset_noNA0.humann2_outrem_rownames
fullset_noNA0.humann2_outrem = fullset_noNA0.humann2_outrem+1

evolutionh2_outrem.pca <- mixOmics::pca(fullset_noNA0.humann2_outrem, ncomp = 3, logratio = 'CLR')
plot(evolutionh2_outrem.pca)
evolutionh2_outrem.pca

evolutionh2_outrem.variates <- as.data.frame(evolutionh2_outrem.pca$variates$X)
evolutionh2_outrem.variates <- evolutionh2_outrem.variates %>%
  rownames_to_column("individual_id")
  
evolutionh2_outrem.pca.varmet <- inner_join(evolutionh2_outrem.variates, fullmetadata, by = "individual_id")

evolutionh2_outrem.pca.varmet$Host_Common <- factor(evolutionh2_outrem.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_oral_h2_pc1pc2_allsamples_outrem <- ggplot(as.data.frame(evolutionh2_outrem.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(evolutionh2_outrem.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(evolutionh2_outrem.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7)

pca_oral_h2_pc1pc2_allsamples_outrem

ggsave("../pca_oral_h2_pc1pc2_allsamples_outrem.pdf", plot = pca_oral_h2_pc1pc2_allsamples_outrem, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)


# List out the non-oral sources and controls
sources_oral <- selectmetadata %>%
  filter(str_detect(Host_Genus, "Gut|Skin|Sediment|Control")) %>%
  select(contains("individual_id")) %>%
  as.matrix(.) %>%
  as.vector(.)

sources_calc <- selectmetadata %>%
  filter(str_detect(Host_Genus, "Plaque|Gut|Skin|Sediment|Control")) %>%
  select(contains("individual_id")) %>%
  as.matrix(.) %>%
  as.vector(.)

# Now remove the outliers and remaining extraction/library blank controls
oral_humann2 <- fullset_noNA0.humann2 %>%
  select(-one_of(outliersH2_paths)) %>%
  select(-one_of(sources_oral)) %>%
  select(-starts_with("ERR")) %>%
  select(-starts_with("ARS")) %>%
  select(-starts_with("LIB")) %>%
  select(-starts_with("EXB"))

calc_humann2 <- fullset_noNA0.humann2 %>%
  select(-one_of(outliersH2_paths)) %>%
  select(-one_of(sources_calc)) %>%
  select(-starts_with("SRR")) %>%
  select(-starts_with("ERR")) %>%
  select(-starts_with("ARS")) %>%
  select(-starts_with("LIB")) %>%
  select(-starts_with("EXB"))

```

Remove any additional columns from the datatable that sum to 0 after 
removing the outliers, blanks, and sources
```{r paths_nosources_pca}
# transpose the tables for input to mixOmics and remove samples that have no entries (sum to 0), then remove paths that are present at <0.05% abundance
oral_humann2_filtered <- oral_humann2 %>%
  rownames_to_column("category") %>%
  adorn_totals(., where = "row", name = "Sum") %>%
  gather("individual_id","CPM", 2:ncol(.)) %>%
  spread(category,CPM, fill = 0) %>%
  filter(Sum > 0) %>%
  select(-Sum) %>%
  gather("category","CPM", 2:ncol(.)) %>%
  spread(individual_id,CPM) %>%
  adorn_totals(., where = "col", name = "Sum_category") %>%
  mutate(Percent = Sum_category/sum(Sum_category)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_category","Percent"))) %>%
  gather("individual_id","CPM",2:ncol(.)) %>%
  spread(category,CPM)

oral_humann2_filtered.rownames <- oral_humann2_filtered$individual_id
oral_humann2_filtered <- oral_humann2_filtered %>% select(-individual_id)
oral_humann2_filtered <- as.data.frame(sapply(oral_humann2_filtered, as.numeric))
rownames(oral_humann2_filtered) <- oral_humann2_filtered.rownames


calc_humann2_filtered <- calc_humann2 %>%
  rownames_to_column("category") %>%
  adorn_totals(., where = "row", name = "Sum") %>%
  gather("individual_id","CPM", 2:ncol(.)) %>%
  spread(category,CPM, fill = 0) %>%
  filter(Sum > 0) %>%
  select(-Sum) %>%
  gather("category","CPM", 2:ncol(.)) %>%
  spread(individual_id,CPM) %>%
  adorn_totals(., where = "col", name = "Sum_category") %>%
  mutate(Percent = Sum_category/sum(Sum_category)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_category","Percent"))) %>%
  gather("individual_id","CPM",2:ncol(.)) %>%
  spread(category,CPM)

calc_humann2_filtered.rownames <- calc_humann2_filtered$individual_id
calc_humann2_filtered <- calc_humann2_filtered %>% select(-individual_id)
calc_humann2_filtered <- as.data.frame(sapply(calc_humann2_filtered, as.numeric))
rownames(calc_humann2_filtered) <- calc_humann2_filtered.rownames

```

Perform a PCA to look at the relation of the samples from different hosts to each other
```{r paths_oral_pca}
# add +1 to all values so the CLR can be performed
oral_humann2_filtered <- oral_humann2_filtered + 1
calc_humann2_filtered <- calc_humann2_filtered + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.oralpca <- tune.pca(oral_humann2_filtered, logratio = 'CLR')
tune.oralpca

tune.calcpca <- tune.pca(calc_humann2_filtered, logratio = 'CLR')
tune.calcpca

# perform a PCA to see how the data cluster
oralhumann2.pca <- mixOmics::pca(oral_humann2_filtered, ncomp = 4, logratio = 'CLR')
plot(oralhumann2.pca)

calchumann2.pca <- mixOmics::pca(calc_humann2_filtered, ncomp = 4, logratio = 'CLR')
plot(calchumann2.pca)

```

Plot oral samples PC1-PC2
```{r paths_oral_pca_plot_12}
oralhumann2.pca.variates <- as.data.frame(oralhumann2.pca$variates$X)
oralhumann2.pca.variates <- oralhumann2.pca.variates %>%
  rownames_to_column("individual_id")
  
oralhumann2.pca.varmet <- inner_join(oralhumann2.pca.variates, fullmetadata, by = "individual_id")

oralhumann2.pca.varmet$Host_Common <- factor(oralhumann2.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_oral_h2_pc1pc2 <- ggplot(as.data.frame(oralhumann2.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(oralhumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(oralhumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7)

pca_oral_h2_pc1pc2

# ggsave("../pca_oral_h2_pc1pc2.pdf", plot = pca_oral_h2_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```

Plot oral samples PC3-PC2
```{rpaths_oral_pca_plot_23}
oralhumann2.pca.variates <- as.data.frame(oralhumann2.pca$variates$X)
oralhumann2.pca.variates <- oralhumann2.pca.variates %>%
  rownames_to_column("individual_id")
  
oralhumann2.pca.varmet <- inner_join(oralhumann2.pca.variates, fullmetadata, by = "individual_id")

oralhumann2.pca.varmet$Host_Common <- factor(oralhumann2.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_oral_h2_pc2pc3 <- ggplot(as.data.frame(oralhumann2.pca.varmet), aes(PC3, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC3 - ", round(oralhumann2.pca$explained_variance[3], 2), sep = "")) +
  ylab(paste("PC2 - ", round(oralhumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7)

pca_oral_h2_pc2pc3

# ggsave("../pca_oral_h2_pc2pc3.pdf", plot = pca_oral_h2_pc2pc3, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```

Plot calculus samples PC1-PC2
```{r paths_calculus_pca_plot_12}
calchumann2.pca.variates <- as.data.frame(calchumann2.pca$variates$X)
calchumann2.pca.variates <- calchumann2.pca.variates %>%
  rownames_to_column("individual_id")
  
calchumann2.pca.varmet <- inner_join(calchumann2.pca.variates, fullmetadata, by = "individual_id")

calchumann2.pca.varmet$Host_Common <- factor(calchumann2.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_calc_h2_pc1pc2 <- ggplot(as.data.frame(calchumann2.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(calchumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7)

pca_calc_h2_pc1pc2

# ggsave("../pca_calc_h2_pc1pc2.pdf", plot = pca_calc_h2_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```

Plot calculus samples PC3-PC2
```{r paths_calculus_pca_plot_23}
calchumann2.pca.variates <- as.data.frame(calchumann2.pca$variates$X)
calchumann2.pca.variates <- calchumann2.pca.variates %>%
  rownames_to_column("individual_id")
  
calchumann2.pca.varmet <- inner_join(calchumann2.pca.variates, fullmetadata, by = "individual_id")

calchumann2.pca.varmet$Host_Common <- factor(calchumann2.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_calc_h2_pc2pc3 <- ggplot(as.data.frame(calchumann2.pca.varmet), aes(PC3, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC3 - ", round(calchumann2.pca$explained_variance[3], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7)  

pca_calc_h2_pc2pc3

# ggsave("../pca_calc_h2_pc2pc3.pdf", plot = pca_calc_h2_pc2pc3, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```

Set the sample order by host genus for plotting heatmaps
```{r sample_order}
# set the sample order
sampleorder.oral = as.vector(c("OME002","OME003","OME005","OME007","OME009","ABM006",
                               "DJA002","EBO001","EBO003","MTK001","MTM001",
                               "MTM004","MTM005","MTM008","MTM009","MTM010",
                               "ABM007","ABM008","CDC005","CDC006","CDC008",
                               "CDC009","CDC010","CDC011","Chimp.150519","DJA006","DJA007",
                               "EBO008","EBO009","KNP001","KNP003","KNP004","KNP005",
                               "KNP006","KNP007","KNP009","ElSidron1","ElSidron2","FUM001",
                               "FUM002","GDN001","GOY005","GOY006","PES001","SPM002",
                               "VLD001","DLV001","ECO002", "ECO004","EMN001","MOA001","OAK002",
                               "OAK003","OAK004","OAK005","OFN001","RIG001","PLV001","TAF008","TAF017","TAF018",
                               "ESA003","ESA004","ESA006","ESA007","ESA008","LIS001",
                               "LIS002","LIS003","PYK001","PYK002","PYK003","PYK004",
                               "PYK005","JAE006","JAE007","JAE008","JAE009","JAE010",
                               "JAE012","JAE013","JAE014","JAE015","JAE016","VLC001",
                               "VLC002","VLC003","VLC004","VLC005","VLC008","VLC009",
                               "VLC010","SRR061192","SRR061294","SRR061320","SRR061365",
                               "SRR061562","SRR062298","SRR062299","SRR063517","SRR1804664",
                               "SRR1804823","SRR512767","SRR513165","SRR513768","SRR513775",
                               "SRR513828","SRR514202","SRR514239","SRR514306","SRR514329"))

sampleorder.calc = as.vector(c("OME002","OME003","OME005","OME007","OME009","ABM006",
                               "DJA002","EBO001","EBO003","MTK001","MTM001",
                               "MTM004","MTM005","MTM008","MTM009","MTM010",
                               "ABM007","ABM008","CDC005","CDC006","CDC008",
                               "CDC009","CDC010","CDC011","Chimp.150519","DJA006","DJA007",
                               "EBO008","EBO009","KNP001","KNP003","KNP004","KNP005",
                               "KNP006","KNP007","KNP009","ElSidron1","ElSidron2","FUM001",
                               "FUM002","GDN001","GOY005","GOY006","PES001","SPM002",
                               "VLD001","DLV001","ECO002", "ECO004","EMN001","MOA001","OAK002",
                               "OAK003","OAK004","OAK005","OFN001","RIG001","PLV001","TAF008","TAF017","TAF018",
                               "ESA003","ESA004","ESA006","ESA007","ESA008","LIS001",
                               "LIS002","LIS003","PYK001","PYK002","PYK003","PYK004",
                               "PYK005","JAE006","JAE007","JAE008","JAE009","JAE010",
                               "JAE012","JAE013","JAE014","JAE015","JAE016","VLC001",
                               "VLC002","VLC003","VLC004","VLC005","VLC008","VLC009",
                               "VLC010"))
```

Data reduction for visualization and interpretation, based on the PCA results.
Select the pathways that have the highest absolute value of loadings from the
PCA in components 1 and 2 using an essentially arbitrary cut-off value. Really
it's chosen to make the heatmaps visually interpretable (not so many that you
can't read them), and with the hope that the clusters make some biological
sense. Then plot the values of these pathways reported by HUMAnN2 in heatmaps.
```{r pc1_loadings}
# pull out the pathways with loadings > 0.08 in PC1
# make a list of the factors with loadings >= 0.08 in each minimum residual (MR)
# turn the loadings category of the psych object into a data frame to be able to work w/it using tidyverse
oralhumann2.pca.loadings <- as.data.frame(oralhumann2.pca$loadings)
calchumann2.pca.loadings <- as.data.frame(calchumann2.pca$loadings)

# select out the factors in PC1 and PC2 that have a loading >=0.05
# for oral samples
oral.pcaPC1 <- oralhumann2.pca.loadings %>%
  dplyr::select(contains("PC1")) %>%
  rownames_to_column("pathways") %>%
  filter(X.PC1 >=0.05) %>%
  pull(pathways) 

oral.pcaPC2 <- oralhumann2.pca.loadings %>%
  dplyr::select(X.PC2) %>%
  rownames_to_column("pathways") %>%
  filter(X.PC2 >=0.08) %>%
  pull(pathways) 

# for calc samples
calc.pcaPC1 <- calchumann2.pca.loadings %>%
  dplyr::select(contains("PC1")) %>%
  rownames_to_column("pathways") %>%
  filter(X.PC1 >=0.05) %>%
  pull(pathways) 

calc.pcaPC2 <- calchumann2.pca.loadings %>%
  dplyr::select(X.PC2) %>%
  rownames_to_column("pathways") %>%
  filter(X.PC2 >=0.08) %>%
  pull(pathways) 

# now pull them out of the data.frame and make into a matrix to put into heatmap.2
# oral.pca1_table <- oral_humann2T %>%
#                               as.data.frame(.) %>%
#                               dplyr::select(one_of(oral.pcaPC1))
# 
# oral.pca2_table <- oral_humann2T %>%
#                               as.data.frame(.) %>%
#                               dplyr::select(one_of(oral.pcaPC2))
# 
# 
# calc.pca1_table <- calc_humann2T %>%
#                                as.data.frame(.) %>%
#                                dplyr::select(one_of(calc.pcaPC1))
# 
# calc.pca2_table <- calc_humann2T %>%
#                                as.data.frame(.) %>%
#                                dplyr::select(one_of(calc.pcaPC2))
# 
# 
# # reorder the table so the rows match as above
# oral.pca1_table.ordered <- oral.pca1_table[match(sampleorder.oral, rownames(oral.pca1_table)),]
# oral.pca2_table.ordered <- oral.pca2_table[match(sampleorder.oral, rownames(oral.pca2_table)),]
# oral.pca1_table.ordered <- as.matrix(sapply(oral.pca1_table.ordered, as.numeric))
# oral.pca2_table.ordered <- as.matrix(sapply(oral.pca2_table.ordered, as.numeric))
# 
# calc.pca1_table.ordered <- calc.pca1_table[match(sampleorder.calc, rownames(calc.pca1_table)),]
# calc.pca2_table.ordered <- calc.pca2_table[match(sampleorder.calc, rownames(calc.pca2_table)),]
# calc.pca1_table.ordered <- as.matrix(sapply(calc.pca1_table.ordered, as.numeric))
# calc.pca2_table.ordered <- as.matrix(sapply(calc.pca2_table.ordered, as.numeric))
```

Plot the oral sample heat maps
```{r eval = F}
# log10-transform the matrix heatmap
oral.pca1_table.ordered = oral.pca1_table.ordered +1
oral.pca1_table.orderedLog <- log10(oral.pca1_table.ordered)
rownames(oral.pca1_table.orderedLog) <- sampleorder.oral
heatmap.oral.PC1 <- heatmap.2(oral.pca1_table.orderedLog, Rowv = FALSE, dendrogram = "column",
                              trace = "none", col = viridis_pal(option = "D"),
                              margins = c(16,10), srtCol = 30, cexCol = 0.5, cexRow = 0.5,
                              key = T, keysize = 0.8, main = "Oral Samples PC1 >0.05")

oral.pca2_table.ordered = oral.pca2_table.ordered +1
oral.pca2_table.orderedLog <- log10(oral.pca2_table.ordered)
rownames(oral.pca2_table.orderedLog) <- sampleorder.oral
heatmap.oral.PC2 <- heatmap.2(oral.pca2_table.orderedLog, Rowv = FALSE, dendrogram = "column",
                              trace = "none", col = viridis_pal(option = "D"),
                              margins = c(16,10), srtCol = 30, cexCol = 0.5, cexRow = 0.7,
                              key = T, keysize = 0.8, main = "Oral Samples PC2 >0.08")



```

Plot the calculus sample heat maps
```{r eval = F}
calc.pca1_table.ordered = calc.pca1_table.ordered +1
calc.pca1_table.orderedLog <- log10(calc.pca1_table.ordered)
rownames(calc.pca1_table.orderedLog) <- sampleorder.calc
heatmap.calc.PC1 <- heatmap.2(calc.pca1_table.orderedLog, Rowv = FALSE, dendrogram = "column",
                              trace = "none", col = viridis_pal(option = "D"),
                              margins = c(16,10), srtCol = 30, cexCol = 0.5, cexRow = 0.5,
                              key = T, keysize = 0.8, main = "Calculus Samples PC1 >0.05")

calc.pca2_table.ordered = calc.pca2_table.ordered +1
calc.pca2_table.orderedLog <- log10(calc.pca2_table.ordered)
rownames(calc.pca2_table.orderedLog) <- sampleorder.calc
heatmap.calc.PC2 <- heatmap.2(calc.pca2_table.orderedLog, Rowv = FALSE, dendrogram = "column",
                              trace = "none", col = viridis_pal(option = "D"),
                              margins = c(16,10), srtCol = 30, cexCol = 0.5, cexRow = 0.7,
                              key = T, keysize = 0.8, main = "Calculus Samples PC2 >0.08")
```

While the overall pathway differences weren't very informative, maybe they will
be for specific genera. Look at *Streptococcus* and *Actinomyces.* Start with
*Streptococcus.* Pull out all the pathways that are assigned to any
*Streptococcus* species and see if there are differences in the abundance by
host genus.
```{r Strep_paths}
# First Strep
# select only the rows with Streptococcus, and the metadata to use for stats
streptococcus_paths <- fullset.humann2_metadata %>%
  filter(str_detect(Pathway, "Streptococcus")) %>%
  dplyr::select(-one_of(outliersH2_paths)) %>%
  dplyr::filter(!str_detect(Pathway, "UNINTEGRATED")) %>%
  gather("individual_id","CPM",2:ncol(.)) %>%
  spread(Pathway, CPM)

# remove the control samples and re-organize the table to put metadata in the front
streptococcus_paths <- streptococcus_paths %>%
  inner_join(., selectmetadata, by = "individual_id") %>%
  dplyr::filter(!str_detect(Host_Genus, "Gut|Skin|Sediment|Control")) %>%
  select(Host_Genus, Host_General, Host_Common, everything())

# put the rownames back
rownames(streptococcus_paths) <- streptococcus_paths$Sample

# convert data to a long table
streptococcus_paths_long <- streptococcus_paths %>%
  tidyr::gather("Pathway","CPM", 5:ncol(.)) 

streptococcus_paths_long$CPM <- as.numeric(streptococcus_paths_long$CPM)
# set the metadata columns as factors
streptococcus_paths_long$Host_Genus <- factor(streptococcus_paths_long$Host_Genus, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo", "Plaque"))

# perform a kruskal-wallace test to determine if the pathways are differentially abundant
streptococcus.kw <- kruskal.test(streptococcus_paths_long$CPM, streptococcus_paths_long$Host_Genus)
streptococcus.kw

# perform a post-hoc test to detrmine if the pathways are differentially abundant
pairwise.wilcox.test(streptococcus_paths_long$CPM, streptococcus_paths_long$Host_Genus,
                     p.adjust.method = "fdr")
```

Calculate summary stats by host genus and by pathway
```{r strep_genus_stats}
# compute summary statistics by host genus
streptococcus_stats_group <- group_by(streptococcus_paths_long, Host_Genus) %>%
  summarise(
    mean = mean(CPM, na.rm = TRUE),
    sd = sd(CPM, na.rm = TRUE),
    median = median(CPM, na.rm = TRUE),
    count = n(),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM, na.rm = TRUE)
  )

# compute summary statistics by pathway
streptococcus_stats_pathway <- group_by(streptococcus_paths_long, Pathway) %>%
  summarise(
    mean = mean(CPM, na.rm = TRUE),
    sd = sd(CPM, na.rm = TRUE),
    median = median(CPM, na.rm = TRUE),
    count = n(),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM, na.rm = TRUE)
  )

```

Graph the distribution of pathway abundance in each host genus. The abundance of
*Streptococcus* pathways is lower in Pan and Gorilla than in Homo.
```{r strep_paths_plot}
streptococcus_paths_long$Host_Common <- factor(streptococcus_paths_long$Host_Common, levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)","Plaque"))

# graph distribution of all paths in each genus
strep_paths_graph <- streptococcus_paths_long %>%
  #filter(CPM<10000) %>%
  ggplot(., aes(x=Host_Common, y=CPM, fill = Host_Common)) +
  geom_boxplot() +
  theme_minimal() +
  scale_fill_manual(values = common_colours) +
  theme(axis.text.x = element_text(angle=45, hjust = 1)) +
  scale_y_continuous(trans='log10') +
  ggtitle("Streptococcus pathways")

strep_paths_graph

ggsave("../strep_paths_graph.pdf", plot = strep_paths_graph, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)

# graph distribution of all paths by each genus (NOT recommended - 385 total!)
#streptococcus_paths_long %>%
#  #filter(CPM<10000) %>%
#  ggplot(., aes(x=Pathway, y=CPM, fill = Host_Genus)) +
#  geom_boxplot() +
#  #geom_point(position = position_jitterdodge()) +
#  theme_classic() +
#  #theme(axis.text.x = element_text(angle=45, hjust = 1)) +
#  theme(axis.text.x = element_blank()) +
#  scale_y_continuous(trans='log10') +
#  ggtitle("Streptococcus pathways")
```

Now do the same for Actinomyces. 
```{r actinomyces_paths}
# Now Actinomyces
# select only the rows with Actinomyces, and the metadata to use for stats
actinomyces_paths <- fullset.humann2_metadata %>%
  filter(str_detect(Pathway, "Actinomyces")) %>%
  dplyr::select(-one_of(outliersH2_paths)) %>%
  dplyr::filter(!str_detect(Pathway, "UNINTEGRATED")) %>%
  gather("individual_id","CPM",2:ncol(.)) %>%
  spread(Pathway, CPM)

# remove the control samples and re-organize the table to put metadata in the front
actinomyces_paths <- actinomyces_paths %>%
  inner_join(., selectmetadata, by = "individual_id") %>%
  dplyr::filter(!str_detect(Host_Genus, "Gut|Skin|Sediment|Control")) %>%
  select(Host_Genus, Host_General, Host_Common, everything())

# put the rownames back
rownames(actinomyces_paths) <- actinomyces_paths$Sample

# convert data to a long table
actinomyces_paths_long <- actinomyces_paths %>%
  tidyr::gather("Pathway","CPM", 5:ncol(.)) 

actinomyces_paths_long$CPM <- as.numeric(actinomyces_paths_long$CPM)

# set the metadata columns as factors
actinomyces_paths_long$Host_Genus <- factor(actinomyces_paths_long$Host_Genus, 
                                            levels = c("Alouatta", "Gorilla", "Pan", "Homo", "Plaque"))

# perform a kruskal-wallace test to determine if the pathways are differentially abundant
actinomyces.kw <- kruskal.test(actinomyces_paths_long$CPM, actinomyces_paths_long$Host_Genus)
actinomyces.kw

# perform a post-hoc test to detrmine if the pathways are differentially abundant
pairwise.wilcox.test(actinomyces_paths_long$CPM, actinomyces_paths_long$Host_Genus,
                     p.adjust.method = "fdr")


```

Calculate summary stats by host genus and by pathway
```{r actinomyces_stats_genus}
# compute summary statistics by host genus
actinomyces_stats_group <- group_by(actinomyces_paths_long, Host_Genus) %>%
  summarise(
    mean = mean(CPM, na.rm = TRUE),
    sd = sd(CPM, na.rm = TRUE),
    median = median(CPM, na.rm = TRUE),
    count = n(),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM, na.rm = TRUE)
  )

# compute summary statistics by pathway
actinomyces_stats_pathway <- group_by(actinomyces_paths_long, Pathway) %>%
  summarise(
    mean = mean(CPM, na.rm = TRUE),
    sd = sd(CPM, na.rm = TRUE),
    median = median(CPM, na.rm = TRUE),
    count = n(),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM, na.rm = TRUE)
  )

```

Graph the distribution of pathway abundance in each host genus. The abundance of
*Actinomyces* pathways is similar across host genera.
```{r actino_paths_plot}
actinomyces_paths_long$Host_Common <- factor(actinomyces_paths_long$Host_Common, levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)","Plaque"))

# graph distribution of all paths in each genus
actino_paths_graph <- actinomyces_paths_long %>%
  #filter(CPM<10000) %>%
  ggplot(., aes(x=Host_Common, y=CPM, fill = Host_Common)) +
  geom_boxplot() +
  #geom_point(position = position_jitterdodge()) +
  theme_minimal() +
  scale_fill_manual(values = common_colours) +
  theme(axis.text.x = element_text(angle=45, hjust = 1)) +
  scale_y_continuous(trans='log10') +
  ggtitle("Actinomyces pathways")

actino_paths_graph

ggsave("../actino_paths_graph.pdf", plot = actino_paths_graph, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)

# graph distribution of all paths by each genus (NOT recommended - 566 total!)
#actinomyces_paths_long %>%
#  #filter(CPM<10000) %>%
#  ggplot(., aes(x=Pathway, y=CPM, fill = Host_Genus)) +
#  geom_boxplot() +
#  #geom_point(position = position_jitterdodge()) +
#  theme_classic() +
#  #theme(axis.text.x = element_text(angle=45, hjust = 1)) +
#  theme(axis.text.x = element_blank()) +
#  scale_y_continuous(trans='log10') +
#  ggtitle("Actinomyces pathways")
```


Now look at whether specific pathways related to sugar transport are different
between host genera. The sugar phospho-transferrase system (PTS) is involved in
transporting simple sugars across the bacterial cell membrane for energy
metabolism. There are many different enzymes that do this, and they are
specialized for different sugars. Typically in Gram-positive bacteria there are
4 enzymes involved in transporting specific sugars, where 2 are general and used
by all systems, and 2 are sugar-specific. Studies on S. mutans indicates that
some of these are expressed consitituatively (all the time), while others are
expressed only when the target sugar is present. For a super duper indepth revew
on the PTS, see doi:10.1128/MMBR.00024-06. Genes involved in sugar transport and
metabolism were reported in *S. mutans* to have been under positive selection,
possibly related to expansion of the species during the human agricultural
revolution and increased variety of carbohydrates in the diet doi:
10.1093/molbev/mss278. We chose to look at the presence/abundance and
distribution of PTS genes across host genera using KEGG identifiers. This
information comes from the genefamilies table from HUMAnN2. From a cursory
inspection of the genefamilies table, very few PTS enzymes were specifically
assigned to *Streptococcus*, so we looked at all and did not select out the
*Strep*-specific assignments.

Read in the genefamilies data file, clean up the headers, remove the blanks and controls. 
```{r KOrthologs_genefamilies}
gene_families.humann2 <- fread("../04-analysis/screening/humann2/tables/humann2_genefamilies_ko_20190408.tsv.gz")

# Fix the column headers
names(gene_families.humann2)[1]<-"Pathways"

colnames(gene_families.humann2) <- gsub("_S.*_L.*_R1_.*.fastq.merged.prefixed.hg19unmapped_Abundance-CPM","", colnames(gene_families.humann2))
colnames(gene_families.humann2) <- gsub("_S.*_L.*_R1_.*.fastq.merged_Abundance-CPM","", colnames(gene_families.humann2))
colnames(gene_families.humann2) <- gsub("_S.*_L.*_R1_.*.fastq.combined.fq.prefixed.extractunmapped.bam_Abundance-CPM","", colnames(gene_families.humann2))

# remove problematic sample LIB017.A0101
gene_families.humann2 <- gene_families.humann2 %>%
  dplyr::select(-LIB017.A0101)

# Finish fixing the column headers
colnames(gene_families.humann2) <- gsub(".A0101.SG1.2","", colnames(gene_families.humann2))
colnames(gene_families.humann2) <- gsub(".A0101","", colnames(gene_families.humann2))
colnames(gene_families.humann2) <- gsub(".A0102","", colnames(gene_families.humann2))
colnames(gene_families.humann2) <- gsub(".B0101","", colnames(gene_families.humann2))

# Now remove the outliers and remaining extraction/library blank controls
# these are defined early on in the main script
gene_families.humann2_filtered <- gene_families.humann2 %>%
  dplyr::select(-starts_with("EXB")) %>%
  dplyr::select(-starts_with("LIB")) %>%
  dplyr::select(-one_of(outliersH2_paths)) %>%
  dplyr::select(-one_of(sources_oral)) %>%
  filter(!str_detect(Pathways, "\\|"))
```

How many reads were assigned and how many were not assigned in each sample, and does the average % of assigned reads differ by host genus?
```{bash, eval = F}
# this table is huge so we need to remove the species-level assignments so it doesn't time out opening for filtering in RStudio on our laptop
zcat /Users/velsko/projects1/microbiome_calculus/evolution/04-analysis/screening/humann2/tables/humann2_genefamilies_20190408.tsv.gz | grep -v "|" > /Users/velsko/projects1/microbiome_calculus/evolution/04-analysis/screening/humann2/tables/humann2_genefamilies_20190408_nospecies.tsv # 1.1GB
gzip /Users/velsko/projects1/microbiome_calculus/evolution/04-analysis/screening/humann2/tables/humann2_genefamilies_20190408_nospecies.tsv # 105MB

```


```{r KO_unassigned}
# The table created by this first part is saved as an RData object that can be loaded instead to save time

# read in the table that doesn't have the KOs collected together, so we can get the counts of assigned and unassigned reads for each sample
# This takes a very long time to load and now times out
# gene_families.humann2.uniref <- fread("grep -v '|' /Users/velsko/projects1/microbiome_calculus/evolution/04-analysis/screening/humann2/tables/humann2_genefamilies_20190408.tsv.gz")

# this is the table created in the bash section just above
gene_families.humann2.uniref <- fread("../04-analysis/screening/humann2/tables/humann2_genefamilies_20190408_nospecies.tsv.gz", sep = "\t")

# /Users/velsko/Dropbox (Personal)/MPI-SHH/evolution_fxn/final_github_evolution/Anthropoid_Calculus_Microbiome_Evolution


# Fix the column headers
# names(gene_families.humann2.uniref)[1]<-"Pathways"
gene_families.humann2.uniref <- gene_families.humann2.uniref %>%
  rename(GeneFamily = `# Gene Family`)

colnames(gene_families.humann2.uniref) <- gsub("_S.*_L.*_R1_.*.fastq.merged.prefixed.hg19unmapped_Abundance-CPM","", colnames(gene_families.humann2.uniref))
colnames(gene_families.humann2.uniref) <- gsub("_S.*_L.*_R1_.*.fastq.merged_Abundance-CPM","", colnames(gene_families.humann2.uniref))
colnames(gene_families.humann2.uniref) <- gsub("_S.*_L.*_R1_.*.fastq.combined.fq.prefixed.extractunmapped.bam_Abundance-CPM","", colnames(gene_families.humann2.uniref))

# remove problematic sample LIB017.A0101
gene_families.humann2.uniref <- gene_families.humann2.uniref %>%
  dplyr::select(-LIB017.A0101)

# Finish fixing the column headers
colnames(gene_families.humann2.uniref) <- gsub(".A0101.SG1.2","", colnames(gene_families.humann2.uniref))
colnames(gene_families.humann2.uniref) <- gsub(".A0101","", colnames(gene_families.humann2.uniref))
colnames(gene_families.humann2.uniref) <- gsub(".A0102","", colnames(gene_families.humann2.uniref))
colnames(gene_families.humann2.uniref) <- gsub(".B0101","", colnames(gene_families.humann2.uniref))

# save(gene_families.humann2.uniref, file = "../04-analysis/screening/humann2/tables/gene_families.humann2.uniref.RData") # this file is 167MB, so still way too big for github
# 
# # load the table as an RData object
# load("../04-analysis/screening/humann2/tables/gene_families.humann2.uniref.RData")

# Now remove the outliers and remaining extraction/library blank controls
# these are defined early on in the main script
# add a row with column totals to compare average number of mapped reads per host genus
gene_families.humann2.uniref_filtered <- gene_families.humann2.uniref %>%
  dplyr::select(-one_of(outliersH2_paths)) %>%
  dplyr::select(-one_of(sources_oral)) %>%
  dplyr::select(-starts_with("EXB")) %>%
  dplyr::select(-starts_with("LIB")) %>%
  adorn_totals(., where = "row", name = "Total_reads")

# select out only the UNMAPPED and total reads rows
read_assignments.humann2.uniref <- gene_families.humann2.uniref_filtered %>%
  dplyr::filter(str_detect(GeneFamily, "UNMAPPED|Total_reads")) %>%
  gather("individual_id","CPM", 2:ncol(.)) %>%
  spread(GeneFamily, CPM, 2:ncol(.)) %>%
  mutate(Total_assigned_reads = 1000000-UNMAPPED) %>%
  inner_join(., fullmetadata) %>%
  select(individual_id,Host_Genus,Host_Common,Host_General,Total_assigned_reads, UNMAPPED) %>%
  rename(SampleID = individual_id)

```

Plot the number of mapped and unmapped reads in copies per million (CPM)
```{r KO_assignment_plots}
# Now remove the outlier and control samples
read_assignments.humann2.uniref.outrem <- read_assignments.humann2.uniref %>%
  filter(!str_detect(SampleID, outliersH2_paths))

# Set the host genus as a factor for plotting
read_assignments.humann2.uniref$Host_Common <- factor(read_assignments.humann2.uniref$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))

read_assignments.humann2.uniref$Host_Genus <- factor(read_assignments.humann2.uniref$Host_Genus, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo",
                                          "Plaque", "Gut", "Skin", "Sediment","Control"))

read_assignments.humann2.uniref.outrem$Host_Common <- factor(read_assignments.humann2.uniref.outrem$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))

read_assignments.humann2.uniref.outrem$Host_Genus <- factor(read_assignments.humann2.uniref.outrem$Host_Genus, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo",
                                          "Plaque", "Gut", "Skin", "Sediment","Control"))

# graph the total number of assigned reads by host genus including outliers
read_assignments.humann2.uniref %>%
  ggplot(., aes(x=Host_Genus, y=Total_assigned_reads, fill = Host_Genus)) +
  geom_violin() +
  geom_point() +
  theme_classic() +
  scale_fill_manual(values = genus_colours) +
  theme(axis.text.x = element_text(hjust = 1)) +
  #scale_y_continuous(trans='log10') +
  ylab("Counts") +
  ggtitle("Total number of reads assigned by HUMAnN2")

# graph the total number of not assigned reads by host genus including outliers
read_assignments.humann2.uniref %>%
  ggplot(., aes(x=Host_Genus, y=UNMAPPED, fill = Host_Genus)) +
  geom_violin() +
  geom_point() +
  theme_classic() +
  scale_fill_manual(values = genus_colours) +
  theme(axis.text.x = element_text(hjust = 1)) +
  #scale_y_continuous(trans='log10') +
  ylab("Counts") +
  ggtitle("Total number of reads NOT assigned by HUMAnN2")

# graph the total number of assigned reads by host genus with outliers removed
h2_read_assigned_plot <- read_assignments.humann2.uniref.outrem %>%
  ggplot(., aes(x=Host_Genus, y=Total_assigned_reads, fill = Host_Genus)) +
  geom_violin() +
  geom_point() +
  theme_classic() +
  scale_fill_manual(values = genus_colours) +
  theme(axis.text.x = element_text(hjust = 1)) +
  #scale_y_continuous(trans='log10') +
  ylab("Counts") +
  ggtitle("Total number of reads assigned by HUMAnN2, outliers removed")

h2_read_assigned_plot

# ggsave("../h2_read_assigned_plot.pdf", plot = h2_read_assigned_plot, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)
# 

# graph the total number of not assigned reads by host genus with outliers removed
read_assignments.humann2.uniref.outrem %>%
  ggplot(., aes(x=Host_Genus, y=UNMAPPED, fill = Host_Genus)) +
  geom_violin() +
  geom_point() +
  theme_classic() +
  scale_fill_manual(values = genus_colours) +
  theme(axis.text.x = element_text(hjust = 1)) +
  #scale_y_continuous(trans='log10') +
  ylab("Counts") +
  ggtitle("Total number of reads NOT assigned by HUMAnN2, outliers removed")


# and graph the percent of reads assigned by each host genus, excluding outliers
percent_humann2_assigned_outrem <- read_assignments.humann2.uniref %>%
  mutate(Percent_assigned = (Total_assigned_reads/(Total_assigned_reads + UNMAPPED)*100)) %>%
  ggplot(., aes(x=Host_Common, y=Percent_assigned)) +
    geom_boxplot() +
  geom_point(size = 3, aes(colour = Host_Common, shape = Host_Common)) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ylab("Percent") +
  ggtitle("Percent of reads assigned to a UniRef90 category, outliers removed")

percent_humann2_assigned_outrem

# ggsave("../percent_humann2_assigned_outrem.pdf", plot = percent_humann2_assigned_outrem, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# and graph the percent of reads assigned by each host genus, excluding outliers
percent_humann2_assigned_outrem <- read_assignments.humann2.uniref %>%
  mutate(Percent_assigned = (Total_assigned_reads/(Total_assigned_reads + UNMAPPED)*100)) %>%
  ggplot(., aes(x=Host_Common, y=Percent_assigned)) +
    geom_boxplot() +
  geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
  scale_colour_manual(values = general_colors) +
  scale_shape_manual(values = general_shapes) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ylab("Percent") +
  ggtitle("Percent of reads assigned to a UniRef90 category, outliers removed")

percent_humann2_assigned_outrem

# ggsave("../percent_humann2_assigned_outrem.pdf", plot = percent_humann2_assigned_outrem, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```

And now finally let's make a table of summary statistics to compare the host genera
```{r KO_genus_stats}
# compute summary statistics by pathway for Homo
Plaque_h2_assignment_stats <- read_assignments.humann2.uniref.outrem %>%
  gather("mapping","CPM", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Plaque")) %>%
  group_by(mapping) %>%
  summarise(
    mean = mean(CPM),
    sd = sd(CPM),
    median = median(CPM),
    percent = (mean(CPM)/1000000*100),
    count = n(),
    prevalence = sum(CPM >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Plaque_h2_assignment_stats['genus'] <- 'Plaque'

# compute summary statistics by pathway for Homo
Homo_h2_assignment_stats <- read_assignments.humann2.uniref.outrem %>%
  gather("mapping","CPM", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Homo")) %>%
  group_by(mapping) %>%
  summarise(
    mean = mean(CPM),
    sd = sd(CPM),
    median = median(CPM),
    percent = (mean(CPM)/1000000*100),
    count = n(),
    prevalence = sum(CPM >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Homo_h2_assignment_stats['genus'] <- 'Homo'

# compute summary statistics by pathway for Homo
Pan_h2_assignment_stats <- read_assignments.humann2.uniref.outrem %>%
  gather("mapping","CPM", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Pan")) %>%
  group_by(mapping) %>%
  summarise(
    mean = mean(CPM),
    sd = sd(CPM),
    median = median(CPM),
    percent = (mean(CPM)/1000000*100),
    count = n(),
    prevalence = sum(CPM >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM)
  )  

# add a column that specificies the genus for merging all 4 genus tables together
Pan_h2_assignment_stats['genus'] <- 'Pan'

# compute summary statistics by pathway for Gorilla
Gorilla_h2_assignment_stats <- read_assignments.humann2.uniref.outrem %>%
  gather("mapping","CPM", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Gorilla")) %>%
  group_by(mapping) %>%
  summarise(
    mean = mean(CPM),
    sd = sd(CPM),
    median = median(CPM),
    percent = (mean(CPM)/1000000*100),
    count = n(),
    prevalence = sum(CPM >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM)
  )  

# add a column that specificies the genus for merging all 4 genus tables together
Gorilla_h2_assignment_stats['genus'] <- 'Gorilla'

# compute summary statistics by pathway for Alouatta
Alouatta_h2_assignment_stats <- read_assignments.humann2.uniref.outrem %>%
  gather("mapping","CPM", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Alouatta")) %>%
  group_by(mapping) %>%
  summarise(
    mean = mean(CPM),
    sd = sd(CPM),
    median = median(CPM),
    percent = (mean(CPM)/1000000*100),
    count = n(),
    prevalence = sum(CPM >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(CPM)
  )  

# add a column that specificies the genus for merging all 4 genus tables together
Alouatta_h2_assignment_stats['genus'] <- 'Alouatta'


# Combine the stats columns and use this to look at the differenes between 
#host genera for pathways of interest
allh2_assignment_stats <- bind_rows(Plaque_h2_assignment_stats, Homo_h2_assignment_stats, Pan_h2_assignment_stats, Gorilla_h2_assignment_stats, Alouatta_h2_assignment_stats)

```

##KEGG ortholog investigation
Now we want to know the % of UniRef90 assignments that were grouped into KOs, we
want to know if the KOs separate the host genera, and if there are more
assignments to KOs in the Carbohydrates pathways in *Homo* than in *Alouatta*, *Pan*,
and *Gorilla*.
```{r pct_assignments}
# What percent of UniRef90 assignments were converted to KOs?
gene_families.humann2_KOs <- gene_families.humann2 %>%
  filter(!str_detect(Pathways, "\\|")) %>%
  adorn_totals(., where = "row", name = "Total") %>%
  gather("individual_id","CPM", 2:ncol(.)) %>%
  spread(Pathways, CPM) %>%
  mutate(Total_grouped = Total-(UNMAPPED+UNGROUPED)) %>%
  mutate(Proportion_grouped = (UNMAPPED+UNGROUPED)/Total) %>%
  inner_join(., selectmetadata, by = "individual_id") 

gene_families.humann2_KOs$Host_Common <- factor(gene_families.humann2_KOs$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))

gene_families.humann2_KOs$Host_Genus <- factor(gene_families.humann2_KOs$Host_Genus, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo",
                                          "Plaque", "Gut", "Skin", "Sediment","Control"))

gene_families.humann2_KOs_plot <- ggplot(gene_families.humann2_KOs %>% filter(!str_detect(Host_Genus, "Control|Gut|Sediment|Skin")), aes(x=Host_Common, y=Proportion_grouped)) +
    geom_boxplot() +
    geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
    scale_colour_manual(values = general_colors) +
    scale_shape_manual(values = general_shapes) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    ylab("Proportion of assignments grouped into KOs") +
    ggtitle("Proportion of HUMAnN2 assignments grouped into KOs")

gene_families.humann2_KOs_plot

outliersH2_KOs <- c("TAF016","BAN001","ECO006","GUE001","SpyNew",
                    "BSQ001","VLD001","CDC007","SPM001","SpyOld",
                    "MTM012","MTM003","MTS001","MTM013","DLV002")

outliersH2_KOs_collapse <- str_c(outliersH2_KOs, collapse = "|")

gene_families.humann2_KOs_plot <- ggplot(gene_families.humann2_KOs %>% filter(!str_detect(Host_Genus, "Control|Gut|Sediment|Skin")) %>% filter(!str_detect(individual_id, outliersH2_KOs_collapse)), aes(x=Host_Common, y=Proportion_grouped)) +
    geom_boxplot() +
    geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
    scale_colour_manual(values = general_colors) +
    scale_shape_manual(values = general_shapes) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    ylab("Proportion of assignments grouped into KOs") +
    ggtitle("Proportion of HUMAnN2 assignments grouped into KOs")

gene_families.humann2_KOs_plot

ggsave("../gene_families.humann2_KOs.pdf", plot = gene_families.humann2_KOs_plot, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)

```

Now the PCA on KOs
```{r KO_pca}
# includes outliers and controls
gene_families.humann2_mat <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         gather("sample", "CPM", 2:ncol(.)) %>%
                                         spread(Pathways, CPM)


rownames(gene_families.humann2_mat) <- gene_families.humann2_mat$sample

gene_families.humann2_mat <- as.matrix(gene_families.humann2_mat %>% select(-sample))

# add +1 to all values so the CLR can be performed
gene_families.humann2_mat <- gene_families.humann2_mat + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.H2KOpca <- tune.pca(gene_families.humann2_mat, logratio = 'CLR')
tune.H2KOpca

# perform a PCA to see how the data cluster
H2KO.pca <- mixOmics::pca(gene_families.humann2_mat, ncomp = 3, logratio = 'CLR')
plot(H2KO.pca)

# Select out the PC variates into a new table and add the metadata for plotting
H2KO.pca.variates <- as.data.frame(H2KO.pca$variates$X)
H2KO.pca.variates <- H2KO.pca.variates %>%
  rownames_to_column("individual_id")
  
H2KO.pca.varmet <- inner_join(H2KO.pca.variates, fullmetadata, by = "individual_id")

H2KO.pca.varmet$Host_Common <- factor(H2KO.pca.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_H2KO_pc1pc2 <- ggplot(H2KO.pca.varmet, aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(calchumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) 

pca_H2KO_pc1pc2

ggsave("../pca_H2KO_pc1pc2.pdf", plot = pca_H2KO_pc1pc2, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)

pca_H2KO_pc1pc2_label <- ggplot(H2KO.pca.varmet, aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_text(aes(label = H2KO.pca.varmet$individual_id), size = 2) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(calchumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) 

pca_H2KO_pc1pc2_label


# PC2-PC3
pca_H2KO_pc3pc2 <- ggplot(H2KO.pca.varmet, aes(PC3, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC3 - ", round(calchumann2.pca$explained_variance[3], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) 

pca_H2KO_pc3pc2

pca_H2KO_pc3pc2_label <- ggplot(H2KO.pca.varmet, aes(PC3, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_text(aes(label = H2KO.pca.varmet$individual_id), size = 2) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC3 - ", round(calchumann2.pca$explained_variance[3], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) 

pca_H2KO_pc3pc2_label

```

And repeat without outliers
```{r KO_pca_outrem}

outliersH2_KOs <- c("TAF016","BAN001","ECO006","GUE001","SpyNew",
                    "BSQ001","VLD001","CDC007","SPM001","SpyOld",
                    "MTM012","MTM003","MTS001","MTM013","DLV002")
  

# excludes outliers and includes controls
gene_families.humann2_mat <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         gather("sample", "CPM", 2:ncol(.)) %>%
                                         spread(Pathways, CPM)


rownames(gene_families.humann2_mat) <- gene_families.humann2_mat$sample

gene_families.humann2_mat <- as.matrix(gene_families.humann2_mat %>% select(-sample))

# add +1 to all values so the CLR can be performed
gene_families.humann2_mat <- gene_families.humann2_mat + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.H2KOpca <- tune.pca(gene_families.humann2_mat, logratio = 'CLR')
tune.H2KOpca

# perform a PCA to see how the data cluster
H2KO.pca <- mixOmics::pca(gene_families.humann2_mat, ncomp = 3, logratio = 'CLR')
plot(H2KO.pca)

# Select out the PC variates into a new table and add the metadata for plotting
H2KO.pca.variates <- as.data.frame(H2KO.pca$variates$X)
H2KO.pca.variates <- H2KO.pca.variates %>%
  rownames_to_column("individual_id")
  
H2KO.pca.varmet <- inner_join(H2KO.pca.variates, fullmetadata, by = "individual_id")

H2KO.pca.varmet$Host_Common <- factor(H2KO.pca.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_H2KO_pc1pc2_outrem <- ggplot(H2KO.pca.varmet, aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(calchumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) 

pca_H2KO_pc1pc2_outrem

# ggsave("../pca_H2KO_pc1pc2_outrem.pdf", plot = pca_H2KO_pc1pc2_outrem, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

ggplot(H2KO.pca.varmet, aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_text(aes(label = H2KO.pca.varmet$individual_id), size = 2) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(calchumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) 


```


And repeat without outliers and controls
```{r KO_pca_outrem_outcont}
# excludes outliers and controls
gene_families.humann2_mat <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_oral)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("sample", "CPM", 2:ncol(.)) %>%
                                         spread(Pathways, CPM)

rownames(gene_families.humann2_mat) <- gene_families.humann2_mat$sample

gene_families.humann2_mat <- as.matrix(gene_families.humann2_mat %>% select(-sample))

# add +1 to all values so the CLR can be performed
gene_families.humann2_mat <- gene_families.humann2_mat + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.H2KOpca <- tune.pca(gene_families.humann2_mat, logratio = 'CLR')
tune.H2KOpca

# perform a PCA to see how the data cluster
H2KO.pca <- mixOmics::pca(gene_families.humann2_mat, ncomp = 3, logratio = 'CLR')
plot(H2KO.pca)

# Select out the PC variates into a new table and add the metadata for plotting
H2KO.pca.variates <- as.data.frame(H2KO.pca$variates$X)
H2KO.pca.variates <- H2KO.pca.variates %>%
  rownames_to_column("individual_id")
  
H2KO.pca.varmet <- inner_join(H2KO.pca.variates, fullmetadata, by = "individual_id")

H2KO.pca.varmet$Host_Common <- factor(H2KO.pca.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_H2KO_pc1pc2_outrem_all <- ggplot(H2KO.pca.varmet, aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(calchumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) 

pca_H2KO_pc1pc2_outrem_all

# ggsave("../pca_H2KO_pc1pc2_outrem_all.pdf", plot = pca_H2KO_pc1pc2_outrem_all, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```


And repeat without outliers, controls, and plaque
```{r KO_pcas_outrem_outcont_outpq}
# excludes outliers and controls and plaque
gene_families.humann2_mat <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.01) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(gene_families.humann2_mat) <- gene_families.humann2_mat$individual_id
gene_families.humann2_mat <- gene_families.humann2_mat %>% select(-individual_id)
gene_families.humann2_mat <- gene_families.humann2_mat + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.H2KOpca <- tune.pca(gene_families.humann2_mat, logratio = 'CLR')
tune.H2KOpca

# perform a PCA to see how the data cluster
H2KO.pca <- mixOmics::pca(gene_families.humann2_mat, ncomp = 2, logratio = 'CLR')
plot(H2KO.pca)

# Select out the PC variates into a new table and add the metadata for plotting
H2KO.pca.variates <- as.data.frame(H2KO.pca$variates$X)
H2KO.pca.variates <- H2KO.pca.variates %>%
  rownames_to_column("individual_id")
  
H2KO.pca.varmet <- inner_join(H2KO.pca.variates, fullmetadata, by = "individual_id")

pca_H2KO_pc1pc2_outrem_all_nopq <- ggplot(H2KO.pca.varmet, aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(calchumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) 

pca_H2KO_pc1pc2_outrem_all_nopq

# ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```


We need to do this again without modern human samples to make sure they aren't
responsible for biasing the differenes between *Homo* and *Pan/Gorilla/Alouatta*.

```{r KO_pcas_outrem_outcont_outpq_nomod}
# excludes outliers and controls and plaque
gene_families.humann2_mat_nomod <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         select(-starts_with("JAE")) %>%
                                         select(-starts_with("VLC")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.01) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(gene_families.humann2_mat_nomod) <- gene_families.humann2_mat_nomod$individual_id
gene_families.humann2_mat_nomod <- gene_families.humann2_mat_nomod %>% select(-individual_id)
gene_families.humann2_mat_nomod <- gene_families.humann2_mat_nomod + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.H2KO_nomod_pca <- tune.pca(gene_families.humann2_mat_nomod, logratio = 'CLR')
tune.H2KO_nomod_pca

# perform a PCA to see how the data cluster
H2KO_nomod.pca <- mixOmics::pca(gene_families.humann2_mat_nomod, ncomp = 4, logratio = 'CLR')
plot(H2KO_nomod.pca)

# Select out the PC variates into a new table and add the metadata for plotting
H2KO_nomod.pca.variates <- as.data.frame(H2KO_nomod.pca$variates$X)
H2KO_nomod.pca.variates <- H2KO_nomod.pca.variates %>%
  rownames_to_column("individual_id")
  
H2KO_nomod.pca.varmet <- inner_join(H2KO_nomod.pca.variates, fullmetadata, by = "individual_id")

pca_H2KO_pc1pc2_outrem_all_nopq_nomod <- ggplot(H2KO_nomod.pca.varmet, aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(calchumann2.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(calchumann2.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) +
  scale_y_reverse()

pca_H2KO_pc1pc2_outrem_all_nopq_nomod

# ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_nomod.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_nomod, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```


Now make bi-plots to see which orthologs contribute the most to sample
separation.
```{r KO_pca_biplots}
# and a bi-plot on the calculus-only PCA to see if the KOs that drive separation are related to those that drive separation in SEED (from AADDER)

# Borrowed from Alex's coprolite CAZymes bi-plot script
# Extract coordinates from mixOmics object and convert into long table dataframe
H2KO.pca_df <- as.data.frame(H2KO.pca$x) %>%
  rownames_to_column("individual_id") %>%
  # Add meta data
  inner_join(fullmetadata, by="individual_id") %>%
  select(individual_id,Host_Genus,Host_Common,PC1,PC2)

# Extract explained variance
pca_expl_vars <- paste0(round(H2KO.pca$explained_variance * 100, 2), "%")

# Extract loadings (eigenvectors) from PC analysis
pca_loadings_df <- as.data.frame(H2KO.pca$loadings$X) %>%
  rownames_to_column(var = "KOrtholog")

# Identify the 10 pathways that have highest positive and negative loadings in PC1 and PC2
pc1_pws <- pca_loadings_df %>%
  top_n(10, PC1)

pc1_neg <- pca_loadings_df %>%
  top_n(-10, PC1) 

pc2_pws <- pca_loadings_df %>%
  top_n(10, PC2)

pc2_neg <- pca_loadings_df %>%
  top_n(-10, PC2) 


# Calculate correction factor for coordinate systems of principal coordinates
# so that they have the same units as the rotations of PC1 and PC2; from
# ggfortify::autoplot
corr_lam <- H2KO.pca$sdev[c("PC1", "PC2")] * sqrt(nrow(H2KO.pca_df))

# make Host Common a factor for plotting
H2KO.pca_df$Host_Common <- factor(H2KO.pca_df$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Neanderthal)", "Homo (Modern Human)", "Plaque"))


# Plot the top 10 negative and positive paths that are strongest loading in PC1 
pca_H2KO_pc1pc2_outrem_all_nopq_pc1bi <- H2KO.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc1_pws,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pc1_neg,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 1, 2, 6)) +
  theme_minimal() + theme(text = element_text(size = 14))

pca_H2KO_pc1pc2_outrem_all_nopq_pc1bi

# ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_pc1bi.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_pc1bi, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

pca_H2KO_pc1pc2_outrem_all_nopq_pc1bi_label <- H2KO.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc1_pws,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc1_pws,
                   aes(x = PC1, y = PC2, label = KOrtholog),
                   size = 1.5, colour = "grey20") +
    geom_segment(data = pc1_neg,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc1_neg,
                   aes(x = PC1, y = PC2, label = KOrtholog),
                   size = 1.5, colour = "grey20", force = 4) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 1, 2, 6)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(text = element_text(size = 14))

pca_H2KO_pc1pc2_outrem_all_nopq_pc1bi_label

# ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_pc1bi_label.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_pc1bi_label, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)


# Plot the top 10 negative and positive paths that are strongest loading in PC2
pca_H2KO_pc1pc2_outrem_all_nopq_pc2bi <- H2KO.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc2_pws,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pc2_neg,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 1, 2, 6)) +
  theme_minimal() + theme(text = element_text(size = 14))

pca_H2KO_pc1pc2_outrem_all_nopq_pc2bi

# ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_pc2bi.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_pc2bi, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

pca_H2KO_pc1pc2_outrem_all_nopq_pc2bi_label <- H2KO.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc2_pws,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc2_pws,
                   aes(x = PC1, y = PC2, label = KOrtholog),
                   size = 1.5, colour = "grey20") +
    geom_segment(data = pc2_neg,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc2_neg,
                   aes(x = PC1, y = PC2, label = KOrtholog),
                   size = 1.5, colour = "grey20", force = 4) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 1, 2, 6)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(text = element_text(size = 14))

pca_H2KO_pc1pc2_outrem_all_nopq_pc2bi_label

# ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_pc2bi_label.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_pc2bi_label, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```


Now make bi-plots with no modern humans, to see if the trends we observed above
are characteristic of all *Homo* or if they're drived by modern samples.
```{r KO_pca_biplots_nomod}
# and a bi-plot on the calculus-only PCA to see if the KOs that drive separation are related to those that drive separation in SEED (from AADDER)

# Borrowed from Alex's coprolite CAZymes bi-plot script
# Extract coordinates from mixOmics object and convert into long table dataframe
H2KO_nomod.pca_df <- as.data.frame(H2KO_nomod.pca$x) %>%
  rownames_to_column("individual_id") %>%
  # Add meta data
  inner_join(fullmetadata, by="individual_id") %>%
  select(individual_id,Host_Genus,Host_Common,PC1,PC2)

# Extract explained variance
pca_expl_vars <- paste0(round(H2KO_nomod.pca$explained_variance * 100, 2), "%")

# Extract loadings (eigenvectors) from PC analysis
pca_loadings_df <- as.data.frame(H2KO_nomod.pca$loadings$X) %>%
  rownames_to_column(var = "KOrtholog")

# Identify the 10 pathways that have highest positive and negative loadings in PC1 and PC2
pc1_pws_nomod <- pca_loadings_df %>%
  top_n(10, PC1)

pc1_neg_nomod <- pca_loadings_df %>%
  top_n(-10, PC1) 

pc2_pws_nomod <- pca_loadings_df %>%
  top_n(10, PC2)

pc2_neg_nomod <- pca_loadings_df %>%
  top_n(-10, PC2) 


# Calculate correction factor for coordinate systems of principal coordinates
# so that they have the same units as the rotations of PC1 and PC2; from
# ggfortify::autoplot
corr_lam <- H2KO_nomod.pca$sdev[c("PC1", "PC2")] * sqrt(nrow(H2KO_nomod.pca_df))

# make Host Genus a factor for plotting
H2KO.pca_df$Host_Common <- factor(H2KO.pca_df$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Neanderthal)", "Homo (Modern Human)", "Plaque"))


# Plot the top 10 negative and positive paths that are strongest loading in PC1 
pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc1bi <- H2KO_nomod.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc1_pws_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pc1_neg_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 6, 2, 1)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  scale_y_reverse()

pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc1bi

ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc1bi.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc1bi, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)

pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc1bi_label <- H2KO_nomod.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc1_pws_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc1_pws_nomod,
                   aes(x = PC1, y = PC2, label = KOrtholog),
                   size = 1.5, colour = "grey20") +
    geom_segment(data = pc1_neg_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc1_neg_nomod,
                   aes(x = PC1, y = PC2, label = KOrtholog),
                   size = 1.5, colour = "grey20", force = 4) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 6, 2, 1)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(text = element_text(size = 14)) +
  scale_y_reverse()

pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc1bi_label

ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc1bi_label.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc1bi_label, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)


# Plot the top 10 negative and positive paths that are strongest loading in PC2
pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc2bi <- H2KO_nomod.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc2_pws_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pc2_neg_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 6, 2, 1)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  scale_y_reverse()

pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc2bi

ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc2bi.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc2bi, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)

pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc2bi_label <- H2KO_nomod.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc2_pws_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc2_pws_nomod,
                   aes(x = PC1, y = PC2, label = KOrtholog),
                   size = 1.5, colour = "grey20") +
    geom_segment(data = pc2_neg_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc2_neg_nomod,
                   aes(x = PC1, y = PC2, label = KOrtholog),
                   size = 1.5, colour = "grey20", force = 4) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 6, 2, 1)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(text = element_text(size = 14)) +
  scale_y_reverse()

pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc2bi_label

ggsave("../pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc2bi_label.pdf", plot = pca_H2KO_pc1pc2_outrem_all_nopq_nomod_pc2bi_label, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)

```


We want to see if there are specific species contributing the KOs that are the top 10 positive and negative in PC1 and PC2. Even though there is no bias for Carbohydrate categories in Homo like we see in SEED, is there still a bias of Strep contributing the KOs that separate Homo?
```{r KO_species_contribution, eval=F}
top10_KOs <- bind_rows(pc1_pws,pc1_neg)
top10_KOs <- bind_rows(top10_KOs,pc2_pws)
top10_KOs <- bind_rows(top10_KOs,pc2_neg)
top10_KOs <- distinct(top10_KOs %>% select(KOrtholog))
top10_KOs <- str_c(top10_KOs$KOrtholog, collapse = "|")

gene_families.humann2$Pathways <- gsub('\\.', '\\|', gene_families.humann2$Pathways)
gene_families.humann2_top10 <- gene_families.humann2 %>%
  rename(Taxonomy = Pathways) %>%
  select(-one_of(outliersH2_paths)) %>%
  select(-one_of(sources_oral)) %>%
  select(-starts_with("EXB")) %>%
  select(-starts_with("LIB")) %>%
  filter(str_detect(Taxonomy, top10_KOs)) %>%
  adorn_totals(where = "col", name = "Species_total") %>%
  filter(Species_total != 0) %>%
  # gather("individual_id","CPM",2:ncol(.)) %>%
  # spread(Taxonomy,CPM) %>%
  # filter(Species_total == 0)
  gather("individual_id","CPM",2:ncol(.)) %>%
  mutate(KOrtholog = sapply(Taxonomy, function(f) {
                                  unlist(str_split(f, "\\|"))[1]
                           })) %>%
   mutate(Genus = sapply(Taxonomy, function(f) {
                                  unlist(str_split(f, "\\|"))[2]
                           })) %>%
  mutate(Species = sapply(Taxonomy, function(f) {
                                  unlist(str_split(f, "\\|"))[3]
                           })) %>%
  inner_join(., fullmetadata, by = "individual_id") %>%
  select(individual_id,Host_Common,Host_Genus,KOrtholog,Genus,Species,CPM)



pc1_pws_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc1_pws[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_stats_homo['KOrtholog'] <- pc1_pws[1,1]
pc1_pws_stats_homo['Host_Genus'] <- 'Homo'

pc1_pws_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc1_pws[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_stats_pan['KOrtholog'] <- pc1_pws[1,1]
pc1_pws_stats_pan['Host_Genus'] <- 'Pan'

pc1_pws_stats <- bind_rows(pc1_pws_stats_homo,pc1_pws_stats_pan)

pc1_pws_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc1_pws[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_stats_gorilla['KOrtholog'] <- pc1_pws[1,1]
pc1_pws_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc1_pws_stats <- bind_rows(pc1_pws_stats,pc1_pws_stats_gorilla)

pc1_pws_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc1_pws[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_stats_alouatta['KOrtholog'] <- pc1_pws[1,1]
pc1_pws_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc1_pws_stats <- bind_rows(pc1_pws_stats,pc1_pws_stats_alouatta)

pc1_neg_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc1_neg[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_stats_homo['KOrtholog'] <- pc1_neg[1,1]
pc1_neg_stats_homo['Host_Genus'] <- 'Homo'

pc1_neg_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc1_neg[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_stats_pan['KOrtholog'] <- pc1_neg[1,1]
pc1_neg_stats_pan['Host_Genus'] <- 'Pan'

pc1_neg_stats <- bind_rows(pc1_neg_stats_homo,pc1_neg_stats_pan)

pc1_neg_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc1_neg[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_stats_gorilla['KOrtholog'] <- pc1_neg[1,1]
pc1_neg_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc1_neg_stats <- bind_rows(pc1_neg_stats,pc1_neg_stats_gorilla)

pc1_neg_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc1_neg[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_stats_alouatta['KOrtholog'] <- pc1_neg[1,1]
pc1_neg_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc1_neg_stats <- bind_rows(pc1_neg_stats,pc1_neg_stats_alouatta)


pc2_pws_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc2_pws[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_stats_homo['KOrtholog'] <- pc2_pws[1,1]
pc2_pws_stats_homo['Host_Genus'] <- 'Homo'

pc2_pws_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc2_pws[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_stats_pan['KOrtholog'] <- pc2_pws[1,1]
pc2_pws_stats_pan['Host_Genus'] <- 'Pan'

pc2_pws_stats <- bind_rows(pc2_pws_stats_homo,pc2_pws_stats_pan)

pc2_pws_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc2_pws[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_stats_gorilla['KOrtholog'] <- pc2_pws[1,1]
pc2_pws_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc2_pws_stats <- bind_rows(pc2_pws_stats,pc2_pws_stats_gorilla)

pc2_pws_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc2_pws[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_stats_alouatta['KOrtholog'] <- pc2_pws[1,1]
pc2_pws_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc2_pws_stats <- bind_rows(pc2_pws_stats,pc2_pws_stats_alouatta)



pc2_neg_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc2_neg[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_stats_homo['KOrtholog'] <- pc2_neg[1,1]
pc2_neg_stats_homo['Host_Genus'] <- 'Homo'

pc2_neg_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc2_neg[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_stats_pan['KOrtholog'] <- pc2_neg[1,1]
pc2_neg_stats_pan['Host_Genus'] <- 'Pan'

pc2_neg_stats <- bind_rows(pc2_neg_stats_homo,pc2_neg_stats_pan)

pc2_neg_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc2_neg[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_stats_gorilla['KOrtholog'] <- pc2_neg[1,1]
pc2_neg_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc2_neg_stats <- bind_rows(pc2_neg_stats,pc2_neg_stats_gorilla)

pc2_neg_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc2_neg[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_stats_alouatta['KOrtholog'] <- pc2_neg[1,1]
pc2_neg_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc2_neg_stats <- bind_rows(pc2_neg_stats,pc2_neg_stats_alouatta)

# now add orthologs 2-10 to each table by changing the row number in the sections below
pc1_pws_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc1_pws[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_stats_homo['KOrtholog'] <- pc1_pws[10,1]
pc1_pws_stats_homo['Host_Genus'] <- 'Homo'

pc1_pws_stats <- bind_rows(pc1_pws_stats, pc1_pws_stats_homo)

pc1_pws_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc1_pws[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_stats_pan['KOrtholog'] <- pc1_pws[10,1]
pc1_pws_stats_pan['Host_Genus'] <- 'Pan'

pc1_pws_stats <- bind_rows(pc1_pws_stats,pc1_pws_stats_pan)

pc1_pws_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc1_pws[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_stats_gorilla['KOrtholog'] <- pc1_pws[10,1]
pc1_pws_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc1_pws_stats <- bind_rows(pc1_pws_stats,pc1_pws_stats_gorilla)

pc1_pws_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc1_pws[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_stats_alouatta['KOrtholog'] <- pc1_pws[10,1]
pc1_pws_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc1_pws_stats <- bind_rows(pc1_pws_stats,pc1_pws_stats_alouatta)



pc1_neg_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc1_neg[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_stats_homo['KOrtholog'] <- pc1_neg[10,1]
pc1_neg_stats_homo['Host_Genus'] <- 'Homo'

pc1_neg_stats <- bind_rows(pc1_neg_stats, pc1_neg_stats_homo)

pc1_neg_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc1_neg[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_stats_pan['KOrtholog'] <- pc1_neg[10,1]
pc1_neg_stats_pan['Host_Genus'] <- 'Pan'

pc1_neg_stats <- bind_rows(pc1_neg_stats,pc1_neg_stats_pan)

pc1_neg_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc1_neg[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_stats_gorilla['KOrtholog'] <- pc1_neg[10,1]
pc1_neg_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc1_neg_stats <- bind_rows(pc1_neg_stats,pc1_neg_stats_gorilla)

pc1_neg_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc1_neg[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_stats_alouatta['KOrtholog'] <- pc1_neg[10,1]
pc1_neg_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc1_neg_stats <- bind_rows(pc1_neg_stats,pc1_neg_stats_alouatta)


pc2_pws_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc2_pws[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM))  %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_stats_homo['KOrtholog'] <- pc2_pws[10,1]
pc2_pws_stats_homo['Host_Genus'] <- 'Homo'

pc2_pws_stats <- bind_rows(pc2_pws_stats, pc2_pws_stats_homo)

pc2_pws_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc2_pws[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_stats_pan['KOrtholog'] <- pc2_pws[10,1]
pc2_pws_stats_pan['Host_Genus'] <- 'Pan'

pc2_pws_stats <- bind_rows(pc2_pws_stats,pc2_pws_stats_pan)

pc2_pws_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc2_pws[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_stats_gorilla['KOrtholog'] <- pc2_pws[10,1]
pc2_pws_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc2_pws_stats <- bind_rows(pc2_pws_stats,pc2_pws_stats_gorilla)

pc2_pws_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc2_pws[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_stats_alouatta['KOrtholog'] <- pc2_pws[10,1]
pc2_pws_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc2_pws_stats <- bind_rows(pc2_pws_stats,pc2_pws_stats_alouatta)


pc2_neg_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc2_neg[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_stats_homo['KOrtholog'] <- pc2_neg[10,1]
pc2_neg_stats_homo['Host_Genus'] <- 'Homo'

pc2_neg_stats <- bind_rows(pc2_neg_stats, pc2_neg_stats_homo)

pc2_neg_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc2_neg[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_stats_pan['KOrtholog'] <- pc2_neg[10,1]
pc2_neg_stats_pan['Host_Genus'] <- 'Pan'

pc2_neg_stats <- bind_rows(pc2_neg_stats,pc2_neg_stats_pan)

pc2_neg_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc2_neg[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_stats_gorilla['KOrtholog'] <- pc2_neg[10,1]
pc2_neg_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc2_neg_stats <- bind_rows(pc2_neg_stats,pc2_neg_stats_gorilla)

pc2_neg_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc2_neg[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_stats_alouatta['KOrtholog'] <- pc2_neg[10,1]
pc2_neg_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc2_neg_stats <- bind_rows(pc2_neg_stats,pc2_neg_stats_alouatta)

write.table(pc1_pws_stats, "../04-analysis/screening/humann2/tables/pc1_pws_stats.tsv", quote=F, sep="\t", row.names = F)
write.table(pc1_neg_stats, "../04-analysis/screening/humann2/tables/pc1_neg_stats.tsv", quote=F, sep="\t", row.names = F)
write.table(pc2_pws_stats, "../04-analysis/screening/humann2/tables/pc2_pws_stats.tsv", quote=F, sep="\t", row.names = F)
write.table(pc2_neg_stats, "../04-analysis/screening/humann2/tables/pc2_neg_stats.tsv", quote=F, sep="\t", row.names = F)

```

Read in the files made above so you don't have to make them again, 
and plot the species distributions of KOrthologs in each host genus.
```{r KO_species_contribution_graphs}
# Read in the files created and written out in the last code chunk
pc1_pws_stats <- fread("../04-analysis/screening/humann2/tables/pc1_pws_stats.tsv")
pc1_neg_stats <- fread("../04-analysis/screening/humann2/tables/pc1_neg_stats.tsv")
pc2_pws_stats <- fread("../04-analysis/screening/humann2/tables/pc2_pws_stats.tsv")
pc2_neg_stats <- fread("../04-analysis/screening/humann2/tables/pc2_neg_stats.tsv")


pc1_pws_stats <- pc1_pws_stats %>% select(KOrtholog,Host_Genus,Genus,Sum,Percent)
pc1_pws_stats$Host_Genus <- factor(pc1_pws_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc1_neg_stats <- pc1_neg_stats %>% select(KOrtholog,Host_Genus,Genus,Sum,Percent)
pc1_neg_stats$Host_Genus <- factor(pc1_neg_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc2_pws_stats <- pc2_pws_stats %>% select(KOrtholog,Host_Genus,Genus,Sum,Percent)
pc2_pws_stats$Host_Genus <- factor(pc2_pws_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc2_neg_stats <- pc2_neg_stats %>% select(KOrtholog,Host_Genus,Genus,Sum,Percent)
pc2_neg_stats$Host_Genus <- factor(pc2_neg_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))

# remove all species with <12% contribution and fill in that % with Other (g__Other)
pc1_top10 <- pc1_pws$KOrtholog

pc1_pws_stats_extra <- lapply(pc1_top10, function(eclass) {
 fisher_subsistence <- pc1_pws_stats %>%
   group_by(Host_Genus) %>%
   filter(KOrtholog == eclass) %>%
   filter(Percent <=12) %>%
#   filter(Percent >12) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc1_pws_stats_extra['KOrtholog'] <- c("K00526","K00526","K00526","K00526",
                                      "K00658","K00658","K00658","K00658",
                                      "K00674","K00674","K00674","K00674",
                                      "K01092","K01092","K01092","K01092",
                                      "K01657","K01657","K01657","K01657",
                                      "K01696","K01696","K01696","K01696",
                                      "K01902","K01902","K01902","K01902",
                                      "K01903","K01903","K01903","K01903",
                                      "K01961","K01961","K01961","K01961",
                                      "K07173","K07173","K07173","K07173")
pc1_pws_stats_extra['Sum'] <- NA
pc1_pws_stats_extra['Genus'] <- 'g__Other'
pc1_pws_stats_extra <- pc1_pws_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(KOrtholog,Host_Genus,Genus,Sum,Percent)

pc1_pws_stats_extra <- pc1_pws_stats_extra %>%
  bind_rows(., pc1_pws_stats %>%
              filter(Percent >12))

pc1_top10 <- pc1_neg$KOrtholog

pc1_neg_stats_extra <- lapply(pc1_top10, function(eclass) {
 fisher_subsistence <- pc1_neg_stats %>%
   group_by(Host_Genus) %>%
   filter(KOrtholog == eclass) %>%
   filter(Percent <=12) %>%
#   filter(Percent >12) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc1_neg_stats_extra['KOrtholog'] <- c("K01999","K01999","K01999","K01999",
                                      "K02005","K02005","K02005","K02005",
                                      "K02034","K02034","K02034","K02034",
                                      "K02117","K02117","K02117","K02117",
                                      "K03205","K03205","K03205","K03205",
                                      "K06147","K06147","K06147","K06147",
                                      "K07052","K07052","K07052","K07052",
                                      "K07133","K07133","K07133","K07133",
                                      "K16785","K16785","K16785","K16785",
                                      "K16926","K16926","K16926","K16926")
pc1_neg_stats_extra['Sum'] <- NA
pc1_neg_stats_extra['Genus'] <- 'g__Other'
pc1_neg_stats_extra <- pc1_neg_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(KOrtholog,Host_Genus,Genus,Sum,Percent)

pc1_neg_stats_extra <- pc1_neg_stats_extra %>%
  bind_rows(., pc1_neg_stats %>%
              filter(Percent >12))

pc2_top10 <- pc2_pws$KOrtholog

pc2_pws_stats_extra <- lapply(pc2_top10, function(eclass) {
 fisher_subsistence <- pc2_pws_stats %>%
   group_by(Host_Genus) %>%
   filter(KOrtholog == eclass) %>%
   filter(Percent <=12) %>%
#   filter(Percent >12) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc2_pws_stats_extra['KOrtholog'] <- c("K00027","K00027","K00027","K00027",
                                      "K00794","K00794","K00794","K00794",
                                      "K01507","K01507","K01507","K01507",
                                      "K01610","K01610","K01610","K01610",
                                      "K01759","K01759","K01759","K01759",
                                      "K02012","K02012","K02012","K02012",
                                      "K03043","K03043","K03043","K03043",
                                      "K03169","K03169","K03169","K03169",
                                      "K03205","K03205","K03205","K03205",
                                      "K03569","K03569","K03569","K03569")
pc2_pws_stats_extra['Sum'] <- NA
pc2_pws_stats_extra['Genus'] <- 'g__Other'
pc2_pws_stats_extra <- pc2_pws_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(KOrtholog,Host_Genus,Genus,Sum,Percent)

pc2_pws_stats_extra <- pc2_pws_stats_extra %>%
  bind_rows(., pc2_pws_stats %>%
              filter(Percent >12))

pc2_top10 <- pc2_neg$KOrtholog

pc2_neg_stats_extra <- lapply(pc2_top10, function(eclass) {
 fisher_subsistence <- pc2_neg_stats %>%
   group_by(Host_Genus) %>%
   filter(KOrtholog == eclass) %>%
   filter(Percent <=12) %>%
#   filter(Percent >12) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc2_neg_stats_extra['KOrtholog'] <- c("K01104","K01104","K01104","K01104",
                                      "K01589","K01589","K01589","K01589",
                                      "K01990","K01990","K01990","K01990",
                                      "K02027","K02027","K02027","K02027",
                                      "K02440","K02440","K02440","K02440",
                                      "K03402","K03402","K03402","K03402",
                                      "K03574","K03574","K03574","K03574",
                                      "K07052","K07052","K07052","K07052",
                                      "K07729","K07729","K07729","K07729",
                                      "K15634","K15634","K15634","K15634")
pc2_neg_stats_extra['Sum'] <- NA
pc2_neg_stats_extra['Genus'] <- 'g__Other'
pc2_neg_stats_extra <- pc2_neg_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(KOrtholog,Host_Genus,Genus,Sum,Percent)

pc2_neg_stats_extra <- pc2_neg_stats_extra %>%
  bind_rows(., pc2_neg_stats %>%
              filter(Percent >12))

# set the species as factor for ordering the colors with g__Other at the top and unclassified at the bottom
# these orders were set after looking at the graphs with no ordering specified
pc1_pws_stats_extra$Genus <- factor(pc1_pws_stats_extra$Genus, levels = c("g__Other",
"g__Aggregatibacter",
"g__Bacillus",
"g__Cryptobacterium",
"g__Fusobacterium",
"g__Neisseria",
"g__Olsenella",
"g__Paracoccus",
"g__Porphyromonas",
"g__Propionibacterium",
"g__Rothia",
"g__Streptococcus",
"g__Streptomyces",
"g__Tannerella",
"unclassified"))

pc1_neg_stats_extra$Genus <- factor(pc1_neg_stats_extra$Genus, levels = c("g__Other",
"g__Aggregatibacter",
"g__Eggerthia",
"g__Filifactor",
"g__Fretibacterium",
"g__Fusobacterium",
"g__Johnsonella",
"g__Olsenella",
"g__Parvimonas",
"g__Peptostreptococcus",
"g__Porphyromonas",
"g__Propionibacterium",
"g__Rhodopseudomonas",
"g__Rothia",
"g__Slackia",
"g__Streptococcus",
"g__Tannerella",
"unclassified"))

pc2_pws_stats_extra$Genus <- factor(pc2_pws_stats_extra$Genus, levels = c("g__Other",
"g__Aggregatibacter",
"g__Bacteroidetes_noname",
"g__Eggerthia",
"g__Filifactor",
"g__Fretibacterium",
"g__Fusobacterium",
"g__Haemophilus",
"g__Johnsonella",
"g__Neisseria",
"g__Olsenella",
"g__Parvimonas",
"g__Porphyromonas",
"g__Propionibacterium",
"g__Rothia",
"g__Streptococcus",
"g__Tannerella",
"unclassified"))


pc2_neg_stats_extra$Genus <- factor(pc2_neg_stats_extra$Genus, levels = c("g__Other",
"g__Aggregatibacter",
"g__Bacillus",
"g__Bifidobacterium",
"g__Erysipelothrix",
"g__Fretibacterium",
"g__Lysinibacillus",
"g__Olsenella",
"g__Porphyromonas",
"g__Propionibacterium",
"g__Rothia",
"g__Streptococcus",
"g__Tannerella",
"unclassified"))



# make bar charts of the top species
pc1_pws_stats_extra_kortholog_plot <- pc1_pws_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#FA815FFF","#FCFDBFFF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~KOrtholog, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC1 positive") +
    theme(title = element_text(size=10))

pc1_pws_stats_extra_kortholog_plot

# ggsave("../pc1_pws_stats_extra_kortholog_plot.pdf", plot = pc1_pws_stats_extra_kortholog_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc1_pws_stats_extra_hostgenus_plot <- pc1_pws_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=KOrtholog, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#FA815FFF","#FCFDBFFF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC1 positive") +
    theme(title = element_text(size=10))

pc1_pws_stats_extra_hostgenus_plot

# ggsave("../pc1_pws_stats_extra_hostgenus_plot.pdf", plot = pc1_pws_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc1_neg_stats_extra_kortholog_plot <- pc1_neg_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
#  filter(Percent > 12) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF",
                              "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                              "#287D8EFF","#3CBC75FF","#DCE318FF",
                              "#400F73FF","#AB337CFF","#FCFDBFFF",
                              "#3F4788FF","#238A8DFF","#56C667FF","#FDE725FF",
                              "#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~KOrtholog, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC1 negative") +
    theme(title = element_text(size=10))

pc1_neg_stats_extra_kortholog_plot

# ggsave("../pc1_neg_stats_extra_kortholog_plot.pdf", plot = pc1_neg_stats_extra_kortholog_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc1_neg_stats_extra_hostgenus_plot <- pc1_neg_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
#  filter(Percent > 12) %>%
ggplot(., aes(x=KOrtholog, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF",
                              "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                              "#287D8EFF","#3CBC75FF","#DCE318FF",
                              "#400F73FF","#AB337CFF","#FCFDBFFF",
                              "#3F4788FF","#238A8DFF","#56C667FF","#FDE725FF",
                              "#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC1 negative") +
    theme(title = element_text(size=10))

pc1_neg_stats_extra_hostgenus_plot

# ggsave("../pc1_neg_stats_extra_hostgenus_plot.pdf", plot = pc1_neg_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_pws_stats_extra_kortholog_plot <- pc2_pws_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
#  filter(Percent > 12) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                               "#FDE725FF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~KOrtholog, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC2 positive") +
    theme(title = element_text(size=10))

pc2_pws_stats_extra_kortholog_plot

ggsave("../pc2_pws_stats_extra_kortholog_plot.pdf", plot = pc2_pws_stats_extra_kortholog_plot, device = "pdf",
       scale = 1, width = 8, height = 5, units = c("in"),
       dpi = 300)


pc2_pws_stats_extra_hostgenus_plot <- pc2_pws_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
#  filter(Percent > 12) %>%
ggplot(., aes(x=KOrtholog, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                               "#FDE725FF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC2 positive") +
    theme(title = element_text(size=10))

pc2_pws_stats_extra_hostgenus_plot

# ggsave("../pc2_pws_stats_extra_hostgenus_plot.pdf", plot = pc2_pws_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_neg_stats_extra_kortholog_plot <- pc2_neg_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
#  filter(Percent > 12) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~KOrtholog, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC2 negative") +
    theme(title = element_text(size=10))

pc2_neg_stats_extra_kortholog_plot

# ggsave("../pc2_neg_stats_extra_kortholog_plot.pdf", plot = pc2_neg_stats_extra_kortholog_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_neg_stats_extra_hostgenus_plot <- pc2_neg_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
#  filter(Percent > 12) %>%
ggplot(., aes(x=KOrtholog, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC2 negative") +
    theme(title = element_text(size=10))

pc2_neg_stats_extra_hostgenus_plot

# ggsave("../pc2_neg_stats_extra_hostgenus_plot.pdf", plot = pc2_neg_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

```

Now let's make heat maps to see the distribution of reads assigned to each KO in each sample. Are there host genus-specific differences in the abunance of each KO?

```{r sample_order_heatmaps}
sampleorder.calc = as.vector(c("OME002","OME003","OME005","OME007","OME009","ABM006",
                               "DJA002","EBO001","EBO003","MTK001","MTM001",
                               "MTM004","MTM005","MTM008","MTM009","MTM010",
                               "ABM007","ABM008","CDC005","CDC006","CDC008",
                               "CDC009","CDC010","CDC011","Chimp.150519","DJA006","DJA007",
                               "EBO008","EBO009","KNP001","KNP003","KNP004","KNP005",
                               "KNP006","KNP007","KNP009","ElSidron1","ElSidron2","FUM001",
                               "FUM002","GDN001","GOY005","GOY006","PES001","SPM002",
                               "DLV001","ECO002", "ECO004","EMN001","MOA001","OAK002",
                               "OAK003","OAK004","OAK005","OFN001","RIG001","PLV001","TAF008","TAF017","TAF018",
                               "ESA003","ESA004","ESA006","ESA007","ESA008","LIS001",
                               "LIS002","LIS003","PYK001","PYK002","PYK003","PYK004",
                               "PYK005","JAE006","JAE007","JAE008","JAE009","JAE010",
                               "JAE012","JAE013","JAE014","JAE015","JAE016","VLC001",
                               "VLC002","VLC003","VLC004","VLC005","VLC008","VLC009",
                               "VLC010"))

sampleorder.nomod.calc = as.vector(c("OME002","OME003","OME005","OME007","OME009","ABM006",
                               "DJA002","EBO001","EBO003","MTK001","MTM001",
                               "MTM004","MTM005","MTM008","MTM009","MTM010",
                               "ABM007","ABM008","CDC005","CDC006","CDC008",
                               "CDC009","CDC010","CDC011","Chimp.150519","DJA006","DJA007",
                               "EBO008","EBO009","KNP001","KNP003","KNP004","KNP005",
                               "KNP006","KNP007","KNP009","ElSidron1","ElSidron2","FUM001",
                               "FUM002","GDN001","GOY005","GOY006","PES001","SPM002",
                               "DLV001","ECO002", "ECO004","EMN001","MOA001","OAK002",
                               "OAK003","OAK004","OAK005","OFN001","RIG001","PLV001","TAF008","TAF017","TAF018",
                               "ESA003","ESA004","ESA006","ESA007","ESA008","LIS001",
                               "LIS002","LIS003","PYK001","PYK002","PYK003","PYK004",
                               "PYK005"))
```

```{r individual_heatmaps_pc1p}
gene_families.humann2_pc1p <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         rename(KOrtholog = Pathways) %>%
                                         inner_join(., pc1_pws) %>%
                                         select(-c("PC1","PC2")) %>%
                                         spread(individual_id,CPM) 

rownames(gene_families.humann2_pc1p) <- gene_families.humann2_pc1p$KOrtholog
gene_families.humann2_pc1p <- as.matrix(gene_families.humann2_pc1p %>% select(-KOrtholog))

# CLR-transform the matrix for plotting the heatmap
gene_families.humann2_pc1p_clr = gene_families.humann2_pc1p+1
gene_families.humann2_pc1p_clr <- as.matrix(logratio.transfo(gene_families.humann2_pc1p_clr, logratio = "CLR"))

gene_families.humann2_pc1p_clr <- as.data.frame.matrix(gene_families.humann2_pc1p_clr)
gene_families.humann2_pc1p_clr <- gene_families.humann2_pc1p_clr %>%
  rownames_to_column("KOrtholog") %>%
  select(KOrtholog, everything()) %>%
  gather("individual_id","Counts", 2:ncol(.)) %>%
  spread(KOrtholog,Counts)

rownames(gene_families.humann2_pc1p_clr) <- gene_families.humann2_pc1p_clr$individual_id
gene_families.humann2_pc1p_clr <- gene_families.humann2_pc1p_clr %>% select(-individual_id)

# reorder the table so the rows match as above
gene_families.humann2_pc1p_clr.ordered <- gene_families.humann2_pc1p_clr[match(sampleorder.calc, rownames(gene_families.humann2_pc1p_clr)),]

gene_families.humann2_pc1p_clr.ordered <- as.matrix(gene_families.humann2_pc1p_clr.ordered)
rownames(gene_families.humann2_pc1p_clr.ordered) <- sampleorder.calc



# the heatmaps don't plot properly in an Rmd file, they need to be plotted separately from a new R script file
# gene_families.humann2_pc1p_clr.ordered_heatmap <- heatmap.2(gene_families.humann2_pc1p_clr.ordered, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading KEGG orthologs PC1 positive")
```


```{r individual_heatmaps_pc1n}
gene_families.humann2_pc1n <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         rename(KOrtholog = Pathways) %>%
                                         inner_join(., pc1_neg) %>%
                                         select(-c("PC1","PC2")) %>%
                                         spread(individual_id,CPM) 

rownames(gene_families.humann2_pc1n) <- gene_families.humann2_pc1n$KOrtholog
gene_families.humann2_pc1n <- as.matrix(gene_families.humann2_pc1n %>% select(-KOrtholog))

# CLR-transform the matrix for plotting the heatmap
gene_families.humann2_pc1n_clr = gene_families.humann2_pc1n+1
gene_families.humann2_pc1n_clr <- as.matrix(logratio.transfo(gene_families.humann2_pc1n_clr, logratio = "CLR"))

gene_families.humann2_pc1n_clr <- as.data.frame.matrix(gene_families.humann2_pc1n_clr)
gene_families.humann2_pc1n_clr <- gene_families.humann2_pc1n_clr %>%
  rownames_to_column("KOrtholog") %>%
  select(KOrtholog, everything()) %>%
  gather("individual_id","Counts", 2:ncol(.)) %>%
  spread(KOrtholog,Counts)

rownames(gene_families.humann2_pc1n_clr) <- gene_families.humann2_pc1n_clr$individual_id
gene_families.humann2_pc1n_clr <- gene_families.humann2_pc1n_clr %>% select(-individual_id)

# reorder the table so the rows match as above
gene_families.humann2_pc1n_clr.ordered <- gene_families.humann2_pc1n_clr[match(sampleorder.calc, rownames(gene_families.humann2_pc1n_clr)),]

gene_families.humann2_pc1n_clr.ordered <- as.matrix(gene_families.humann2_pc1n_clr.ordered)
rownames(gene_families.humann2_pc1n_clr.ordered) <- sampleorder.calc


# the heatmaps don't plot properly in an Rmd file, they need to be plotted separately from a new R script file
# gene_families.humann2_pc1n_clr.ordered_heatmap <- heatmap.2(gene_families.humann2_pc1n_clr.ordered, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading KEGG orthologs PC1 negative")
```


```{r individual_heatmaps_pc2p}
gene_families.humann2_pc2p <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         rename(KOrtholog = Pathways) %>%
                                         inner_join(., pc2_pws) %>%
                                         select(-c("PC1","PC2")) %>%
                                         spread(individual_id,CPM) 

rownames(gene_families.humann2_pc2p) <- gene_families.humann2_pc2p$KOrtholog
gene_families.humann2_pc2p <- as.matrix(gene_families.humann2_pc2p %>% select(-KOrtholog))

# CLR-transform the matrix for plotting the heatmap
gene_families.humann2_pc2p_clr = gene_families.humann2_pc2p+1
gene_families.humann2_pc2p_clr <- as.matrix(logratio.transfo(gene_families.humann2_pc2p_clr, logratio = "CLR"))

gene_families.humann2_pc2p_clr <- as.data.frame.matrix(gene_families.humann2_pc2p_clr)
gene_families.humann2_pc2p_clr <- gene_families.humann2_pc2p_clr %>%
  rownames_to_column("KOrtholog") %>%
  select(KOrtholog, everything()) %>%
  gather("individual_id","Counts", 2:ncol(.)) %>%
  spread(KOrtholog,Counts)

rownames(gene_families.humann2_pc2p_clr) <- gene_families.humann2_pc2p_clr$individual_id
gene_families.humann2_pc2p_clr <- gene_families.humann2_pc2p_clr %>% select(-individual_id)

# reorder the table so the rows match as above
gene_families.humann2_pc2p_clr.ordered <- gene_families.humann2_pc2p_clr[match(sampleorder.calc, rownames(gene_families.humann2_pc2p_clr)),]

gene_families.humann2_pc2p_clr.ordered <- as.matrix(gene_families.humann2_pc2p_clr.ordered)
rownames(gene_families.humann2_pc2p_clr.ordered) <- sampleorder.calc

# the heatmaps don't plot properly in an Rmd file, they need to be plotted separately from a new R script file
# gene_families.humann2_pc2p_clr.ordered_heatmap <- heatmap.2(gene_families.humann2_pc2p_clr.ordered, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading KEGG orthologs PC2 positive")
```


```{r individual_heatmaps_pc2n}
gene_families.humann2_pc2n <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         rename(KOrtholog = Pathways) %>%
                                         inner_join(., pc2_neg) %>%
                                         select(-c("PC1","PC2")) %>%
                                         spread(individual_id,CPM) 

rownames(gene_families.humann2_pc2n) <- gene_families.humann2_pc2n$KOrtholog
gene_families.humann2_pc2n <- as.matrix(gene_families.humann2_pc2n %>% select(-KOrtholog))

# CLR-transform the matrix for plotting the heatmap
gene_families.humann2_pc2n_clr = gene_families.humann2_pc2n+1
gene_families.humann2_pc2n_clr <- as.matrix(logratio.transfo(gene_families.humann2_pc2n_clr, logratio = "CLR"))

gene_families.humann2_pc2n_clr <- as.data.frame.matrix(gene_families.humann2_pc2n_clr)
gene_families.humann2_pc2n_clr <- gene_families.humann2_pc2n_clr %>%
  rownames_to_column("KOrtholog") %>%
  select(KOrtholog, everything()) %>%
  gather("individual_id","Counts", 2:ncol(.)) %>%
  spread(KOrtholog,Counts)

rownames(gene_families.humann2_pc2n_clr) <- gene_families.humann2_pc2n_clr$individual_id
gene_families.humann2_pc2n_clr <- gene_families.humann2_pc2n_clr %>% select(-individual_id)

# reorder the table so the rows match as above
gene_families.humann2_pc2n_clr.ordered <- gene_families.humann2_pc2n_clr[match(sampleorder.calc, rownames(gene_families.humann2_pc2n_clr)),]

gene_families.humann2_pc2n_clr.ordered <- as.matrix(gene_families.humann2_pc2n_clr.ordered)
rownames(gene_families.humann2_pc2n_clr.ordered) <- sampleorder.calc

# the heatmaps don't plot properly in an Rmd file, they need to be plotted separately from a new R script file
# gene_families.humann2_pc2n_clr.ordered_heatmap <- heatmap.2(gene_families.humann2_pc2n_clr.ordered, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading KEGG orthologs PC2 negative")
```

```{r individual_heatmaps_pc1p_nomod}
gene_families.humann2_pc1p_nomod <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         rename(KOrtholog = Pathways) %>%
                                         inner_join(., pc1_pws_nomod) %>%
                                         select(-c("PC1","PC2")) %>%
                                         spread(individual_id,CPM) %>%
                                         select(-one_of(c("JAE","VLC")))

rownames(gene_families.humann2_pc1p_nomod) <- gene_families.humann2_pc1p_nomod$KOrtholog
gene_families.humann2_pc1p_nomod <- as.matrix(gene_families.humann2_pc1p_nomod %>% select(-KOrtholog))

# CLR-transform the matrix for plotting the heatmap
gene_families.humann2_pc1p_nomod_clr = gene_families.humann2_pc1p_nomod+1
gene_families.humann2_pc1p_nomod_clr <- as.matrix(logratio.transfo(gene_families.humann2_pc1p_nomod_clr, logratio = "CLR"))

gene_families.humann2_pc1p_nomod_clr <- as.data.frame.matrix(gene_families.humann2_pc1p_nomod_clr)
gene_families.humann2_pc1p_nomod_clr <- gene_families.humann2_pc1p_nomod_clr %>%
  rownames_to_column("KOrtholog") %>%
  select(KOrtholog, everything()) %>%
  gather("individual_id","Counts", 2:ncol(.)) %>%
  spread(KOrtholog,Counts)

rownames(gene_families.humann2_pc1p_nomod_clr) <- gene_families.humann2_pc1p_nomod_clr$individual_id
gene_families.humann2_pc1p_nomod_clr <- gene_families.humann2_pc1p_nomod_clr %>% select(-individual_id)

# reorder the table so the rows match as above
gene_families.humann2_pc1p_nomod_clr.ordered <- gene_families.humann2_pc1p_nomod_clr[match(sampleorder.nomod.calc, rownames(gene_families.humann2_pc1p_nomod_clr)),]

gene_families.humann2_pc1p_nomod_clr.ordered <- as.matrix(gene_families.humann2_pc1p_nomod_clr.ordered)
rownames(gene_families.humann2_pc1p_nomod_clr.ordered) <- sampleorder.nomod.calc

# the heatmaps don't plot properly in an Rmd file, they need to be plotted separately from a new R script file
# gene_families.humann2_pc1p_nomod_clr.ordered_heatmap <- heatmap.2(gene_families.humann2_pc1p_nomod_clr.ordered, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading KEGG orthologs PC1 positive")
```


```{r individual_heatmaps_pc1n_nomod}
gene_families.humann2_pc1n_nomod <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         rename(KOrtholog = Pathways) %>%
                                         inner_join(., pc1_neg_nomod) %>%
                                         select(-c("PC1","PC2")) %>%
                                         spread(individual_id,CPM) %>%
                                         select(-one_of(c("JAE","VLC")))

rownames(gene_families.humann2_pc1n_nomod) <- gene_families.humann2_pc1n_nomod$KOrtholog
gene_families.humann2_pc1n_nomod <- as.matrix(gene_families.humann2_pc1n_nomod %>% select(-KOrtholog))

# CLR-transform the matrix for plotting the heatmap
gene_families.humann2_pc1n_nomod_clr = gene_families.humann2_pc1n_nomod+1
gene_families.humann2_pc1n_nomod_clr <- as.matrix(logratio.transfo(gene_families.humann2_pc1n_nomod_clr, logratio = "CLR"))

gene_families.humann2_pc1n_nomod_clr <- as.data.frame.matrix(gene_families.humann2_pc1n_nomod_clr)
gene_families.humann2_pc1n_nomod_clr <- gene_families.humann2_pc1n_nomod_clr %>%
  rownames_to_column("KOrtholog") %>%
  select(KOrtholog, everything()) %>%
  gather("individual_id","Counts", 2:ncol(.)) %>%
  spread(KOrtholog,Counts)

rownames(gene_families.humann2_pc1n_nomod_clr) <- gene_families.humann2_pc1n_nomod_clr$individual_id
gene_families.humann2_pc1n_nomod_clr <- gene_families.humann2_pc1n_nomod_clr %>% select(-individual_id)

# reorder the table so the rows match as above
gene_families.humann2_pc1n_nomod_clr.ordered <- gene_families.humann2_pc1n_nomod_clr[match(sampleorder.nomod.calc, rownames(gene_families.humann2_pc1n_nomod_clr)),]

gene_families.humann2_pc1n_nomod_clr.ordered <- as.matrix(gene_families.humann2_pc1n_nomod_clr.ordered)
rownames(gene_families.humann2_pc1n_nomod_clr.ordered) <- sampleorder.nomod.calc

# the heatmaps don't plot properly in an Rmd file, they need to be plotted separately from a new R script file
# gene_families.humann2_pc1n_nomod_clr.ordered_heatmap <- heatmap.2(gene_families.humann2_pc1n_nomod_clr.ordered, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading KEGG orthologs PC1 negative")
```


```{r individual_heatmaps_pc2p_nomod}
gene_families.humann2_pc2p_nomod <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         rename(KOrtholog = Pathways) %>%
                                         inner_join(., pc2_pws_nomod) %>%
                                         select(-c("PC1","PC2")) %>%
                                         spread(individual_id,CPM) %>%
                                         select(-one_of(c("JAE","VLC")))

rownames(gene_families.humann2_pc2p_nomod) <- gene_families.humann2_pc2p_nomod$KOrtholog
gene_families.humann2_pc2p_nomod <- as.matrix(gene_families.humann2_pc2p_nomod %>% select(-KOrtholog))

# CLR-transform the matrix for plotting the heatmap
gene_families.humann2_pc2p_nomod_clr = gene_families.humann2_pc2p_nomod+1
gene_families.humann2_pc2p_nomod_clr <- as.matrix(logratio.transfo(gene_families.humann2_pc2p_nomod_clr, logratio = "CLR"))
#gene_families.humann2_pc2p_nomod_clr <- logratio.transfo(gene_families.humann2_pc2p_nomod, logratio = "CLR") %>% as.matrix(.)

gene_families.humann2_pc2p_nomod_clr <- as.data.frame.matrix(gene_families.humann2_pc2p_nomod_clr)
gene_families.humann2_pc2p_nomod_clr <- gene_families.humann2_pc2p_nomod_clr %>%
  rownames_to_column("KOrtholog") %>%
  select(KOrtholog, everything()) %>%
  gather("individual_id","Counts", 2:ncol(.)) %>%
  spread(KOrtholog,Counts)

rownames(gene_families.humann2_pc2p_nomod_clr) <- gene_families.humann2_pc2p_nomod_clr$individual_id
gene_families.humann2_pc2p_nomod_clr <- gene_families.humann2_pc2p_nomod_clr %>% select(-individual_id)

# reorder the table so the rows match as above
gene_families.humann2_pc2p_nomod_clr.ordered <- gene_families.humann2_pc2p_nomod_clr[match(sampleorder.nomod.calc, rownames(gene_families.humann2_pc2p_nomod_clr)),]

gene_families.humann2_pc2p_nomod_clr.ordered <- as.matrix(gene_families.humann2_pc2p_nomod_clr.ordered)
rownames(gene_families.humann2_pc2p_nomod_clr.ordered) <- sampleorder.nomod.calc


# the heatmaps don't plot properly in an Rmd file, they need to be plotted separately from a new R script file
# gene_families.humann2_pc2p_nomod_clr.ordered_heatmap <- heatmap.2(gene_families.humann2_pc2p_nomod_clr.ordered, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading KEGG orthologs PC2 positive")
```


```{r individual_heatmaps_pc2n_nomod}
gene_families.humann2_pc2n_nomod <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\||UNMAPPED|UNINTEGRATED")) %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         rename(KOrtholog = Pathways) %>%
                                         inner_join(., pc2_neg_nomod) %>%
                                         select(-c("PC1","PC2")) %>%
                                         spread(individual_id,CPM) %>%
                                         select(-one_of(c("JAE","VLC")))

rownames(gene_families.humann2_pc2n_nomod) <- gene_families.humann2_pc2n_nomod$KOrtholog
gene_families.humann2_pc2n_nomod <- as.matrix(gene_families.humann2_pc2n_nomod %>% select(-KOrtholog))

# CLR-transform the matrix for plotting the heatmap
gene_families.humann2_pc2n_nomod_clr = gene_families.humann2_pc2n_nomod+1
gene_families.humann2_pc2n_nomod_clr <- as.matrix(logratio.transfo(gene_families.humann2_pc2n_nomod_clr, logratio = "CLR"))

gene_families.humann2_pc2n_nomod_clr <- as.data.frame.matrix(gene_families.humann2_pc2n_nomod_clr)
gene_families.humann2_pc2n_nomod_clr <- gene_families.humann2_pc2n_nomod_clr %>%
  rownames_to_column("KOrtholog") %>%
  select(KOrtholog, everything()) %>%
  gather("individual_id","Counts", 2:ncol(.)) %>%
  spread(KOrtholog,Counts)

rownames(gene_families.humann2_pc2n_nomod_clr) <- gene_families.humann2_pc2n_nomod_clr$individual_id
gene_families.humann2_pc2n_nomod_clr <- gene_families.humann2_pc2n_nomod_clr %>% select(-individual_id)

# reorder the table so the rows match as above
gene_families.humann2_pc2n_nomod_clr.ordered <- gene_families.humann2_pc2n_nomod_clr[match(sampleorder.nomod.calc, rownames(gene_families.humann2_pc2n_nomod_clr)),]

gene_families.humann2_pc2n_nomod_clr.ordered <- as.matrix(gene_families.humann2_pc2n_nomod_clr.ordered)
rownames(gene_families.humann2_pc2n_nomod_clr.ordered) <- sampleorder.nomod.calc

# the heatmaps don't plot properly in an Rmd file, they need to be plotted separately from a new R script file
# gene_families.humann2_pc2n_nomod_clr.ordered_heatmap <- heatmap.2(gene_families.humann2_pc2n_nomod_clr.ordered, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading KEGG orthologs PC2 negative")
```

Since we see that most of the contribution to the PC2 negative KOs in 
*Homo* is from *Strep*, is this still true without the modern humans?
```{r KO_species_contribution_nomod, eval=F}

top10_KOs <- bind_rows(pc1_pws_nomod,pc1_neg_nomod)
top10_KOs <- bind_rows(top10_KOs,pc2_pws_nomod)
top10_KOs <- bind_rows(top10_KOs,pc2_neg_nomod)
top10_KOs <- distinct(top10_KOs %>% select(KOrtholog))
top10_KOs <- str_c(top10_KOs$KOrtholog, collapse = "|")


# gene_families.humann2_top10 <- gene_families.humann2 %>%
#   rename(Taxonomy = Pathways) %>%
#   select(one_of("Taxonomy","DLV001")) %>%
#   filter(str_detect(Taxonomy, "K01903"))

gene_families.humann2_check <- head(gene_families.humann2)
gene_families.humann2_check$Pathways <- gsub('\\.', '\\|', gene_families.humann2_check$Pathways)


gene_families.humann2$Pathways <- gsub('\\.', '\\|', gene_families.humann2$Pathways)
gene_families.humann2_top10 <- gene_families.humann2 %>%
  rename(Taxonomy = Pathways) %>%
  select(-one_of(outliersH2_paths)) %>%
  select(-one_of(sources_oral)) %>%
  select(-starts_with("EXB")) %>%
  select(-starts_with("LIB")) %>%
  select(-starts_with("JAE")) %>%
  select(-starts_with("VLC")) %>%
  filter(str_detect(Taxonomy, top10_KOs)) %>%
  adorn_totals(where = "col", name = "Species_total") %>%
  filter(Species_total != 0) %>%
  # gather("individual_id","CPM",2:ncol(.)) %>%
  # spread(Taxonomy,CPM) %>%
  # filter(Species_total == 0)
  gather("individual_id","CPM",2:ncol(.)) %>%
  mutate(KOrtholog = sapply(Taxonomy, function(f) {
                                  unlist(str_split(f, "\\|"))[1]
                           })) %>%
   mutate(Genus = sapply(Taxonomy, function(f) {
                                  unlist(str_split(f, "\\|"))[2]
                           })) %>%
  mutate(Species = sapply(Taxonomy, function(f) {
                                  unlist(str_split(f, "\\|"))[3]
                           })) %>%
  inner_join(., fullmetadata, by = "individual_id") %>%
  select(individual_id,Host_Common,Host_Genus,KOrtholog,Genus,Species,CPM)



pc1_pws_nomod_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc1_pws_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_nomod_stats_homo['KOrtholog'] <- pc1_pws_nomod[1,1]
pc1_pws_nomod_stats_homo['Host_Genus'] <- 'Homo'

pc1_pws_nomod_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc1_pws_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_nomod_stats_pan['KOrtholog'] <- pc1_pws_nomod[1,1]
pc1_pws_nomod_stats_pan['Host_Genus'] <- 'Pan'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats_homo,pc1_pws_nomod_stats_pan)

pc1_pws_nomod_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc1_pws_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_nomod_stats_gorilla['KOrtholog'] <- pc1_pws_nomod[1,1]
pc1_pws_nomod_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats,pc1_pws_nomod_stats_gorilla)

pc1_pws_nomod_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc1_pws_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_nomod_stats_alouatta['KOrtholog'] <- pc1_pws_nomod[1,1]
pc1_pws_nomod_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats,pc1_pws_nomod_stats_alouatta)

pc1_neg_nomod_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc1_neg_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_nomod_stats_homo['KOrtholog'] <- pc1_neg_nomod[1,1]
pc1_neg_nomod_stats_homo['Host_Genus'] <- 'Homo'

pc1_neg_nomod_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc1_neg_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_nomod_stats_pan['KOrtholog'] <- pc1_neg_nomod[1,1]
pc1_neg_nomod_stats_pan['Host_Genus'] <- 'Pan'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats_homo,pc1_neg_nomod_stats_pan)

pc1_neg_nomod_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc1_neg_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_nomod_stats_gorilla['KOrtholog'] <- pc1_neg_nomod[1,1]
pc1_neg_nomod_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats,pc1_neg_nomod_stats_gorilla)

pc1_neg_nomod_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc1_neg_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_nomod_stats_alouatta['KOrtholog'] <- pc1_neg_nomod[1,1]
pc1_neg_nomod_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats,pc1_neg_nomod_stats_alouatta)


pc2_pws_nomod_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc2_pws_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_nomod_stats_homo['KOrtholog'] <- pc2_pws_nomod[1,1]
pc2_pws_nomod_stats_homo['Host_Genus'] <- 'Homo'

pc2_pws_nomod_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc2_pws_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_nomod_stats_pan['KOrtholog'] <- pc2_pws_nomod[1,1]
pc2_pws_nomod_stats_pan['Host_Genus'] <- 'Pan'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats_homo,pc2_pws_nomod_stats_pan)

pc2_pws_nomod_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc2_pws_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_nomod_stats_gorilla['KOrtholog'] <- pc2_pws_nomod[1,1]
pc2_pws_nomod_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats,pc2_pws_nomod_stats_gorilla)

pc2_pws_nomod_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc2_pws_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_nomod_stats_alouatta['KOrtholog'] <- pc2_pws_nomod[1,1]
pc2_pws_nomod_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats,pc2_pws_nomod_stats_alouatta)



pc2_neg_nomod_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc2_neg_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_nomod_stats_homo['KOrtholog'] <- pc2_neg_nomod[1,1]
pc2_neg_nomod_stats_homo['Host_Genus'] <- 'Homo'

pc2_neg_nomod_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc2_neg_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_nomod_stats_pan['KOrtholog'] <- pc2_neg_nomod[1,1]
pc2_neg_nomod_stats_pan['Host_Genus'] <- 'Pan'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats_homo,pc2_neg_nomod_stats_pan)

pc2_neg_nomod_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc2_neg_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_nomod_stats_gorilla['KOrtholog'] <- pc2_neg_nomod[1,1]
pc2_neg_nomod_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats,pc2_neg_nomod_stats_gorilla)

pc2_neg_nomod_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc2_neg_nomod[1,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_nomod_stats_alouatta['KOrtholog'] <- pc2_neg_nomod[1,1]
pc2_neg_nomod_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats,pc2_neg_nomod_stats_alouatta)

# now add orthologs 2-10 to each table by chaning the row number in the sections below
pc1_pws_nomod_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc1_pws_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_nomod_stats_homo['KOrtholog'] <- pc1_pws_nomod[10,1]
pc1_pws_nomod_stats_homo['Host_Genus'] <- 'Homo'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats, pc1_pws_nomod_stats_homo)

pc1_pws_nomod_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc1_pws_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_nomod_stats_pan['KOrtholog'] <- pc1_pws_nomod[10,1]
pc1_pws_nomod_stats_pan['Host_Genus'] <- 'Pan'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats,pc1_pws_nomod_stats_pan)

pc1_pws_nomod_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc1_pws_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_nomod_stats_gorilla['KOrtholog'] <- pc1_pws_nomod[10,1]
pc1_pws_nomod_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats,pc1_pws_nomod_stats_gorilla)

pc1_pws_nomod_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc1_pws_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_pws_nomod_stats_alouatta['KOrtholog'] <- pc1_pws_nomod[10,1]
pc1_pws_nomod_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats,pc1_pws_nomod_stats_alouatta)



pc1_neg_nomod_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc1_neg_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_nomod_stats_homo['KOrtholog'] <- pc1_neg_nomod[10,1]
pc1_neg_nomod_stats_homo['Host_Genus'] <- 'Homo'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats, pc1_neg_nomod_stats_homo)

pc1_neg_nomod_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc1_neg_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_nomod_stats_pan['KOrtholog'] <- pc1_neg_nomod[10,1]
pc1_neg_nomod_stats_pan['Host_Genus'] <- 'Pan'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats,pc1_neg_nomod_stats_pan)

pc1_neg_nomod_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc1_neg_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_nomod_stats_gorilla['KOrtholog'] <- pc1_neg_nomod[10,1]
pc1_neg_nomod_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats,pc1_neg_nomod_stats_gorilla)

pc1_neg_nomod_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc1_neg_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc1_neg_nomod_stats_alouatta['KOrtholog'] <- pc1_neg_nomod[10,1]
pc1_neg_nomod_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats,pc1_neg_nomod_stats_alouatta)


pc2_pws_nomod_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc2_pws_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM))  %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_nomod_stats_homo['KOrtholog'] <- pc2_pws_nomod[10,1]
pc2_pws_nomod_stats_homo['Host_Genus'] <- 'Homo'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats, pc2_pws_nomod_stats_homo)

pc2_pws_nomod_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc2_pws_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_nomod_stats_pan['KOrtholog'] <- pc2_pws_nomod[10,1]
pc2_pws_nomod_stats_pan['Host_Genus'] <- 'Pan'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats,pc2_pws_nomod_stats_pan)

pc2_pws_nomod_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc2_pws_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_nomod_stats_gorilla['KOrtholog'] <- pc2_pws_nomod[10,1]
pc2_pws_nomod_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats,pc2_pws_nomod_stats_gorilla)

pc2_pws_nomod_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc2_pws_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_pws_nomod_stats_alouatta['KOrtholog'] <- pc2_pws_nomod[10,1]
pc2_pws_nomod_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats,pc2_pws_nomod_stats_alouatta)


pc2_neg_nomod_stats_homo <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Homo"))%>%
  filter(str_detect(KOrtholog, pc2_neg_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_nomod_stats_homo['KOrtholog'] <- pc2_neg_nomod[10,1]
pc2_neg_nomod_stats_homo['Host_Genus'] <- 'Homo'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats, pc2_neg_nomod_stats_homo)

pc2_neg_nomod_stats_pan <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Pan"))%>%
  filter(str_detect(KOrtholog, pc2_neg_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_nomod_stats_pan['KOrtholog'] <- pc2_neg_nomod[10,1]
pc2_neg_nomod_stats_pan['Host_Genus'] <- 'Pan'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats,pc2_neg_nomod_stats_pan)

pc2_neg_nomod_stats_gorilla <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Gorilla"))%>%
  filter(str_detect(KOrtholog, pc2_neg_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_nomod_stats_gorilla['KOrtholog'] <- pc2_neg_nomod[10,1]
pc2_neg_nomod_stats_gorilla['Host_Genus'] <- 'Gorilla'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats,pc2_neg_nomod_stats_gorilla)

pc2_neg_nomod_stats_alouatta <- gene_families.humann2_top10 %>%
  filter(str_detect(Host_Genus, "Alouatta"))%>%
  filter(str_detect(KOrtholog, pc2_neg_nomod[10,1])) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(CPM)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
pc2_neg_nomod_stats_alouatta['KOrtholog'] <- pc2_neg_nomod[10,1]
pc2_neg_nomod_stats_alouatta['Host_Genus'] <- 'Alouatta'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats,pc2_neg_nomod_stats_alouatta)


write.table(pc1_pws_nomod_stats, "../04-analysis/screening/humann2/tables/pc1_pws_nomod_stats.tsv", quote=F, sep="\t", row.names = F)
write.table(pc1_neg_nomod_stats, "../04-analysis/screening/humann2/tables/pc1_neg_nomod_stats.tsv", quote=F, sep="\t", row.names = F)
write.table(pc2_pws_nomod_stats, "../04-analysis/screening/humann2/tables/pc2_pws_nomod_stats.tsv", quote=F, sep="\t", row.names = F)
write.table(pc2_neg_nomod_stats, "../04-analysis/screening/humann2/tables/pc2_neg_nomod_stats.tsv", quote=F, sep="\t", row.names = F)

```



Read in the files made above so you don't have to make them again, and plot the
species distributions of KOrthologs in each host genus without modern humans.
```{r KO_species_contribution_nomod_graphs}
pc1_pws_nomod_stats <- fread("../04-analysis/screening/humann2/tables/pc1_pws_nomod_stats.tsv")
pc1_neg_nomod_stats <- fread("../04-analysis/screening/humann2/tables/pc1_neg_nomod_stats.tsv")
pc2_pws_nomod_stats <- fread("../04-analysis/screening/humann2/tables/pc2_pws_nomod_stats.tsv")
pc2_neg_nomod_stats <- fread("../04-analysis/screening/humann2/tables/pc2_neg_nomod_stats.tsv")

pc1_pws_nomod_stats <- pc1_pws_nomod_stats %>% select(KOrtholog,Host_Genus,Genus,Sum,Percent)
pc1_pws_nomod_stats$Host_Genus <- factor(pc1_pws_nomod_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc1_neg_nomod_stats <- pc1_neg_nomod_stats %>% select(KOrtholog,Host_Genus,Genus,Sum,Percent)
pc1_neg_nomod_stats$Host_Genus <- factor(pc1_neg_nomod_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc2_pws_nomod_stats <- pc2_pws_nomod_stats %>% select(KOrtholog,Host_Genus,Genus,Sum,Percent)
pc2_pws_nomod_stats$Host_Genus <- factor(pc2_pws_nomod_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc2_neg_nomod_stats <- pc2_neg_nomod_stats %>% select(KOrtholog,Host_Genus,Genus,Sum,Percent)
pc2_neg_nomod_stats$Host_Genus <- factor(pc2_neg_nomod_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))

# remove all species with <12% contribution and fill in that % with Other (g__Other)
pc1_top10 <- pc1_pws_nomod$KOrtholog

pc1_pws_nomod_stats_extra <- lapply(pc1_top10, function(eclass) {
 fisher_subsistence <- pc1_pws_nomod_stats %>%
   group_by(Host_Genus) %>%
   filter(KOrtholog == eclass) %>%
   filter(Percent <=12) %>%
#   filter(Percent >12) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc1_pws_nomod_stats_extra['KOrtholog'] <- c("K00526","K00526","K00526","K00526",
                                      "K00658","K00658","K00658","K00658",
                                      "K00674","K00674","K00674","K00674",
                                      "K01657","K01657","K01657","K01657",
                                      "K01687","K01687","K01687","K01687",
                                      "K01696","K01696","K01696","K01696",
                                      "K01902","K01902","K01902","K01902",
                                      "K01903","K01903","K01903","K01903",
                                      "K01961","K01961","K01961","K01961",
                                      "K07173","K07173","K07173","K07173")
pc1_pws_nomod_stats_extra['Sum'] <- NA
pc1_pws_nomod_stats_extra['Genus'] <- 'g__Other'
pc1_pws_nomod_stats_extra <- pc1_pws_nomod_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(KOrtholog,Host_Genus,Genus,Sum,Percent)

pc1_pws_nomod_stats_extra <- pc1_pws_nomod_stats_extra %>%
  bind_rows(., pc1_pws_nomod_stats %>%
              filter(Percent >12))

pc1_top10 <- pc1_neg_nomod$KOrtholog

pc1_neg_nomod_stats_extra <- lapply(pc1_top10, function(eclass) {
 fisher_subsistence <- pc1_neg_nomod_stats %>%
   group_by(Host_Genus) %>%
   filter(KOrtholog == eclass) %>%
   filter(Percent <=12) %>%
#   filter(Percent >12) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc1_neg_nomod_stats_extra['KOrtholog'] <- c("K01999","K01999","K01999","K01999",
                                      "K02005","K02005","K02005","K02005",
                                      "K02034","K02034","K02034","K02034",
                                      "K02117","K02117","K02117","K02117",
                                      "K03205","K03205","K03205","K03205",
                                      "K07052","K07052","K07052","K07052",
                                      "K07133","K07133","K07133","K07133",
                                      "K07481","K07481","K07481","K07481",
                                      "K16785","K16785","K16785","K16785",
                                      "K16926","K16926","K16926","K16926")
pc1_neg_nomod_stats_extra['Sum'] <- NA
pc1_neg_nomod_stats_extra['Genus'] <- 'g__Other'
pc1_neg_nomod_stats_extra <- pc1_neg_nomod_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(KOrtholog,Host_Genus,Genus,Sum,Percent)

pc1_neg_nomod_stats_extra <- pc1_neg_nomod_stats_extra %>%
  bind_rows(., pc1_neg_nomod_stats %>%
              filter(Percent >12))

pc2_top10 <- pc2_pws_nomod$KOrtholog

pc2_pws_nomod_stats_extra <- lapply(pc2_top10, function(eclass) {
 fisher_subsistence <- pc2_pws_nomod_stats %>%
   group_by(Host_Genus) %>%
   filter(KOrtholog == eclass) %>%
   filter(Percent <=12) %>%
#   filter(Percent >12) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc2_pws_nomod_stats_extra['KOrtholog'] <- c("K01104","K01104","K01104","K01104",
                                      "K01496","K01496","K01496","K01496",
                                      "K01990","K01990","K01990","K01990",
                                      "K02026","K02026","K02026","K02026",
                                      "K02027","K02027","K02027","K02027",
                                      "K02440","K02440","K02440","K02440",
                                      "K03402","K03402","K03402","K03402",
                                      "K03574","K03574","K03574","K03574",
                                      "K07052","K07052","K07052","K07052",
                                      "K15634","K15634","K15634","K15634")
pc2_pws_nomod_stats_extra['Sum'] <- NA
pc2_pws_nomod_stats_extra['Genus'] <- 'g__Other'
pc2_pws_nomod_stats_extra <- pc2_pws_nomod_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(KOrtholog,Host_Genus,Genus,Sum,Percent)

pc2_pws_nomod_stats_extra <- pc2_pws_nomod_stats_extra %>%
  bind_rows(., pc2_pws_nomod_stats %>%
              filter(Percent >12))

pc2_top10 <- pc2_neg_nomod$KOrtholog

pc2_neg_nomod_stats_extra <- lapply(pc2_top10, function(eclass) {
 fisher_subsistence <- pc2_neg_nomod_stats %>%
   group_by(Host_Genus) %>%
   filter(KOrtholog == eclass) %>%
   filter(Percent <=12) %>%
#   filter(Percent >12) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc2_neg_nomod_stats_extra['KOrtholog'] <- c("K00027","K00027","K00027","K00027",
                                      "K00615","K00615","K00615","K00615",
                                      "K01610","K01610","K01610","K01610",
                                      "K01759","K01759","K01759","K01759",
                                      "K01955","K01955","K01955","K01955",
                                      "K02012","K02012","K02012","K02012",
                                      "K03043","K03043","K03043","K03043",
                                      "K03046","K03046","K03046","K03046",
                                      "K03169","K03169","K03169","K03169",
                                      "K03569","K03569","K03569","K03569")
pc2_neg_nomod_stats_extra['Sum'] <- NA
pc2_neg_nomod_stats_extra['Genus'] <- 'g__Other'
pc2_neg_nomod_stats_extra <- pc2_neg_nomod_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(KOrtholog,Host_Genus,Genus,Sum,Percent)

pc2_neg_nomod_stats_extra <- pc2_neg_nomod_stats_extra %>%
  bind_rows(., pc2_neg_nomod_stats %>%
              filter(Percent >12))

# set the species as factor for ordering the colors with g__Other at the top and unclassified at the bottom
# these orders are set after looking at the graphs with no ordering specified
pc1_pws_nomod_stats_extra$Genus <- factor(pc1_pws_nomod_stats_extra$Genus, levels = c("g__Other",
"g__Aggregatibacter",
"g__Cryptobacterium",
"g__Fusobacterium",
"g__Neisseria",
"g__Olsenella",
"g__Paracoccus",
"g__Porphyromonas",
"g__Propionibacterium",
"g__Rothia",
"g__Streptococcus",
"g__Streptomyces",
"g__Tannerella",
"unclassified"))

pc1_neg_nomod_stats_extra$Genus <- factor(pc1_neg_nomod_stats_extra$Genus, levels = c("g__Other",
"g__Aggregatibacter",
"g__Eggerthia",
"g__Eikenella",
"g__Filifactor",
"g__Fretibacterium",
"g__Fusobacterium",
"g__Johnsonella",
"g__Neisseria",
"g__Olsenella",
"g__Parvimonas",
"g__Peptostreptococcus",
"g__Porphyromonas",
"g__Propionibacterium",
"g__Pseudomonas",
"g__Rhodopseudomonas",
"g__Rothia",
"g__Slackia",
"g__Streptococcus",
"g__Tannerella",
"unclassified"))

pc2_pws_nomod_stats_extra$Genus <- factor(pc2_pws_nomod_stats_extra$Genus, levels = c("g__Other",
"g__Aggregatibacter",
"g__Bacillus",
"g__Bifidobacterium",
"g__Fretibacterium",
"g__Neisseria",
"g__Olsenella",
"g__Porphyromonas",
"g__Propionibacterium",
"g__Rothia",
"g__Streptococcus",
"g__Tannerella",
"unclassified"))

pc2_neg_nomod_stats_extra$Genus <- factor(pc2_neg_nomod_stats_extra$Genus, levels = c("g__Other",
"g__Aggregatibacter",
"g__Bacteroidetes_noname",
"g__Campylobacter",
"g__Fretibacterium",
"g__Fusobacterium",
"g__Methanobrevibacter",
"g__Mogibacterium",
"g__Neisseria",
"g__Olsenella",
"g__Porphyromonas",
"g__Propionibacterium",
"g__Streptococcus",
"g__Tannerella",
"unclassified"))


# make bar charts of the top species
pc1_pws_nomod_stats_extra_kortholog_plot <- pc1_pws_nomod_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#FA815FFF","#FCFDBFFF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~KOrtholog, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC1 positive") +
    theme(title = element_text(size=10))

pc1_pws_nomod_stats_extra_kortholog_plot

# ggsave("../pc1_pws_nomod_stats_extra_kortholog_plot.pdf", plot = pc1_pws_nomod_stats_extra_kortholog_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc1_pws_nomod_stats_extra_hostgenus_plot <- pc1_pws_nomod_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=KOrtholog, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#FA815FFF","#FCFDBFFF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC1 positive") +
    theme(title = element_text(size=10))

pc1_pws_nomod_stats_extra_hostgenus_plot

ggsave("../pc1_pws_nomod_stats_extra_hostgenus_plot.pdf", plot = pc1_pws_nomod_stats_extra_hostgenus_plot, device = "pdf",
       scale = 1, width = 8, height = 5, units = c("in"),
       dpi = 300)


pc1_neg_nomod_stats_extra_kortholog_plot <- pc1_neg_nomod_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                              "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                              "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                              "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                              "#3F4788FF","#238A8DFF","#56C667FF","#FDE725FF",
                              "#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~KOrtholog, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC1 negative") +
    theme(title = element_text(size=10))

pc1_neg_nomod_stats_extra_kortholog_plot

# ggsave("../pc1_neg_nomod_stats_extra_kortholog_plot.pdf", plot = pc1_neg_nomod_stats_extra_kortholog_plot, device = "pdf",
#        scale = 1, width = 10, height = 5, units = c("in"),
#        dpi = 300)


pc1_neg_nomod_stats_extra_hostgenus_plot <- pc1_neg_nomod_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=KOrtholog, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
 scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                              "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                              "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                              "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                              "#3F4788FF","#238A8DFF","#56C667FF","#FDE725FF",
                              "#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC1 negative") +
    theme(title = element_text(size=10))

pc1_neg_nomod_stats_extra_hostgenus_plot

# ggsave("../pc1_neg_nomod_stats_extra_hostgenus_plot.pdf", plot = pc1_neg_nomod_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 10, height = 5, units = c("in"),
#        dpi = 300)

pc2_pws_nomod_stats_extra_kortholog_plot <- pc2_pws_nomod_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~KOrtholog, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC2 positive") +
    theme(title = element_text(size=10))

pc2_pws_nomod_stats_extra_kortholog_plot

# ggsave("../pc2_pws_nomod_stats_extra_kortholog_plot.pdf", plot = pc2_pws_nomod_stats_extra_kortholog_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_pws_nomod_stats_extra_hostgenus_plot <- pc2_pws_nomod_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=KOrtholog, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC2 positive") +
    theme(title = element_text(size=10))

pc2_pws_nomod_stats_extra_hostgenus_plot

# ggsave("../pc2_pws_nomod_stats_extra_hostgenus_plot.pdf", plot = pc2_pws_nomod_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_neg_nomod_stats_extra_kortholog_plot <- pc2_neg_nomod_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","orange",
                               "#6B1D81FF","#D6456CFF","pink",
                               "navyblue","#287D8EFF","#3CBC75FF",
                               "#400F73FF","#AB337CFF","#FCFDBFFF",
                               "#FDE725FF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~KOrtholog, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC2 negative") +
    theme(title = element_text(size=10))

pc2_neg_nomod_stats_extra_kortholog_plot

# ggsave("../pc2_neg_nomod_stats_extra_kortholog_plot.pdf", plot = pc2_neg_nomod_stats_extra_kortholog_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_neg_nomod_stats_extra_hostgenus_plot <- pc2_neg_nomod_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=KOrtholog, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#440154FF","#39558CFF","#1F968BFF","orange",
                               "#6B1D81FF","#D6456CFF","pink",
                               "navyblue","#287D8EFF","#3CBC75FF",
                               "#400F73FF","#AB337CFF","#FCFDBFFF",
                               "#FDE725FF","#bdbdbd")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow = 2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to KEGG Orthologs, PC2 negative") +
    theme(title = element_text(size=10))

pc2_neg_nomod_stats_extra_hostgenus_plot

# ggsave("../pc2_neg_nomod_stats_extra_hostgenus_plot.pdf", plot = pc2_neg_nomod_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

```


Now look at just the KOs that are in the 15 KEGG Carbohydrate pathways. Are they
more prevalent/abundant in *Homo* than *Alouatta/Gorilla/Pan*? Do they separate the
host genera in a PCA?
```{r KO_counts}
# read in the file with KEGG carbohydrate path KOs
kegg_carb_kos <- fread("../04-analysis/screening/humann2/tables/kegg_carbohydrates.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_carbs <- distinct(kegg_carb_kos, KOrtholog)
unique_KOs_carbs <- str_c(unique_KOs_carbs$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG carbohydrate paths
humann2_KEGG_carbs <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_carbs))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_oral)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB"))

# to calculate the number of KOs in each population we need a prevalence table
prev_ko_carbs <- humann2_KEGG_carbs %>%
  gather("sample","CPM", 2:ncol(.)) %>%
  rename(., Present = CPM)

# make Relab binary to make a presence/absence table
prev_ko_carbs$Present[prev_ko_carbs$Present>0] <- 1

# convert to a wide table
prev_ko_carbs <- prev_ko_carbs %>%
                 spread(Pathways,Present)
rownames(prev_ko_carbs) <- prev_ko_carbs$sample
prev_ko_carbs <- prev_ko_carbs %>% select(-sample)

# calculate observed species for each sample with janitor 'adorn_totals'
prev_ko_carbs_totals <- prev_ko_carbs %>%
  adorn_totals(., where = "col", name = "Total_ko") %>%
  select(Total_ko) %>%
  rownames_to_column("individual_id") %>%
  inner_join(., fullmetadata) %>%
  select(individual_id,Host_Genus,Host_Common,Host_General,Total_ko)


# set the subsistence as factors for plotting
prev_ko_carbs_totals$Host_Common <- factor(prev_ko_carbs_totals$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))


# make box plots by subsistence strategy and population
kegg_carbo_ko_totals <- ggplot(prev_ko_carbs_totals, aes(x=Host_Common, y = Total_ko)) +
    geom_boxplot() +
    geom_point(size = 3, aes(colour = Host_Common, shape = Host_Common)) +
    scale_colour_manual(values = common_colours) +
    scale_shape_manual(values = common_shapes) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    ylab("Number of KEGG Orthologs") +
    ggtitle("Number of KEGG Carbohydrate Pathway Orthologs")

kegg_carbo_ko_totals

kegg_carbo_ko_totals <- ggplot(prev_ko_carbs_totals, aes(x=Host_Common, y = Total_ko)) +
    geom_boxplot() +
    geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
    scale_colour_manual(values = general_colors) +
    scale_shape_manual(values = general_shapes) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    ylab("Number of KEGG Orthologs") +
    ggtitle("Number of KEGG Carbohydrate Pathway Orthologs")

kegg_carbo_ko_totals

# ggsave("../kegg_carbo_ko_totals.pdf", plot = kegg_carbo_ko_totals, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# and let's determine the median and SD, and see if there are statistically significant differences between the groups, with Homo Modern and Neanderthal separate and combined
prev_ko_carbs_totals_stats <- prev_ko_carbs_totals %>%
  group_by(Host_Common) %>%
  summarise(
    mean = mean(Total_ko),
    sd = sd(Total_ko),
    median = median(Total_ko),
    count = n(),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Total_ko)
  )

prev_ko_carbs_totals_stats <- prev_ko_carbs_totals_stats %>% rename(Host_Genus = Host_Common)

prev_ko_carbs_totals_stats <- bind_rows(prev_ko_carbs_totals_stats, prev_ko_carbs_totals %>%
  group_by(Host_Genus) %>%
    summarise(
    mean = mean(Total_ko),
    sd = sd(Total_ko),
    median = median(Total_ko),
    count = n(),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Total_ko))%>% 
    filter(str_detect(Host_Genus, "Homo")))

prev_ko_carbs_totals_stats$Host_Genus <- factor(prev_ko_carbs_totals_stats$Host_Genus, levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)","Homo","Plaque"))

pairwise.wilcox.test(prev_ko_carbs_totals$Total_ko, prev_ko_carbs_totals$Host_Genus,
                     p.adjust.method = "fdr")
pairwise.wilcox.test(prev_ko_carbs_totals$Total_ko, prev_ko_carbs_totals$Host_Common,
                     p.adjust.method = "fdr")

prev_ko_carbs_totals_nomod <- prev_ko_carbs_totals %>% filter(!str_detect(Host_General, "ModernDayHuman"))
pairwise.wilcox.test(prev_ko_carbs_totals_nomod$Total_ko, prev_ko_carbs_totals_nomod$Host_Common,
                     p.adjust.method = "fdr")

```

And repeat the section above but for abundance
```{r KO_abundance}
# select out the KEGG orthologs that are involved in the KEGG carbohydrate paths
humann2_KEGG_carbs <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_carbs))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_oral)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB"))

# to calculate the number of KOs in each population we need a abundalence table
abund_ko_carbs <- humann2_KEGG_carbs %>%
  gather("sample","CPM", 2:ncol(.)) 

# convert to a wide table
abund_ko_carbs <- abund_ko_carbs %>%
                 spread(Pathways,CPM)
rownames(abund_ko_carbs) <- abund_ko_carbs$sample
abund_ko_carbs <- abund_ko_carbs %>% select(-sample)

# calculate observed species for each sample with janitor 'adorn_totals'
abund_ko_carbs_totals <- abund_ko_carbs %>%
  adorn_totals(., where = "col", name = "Total_ko") %>%
  select(Total_ko) %>%
  rownames_to_column("individual_id") %>%
  inner_join(., fullmetadata) %>%
  select(individual_id,Host_Genus,Host_Common,Host_General,Total_ko)


# set the subsistence as factors for plotting
abund_ko_carbs_totals$Host_Common <- factor(abund_ko_carbs_totals$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))


# make box plots by subsistence strategy and population
kegg_carbo_ko_totals_abund <- ggplot(abund_ko_carbs_totals, aes(x=Host_Common, y = Total_ko)) +
    geom_boxplot() +
    geom_point(size = 3, aes(colour = Host_Common, shape = Host_Common)) +
    scale_colour_manual(values = common_colours) +
    scale_shape_manual(values = common_shapes) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    ylab("Copies per million") +
    ggtitle("Abundance of KEGG Carbohydrate Pathway Orthologs")

kegg_carbo_ko_totals_abund

kegg_carbo_ko_totals_abund <- ggplot(abund_ko_carbs_totals, aes(x=Host_Common, y = Total_ko)) +
    geom_boxplot() +
    geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
    scale_colour_manual(values = general_colors) +
    scale_shape_manual(values = general_shapes) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    ylab("Copies per million") +
    ggtitle("Abundance of KEGG Carbohydrate Pathway Orthologs")

kegg_carbo_ko_totals_abund


# ggsave("../kegg_carbo_ko_totals_abund.pdf", plot = kegg_carbo_ko_totals_abund, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

humann2_kegg_carbs <- plot_grid(kegg_carbo_ko_totals, kegg_carbo_ko_totals_abund,
          nrow=1, labels = c("A", "B"), rel_widths = c(1, 1.33))

#save_plot(kegg_carbo_ko_totals, kegg_carbo_ko_totals_abund, plot, nrow = 1, labels = c("A", "B"), rel_widths = c(1, 1.33))

# and let's determine the median and SD, and see if there are statistically significant differences between the groups, with Homo Modern and Neanderthal separate and combined
abund_ko_carbs_totals_stats <- abund_ko_carbs_totals %>%
  group_by(Host_Common) %>%
  summarise(
    mean = mean(Total_ko),
    sd = sd(Total_ko),
    median = median(Total_ko),
    count = n(),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Total_ko)
  )

abund_ko_carbs_totals_stats <- abund_ko_carbs_totals_stats %>% rename(Host_Genus = Host_Common)

abund_ko_carbs_totals_stats <- bind_rows(abund_ko_carbs_totals_stats, abund_ko_carbs_totals %>%
  group_by(Host_Genus) %>%
    summarise(
    mean = mean(Total_ko),
    sd = sd(Total_ko),
    median = median(Total_ko),
    count = n(),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Total_ko))%>% 
    filter(str_detect(Host_Genus, "Homo")))

abund_ko_carbs_totals_stats$Host_Genus <- factor(abund_ko_carbs_totals_stats$Host_Genus, levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)","Homo","Plaque"))

pairwise.wilcox.test(abund_ko_carbs_totals$Total_ko, abund_ko_carbs_totals$Host_Genus,
                     p.adjust.method = "fdr")
pairwise.wilcox.test(abund_ko_carbs_totals$Total_ko, abund_ko_carbs_totals$Host_Common,
                     p.adjust.method = "fdr")

abund_ko_carbs_totals_nomod <- abund_ko_carbs_totals %>% filter(!str_detect(Host_General, "ModernDayHuman"))
pairwise.wilcox.test(abund_ko_carbs_totals_nomod$Total_ko, abund_ko_carbs_totals_nomod$Host_Common,
                     p.adjust.method = "fdr")

```

Now run and plot a PCA on just the KOs in selected pathways. Do KOs in each
pathway separate the host genera? Are any pathways better at separating the
groups than others? Note - this does not appear to be the case. Most separate
*Gorilla/Alouatta* from *Homo/Pan* along PC2, with *Homo* spread across PC1 and
*Pan* at one end of the *Homo* PC1 spread.
```{r pca_ko_aminoacids}
# read in the files with KOs for the pathways we'll look at
# Amino acids
kegg_aminoacids_kos <- fread("../04-analysis/screening/humann2/tables/kegg_amino_acids.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_aminoacids <- distinct(kegg_aminoacids_kos, KOrtholog)
unique_KOs_aminoacids <- str_c(unique_KOs_aminoacids$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG amino acids paths
humann2_KEGG_aminoacids <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_aminoacids))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

aminoacids.kegg <- humann2_KEGG_aminoacids %>%
  gather("KOrtholog","Counts",2:ncol(.)) %>%
  spread(individual_id,Counts) %>%
  select(KOrtholog) %>%
  inner_join(., kegg_aminoacids_kos)

# write.table(aminoacids.kegg, "../04-analysis/screening/humann2/tables/aminoacids.kegg.tsv", sep = "\t", quote = F, row.names = F)

rownames(humann2_KEGG_aminoacids) <- humann2_KEGG_aminoacids$individual_id
humann2_KEGG_aminoacids <- humann2_KEGG_aminoacids %>% select(-individual_id)
humann2_KEGG_aminoacids <- humann2_KEGG_aminoacids + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_aminoacids <- tune.pca(humann2_KEGG_aminoacids, logratio = 'CLR')
tune.h2_KO_aminoacids

# perform a PCA to see how the data cluster
h2_KO_aminoacids.pca <- mixOmics::pca(humann2_KEGG_aminoacids, ncomp = 3, logratio = 'CLR')
plot(h2_KO_aminoacids.pca)

h2_KO_aminoacids.pca.variates <- as.data.frame(h2_KO_aminoacids.pca$variates$X)
h2_KO_aminoacids.pca.variates <- h2_KO_aminoacids.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_aminoacids.pca.varmet <- inner_join(h2_KO_aminoacids.pca.variates, fullmetadata, by = "individual_id")

h2_KO_aminoacids.pca.varmet$Host_Common <- factor(h2_KO_aminoacids.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_aminoacids_pc1pc2 <- ggplot(as.data.frame(h2_KO_aminoacids.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_aminoacids.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_aminoacids.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Amino acid pathway orthologs")

pca_h2_KO_aminoacids_pc1pc2

# ggsave("../pca_h2_KO_aminoacids_pc1pc2.pdf", plot = pca_h2_KO_aminoacids_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_aminoacids.pca.mat <- as.data.frame.matrix(h2_KO_aminoacids.pca$X)

# Euclidean distance
h2_KO_aminoacids.pca.mat.euc <- vegdist(h2_KO_aminoacids.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_aminoacids.pca.mat.euc.adonis <- adonis2(h2_KO_aminoacids.pca.mat.euc ~ Host_Genus, h2_KO_aminoacids.pca.varmet, perm=999)
h2_KO_aminoacids.pca.mat.euc.adonis


# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
homo.list <- h2_KO_aminoacids.pca.varmet %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  pull(individual_id)

pan.list <- h2_KO_aminoacids.pca.varmet %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  pull(individual_id)

gorilla.list <- h2_KO_aminoacids.pca.varmet %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  pull(individual_id)

alouatta.list <- h2_KO_aminoacids.pca.varmet %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  pull(individual_id)


centroids.aminoacid.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_aminoacids.pca.mat.euc, homo.list, pan.list))
centroids.aminoacid.KO_L4
centroids.aminoacid.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_aminoacids.pca.mat.euc, homo.list, gorilla.list))
centroids.aminoacid.KO_L4
centroids.aminoacid.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_aminoacids.pca.mat.euc, homo.list, alouatta.list))
centroids.aminoacid.KO_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.aminoacid.KO_L4 <- data.frame(Distance = c(11.29433,11.02668,16.05626),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_aminoacids.pca.mat.euc, h2_KO_aminoacids.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '11.29433'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '11.02668'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '16.05626'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.aminoacids <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Amino acid pathway orthologs")

individual_centroids.list_plot.aminoacids
```

```{r pca_ko_carbs}
# Carbohydrates
kegg_carb_kos <- fread("../04-analysis/screening/humann2/tables/kegg_carbohydrates.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_carbs <- distinct(kegg_carb_kos, KOrtholog)
unique_KOs_carbs <- str_c(unique_KOs_carbs$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG carbohydrate paths
humann2_KEGG_carbs <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_carbs))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB"))  %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

carbs.kegg <- humann2_KEGG_carbs %>%
  gather("KOrtholog","Counts",2:ncol(.)) %>%
  spread(individual_id,Counts) %>%
  select(KOrtholog) %>%
  inner_join(., kegg_carb_kos) 

# write.table(carbs.kegg, "../04-analysis/screening/humann2/tables/carbs.kegg.tsv", sep = "\t", quote = F, row.names = F)

rownames(humann2_KEGG_carbs) <- humann2_KEGG_carbs$individual_id
humann2_KEGG_carbs <- humann2_KEGG_carbs %>% select(-individual_id)
humann2_KEGG_carbs <- humann2_KEGG_carbs + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_carbs <- tune.pca(humann2_KEGG_carbs, logratio = 'CLR')
tune.h2_KO_carbs

# perform a PCA to see how the data cluster
h2_KO_carbs.pca <- mixOmics::pca(humann2_KEGG_carbs, ncomp = 3, logratio = 'CLR')
plot(h2_KO_carbs.pca)

h2_KO_carbs.pca.variates <- as.data.frame(h2_KO_carbs.pca$variates$X)
h2_KO_carbs.pca.variates <- h2_KO_carbs.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_carbs.pca.varmet <- inner_join(h2_KO_carbs.pca.variates, fullmetadata, by = "individual_id")

h2_KO_carbs.pca.varmet$Host_Common <- factor(h2_KO_carbs.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_carbs_pc1pc2 <- ggplot(as.data.frame(h2_KO_carbs.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_carbs.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_carbs.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Carbohydrate pathway orthologs")

pca_h2_KO_carbs_pc1pc2

# ggsave("../pca_h2_KO_carbs_pc1pc2.pdf", plot = pca_h2_KO_carbs_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_carbs.pca.mat <- as.data.frame.matrix(h2_KO_carbs.pca$X)

# Euclidean distance
h2_KO_carbs.pca.mat.euc <- vegdist(h2_KO_carbs.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_carbs.pca.mat.euc.adonis <- adonis2(h2_KO_carbs.pca.mat.euc ~ Host_Genus, h2_KO_carbs.pca.varmet, perm=999)
h2_KO_carbs.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.carbs.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_carbs.pca.mat.euc, homo.list, pan.list))
centroids.carbs.KO_L4
centroids.carbs.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_carbs.pca.mat.euc, homo.list, gorilla.list))
centroids.carbs.KO_L4
centroids.carbs.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_carbs.pca.mat.euc, homo.list, alouatta.list))
centroids.carbs.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_carbs.pca.mat.euc, h2_KO_carbs.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '12.70941'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '12.65105'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '19.1258'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.carbs <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Carbohydrate pathway orthologs")

individual_centroids.list_plot.carbs
```

```{r pca_ko_cofvit}
# Cofactors and vitamins
kegg_cofvit_kos <- fread("../04-analysis/screening/humann2/tables/kegg_cofactors_vitamins.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_cofvit <- distinct(kegg_cofvit_kos, KOrtholog)
unique_KOs_cofvit <- str_c(unique_KOs_cofvit$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG Cofactors and Vitamins paths
humann2_KEGG_cofvit <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_cofvit))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(humann2_KEGG_cofvit) <- humann2_KEGG_cofvit$individual_id
humann2_KEGG_cofvit <- humann2_KEGG_cofvit %>% select(-individual_id)
humann2_KEGG_cofvit <- humann2_KEGG_cofvit + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_cofvit <- tune.pca(humann2_KEGG_cofvit, logratio = 'CLR')
tune.h2_KO_cofvit

# perform a PCA to see how the data cluster
h2_KO_cofvit.pca <- mixOmics::pca(humann2_KEGG_cofvit, ncomp = 3, logratio = 'CLR')
plot(h2_KO_cofvit.pca)

h2_KO_cofvit.pca.variates <- as.data.frame(h2_KO_cofvit.pca$variates$X)
h2_KO_cofvit.pca.variates <- h2_KO_cofvit.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_cofvit.pca.varmet <- inner_join(h2_KO_cofvit.pca.variates, fullmetadata, by = "individual_id")

h2_KO_cofvit.pca.varmet$Host_Common <- factor(h2_KO_cofvit.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_cofvit_pc1pc2 <- ggplot(as.data.frame(h2_KO_cofvit.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_cofvit.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_cofvit.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Cofactors and vitamins pathway orthologs")

pca_h2_KO_cofvit_pc1pc2

# ggsave("../pca_h2_KO_cofvit_pc1pc2.pdf", plot = pca_h2_KO_cofvit_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_cofvit.pca.mat <- as.data.frame.matrix(h2_KO_cofvit.pca$X)

# Euclidean distance
h2_KO_cofvit.pca.mat.euc <- vegdist(h2_KO_cofvit.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_cofvit.pca.mat.euc.adonis <- adonis2(h2_KO_cofvit.pca.mat.euc ~ Host_Genus, h2_KO_cofvit.pca.varmet, perm=999)
h2_KO_cofvit.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.cofvit.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_cofvit.pca.mat.euc, homo.list, pan.list))
centroids.cofvit.KO_L4
centroids.cofvit.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_cofvit.pca.mat.euc, homo.list, gorilla.list))
centroids.cofvit.KO_L4
centroids.cofvit.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_cofvit.pca.mat.euc, homo.list, alouatta.list))
centroids.cofvit.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_cofvit.pca.mat.euc, h2_KO_cofvit.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '8.671386'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '10.58335'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '15.32876'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.cofvit <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Cofactors and vitamins pathway orthologs")

individual_centroids.list_plot.cofvit
```

```{r pca_ko_energy}
# read in the files with KOs for the pathways we'll look at
# Energy
kegg_energy_kos <- fread("../04-analysis/screening/humann2/tables/kegg_energy.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_energy <- distinct(kegg_energy_kos, KOrtholog)
unique_KOs_energy <- str_c(unique_KOs_energy$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG Energy paths
humann2_KEGG_energy <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_energy))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(humann2_KEGG_energy) <- humann2_KEGG_energy$individual_id
humann2_KEGG_energy <- humann2_KEGG_energy %>% select(-individual_id)
humann2_KEGG_energy <- humann2_KEGG_energy + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_energy <- tune.pca(humann2_KEGG_energy, logratio = 'CLR')
tune.h2_KO_energy

# perform a PCA to see how the data cluster
h2_KO_energy.pca <- mixOmics::pca(humann2_KEGG_energy, ncomp = 3, logratio = 'CLR')
plot(h2_KO_energy.pca)

h2_KO_energy.pca.variates <- as.data.frame(h2_KO_energy.pca$variates$X)
h2_KO_energy.pca.variates <- h2_KO_energy.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_energy.pca.varmet <- inner_join(h2_KO_energy.pca.variates, fullmetadata, by = "individual_id")

h2_KO_energy.pca.varmet$Host_Common <- factor(h2_KO_energy.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_energy_pc1pc2 <- ggplot(as.data.frame(h2_KO_energy.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_energy.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_energy.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Energy pathway orthologs")

pca_h2_KO_energy_pc1pc2

# ggsave("../pca_h2_KO_energy_pc1pc2.pdf", plot = pca_h2_KO_energy_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_energy.pca.mat <- as.data.frame.matrix(h2_KO_energy.pca$X)

# Euclidean distance
h2_KO_energy.pca.mat.euc <- vegdist(h2_KO_energy.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_energy.pca.mat.euc.adonis <- adonis2(h2_KO_energy.pca.mat.euc ~ Host_Genus, h2_KO_energy.pca.varmet, perm=999)
h2_KO_energy.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.energy.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_energy.pca.mat.euc, homo.list, pan.list))
centroids.energy.KO_L4
centroids.energy.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_energy.pca.mat.euc, homo.list, gorilla.list))
centroids.energy.KO_L4
centroids.energy.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_energy.pca.mat.euc, homo.list, alouatta.list))
centroids.energy.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_energy.pca.mat.euc, h2_KO_energy.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '9.783343'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '10.83156'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '13.38561'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.energy <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Energy pathway orthologs")

individual_centroids.list_plot.energy
```

```{r pca_ko_glycans}
# Glycans
kegg_glycans_kos <- fread("../04-analysis/screening/humann2/tables/kegg_glycans.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_glycans <- distinct(kegg_glycans_kos, KOrtholog)
unique_KOs_glycans <- str_c(unique_KOs_glycans$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG Glycans paths
humann2_KEGG_glycans <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_glycans))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(humann2_KEGG_glycans) <- humann2_KEGG_glycans$individual_id
humann2_KEGG_glycans <- humann2_KEGG_glycans %>% select(-individual_id)
humann2_KEGG_glycans <- humann2_KEGG_glycans + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_glycans <- tune.pca(humann2_KEGG_glycans, logratio = 'CLR')
tune.h2_KO_glycans

# perform a PCA to see how the data cluster
h2_KO_glycans.pca <- mixOmics::pca(humann2_KEGG_glycans, ncomp = 3, logratio = 'CLR')
plot(h2_KO_glycans.pca)

h2_KO_glycans.pca.variates <- as.data.frame(h2_KO_glycans.pca$variates$X)
h2_KO_glycans.pca.variates <- h2_KO_glycans.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_glycans.pca.varmet <- inner_join(h2_KO_glycans.pca.variates, fullmetadata, by = "individual_id")

h2_KO_glycans.pca.varmet$Host_Common <- factor(h2_KO_glycans.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_glycans_pc1pc2 <- ggplot(as.data.frame(h2_KO_glycans.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_glycans.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_glycans.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Glycans pathway orthologs")

pca_h2_KO_glycans_pc1pc2

# ggsave("../pca_h2_KO_glycans_pc1pc2.pdf", plot = pca_h2_KO_glycans_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_glycans.pca.mat <- as.data.frame.matrix(h2_KO_glycans.pca$X)

# Euclidean distance
h2_KO_glycans.pca.mat.euc <- vegdist(h2_KO_glycans.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_glycans.pca.mat.euc.adonis <- adonis2(h2_KO_glycans.pca.mat.euc ~ Host_Genus, h2_KO_glycans.pca.varmet, perm=999)
h2_KO_glycans.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.glycans.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_glycans.pca.mat.euc, homo.list, pan.list))
centroids.glycans.KO_L4
centroids.glycans.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_glycans.pca.mat.euc, homo.list, gorilla.list))
centroids.glycans.KO_L4
centroids.glycans.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_glycans.pca.mat.euc, homo.list, alouatta.list))
centroids.glycans.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_glycans.pca.mat.euc, h2_KO_glycans.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '5.765331'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '5.578288'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '9.658647'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.glycans <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Carbohydrate pathway orthologs")

individual_centroids.list_plot.glycans
```

```{r pca_ko_lipids}
# Lipids
kegg_lipids_kos <- fread("../04-analysis/screening/humann2/tables/kegg_lipids.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_lipids <- distinct(kegg_lipids_kos, KOrtholog)
unique_KOs_lipids <- str_c(unique_KOs_lipids$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG Lipids paths
humann2_KEGG_lipids <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_lipids))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

lipids.kegg <- humann2_KEGG_lipids %>%
  gather("KOrtholog","Counts",2:ncol(.)) %>%
  spread(individual_id,Counts) %>%
  select(KOrtholog) %>%
  inner_join(., kegg_lipids_kos) 

# write.table(lipids.kegg, "../04-analysis/screening/humann2/tables/lipids.kegg.tsv", sep = "\t", quote = F, row.names = F)

rownames(humann2_KEGG_lipids) <- humann2_KEGG_lipids$individual_id
humann2_KEGG_lipids <- humann2_KEGG_lipids %>% select(-individual_id)
humann2_KEGG_lipids <- humann2_KEGG_lipids + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_lipids <- tune.pca(humann2_KEGG_lipids, logratio = 'CLR')
tune.h2_KO_lipids

# perform a PCA to see how the data cluster
h2_KO_lipids.pca <- mixOmics::pca(humann2_KEGG_lipids, ncomp = 3, logratio = 'CLR')
plot(h2_KO_lipids.pca)

h2_KO_lipids.pca.variates <- as.data.frame(h2_KO_lipids.pca$variates$X)
h2_KO_lipids.pca.variates <- h2_KO_lipids.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_lipids.pca.varmet <- inner_join(h2_KO_lipids.pca.variates, fullmetadata, by = "individual_id")

h2_KO_lipids.pca.varmet$Host_Common <- factor(h2_KO_lipids.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_lipids_pc1pc2 <- ggplot(as.data.frame(h2_KO_lipids.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_lipids.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_lipids.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Lipids pathway orthologs")

pca_h2_KO_lipids_pc1pc2

ggsave("../pca_h2_KO_lipids_pc1pc2.pdf", plot = pca_h2_KO_lipids_pc1pc2, device = "pdf",
       scale = 1, width = 7, height = 5, units = c("in"),
       dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_lipids.pca.mat <- as.data.frame.matrix(h2_KO_lipids.pca$X)

# Euclidean distance
h2_KO_lipids.pca.mat.euc <- vegdist(h2_KO_lipids.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_lipids.pca.mat.euc.adonis <- adonis2(h2_KO_lipids.pca.mat.euc ~ Host_Genus, h2_KO_lipids.pca.varmet, perm=999)
h2_KO_lipids.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.lipids.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_lipids.pca.mat.euc, homo.list, pan.list))
centroids.lipids.KO_L4
centroids.lipids.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_lipids.pca.mat.euc, homo.list, gorilla.list))
centroids.lipids.KO_L4
centroids.lipids.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_lipids.pca.mat.euc, homo.list, alouatta.list))
centroids.lipids.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_lipids.pca.mat.euc, h2_KO_lipids.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '7.343378'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '7.497114'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '10.80786'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.lipids <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Lipids pathway orthologs")

individual_centroids.list_plot.lipids
```

```{r pca_ko_membrane}
# Membrane transport
kegg_membrane_kos <- fread("../04-analysis/screening/humann2/tables/kegg_membrane_transport.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_membrane <- distinct(kegg_membrane_kos, KOrtholog)
unique_KOs_membrane <- str_c(unique_KOs_membrane$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG Membrane transport paths
humann2_KEGG_membrane <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_membrane))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(humann2_KEGG_membrane) <- humann2_KEGG_membrane$individual_id
humann2_KEGG_membrane <- humann2_KEGG_membrane %>% select(-individual_id)
humann2_KEGG_membrane <- humann2_KEGG_membrane + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_membrane <- tune.pca(humann2_KEGG_membrane, logratio = 'CLR')
tune.h2_KO_membrane

# perform a PCA to see how the data cluster
h2_KO_membrane.pca <- mixOmics::pca(humann2_KEGG_membrane, ncomp = 3, logratio = 'CLR')
plot(h2_KO_membrane.pca)

h2_KO_membrane.pca.variates <- as.data.frame(h2_KO_membrane.pca$variates$X)
h2_KO_membrane.pca.variates <- h2_KO_membrane.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_membrane.pca.varmet <- inner_join(h2_KO_membrane.pca.variates, fullmetadata, by = "individual_id")

h2_KO_membrane.pca.varmet$Host_Common <- factor(h2_KO_membrane.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_membrane_pc1pc2 <- ggplot(as.data.frame(h2_KO_membrane.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_membrane.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_membrane.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Membrane pathway orthologs")

pca_h2_KO_membrane_pc1pc2

# ggsave("../pca_h2_KO_membrane_pc1pc2.pdf", plot = pca_h2_KO_membrane_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_membrane.pca.mat <- as.data.frame.matrix(h2_KO_membrane.pca$X)

# Euclidean distance
h2_KO_membrane.pca.mat.euc <- vegdist(h2_KO_membrane.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_membrane.pca.mat.euc.adonis <- adonis2(h2_KO_membrane.pca.mat.euc ~ Host_Genus, h2_KO_membrane.pca.varmet, perm=999)
h2_KO_membrane.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.membrane.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_membrane.pca.mat.euc, homo.list, pan.list))
centroids.membrane.KO_L4
centroids.membrane.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_membrane.pca.mat.euc, homo.list, gorilla.list))
centroids.membrane.KO_L4
centroids.membrane.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_membrane.pca.mat.euc, homo.list, alouatta.list))
centroids.membrane.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_membrane.pca.mat.euc, h2_KO_membrane.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '11.00175'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '12.41601'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '20.89751'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.membrane <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Membrane pathway orthologs")

individual_centroids.list_plot.membrane
```

```{r pca_ko_motility}
# Motility
kegg_motility_kos <- fread("../04-analysis/screening/humann2/tables/kegg_motility.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_motility <- distinct(kegg_motility_kos, KOrtholog)
unique_KOs_motility <- str_c(unique_KOs_motility$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG Motility paths
humann2_KEGG_motility <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_motility))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(humann2_KEGG_motility) <- humann2_KEGG_motility$individual_id
humann2_KEGG_motility <- humann2_KEGG_motility %>% select(-individual_id)
humann2_KEGG_motility <- humann2_KEGG_motility + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_motility <- tune.pca(humann2_KEGG_motility, logratio = 'CLR')
tune.h2_KO_motility

# perform a PCA to see how the data cluster
h2_KO_motility.pca <- mixOmics::pca(humann2_KEGG_motility, ncomp = 3, logratio = 'CLR')
plot(h2_KO_motility.pca)

h2_KO_motility.pca.variates <- as.data.frame(h2_KO_motility.pca$variates$X)
h2_KO_motility.pca.variates <- h2_KO_motility.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_motility.pca.varmet <- inner_join(h2_KO_motility.pca.variates, fullmetadata, by = "individual_id")

h2_KO_motility.pca.varmet$Host_Common <- factor(h2_KO_motility.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_motility_pc1pc2 <- ggplot(as.data.frame(h2_KO_motility.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_motility.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_motility.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Motility pathway orthologs")

pca_h2_KO_motility_pc1pc2

# ggsave("../pca_h2_KO_motility_pc1pc2.pdf", plot = pca_h2_KO_motility_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_motility.pca.mat <- as.data.frame.matrix(h2_KO_motility.pca$X)

# Euclidean distance
h2_KO_motility.pca.mat.euc <- vegdist(h2_KO_motility.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_motility.pca.mat.euc.adonis <- adonis2(h2_KO_motility.pca.mat.euc ~ Host_Genus, h2_KO_motility.pca.varmet, perm=999)
h2_KO_motility.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.motility.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_motility.pca.mat.euc, homo.list, pan.list))
centroids.motility.KO_L4
centroids.motility.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_motility.pca.mat.euc, homo.list, gorilla.list))
centroids.motility.KO_L4
centroids.motility.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_motility.pca.mat.euc, homo.list, alouatta.list))
centroids.motility.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_motility.pca.mat.euc, h2_KO_motility.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '3.354747'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '3.103071'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '4.076227'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.motility <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Motility pathway orthologs")

individual_centroids.list_plot.motility
```

```{r pca_ko_nucleotide}
# Nucleotides
kegg_nucleotides_kos <- fread("../04-analysis/screening/humann2/tables/kegg_nucleotides.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_nucleotide <- distinct(kegg_nucleotides_kos, KOrtholog)
unique_KOs_nucleotide <- str_c(unique_KOs_nucleotide$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG Nucleotides paths
humann2_KEGG_nucleotide <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_nucleotide))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(humann2_KEGG_nucleotide) <- humann2_KEGG_nucleotide$individual_id
humann2_KEGG_nucleotide <- humann2_KEGG_nucleotide %>% select(-individual_id)
humann2_KEGG_nucleotide <- humann2_KEGG_nucleotide + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_nucleotide <- tune.pca(humann2_KEGG_nucleotide, logratio = 'CLR')
tune.h2_KO_nucleotide

# perform a PCA to see how the data cluster
h2_KO_nucleotide.pca <- mixOmics::pca(humann2_KEGG_nucleotide, ncomp = 4, logratio = 'CLR')
plot(h2_KO_nucleotide.pca)

h2_KO_nucleotide.pca.variates <- as.data.frame(h2_KO_nucleotide.pca$variates$X)
h2_KO_nucleotide.pca.variates <- h2_KO_nucleotide.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_nucleotide.pca.varmet <- inner_join(h2_KO_nucleotide.pca.variates, fullmetadata, by = "individual_id")

h2_KO_nucleotide.pca.varmet$Host_Common <- factor(h2_KO_nucleotide.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_nucleotide_pc1pc2 <- ggplot(as.data.frame(h2_KO_nucleotide.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_nucleotide.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_nucleotide.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Nucleotides pathway orthologs")

pca_h2_KO_nucleotide_pc1pc2

# ggsave("../pca_h2_KO_nucleotide_pc1pc2.pdf", plot = pca_h2_KO_nucleotide_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_nucleotide.pca.mat <- as.data.frame.matrix(h2_KO_nucleotide.pca$X)

# Euclidean distance
h2_KO_nucleotide.pca.mat.euc <- vegdist(h2_KO_nucleotide.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_nucleotide.pca.mat.euc.adonis <- adonis2(h2_KO_nucleotide.pca.mat.euc ~ Host_Genus, h2_KO_nucleotide.pca.varmet, perm=999)
h2_KO_nucleotide.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.nucleotide.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_nucleotide.pca.mat.euc, homo.list, pan.list))
centroids.nucleotide.KO_L4
centroids.nucleotide.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_nucleotide.pca.mat.euc, homo.list, gorilla.list))
centroids.nucleotide.KO_L4
centroids.nucleotide.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_nucleotide.pca.mat.euc, homo.list, alouatta.list))
centroids.nucleotide.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_nucleotide.pca.mat.euc, h2_KO_nucleotide.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '5.98507'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '5.694573'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '8.717973'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.nucleotide <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Nucleotides pathway orthologs")

individual_centroids.list_plot.nucleotide
```

```{r pca_ko_replication}
# Replication and DNA repair
kegg_replic_kos <- fread("../04-analysis/screening/humann2/tables/kegg_replication_repair.tsv")

# list out the unique KOs to pull from the full genefamilies table
unique_KOs_replic <- distinct(kegg_replic_kos, KOrtholog)
unique_KOs_replic <- str_c(unique_KOs_replic$KOrtholog, collapse = "|")

# select out the KEGG orthologs that are involved in the KEGG Cofactors and Vitamins paths
humann2_KEGG_replic <- gene_families.humann2 %>%
                                         filter(!str_detect(Pathways, "\\|")) %>%
                                         filter(str_detect(Pathways, unique_KOs_replic))  %>%
                                         select(-one_of(outliersH2_KOs)) %>%
                                         select(-one_of(sources_calc)) %>%
                                         select(-starts_with("EXB")) %>%
                                         select(-starts_with("LIB")) %>%
                                         adorn_totals(where = "col", name = "Sum_KOs") %>%
                                         mutate(Percent = Sum_KOs/sum(Sum_KOs)*100) %>%
                                         filter(Percent > 0.05) %>%
                                         select(-one_of(c("Sum_KOs","Percent"))) %>%
                                         gather("individual_id","CPM",2:ncol(.)) %>%
                                         spread(Pathways,CPM)

rownames(humann2_KEGG_replic) <- humann2_KEGG_replic$individual_id
humann2_KEGG_replic <- humann2_KEGG_replic %>% select(-individual_id)
humann2_KEGG_replic <- humann2_KEGG_replic + 1

# prepare to run a PCA
# check the number of components to retain by tuning the PCA
tune.h2_KO_replic <- tune.pca(humann2_KEGG_replic, logratio = 'CLR')
tune.h2_KO_replic

# perform a PCA to see how the data cluster
h2_KO_replic.pca <- mixOmics::pca(humann2_KEGG_replic, ncomp = 3, logratio = 'CLR')
plot(h2_KO_replic.pca)

h2_KO_replic.pca.variates <- as.data.frame(h2_KO_replic.pca$variates$X)
h2_KO_replic.pca.variates <- h2_KO_replic.pca.variates %>%
  rownames_to_column("individual_id")
  
h2_KO_replic.pca.varmet <- inner_join(h2_KO_replic.pca.variates, fullmetadata, by = "individual_id")

h2_KO_replic.pca.varmet$Host_Common <- factor(h2_KO_replic.pca.varmet$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                            "Plaque", "Gut", "Skin", "Sediment","EnvironmentalControl","ExtractionControl", "LibraryControl"))

pca_h2_KO_replic_pc1pc2 <- ggplot(as.data.frame(h2_KO_replic.pca.varmet), aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(h2_KO_replic.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(h2_KO_replic.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 7) + 
  ggtitle("Replication & DNA repair pathway orthologs")

pca_h2_KO_replic_pc1pc2

# ggsave("../pca_h2_KO_replic_pc1pc2.pdf", plot = pca_h2_KO_replic_pc1pc2, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# use adonis to test for significant differences between groups
h2_KO_replic.pca.mat <- as.data.frame.matrix(h2_KO_replic.pca$X)

# Euclidean distance
h2_KO_replic.pca.mat.euc <- vegdist(h2_KO_replic.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
h2_KO_replic.pca.mat.euc.adonis <- adonis2(h2_KO_replic.pca.mat.euc ~ Host_Genus, h2_KO_replic.pca.varmet, perm=999)
h2_KO_replic.pca.mat.euc.adonis

# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.replication.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_replic.pca.mat.euc, homo.list, pan.list))
centroids.replication.KO_L4
centroids.replication.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_replic.pca.mat.euc, homo.list, gorilla.list))
centroids.replication.KO_L4
centroids.replication.KO_L4 <- as.data.frame(dist_between_centroids(h2_KO_replic.pca.mat.euc, homo.list, alouatta.list))
centroids.replication.KO_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(h2_KO_replic.pca.mat.euc, h2_KO_replic.pca.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., individual_id = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(individual_id,Host_Genus,CentroidGroup,CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['xaxis'] <- c(1:59)
individual_centroids.homo['yaxis'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['xaxis'] <- c(61:80)
individual_centroids.pan['yaxis'] <- '4.032495'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['xaxis'] <- c(82:106)
individual_centroids.gorilla['yaxis'] <- '3.761379'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['xaxis'] <- c(108:112)
individual_centroids.alouatta['yaxis'] <- '7.119739'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$CentroidDistance <- as.numeric(individual_centroids.list$CentroidDistance)
individual_centroids.list$yaxis <- as.numeric(individual_centroids.list$yaxis)

# Plot
individual_centroids.list_plot.replication <- ggplot(individual_centroids.list, aes(x=xaxis, y=yaxis, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=yaxis-CentroidDistance, ymax=yaxis+CentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) + 
  ggtitle("Replication & DNA repair pathway orthologs")

individual_centroids.list_plot.replication
```
