---
title: "Deep Evolution Functional Annotation and Analysis with AADDER"
output: html_notebook
---
# Preparation

Start by using decontam to remove the SEED categories that appear to come from contaminants, using the program decontam. 
First load the libraries you'll use
```{r}
library("data.table")
library(decontam)
library(psych)
library(mixOmics)
library(vegan)
library(reshape2)
library(usedist)
library(janitor)
library(tidyverse)
library(phyloseq)
library(gplots)
library(ggrepel)
library(viridis)

```

Set the working directory of the notebook to the directory of the notebook itself.  

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath("."))
```

Read in the data and metadata tables
```{r read_tables}
# read in the full AADDER table and change the first column title so it doesn't start with #
fullset.aadder <- fread("../04-analysis/screening/aadder/tables/AADDER_refseq_seed_85pc_supp0.01_20190513-ex_seedPath_to_count_summarised.txt.gz")
fullset.aadder <- dplyr::rename(fullset.aadder, Category = `#Datasets`)

# Fix the column headers
colnames(fullset.aadder) <- gsub("_S.*_L.*_R1_.*.fastq.merged.prefixed.hg19unmapped.blastn.sam.gz","", colnames(fullset.aadder))
colnames(fullset.aadder) <- gsub("_S.*_L.*_R1_.*.fastq.combined.fq.prefixed.extractunmapped.bam.blastn.sam.gz","", colnames(fullset.aadder))
colnames(fullset.aadder) <- gsub("_S.*_L.*_R1_.*.fastq.merged.blastn.sam.gz","", colnames(fullset.aadder))

# transpose the data
decontam_freq_in.aadder <- fullset.aadder %>%
  gather("Sample","Counts", 2:ncol(.)) %>%
  spread(Category, Counts)

# read in the metadata
fullmetadata <- read_tsv("../00-documentation.backup/02-calculus_microbiome-deep_evolution-individualscontrolssources_metadata_20190523.tsv")
fullmetadata <- rename(fullmetadata, SampleID = `#SampleID`)

filtermetadata <- fullmetadata %>%
  as.data.frame(.) %>%
  filter(!grepl("SRR|ERR", SampleID)) %>%
  filter(SumLibraryConcentration_copies_ul > 0)

```

# Decontamination with Decontam
Prepare the matrix for decontam
```{r decontam_data_prep}
# list out the samples that have all 0 or NA values to remove from the input matrix
NA_filter_list <- fullmetadata %>% as.data.frame(.) %>% filter((is.na(SumLibraryConcentration_copies_ul) | SumLibraryConcentration_copies_ul <= 0)) %>% select(SampleID) %>% rename(Sample = SampleID)

# remove NA value samples from the input data frame
aadder.dataframe <- decontam_freq_in.aadder %>%
  anti_join(., NA_filter_list) %>%
  gather("Category","Counts", 2:ncol(.)) %>%
  spread(Sample, Counts) %>%
  select(-LIB025.A0301)

# Select only the SEED entries that go down to level 4, b/c this is the specific protein/enzyme entry
aadder.dataframe <- aadder.dataframe[str_count(aadder.dataframe$Category, ";") == 4, ]
aadder.dataframe <- aadder.dataframe %>%
  gather("Sample","Counts", 2:ncol(.)) %>%
  spread(Category, Counts)

# make a matrix for decontam
aadder_matrix <- data.matrix(aadder.dataframe[2:ncol(aadder.dataframe)])
rownames(aadder_matrix) <- aadder.dataframe$Sample

metadataframe <- data.frame(filtermetadata)
rownames(metadataframe) <- filtermetadata$SampleID
final_metadata <- sample_data(metadataframe)

```

This selected out many more proteins. But check the combined prevalence and
frequency method to see how it differs from the individual methods.
```{r decontam_combined}
# Assign sample-type status to identify control samples (same as was done for prevalence)
final_metadata$is.control <- final_metadata$Sample_or_Control=="Control"

# Assign contaminant status
contam.comb.dna <- isContaminant(aadder_matrix, method="combined", 
                                 conc=final_metadata$SumLibraryConcentration_copies_ul, 
                                 neg=final_metadata$is.control,
                                 threshold = 0.6)

## Check how many contaminants
table(contam.comb.dna$contaminant)

## List of contaminants
cont_comb_dna_list <- contam.comb.dna[which(contam.comb.dna$contaminant=="TRUE"), ]

# Graph it to look at the histogram and pick a suitable probability-values cut-off
ggplot(contam.comb.dna, aes(x=p)) +
  geom_histogram(bins=100) +
  scale_y_log10() +
  geom_hline(yintercept=10, linetype="dashed", color = "red") +
  geom_hline(yintercept=2, linetype="dashed", color = "red") +
  ggtitle('Combined')

```

Well who knows. Both Prevalence (threshold = 0.7) and Combined (threshold =
0.925) cut out similar pathways. Unfortunately we don't have a way to check what
is being removed based on classification like we do with taxonomy, where we have
some idea of which species/genera are realistically expected in an oral sample.
Those threshold values cut out many that fall in heme, sporulation, etc
(categories that might be of interest to explore because of associations with
biofilm maturity or metabolism)...so try combined w/ threshold = 0.6 (slightly
higher than default 0.5). Those paths that are removed at combined threshold =
0.95 but aren't removed with combined threshold = 0.06 are removed later by
filtering low-abundance paths anyway.
List out the pathways that are considered contaminants, to be removed from the
table prior to all analyses. Then remove them from the table to make the
decontaminated input table for analysis.
```{r}
contaminants <- cont_comb_dna_list %>% rownames_to_column("Category") %>% select(Category)
aadder.decontam <- anti_join(fullset.aadder, contaminants)
```

## Sample Clustering and Outlier Removal
Visualization of sample clustering based on removing the SEED categories
identified as contaminants and further filtering very low-abundance categories
(protein/enzyme level). Perform PCAs on unfiltered data followed by the data
filtered in steps, starting with removing decontam-identified contaminants,
plotting PCAs for level 2, level 3, and level 4 categories, and then filtering
out the very low-abundance categories (<0.05% relative abundance) and plotting
PCAs for level 2, level 3, and level 4 categories.

First set the colors and shapes for PCA plots, based on the metadata categories 
Host Genus, Host Common, and Environment. 
```{r colors_shapes}
# set the colors and shapes for PCA plots
env_colours <- c("#1f78b4", 
                 "#6a3d9a", 
                 "#6a3d9a", 
                 "#6a3d9a", 
                 "#33a02c", 
                 "#33a02c", 
                 "#33a02c", 
                 "#33a02c", 
                 "#e31a1c", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#ff7f00", 
                 "#d9d9d9", 
                 "#d9d9d9",
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9", 
                 "#d9d9d9",
                 "#d9d9d9")

names(env_colours) <- c("Howler_Monkey", 
                        "Gorilla_1", 
                        "Gorilla_2", 
                        "Gorilla_3", 
                        "Chimp_1", 
                        "Chimp_2", 
                        "Chimp_3", 
                        "Chimp_4",
                        "Neanderthal", 
                        "PreagriculturalHuman_1", 
                        "PreagriculturalHuman_2", 
                        "PreantibioticHuman_1", 
                        "PreantibioticHuman_2", 
                        "ModernDayHuman_1", 
                        "ModernDayHuman_2", 
                        "ExtractionControl", 
                        "LibraryControl", 
                        "ruralGut", 
                        "urbanGut",
                        "sediment", 
                        "skin", 
                        "subPlaque",
                        "supPlaque",
                        "EnvironmentalControl")

env_shapes <- c(8,
                0,
                1,
                2,
                0,
                1,
                2,
                5,
                11,
                0,
                12,
                1,
                10,
                2,
                6,
                10,
                13,
                7,
                8,
                14,
                3,
                4,
                12,
                0)


names(env_shapes) <- c("Howler_Monkey", 
                       "Gorilla_1", 
                       "Gorilla_2", 
                       "Gorilla_3", 
                       "Chimp_1", 
                       "Chimp_2", 
                       "Chimp_3", 
                       "Chimp_4",
                       "Neanderthal", 
                       "PreagriculturalHuman_1", 
                       "PreagriculturalHuman_2", 
                       "PreantibioticHuman_1", 
                       "PreantibioticHuman_2", 
                       "ModernDayHuman_1", 
                       "ModernDayHuman_2", 
                       "ExtractionControl", 
                       "LibraryControl", 
                       "ruralGut", 
                       "sediment", 
                       "skin", 
                       "subPlaque",
                       "supPlaque", 
                       "urbanGut",
                       "EnvironmentalControl")

common_colours <- c(Alouatta = "#1f78b4", Gorilla = "#6a3d9a", Pan = "#33a02c", 
                    `Homo (Neanderthal)` = "#fdbf6f", 
                    `Homo (Modern Human)` = "#ff7f00", ExtractionControl = "#d9d9d9", 
                    LibraryControl = "#d9d9d9", Plaque = "#d9d9d9", Gut = "#d9d9d9", 
                    Skin = "#d9d9d9", Sediment = "#d9d9d9", EnvironmentalControl = "#d9d9d9")

common_shapes <- c(Alouatta = 8, Gorilla = 0, Pan = 1, 
                   `Homo (Neanderthal)` = 2, `Homo (Modern Human)` = 6,
                   ExtractionControl = 3, LibraryControl = 13, Plaque = 9, 
                   Gut = 7, Skin = 14, Sediment = 10, EnvironmentalControl = 12)


genus_colours <- c(Alouatta = "#1f78b4", Gorilla = "#6a3d9a", Pan = "#33a02c", 
                   Homo = "#ff7f00",  ExtractionBlank = "#d9d9d9", 
                   LibraryBlank ="#d9d9d9", Plaque = "#fdbf6f", Gut = "#d9d9d9", 
                   Skin = "#d9d9d9", Sediment = "#d9d9d9", Control = "#d9d9d9")

genus_shapes <- c(Alouatta = 8, Gorilla = 0, Pan = 1,Homo = 6,
                  ExtractionBlank = 3, LibraryBlank = 13, Plaque = 9, 
                   Gut = 7, Skin = 14, Sediment = 10, Control = 12)

general_colors <- c(Howler = "#1f78b4", Gorilla = "#6a3d9a", Chimp = "#33a02c", 
                    Neanderthal = "#ff7f00", PreagriculturalHuman = "#ff7f00",
                    PreantibioticHuman = "#ff7f00", ModernDayHuman = "#ff7f00", ExtractionControl = "#d9d9d9", 
                    LibraryControl = "#d9d9d9", plaque = "#d9d9d9", gut = "#d9d9d9", 
                    skin = "#d9d9d9", sediment = "#d9d9d9", EnvironmentalControl = "#d9d9d9")

general_shapes <- c(Howler = 8, Gorilla = 0, Chimp = 1, 
                   Neanderthal = 2, PreagriculturalHuman = 0,
                   PreantibioticHuman = 1, ModernDayHuman = 6,
                   ExtractionControl = 3, LibraryControl = 13, plaque = 14, 
                   gut = 7, skin = 9, sediment = 10, EnvironmentalControl = 12)

# define the oral samples
oralSamples <- fullmetadata %>%
  filter(str_detect(Is_Oral,"Calculus|Plaque")) %>%
  select(SampleID)

# define the calculus samples
calcSamples <- fullmetadata %>%
  filter(str_detect(Is_Oral,"Calculus")) %>%
  select(SampleID)

```


## Level 4 SEED classifications (enzymes)
Now b/c SEED levels are cumulative we'll repeat all of these plots with only L4,
the enzymes. This is the data we'll be using for the paper. So we'll make plots
with only L4 entries for all sample groups, all sample groups decontaminated,
all sample groups decontaminated and filtered for only >0.05%, and all sample
groups decontaminated and filtered and outliers removed.
Start with L4 all sample groups. 
```{r full_pca}
# start with the full table and pull out only the level 4 entries
fullset.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
rownames(fullset.aadder_L4) <- fullset.aadder_L4$Category
fullset.aadder_L4 <- fullset.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(fullset.aadder_L4) <- fullset.aadder_L4$SampleID
fullset.aadder_L4 <- as.matrix(fullset.aadder_L4 %>% 
  select(-SampleID))

# add +1 to all values to be able to do a CLR transformation before running PCA
fullset.aadder_L4 <- fullset.aadder_L4+1
fullset.aadder_L4.pca <- mixOmics::pca(fullset.aadder_L4, ncomp = 3, logratio = 'CLR')
fullset.aadder_L4.pca
plot(fullset.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
fullset.aadder_L4.variates <- as.data.frame(fullset.aadder_L4.pca$variates$X)
fullset.aadder_L4.variates <- fullset.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
fullset.aadder_L4.varmet <- inner_join(fullset.aadder_L4.variates, fullmetadata, by = "SampleID")

# set metadata as factors for plotting
fullset.aadder_L4.varmet$Host_Common <- factor(fullset.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_fullset.aadder_L4_pc1pc2 <- ggplot(fullset.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  #geom_text(aes(label = fullset.aadder_L4.varmet$SampleID), size = 2) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(fullset.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(fullset.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 unfiltered PC1-PC2")

pca_fullset.aadder_L4_pc1pc2

# ggsave("../pca_fullset.aadder_L4_pc1pc2.pdf", plot = pca_fullset.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

```


Now L4 all sample groups decontaminated. *Use this section to determine the*
*outlier samples, see the note in italics above the section below for*
*explanation.*
```{r decontam_pca}
#start by creating a L4 only table like in the last step
decontam.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.aadder_L4 <- anti_join(decontam.aadder_L4, contaminants)
rownames(decontam.aadder_L4) <- decontam.aadder_L4$Category
decontam.aadder_L4 <- decontam.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.aadder_L4) <- decontam.aadder_L4$SampleID
decontam.aadder_L4 <- as.matrix(decontam.aadder_L4 %>% 
  select(-SampleID))

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.aadder_L4 <- decontam.aadder_L4+1
decontam.aadder_L4.pca <- mixOmics::pca(decontam.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.aadder_L4.pca
plot(decontam.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.aadder_L4.variates <- as.data.frame(decontam.aadder_L4.pca$variates$X)
decontam.aadder_L4.variates <- decontam.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.aadder_L4.varmet <- inner_join(decontam.aadder_L4.variates, fullmetadata, by = "SampleID")

# set metadata as factors for plotting
decontam.aadder_L4.varmet$Host_Common <- factor(decontam.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.aadder_L4_pc1pc2 <- ggplot(decontam.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  #geom_text(aes(label = decontam.aadder_L4.varmet$SampleID), size = 2) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated PC1-PC2")

pca_decontam.aadder_L4_pc1pc2

# ggsave("../pca_decontam.aadder_L4_pc1pc2.pdf", plot = pca_decontam.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

# with names
pca_decontam.aadder_L4_pc1pc2_names <- ggplot(decontam.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_text(aes(label = decontam.aadder_L4.varmet$SampleID), size = 2) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated PC1-PC2")

pca_decontam.aadder_L4_pc1pc2_names

```

Now L4 all sample groups decontaminated and filter for only >0.05% abundance.
Use this to identify outliers to remove in the next step. *Actually this seems*
*to be making the separation between samples and controls much less distinct,*
*making it more difficult to determine which samples are outliers. Instead of*
*filtering out enzymes <0.05% abundance, we'll keep all enzymes and determine the*
*outliers from the dataset where only the decontam-identified contaminants have*
*been removed.*
```{r decontam_filter_pca}
# make the decontaminated table like in the last step and then filter out enzymes with <0.05% abundance
decontam.filtered.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.filtered.aadder_L4 <- anti_join(decontam.filtered.aadder_L4, contaminants)
rownames(decontam.filtered.aadder_L4) <- decontam.filtered.aadder_L4$Category

decontam.filtered.aadder_L4 <- decontam.filtered.aadder_L4 %>% 
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.01) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.filtered.aadder_L4) <- decontam.filtered.aadder_L4$SampleID
decontam.filtered.aadder_L4 <- as.matrix(decontam.filtered.aadder_L4 %>% 
  select(-SampleID))

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.filtered.aadder_L4 <- decontam.filtered.aadder_L4+1
decontam.filtered.aadder_L4.pca <- mixOmics::pca(decontam.filtered.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.filtered.aadder_L4.pca
plot(decontam.filtered.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.filtered.aadder_L4.variates <- as.data.frame(decontam.filtered.aadder_L4.pca$variates$X)
decontam.filtered.aadder_L4.variates <- decontam.filtered.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.filtered.aadder_L4.varmet <- inner_join(decontam.filtered.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.filtered.aadder_L4.varmet <- decontam.filtered.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.filtered.aadder_L4.varmet$Host_Common <- factor(decontam.filtered.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.filtered.aadder_L4_pc1pc2 <- ggplot(decontam.filtered.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.filtered.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.filtered.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated, filtered PC1-PC2")

pca_decontam.filtered.aadder_L4_pc1pc2

# ggsave("../pca_decontam.filtered.aadder_L4_pc1pc2.pdf", plot = pca_decontam.filtered.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

# identify the outliers by the names
pca_decontam.filtered.aadder_L4_pc1pc2_names <- ggplot(decontam.filtered.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_text(aes(label = decontam.filtered.aadder_L4.varmet$SampleID), size = 2) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.filtered.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.filtered.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated PC1-PC2")

pca_decontam.filtered.aadder_L4_pc1pc2_names

# identify the outliers by the names
pca_decontam.filtered.aadder_L4_pc1pc3_names <- ggplot(decontam.filtered.aadder_L4.varmet, 
                                            aes(PC3, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_text(aes(label = decontam.filtered.aadder_L4.varmet$SampleID), size = 2) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.filtered.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC3 - ", round(decontam.filtered.aadder_L4.pca$explained_variance[3], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated, filtered PC2-PC3")

pca_decontam.filtered.aadder_L4_pc1pc3_names

```

Now L4 all sample groups decontaminated and filtered for only >0.05% abundance
and remove outliers. Use the table generated here to select out just the
Carbohydrate category enzymes for the next step.
```{r decontam_filter_2_pca}
# list the outliers from the last PCA
outliers_L4 <- c("DLV002.A0101","KNP004.A0102","TAF016.B0101","BAN001.A0101","DLV001.A0101",
                 "SpyNew","ECO010.B0101","ECO006.B0101","GUE001.A0101","SPM001.A0101",
                 "CDC007.A0101","WAL001.A0101","LOB001.A0101","BIT001.A0101","MTM013.A0101",
                 "MTS003.A0101","MTM006.A0101","FUM003.A0101","MTM012.A0101","MTM003.A0101",
                 "IBA002.A0101","OAK001.A0101","MTM002.A0101","SPM002.A0101","CDC005.A0101",
                 "OAK001.A0101","DJA006.A0101","MTM009.A0101","BSQ001.A0101","ZAF001.A0101")
outliers_L4 <- str_c(outliers_L4, collapse = "|")

# make the decontaminated table like in the last step and then filter out outliers
decontam.outrem.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.aadder_L4 <- anti_join(decontam.outrem.aadder_L4, contaminants)
rownames(decontam.outrem.aadder_L4) <- decontam.outrem.aadder_L4$Category

decontam.outrem.aadder_L4 <- decontam.outrem.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4))


rownames(decontam.outrem.aadder_L4) <- decontam.outrem.aadder_L4$SampleID
decontam.outrem.aadder_L4 <- as.matrix(decontam.outrem.aadder_L4 %>% 
  select(-SampleID))

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.aadder_L4 <- decontam.outrem.aadder_L4+1
decontam.outrem.aadder_L4.pca <- mixOmics::pca(decontam.outrem.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.aadder_L4.pca
plot(decontam.outrem.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.aadder_L4.variates <- as.data.frame(decontam.outrem.aadder_L4.pca$variates$X)
decontam.outrem.aadder_L4.variates <- decontam.outrem.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.aadder_L4.varmet <- inner_join(decontam.outrem.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.aadder_L4.varmet <- decontam.outrem.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated, outliers removed PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

# identify the outliers by the names
pca_decontam.outrem.aadder_L4_pc1pc2_names <- ggplot(decontam.outrem.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_text(aes(label = decontam.outrem.aadder_L4.varmet$SampleID), size = 2) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated, outliers removed PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.aadder_L4_pc1pc2_names
```

Now L4 only oral samples (calculus and plaque) decontaminated and remove
outliers. Now we will filter out all enzymes with <0.05% relative abundance.
```{r oral_pca}
# make the decontaminated table like in the last step and then filter out outliers
decontam.outrem.oral.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.oral.aadder_L4 <- anti_join(decontam.outrem.oral.aadder_L4, contaminants)
rownames(decontam.outrem.oral.aadder_L4) <- decontam.outrem.oral.aadder_L4$Category

decontam.outrem.oral.aadder_L4 <- decontam.outrem.oral.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., oralSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)


rownames(decontam.outrem.oral.aadder_L4) <- decontam.outrem.oral.aadder_L4$SampleID
decontam.outrem.oral.aadder_L4 <- as.matrix(decontam.outrem.oral.aadder_L4 %>% 
  select(-SampleID))

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.oral.aadder_L4 <- decontam.outrem.oral.aadder_L4+1
decontam.outrem.oral.aadder_L4.pca <- mixOmics::pca(decontam.outrem.oral.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.oral.aadder_L4.pca
plot(decontam.outrem.oral.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.oral.aadder_L4.variates <- as.data.frame(decontam.outrem.oral.aadder_L4.pca$variates$X)
decontam.outrem.oral.aadder_L4.variates <- decontam.outrem.oral.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.oral.aadder_L4.varmet <- inner_join(decontam.outrem.oral.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.oral.aadder_L4.varmet <- decontam.outrem.oral.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.oral.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.oral.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.oral.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.oral.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.oral.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.oral.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated, outrem, filtered - oral - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.oral.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.oral.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.oral.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

```

Now L4 only calculus samples decontaminated and remove outliers. We will filter
out all enzymes with <0.05% relative abundance here as well. This helps with the
bi-plots to get the arrows out of a bundled cluster near the origin so we can
distinguish them. Otherwise there are too many variables and the contribution 
from each is very small.
```{r calculus_pca}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers, filter for >0.05
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)


rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$SampleID
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% 
  select(-SampleID))

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.aadder_L4.pca
plot(decontam.outrem.calc.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.aadder_L4.pca$variates$X)
decontam.outrem.calc.aadder_L4.variates <- decontam.outrem.calc.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.aadder_L4.varmet <- inner_join(decontam.outrem.calc.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.aadder_L4.varmet <- decontam.outrem.calc.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 decontaminated, outrem, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.calc.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

pca_decontam.outrem.calc.aadder_L4_pc1pc2_ellipse <- ggplot(decontam.outrem.calc.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 decontaminated, outrem, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.aadder_L4_pc1pc2_ellipse

# use adonis to test for significant differences between groups
decontam.outrem.calc.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera

# list the samples in each host genus for calculating centroid distance
homo.list <- decontam.outrem.calc.aadder_L4.varmet %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  pull(SampleID)

pan.list <- decontam.outrem.calc.aadder_L4.varmet %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  pull(SampleID)

gorilla.list <- decontam.outrem.calc.aadder_L4.varmet %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  pull(SampleID)

alouatta.list <- decontam.outrem.calc.aadder_L4.varmet %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  pull(SampleID)

# calculate distance between centroids of Homo and each of the other genera
centroids.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.aadder_L4
centroids.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.aadder_L4
centroids.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.aadder_L4 <- data.frame(Distance = c(13.44366,10.9097,16.31472),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.aadder_L4.pca.mat.euc, decontam.outrem.calc.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '13.44366'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '10.9097'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '16.31472'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("Aadder L4 decontaminated, outrem, filtered - calculus - PC1")

individual_centroids.list_plot

```

## SEED category PCAs
Now L4 *Amino Acid* category enzymes calculus samples decontaminated and
filtered for only >0.05% abundance and remove outliers. Use the table generated
here to select out just the selected category enzymes for the next step.
```{r L4_aminoacids}
decontam.outrem.calc.aminoacid.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aminoacid.aadder_L4 <- anti_join(decontam.outrem.calc.aminoacid.aadder_L4, contaminants)

decontam.outrem.calc.aminoacid.aadder_L4 <- decontam.outrem.calc.aminoacid.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.aminoacid.aadder_L4) <- decontam.outrem.calc.aminoacid.aadder_L4$SampleID
decontam.outrem.calc.aminoacid.aadder_L4 <- decontam.outrem.calc.aminoacid.aadder_L4 %>% select(contains("Amino Acid"))

aminoacids.list <- decontam.outrem.calc.aminoacid.aadder_L4 %>% 
   rownames_to_column("SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  mutate(protein = sapply(Category, function(f) {
  unlist(str_split(f, ";"))[4]
  })) %>%
  select(protein)

# write.table(aminoacids.list, "../04-analysis/screening/humann2/tables/aminoacids.seed.tsv", sep = "\t", row.names = F)

decontam.outrem.calc.aminoacid.aadder_L4 <- as.matrix(decontam.outrem.calc.aminoacid.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.aminoacid.aadder_L4 <- decontam.outrem.calc.aminoacid.aadder_L4+1
decontam.outrem.calc.aminoacid.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.aminoacid.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.aminoacid.aadder_L4.pca
plot(decontam.outrem.calc.aminoacid.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.aminoacid.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.aminoacid.aadder_L4.pca$variates$X)
decontam.outrem.calc.aminoacid.aadder_L4.variates <- decontam.outrem.calc.aminoacid.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.aminoacid.aadder_L4.varmet <- inner_join(decontam.outrem.calc.aminoacid.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.aminoacid.aadder_L4.varmet <- decontam.outrem.calc.aminoacid.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.aminoacid.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.aminoacid.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))

decontam.outrem.calc.aminoacid.aadder_L4.varmet$Host_Genus <- factor(decontam.outrem.calc.aminoacid.aadder_L4.varmet$Host_Genus, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo"))

# Plot the PCA
pca_decontam.outrem.calc.aminoacid.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.aminoacid.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.aminoacid.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.aminoacid.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  # stat_ellipse() + # inherit.aes=F, aes(PC1,PC2, color = Host_Genus, shape = Host_Genus)
  ggtitle("Aadder L4 Amino Acids decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.aminoacid.aadder_L4_pc1pc2

ggsave("../pca_decontam.outrem.calc.aminoacid.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.aminoacid.aadder_L4_pc1pc2, device = "pdf",
  scale = 1, width = 7, height = 5, units = c("in"),
  dpi = 300)

pca_decontam.outrem.calc.aminoacid.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.aminoacid.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.aminoacid.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.aminoacid.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() + # inherit.aes=F, aes(PC1,PC2, color = Host_Genus, shape = Host_Genus)
  ggtitle("Aadder L4 Amino Acids decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.aminoacid.aadder_L4_pc1pc2

# use adonis to test for significant differences between groups
decontam.outrem.calc.aminoacid.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.aminoacid.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.aminoacid.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.aminoacid.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
decontam.outrem.calc.aminoacid.aadder_L4.pca.mat.euc.adonis <- adonis2(decontam.outrem.calc.aminoacid.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.aminoacid.aadder_L4.varmet, perm=999)
decontam.outrem.calc.aminoacid.aadder_L4.pca.mat.euc.adonis


# try one-dimensional plots of Euclidian distances between cluster centroids
# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.aminoacid.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.aminoacid.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.aminoacid.aadder_L4
centroids.aminoacid.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.aminoacid.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.aminoacid.aadder_L4
centroids.aminoacid.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.aminoacid.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.aminoacid.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.aminoacid.aadder_L4 <- data.frame(Distance = c(4.931635,2.902647,5.64902),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.aminoacid.aadder_L4.pca.mat.euc, decontam.outrem.calc.aminoacid.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '4.931635'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '2.902647'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '5.64902'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.aminoacids <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Amino acids")

individual_centroids.list_plot.aminoacids
```

Now L4 *Carbohydrate* category enzymes calculus samples decontaminated and
filtered for only >0.05% abundance and remove outliers. Use the table generated
here to select out just the Carbohydrate category enzymes for the next step.
```{r L4_carbohydrate}
decontam.outrem.calc.carb.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.carb.aadder_L4 <- anti_join(decontam.outrem.calc.carb.aadder_L4, contaminants)

decontam.outrem.calc.carb.aadder_L4 <- decontam.outrem.calc.carb.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.carb.aadder_L4) <- decontam.outrem.calc.carb.aadder_L4$SampleID
decontam.outrem.calc.carb.aadder_L4 <- decontam.outrem.calc.carb.aadder_L4 %>% select(contains("Carbohydrates"))

carbs.list <- decontam.outrem.calc.carb.aadder_L4 %>% 
   rownames_to_column("SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  mutate(protein = sapply(Category, function(f) {
  unlist(str_split(f, ";"))[4]
  })) %>%
  select(protein)

# write.table(carbs.list, "../04-analysis/screening/humann2/tables/carbs.seed.tsv", sep = "\t", row.names = F)

decontam.outrem.calc.carb.aadder_L4 <- as.matrix(decontam.outrem.calc.carb.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.carb.aadder_L4 <- decontam.outrem.calc.carb.aadder_L4+1
decontam.outrem.calc.carb.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.carb.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.carb.aadder_L4.pca
plot(decontam.outrem.calc.carb.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.carb.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.carb.aadder_L4.pca$variates$X)
decontam.outrem.calc.carb.aadder_L4.variates <- decontam.outrem.calc.carb.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.carb.aadder_L4.varmet <- inner_join(decontam.outrem.calc.carb.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.carb.aadder_L4.varmet <- decontam.outrem.calc.carb.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.carb.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.carb.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.carb.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.carb.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.carb.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.carb.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 Carbohydrates decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.carb.aadder_L4_pc1pc2

ggsave("../pca_decontam.outrem.calc.carb.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.carb.aadder_L4_pc1pc2, device = "pdf",
  scale = 1, width = 7, height = 5, units = c("in"),
  dpi = 300)

pca_decontam.outrem.calc.carb.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.carb.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.carb.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.carb.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 Carbohydrates decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.carb.aadder_L4_pc1pc2

# use adonis to test for significant differences between groups
decontam.outrem.calc.carb.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.carb.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.carb.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.carb.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.carb.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.carb.aadder_L4.varmet, perm=999)


# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.carb.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.carb.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.carb.aadder_L4
centroids.carb.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.carb.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.carb.aadder_L4
centroids.carb.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.carb.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.carb.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.carb.aadder_L4 <- data.frame(Distance = c(6.344419,6.3488,7.733213),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.carb.aadder_L4.pca.mat.euc, decontam.outrem.calc.carb.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids 

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '6.344419'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '6.3488'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '7.733213'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.carbs <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Carbohydrates")

individual_centroids.list_plot.carbs

```


Now L4 *Cofactors and Vitamins* category enzymes calculus samples decontaminated
and filtered for only >0.05% abundance and remove outliers. Use the table
generated here to select out just the Carbohydrate category enzymes for the next
step.
```{r L4_cofactors_vitamins}
decontam.outrem.calc.cofvit.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.cofvit.aadder_L4 <- anti_join(decontam.outrem.calc.cofvit.aadder_L4, contaminants)

decontam.outrem.calc.cofvit.aadder_L4 <- decontam.outrem.calc.cofvit.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.cofvit.aadder_L4) <- decontam.outrem.calc.cofvit.aadder_L4$SampleID
decontam.outrem.calc.cofvit.aadder_L4 <- decontam.outrem.calc.cofvit.aadder_L4 %>% select(contains("Cofactors"))
decontam.outrem.calc.cofvit.aadder_L4 <- as.matrix(decontam.outrem.calc.cofvit.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.cofvit.aadder_L4 <- decontam.outrem.calc.cofvit.aadder_L4+1
decontam.outrem.calc.cofvit.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.cofvit.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.cofvit.aadder_L4.pca
plot(decontam.outrem.calc.cofvit.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.cofvit.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.cofvit.aadder_L4.pca$variates$X)
decontam.outrem.calc.cofvit.aadder_L4.variates <- decontam.outrem.calc.cofvit.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.cofvit.aadder_L4.varmet <- inner_join(decontam.outrem.calc.cofvit.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.cofvit.aadder_L4.varmet <- decontam.outrem.calc.cofvit.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.cofvit.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.cofvit.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.cofvit.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.cofvit.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.cofvit.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.cofvit.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 Cofactors and Vitamins decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.cofvit.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.calc.cofvit.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.cofvit.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

pca_decontam.outrem.calc.cofvit.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.cofvit.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.cofvit.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.cofvit.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 Cofactors and Vitamins decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.cofvit.aadder_L4_pc1pc2


# use adonis to test for significant differences between groups
decontam.outrem.calc.cofvit.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.cofvit.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.cofvit.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.cofvit.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.cofvit.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.cofvit.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.cofvit.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.cofvit.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.cofvit.aadder_L4
centroids.cofvit.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.cofvit.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.cofvit.aadder_L4
centroids.cofvit.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.cofvit.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.cofvit.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.cofvit.aadder_L4 <- data.frame(Distance = c(4.931635,2.902647,5.64902),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.cofvit.aadder_L4.pca.mat.euc, decontam.outrem.calc.cofvit.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '3.255958'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '2.719241'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '4.277399'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.cofvits <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Cofactors and Vitamins")

individual_centroids.list_plot.cofvits
```

*Won't run*
Now L4 *Energy* category enzymes calculus samples decontaminated and filtered
for only >0.05% abundance and remove outliers. Use the table generated here to
select out just the Carbohydrate category enzymes for the next step.
```{r L4_evergy, eval=F}
decontam.outrem.calc.energy.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.energy.aadder_L4 <- anti_join(decontam.outrem.calc.energy.aadder_L4, contaminants)

decontam.outrem.calc.energy.aadder_L4 <- decontam.outrem.calc.energy.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.energy.aadder_L4) <- decontam.outrem.calc.energy.aadder_L4$SampleID
decontam.outrem.calc.energy.aadder_L4 <- decontam.outrem.calc.energy.aadder_L4 %>% select(contains("Energy"))
decontam.outrem.calc.energy.aadder_L4 <- as.matrix(decontam.outrem.calc.energy.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.energy.aadder_L4 <- decontam.outrem.calc.energy.aadder_L4+1
decontam.outrem.calc.energy.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.energy.aadder_L4, ncomp = 2, logratio = 'CLR')
decontam.outrem.calc.energy.aadder_L4.pca
plot(decontam.outrem.calc.energy.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.energy.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.energy.aadder_L4.pca$variates$X)
decontam.outrem.calc.energy.aadder_L4.variates <- decontam.outrem.calc.energy.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.energy.aadder_L4.varmet <- inner_join(decontam.outrem.calc.energy.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.energy.aadder_L4.varmet <- decontam.outrem.calc.energy.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.energy.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.energy.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.energy.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.energy.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.energy.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.energy.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 Energy decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.energy.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.calc.energy.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.energy.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

pca_decontam.outrem.calc.energy.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.energy.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.energy.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.energy.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 Energy decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.energy.aadder_L4_pc1pc2


# use adonis to test for significant differences between groups
decontam.outrem.calc.energy.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.energy.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.energy.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.energy.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.energy.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.energy.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.energy.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.energy.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.energy.aadder_L4
centroids.energy.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.energy.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.energy.aadder_L4
centroids.energy.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.energy.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.energy.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.energy.aadder_L4 <- data.frame(Distance = c(4.931635,2.902647,5.64902),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.energy.aadder_L4.pca.mat.euc, decontam.outrem.calc.energy.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '4.931635'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '2.902647'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '5.64902'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.energys <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Energy")

individual_centroids.list_plot.energys
```


Now L4 *Cell Wall and Capsule* category enzymes calculus samples decontaminated
and filtered for only >0.05% abundance and remove outliers. Use the table
generated here to select out just the Carbohydrate category enzymes for the next
step.
```{r L4_cellwall}
decontam.outrem.calc.cellwall.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.cellwall.aadder_L4 <- anti_join(decontam.outrem.calc.cellwall.aadder_L4, contaminants)

decontam.outrem.calc.cellwall.aadder_L4 <- decontam.outrem.calc.cellwall.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.cellwall.aadder_L4) <- decontam.outrem.calc.cellwall.aadder_L4$SampleID
decontam.outrem.calc.cellwall.aadder_L4 <- decontam.outrem.calc.cellwall.aadder_L4 %>% select(contains("Wall"))
decontam.outrem.calc.cellwall.aadder_L4 <- as.matrix(decontam.outrem.calc.cellwall.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.cellwall.aadder_L4 <- decontam.outrem.calc.cellwall.aadder_L4+1
decontam.outrem.calc.cellwall.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.cellwall.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.cellwall.aadder_L4.pca
plot(decontam.outrem.calc.cellwall.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.cellwall.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.cellwall.aadder_L4.pca$variates$X)
decontam.outrem.calc.cellwall.aadder_L4.variates <- decontam.outrem.calc.cellwall.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.cellwall.aadder_L4.varmet <- inner_join(decontam.outrem.calc.cellwall.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.cellwall.aadder_L4.varmet <- decontam.outrem.calc.cellwall.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.cellwall.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.cellwall.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.cellwall.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.cellwall.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.cellwall.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.cellwall.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 Cell Wall and Capsule decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.cellwall.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.calc.cellwall.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.cellwall.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

pca_decontam.outrem.calc.cellwall.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.cellwall.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.cellwall.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.cellwall.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 Cell Wall and Capsule decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.cellwall.aadder_L4_pc1pc2


# use adonis to test for significant differences between groups
decontam.outrem.calc.cellwall.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.cellwall.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.cellwall.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.cellwall.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.cellwall.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.cellwall.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.cellwall.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.cellwall.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.cellwall.aadder_L4
centroids.cellwall.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.cellwall.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.cellwall.aadder_L4
centroids.cellwall.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.cellwall.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.cellwall.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.cellwall.aadder_L4 <- data.frame(Distance = c(4.931635,2.902647,5.64902),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.cellwall.aadder_L4.pca.mat.euc, decontam.outrem.calc.cellwall.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '1.161767'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '2.563526'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '3.409023'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.cellwalls <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours)  +
  ggtitle("SEED - Cell Wall")

individual_centroids.list_plot.cellwalls
```


Now L4 *Fatty Acids, Lipids, and Isoprenoids* category enzymes calculus samples
decontaminated and filtered for only >0.05% abundance and remove outliers. Use
the table generated here to select out just the Carbohydrate category enzymes
for the next step.
```{r L4_fattyacids}
decontam.outrem.calc.fattyacids.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.fattyacids.aadder_L4 <- anti_join(decontam.outrem.calc.fattyacids.aadder_L4, contaminants)

decontam.outrem.calc.fattyacids.aadder_L4 <- decontam.outrem.calc.fattyacids.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.fattyacids.aadder_L4) <- decontam.outrem.calc.fattyacids.aadder_L4$SampleID
decontam.outrem.calc.fattyacids.aadder_L4 <- decontam.outrem.calc.fattyacids.aadder_L4 %>% select(contains("Lipids"))

fattyacids.list <- decontam.outrem.calc.fattyacids.aadder_L4 %>% 
   rownames_to_column("SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  mutate(protein = sapply(Category, function(f) {
  unlist(str_split(f, ";"))[4]
  })) %>%
  select(protein)

# write.table(fattyacids.list, "../04-analysis/screening/humann2/tables/fattyacids.seed.tsv", sep = "\t", row.names = F)

decontam.outrem.calc.fattyacids.aadder_L4 <- as.matrix(decontam.outrem.calc.fattyacids.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.fattyacids.aadder_L4 <- decontam.outrem.calc.fattyacids.aadder_L4+1
decontam.outrem.calc.fattyacids.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.fattyacids.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.fattyacids.aadder_L4.pca
plot(decontam.outrem.calc.fattyacids.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.fattyacids.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.fattyacids.aadder_L4.pca$variates$X)
decontam.outrem.calc.fattyacids.aadder_L4.variates <- decontam.outrem.calc.fattyacids.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.fattyacids.aadder_L4.varmet <- inner_join(decontam.outrem.calc.fattyacids.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.fattyacids.aadder_L4.varmet <- decontam.outrem.calc.fattyacids.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.fattyacids.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.fattyacids.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.fattyacids.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.fattyacids.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.fattyacids.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.fattyacids.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 FAs, Lipids, Isoprenoids decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.fattyacids.aadder_L4_pc1pc2

ggsave("../pca_decontam.outrem.calc.fattyacids.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.fattyacids.aadder_L4_pc1pc2, device = "pdf",
  scale = 1, width = 7, height = 5, units = c("in"),
  dpi = 300)

pca_decontam.outrem.calc.fattyacids.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.fattyacids.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.fattyacids.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.fattyacids.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 FAs, Lipids, Isoprenoids decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.fattyacids.aadder_L4_pc1pc2


# use adonis to test for significant differences between groups
decontam.outrem.calc.fattyacids.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.fattyacids.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.fattyacids.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.fattyacids.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.fattyacids.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.fattyacids.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.fattyacids.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.fattyacids.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.fattyacids.aadder_L4
centroids.fattyacids.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.fattyacids.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.fattyacids.aadder_L4
centroids.fattyacids.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.fattyacids.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.fattyacids.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.fattyacids.aadder_L4 <- data.frame(Distance = c(4.931635,2.902647,5.64902),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.fattyacids.aadder_L4.pca.mat.euc, decontam.outrem.calc.fattyacids.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '1.064002'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '1.152541'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '1.385385'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.fattyacids <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Fatty acids, Lipids, Isoprenoids")

individual_centroids.list_plot.fattyacids
```


Now L4 *Membrane Transport* category enzymes calculus samples decontaminated and
filtered for only >0.05% abundance and remove outliers. Use the table generated
here to select out just the Carbohydrate category enzymes for the next step.
```{r L4_membrane}
decontam.outrem.calc.membrane.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.membrane.aadder_L4 <- anti_join(decontam.outrem.calc.membrane.aadder_L4, contaminants)

decontam.outrem.calc.membrane.aadder_L4 <- decontam.outrem.calc.membrane.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.membrane.aadder_L4) <- decontam.outrem.calc.membrane.aadder_L4$SampleID
decontam.outrem.calc.membrane.aadder_L4 <- decontam.outrem.calc.membrane.aadder_L4 %>% select(contains("Membrane"))
decontam.outrem.calc.membrane.aadder_L4 <- as.matrix(decontam.outrem.calc.membrane.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.membrane.aadder_L4 <- decontam.outrem.calc.membrane.aadder_L4+1
decontam.outrem.calc.membrane.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.membrane.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.membrane.aadder_L4.pca
plot(decontam.outrem.calc.membrane.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.membrane.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.membrane.aadder_L4.pca$variates$X)
decontam.outrem.calc.membrane.aadder_L4.variates <- decontam.outrem.calc.membrane.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.membrane.aadder_L4.varmet <- inner_join(decontam.outrem.calc.membrane.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.membrane.aadder_L4.varmet <- decontam.outrem.calc.membrane.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.membrane.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.membrane.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.membrane.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.membrane.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.membrane.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.membrane.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 Membrane Transport decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.membrane.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.calc.membrane.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.membrane.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

pca_decontam.outrem.calc.membrane.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.membrane.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.membrane.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.membrane.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse()
  ggtitle("Aadder L4 Membrane Transport decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.membrane.aadder_L4_pc1pc2


# use adonis to test for significant differences between groups
decontam.outrem.calc.membrane.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.membrane.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.membrane.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.membrane.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.membrane.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.membrane.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.membrane.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.membrane.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.membrane.aadder_L4
centroids.membrane.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.membrane.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.membrane.aadder_L4
centroids.membrane.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.membrane.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.membrane.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.membrane.aadder_L4 <- data.frame(Distance = c(4.931635,2.902647,5.64902),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.membrane.aadder_L4.pca.mat.euc, decontam.outrem.calc.membrane.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '3.264218'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '2.977712'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '3.031299'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.membrane <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Membrane")

individual_centroids.list_plot.membrane
```


Now L4 *Motility* category enzymes calculus samples decontaminated and filtered
for only >0.05% abundance and remove outliers. Use the table generated here to
select out just the Carbohydrate category enzymes for the next step.
```{r L4_motility}
decontam.outrem.calc.motility.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.motility.aadder_L4 <- anti_join(decontam.outrem.calc.motility.aadder_L4, contaminants)

decontam.outrem.calc.motility.aadder_L4 <- decontam.outrem.calc.motility.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.motility.aadder_L4) <- decontam.outrem.calc.motility.aadder_L4$SampleID
decontam.outrem.calc.motility.aadder_L4 <- decontam.outrem.calc.motility.aadder_L4 %>% select(contains("Motility"))
decontam.outrem.calc.motility.aadder_L4 <- as.matrix(decontam.outrem.calc.motility.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.motility.aadder_L4 <- decontam.outrem.calc.motility.aadder_L4+1
decontam.outrem.calc.motility.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.motility.aadder_L4, ncomp = 2, logratio = 'CLR')
decontam.outrem.calc.motility.aadder_L4.pca
plot(decontam.outrem.calc.motility.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.motility.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.motility.aadder_L4.pca$variates$X)
decontam.outrem.calc.motility.aadder_L4.variates <- decontam.outrem.calc.motility.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.motility.aadder_L4.varmet <- inner_join(decontam.outrem.calc.motility.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.motility.aadder_L4.varmet <- decontam.outrem.calc.motility.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2)

# set metadata as factors for plotting
decontam.outrem.calc.motility.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.motility.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.motility.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.motility.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.motility.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.motility.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 Motility decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.motility.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.calc.motility.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.motility.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

pca_decontam.outrem.calc.motility.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.motility.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.motility.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.motility.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 Motility decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.motility.aadder_L4_pc1pc2


# use adonis to test for significant differences between groups
decontam.outrem.calc.motility.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.motility.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.motility.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.motility.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.motility.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.motility.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.motility.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.motility.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.motility.aadder_L4
centroids.motility.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.motility.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.motility.aadder_L4
centroids.motility.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.motility.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.motility.aadder_L4

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.motility.aadder_L4.pca.mat.euc, decontam.outrem.calc.motility.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '0.5415764'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '0.1175704'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '0.05705411'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.motility <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Motility")

individual_centroids.list_plot.motility
```


Now L4 *Nucleotides and Nucleosides, Nucleotide sugars* category enzymes
calculus samples decontaminated and filtered for only >0.05% abundance and
remove outliers. Use the table generated here to select out just the
Carbohydrate category enzymes for the next step.

```{r L4_nucleotides}
decontam.outrem.calc.nucleotide.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.nucleotide.aadder_L4 <- anti_join(decontam.outrem.calc.nucleotide.aadder_L4, contaminants)

decontam.outrem.calc.nucleotide.aadder_L4 <- decontam.outrem.calc.nucleotide.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.nucleotide.aadder_L4) <- decontam.outrem.calc.nucleotide.aadder_L4$SampleID
decontam.outrem.calc.nucleotide.aadder_L4 <- decontam.outrem.calc.nucleotide.aadder_L4 %>% select(contains("Nucleotide"))
decontam.outrem.calc.nucleotide.aadder_L4 <- as.matrix(decontam.outrem.calc.nucleotide.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.nucleotide.aadder_L4 <- decontam.outrem.calc.nucleotide.aadder_L4+1
decontam.outrem.calc.nucleotide.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.nucleotide.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.nucleotide.aadder_L4.pca
plot(decontam.outrem.calc.nucleotide.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.nucleotide.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.nucleotide.aadder_L4.pca$variates$X)
decontam.outrem.calc.nucleotide.aadder_L4.variates <- decontam.outrem.calc.nucleotide.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.nucleotide.aadder_L4.varmet <- inner_join(decontam.outrem.calc.nucleotide.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.nucleotide.aadder_L4.varmet <- decontam.outrem.calc.nucleotide.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.nucleotide.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.nucleotide.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.nucleotide.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.nucleotide.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.nucleotide.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.nucleotide.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 Nucleotides/sides decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.nucleotide.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.calc.nucleotide.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.nucleotide.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

pca_decontam.outrem.calc.nucleotide.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.nucleotide.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.nucleotide.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.nucleotide.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 Nucleotides/sides decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.nucleotide.aadder_L4_pc1pc2


# use adonis to test for significant differences between groups
decontam.outrem.calc.nucleotide.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.nucleotide.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.nucleotide.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.nucleotide.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.nucleotide.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.nucleotide.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.nucleotide.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.nucleotide.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.nucleotide.aadder_L4
centroids.nucleotide.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.nucleotide.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.nucleotide.aadder_L4
centroids.nucleotide.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.nucleotide.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.nucleotide.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.nucleotide.aadder_L4 <- data.frame(Distance = c(4.931635,2.902647,5.64902),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.nucleotide.aadder_L4.pca.mat.euc, decontam.outrem.calc.nucleotide.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '2.073107'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '2.189981'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '2.709088'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.nucleotide <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Nucleosides, Nucleotides, Nucleotide sugars")

individual_centroids.list_plot.nucleotide
```


Now L4 *Cell Division Cell Cycle, DNA Repair* category enzymes calculus samples
decontaminated and filtered for only >0.05% abundance and remove outliers. Use
the table generated here to select out just the Carbohydrate category enzymes
for the next step.
```{r L4_celldivision}
decontam.outrem.calc.replication.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.replication.aadder_L4 <- anti_join(decontam.outrem.calc.replication.aadder_L4, contaminants)

decontam.outrem.calc.replication.aadder_L4 <- decontam.outrem.calc.replication.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts)

rownames(decontam.outrem.calc.replication.aadder_L4) <- decontam.outrem.calc.replication.aadder_L4$SampleID
decontam.outrem.calc.replication.aadder_L4 <- decontam.outrem.calc.replication.aadder_L4 %>% select(contains("Division"))
decontam.outrem.calc.replication.aadder_L4 <- as.matrix(decontam.outrem.calc.replication.aadder_L4)

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.replication.aadder_L4 <- decontam.outrem.calc.replication.aadder_L4+1
decontam.outrem.calc.replication.aadder_L4.pca <- mixOmics::pca(decontam.outrem.calc.replication.aadder_L4, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.replication.aadder_L4.pca
plot(decontam.outrem.calc.replication.aadder_L4.pca)

# Select out the PC variates into a new table and add the metadata for plotting
decontam.outrem.calc.replication.aadder_L4.variates <- as.data.frame(decontam.outrem.calc.replication.aadder_L4.pca$variates$X)
decontam.outrem.calc.replication.aadder_L4.variates <- decontam.outrem.calc.replication.aadder_L4.variates %>%
  rownames_to_column("SampleID")
  
decontam.outrem.calc.replication.aadder_L4.varmet <- inner_join(decontam.outrem.calc.replication.aadder_L4.variates, fullmetadata, by = "SampleID")
decontam.outrem.calc.replication.aadder_L4.varmet <- decontam.outrem.calc.replication.aadder_L4.varmet %>% 
  select(SampleID,Host_Genus,Host_General,Host_Common,PC1,PC2,PC3)

# set metadata as factors for plotting
decontam.outrem.calc.replication.aadder_L4.varmet$Host_Common <- factor(decontam.outrem.calc.replication.aadder_L4.varmet$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))


# Plot the PCA
pca_decontam.outrem.calc.replication.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.replication.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Common, shape = Host_Common)) +
  geom_point(size = 4) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.replication.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.replication.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  ggtitle("Aadder L4 Cell division/cycle & DNA repiar, decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.replication.aadder_L4_pc1pc2

# ggsave("../pca_decontam.outrem.calc.replication.aadder_L4_pc1pc2.pdf", plot = pca_decontam.outrem.calc.replication.aadder_L4_pc1pc2, device = "pdf",
#   scale = 1, width = 7, height = 5, units = c("in"),
#   dpi = 300)

pca_decontam.outrem.calc.replication.aadder_L4_pc1pc2 <- ggplot(decontam.outrem.calc.replication.aadder_L4.varmet, 
                                            aes(PC1, PC2, colour = Host_Genus, shape = Host_Genus)) +
  geom_point(size = 4) +
  scale_colour_manual(values = genus_colours) +
  scale_shape_manual(values = genus_shapes) +
  xlab(paste("PC1 - ", round(decontam.outrem.calc.replication.aadder_L4.pca$explained_variance[1], 2), sep = "")) +
  ylab(paste("PC2 - ", round(decontam.outrem.calc.replication.aadder_L4.pca$explained_variance[2], 2), sep = "")) +
  theme_minimal(base_size = 14) +
  stat_ellipse() +
  ggtitle("Aadder L4 Cell Wall and Capsule decontaminated, filtered - calculus - PC1-PC2") +
  theme(plot.title = element_text(size = 10))

pca_decontam.outrem.calc.replication.aadder_L4_pc1pc2


# use adonis to test for significant differences between groups
decontam.outrem.calc.replication.aadder_L4.pca.mat <- as.data.frame.matrix(decontam.outrem.calc.replication.aadder_L4.pca$X)

# Euclidean distance
decontam.outrem.calc.replication.aadder_L4.pca.mat.euc <- vegdist(decontam.outrem.calc.replication.aadder_L4.pca.mat, method = "euclidean", na.rm = T)

# need to extract the rotated data matrix and then run euclidean distance calc w/vegan
adonis(decontam.outrem.calc.replication.aadder_L4.pca.mat.euc ~ Host_Genus, decontam.outrem.calc.replication.aadder_L4.varmet, perm=999)

# Make a plot of the distances to centroids to look at overlap between Homo and the other genera
# calculate distance between centroids of Homo and each of the other genera
centroids.replication.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.replication.aadder_L4.pca.mat.euc, homo.list, pan.list))
centroids.replication.aadder_L4
centroids.replication.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.replication.aadder_L4.pca.mat.euc, homo.list, gorilla.list))
centroids.replication.aadder_L4
centroids.replication.aadder_L4 <- as.data.frame(dist_between_centroids(decontam.outrem.calc.replication.aadder_L4.pca.mat.euc, homo.list, alouatta.list))
centroids.replication.aadder_L4

# make a data frame with the Euclidian distance between homo centroid and each genus centroid
centroids.replication.aadder_L4 <- data.frame(Distance = c(4.931635,2.902647,5.64902),
                                            Genus = c("Pan","Gorilla","Alouatta"))

# make a data frame with the distances 
individual_centroids <- dist_to_centroids(decontam.outrem.calc.replication.aadder_L4.pca.mat.euc, decontam.outrem.calc.replication.aadder_L4.varmet$Host_Genus)
individual_centroids <- individual_centroids %>% rename(., SampleID = Item)
individual_centroids <- individual_centroids %>% inner_join(., fullmetadata) %>% select(SampleID,Host_Genus,CentroidGroup,CentroidDistance) %>% rename(WithinCentroidDistance = CentroidDistance)
individual_centroids

individual_centroids.homo <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(CentroidGroup, "Homo"))
individual_centroids.homo['Centroid'] <- c(1:55)
individual_centroids.homo['BetweenCentroidDistance'] <- '0'

individual_centroids.pan <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(CentroidGroup, "Pan"))
individual_centroids.pan['Centroid'] <- c(57:73)
individual_centroids.pan['BetweenCentroidDistance'] <- '1.162747'

individual_centroids.gorilla <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(CentroidGroup, "Gorilla"))
individual_centroids.gorilla['Centroid'] <- c(75:92)
individual_centroids.gorilla['BetweenCentroidDistance'] <- '0.7992966'

individual_centroids.alouatta <- individual_centroids %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(CentroidGroup, "Alouatta"))
individual_centroids.alouatta['Centroid'] <- c(94:98)
individual_centroids.alouatta['BetweenCentroidDistance'] <- '1.634786'

individual_centroids.list <- bind_rows(individual_centroids.homo, individual_centroids.pan)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.gorilla)
individual_centroids.list <- bind_rows(individual_centroids.list, individual_centroids.alouatta)
individual_centroids.list$WithinCentroidDistance <- as.numeric(individual_centroids.list$WithinCentroidDistance)
individual_centroids.list$BetweenCentroidDistance <- as.numeric(individual_centroids.list$BetweenCentroidDistance)

# Plot
individual_centroids.list_plot.replication <- ggplot(individual_centroids.list, aes(x=Centroid, y=BetweenCentroidDistance, group=Host_Genus, color=Host_Genus)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=BetweenCentroidDistance-WithinCentroidDistance, ymax=BetweenCentroidDistance+WithinCentroidDistance)) +
  theme_minimal() +
  scale_color_manual(values = genus_colours) +
  ggtitle("SEED - Cell division/cycle & DNA repair")

individual_centroids.list_plot.replication
```

## Bi-plots of L4 PCAs

Make bi-plots for PC1 on the L4 decontam, outrem pca
```{r pc1_biplots}
# Borrowed from Alex's coprolite CAZymes bi-plot script
# Extract coordinates from mixOmics object and convert into long table dataframe
decontam.outrem.calc.aadder_L4.pca_df <- as.data.frame(decontam.outrem.calc.aadder_L4.pca$x) %>%
  rownames_to_column("SampleID") %>%
  # Add meta data
  inner_join(fullmetadata, by="SampleID") %>%
  select(SampleID,Host_Genus,Host_Common,PC1,PC2,PC3)

# Extract explained variance
pca_expl_vars <- paste0(round(decontam.outrem.calc.aadder_L4.pca$explained_variance * 100, 2), "%")

# Extract loadings (eigenvectors) from PC analysis
pca_loadings_df <- as.data.frame(decontam.outrem.calc.aadder_L4.pca$loadings$X) %>%
  rownames_to_column(var = "seed_id")

# Identify the 10 pathways that have highest positive and negative loadings in PC1
pc1_pws <- pca_loadings_df %>%
  top_n(10, PC1) %>%
  mutate(Protein = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           }))

pc1_neg <- pca_loadings_df %>%
  top_n(-10, PC1) %>%
  mutate(Protein = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           }))

# write.table(pc1_pws, "../pc1_pws.tsv", sep = "\t", quote = F, row.names = F)
# write.table(pc1_neg, "../pc1_neg.tsv", sep = "\t", quote = F, row.names = F)

# Calculate correction factor for coordinate systems of principal coordinates
# so that they have the same units as the rotations of PC1 and PC2; from
# ggfortify::autoplot
corr_lam <- decontam.outrem.calc.aadder_L4.pca$sdev[c("PC1", "PC2")] * sqrt(nrow(decontam.outrem.calc.aadder_L4.pca_df))

# make Host Genus a factor for plotting
decontam.outrem.calc.aadder_L4_nomodern.pca_df$Host_Common <- factor(decontam.outrem.calc.aadder_L4_nomodern.pca_df$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Neanderthal)", "Homo (Modern Human)", "Plaque"))


# Plot the top 10 negative and positive paths that are strongest loading in PC1 
pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc1bi <- decontam.outrem.calc.aadder_L4.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc1_pws,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pc1_neg,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 6, 2, 1)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(legend.position = "top")

pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc1bi

# ggsave("../pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc1bi.pdf", plot = pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc1bi, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc1bi_label <- decontam.outrem.calc.aadder_L4.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc1_pws,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc1_pws,
                   aes(x = PC1, y = PC2, label = Protein),
                   size = 1.5, colour = "grey20") +
    geom_segment(data = pc1_neg,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc1_neg,
                   aes(x = PC1, y = PC2, label = Protein),
                   size = 1.5, colour = "grey20", force = 4) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 6, 2, 1)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(text = element_text(size = 14)) +
  theme(legend.position = "top")

pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc1bi_label

# ggsave("../pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc1bi_label.pdf", plot = pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc1bi_label, device = "pdf",
#        scale = 1, width = 7, height = 25, units = c("in"),
#        dpi = 300)

```


And make bi-plots for PC2 on the L4 decontam, outrem pca
```{r pc2_biplots}
# Borrowed from Alex's coprolite CAZymes bi-plot script
# Extract coordinates from mixOmics object and convert into long table dataframe
decontam.outrem.calc.aadder_L4.pca_df <- as.data.frame(decontam.outrem.calc.aadder_L4.pca$x) %>%
  rownames_to_column("SampleID") %>%
  # Add meta data
  inner_join(fullmetadata, by="SampleID") %>%
  select(SampleID,Host_Genus,Host_Common,PC1,PC2,PC3)

# Extract explained variance
pca_expl_vars <- paste0(round(decontam.outrem.calc.aadder_L4.pca$explained_variance * 100, 2), "%")

# Extract loadings (eigenvectors) from PC analysis
pca_loadings_df <- as.data.frame(decontam.outrem.calc.aadder_L4.pca$loadings$X) %>%
  rownames_to_column(var = "seed_id")

# Identify the 10 pathways that have highest positive and negative loadings in PC1
pc2_pws <- pca_loadings_df %>%
  top_n(10, PC2) %>%
  mutate(Protein = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           }))

pc2_neg <- pca_loadings_df %>%
  top_n(-10, PC2) %>%
  mutate(Protein = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           }))

# write.table(pc2_pws, "../pc2_pws.tsv", sep = "\t", quote = F, row.names = F)
# write.table(pc2_neg, "../pc2_neg.tsv", sep = "\t", quote = F, row.names = F)

# Calculate correction factor for coordinate systems of principal coordinates
# so that they have the same units as the rotations of PC1 and PC2; from
# ggfortify::autoplot
corr_lam <- decontam.outrem.calc.aadder_L4.pca$sdev[c("PC1", "PC2")] * sqrt(nrow(decontam.outrem.calc.aadder_L4.pca_df))

# make Host Genus a factor for plotting
decontam.outrem.calc.aadder_L4_nomodern.pca_df$Host_Common <- factor(decontam.outrem.calc.aadder_L4_nomodern.pca_df$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Neanderthal)", "Homo (Modern Human)", "Plaque"))


# Plot the top 10 negative and positive paths that are strongest loading in PC1 
pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc2bi <- decontam.outrem.calc.aadder_L4.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc2_pws,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pc2_neg,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 6, 2, 1)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(legend.position = "top")

pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc2bi

# ggsave("../pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc2bi.pdf", plot = pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc2bi, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc2bi_label <- decontam.outrem.calc.aadder_L4.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc2_pws,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc2_pws,
                   aes(x = PC1, y = PC2, label = Protein),
                   size = 2, colour = "grey20") +
    geom_segment(data = pc2_neg,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc2_neg,
                   aes(x = PC1, y = PC2, label = Protein),
                   size = 2, colour = "grey20") +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 6, 2, 1)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(legend.position = "top")

pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc2bi_label

# ggsave("../pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc2bi_label.pdf", plot = pca_decontam.outrem.calc.aadder_L4_pc1pc2_pc2bi_label, device = "pdf",
#        scale = 1, width = 17, height = 25, units = c("in"),
#        dpi = 300)

```

And we need to repeat these bi-plots without the modern human samples to ensure
the loadings aren't driven mainly by modern samples. First for PC1.
```{r pc1_biplots_nomodern}
# select out just the enzymes
decontam.outrem.calc.aadder_L4_nomodern <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4_nomodern <- anti_join(decontam.outrem.calc.aadder_L4_nomodern, contaminants)
rownames(decontam.outrem.calc.aadder_L4_nomodern) <- decontam.outrem.calc.aadder_L4_nomodern$Category

# remove outliers, filter for >0.05
decontam.outrem.calc.aadder_L4_nomodern <- decontam.outrem.calc.aadder_L4_nomodern %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,"JAE|VLC"))


rownames(decontam.outrem.calc.aadder_L4_nomodern) <- decontam.outrem.calc.aadder_L4_nomodern$SampleID
decontam.outrem.calc.aadder_L4_nomodern <- as.matrix(decontam.outrem.calc.aadder_L4_nomodern %>% 
  select(-SampleID))

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.aadder_L4_nomodern <- decontam.outrem.calc.aadder_L4_nomodern+1
decontam.outrem.calc.aadder_L4_nomodern.pca <- mixOmics::pca(decontam.outrem.calc.aadder_L4_nomodern, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.aadder_L4_nomodern.pca
plot(decontam.outrem.calc.aadder_L4_nomodern.pca)

# Borrowed from Alex's coprolite CAZymes bi-plot script
# Extract coordinates from mixOmics object and convert into long table dataframe
decontam.outrem.calc.aadder_L4_nomodern.pca_df <- as.data.frame(decontam.outrem.calc.aadder_L4_nomodern.pca$x) %>%
  rownames_to_column("SampleID") %>%
  # Add meta data
  inner_join(fullmetadata, by="SampleID") %>%
  select(SampleID,Host_Genus,Host_Common,PC1,PC2,PC3)

# Extract explained variance
pca_expl_vars <- paste0(round(decontam.outrem.calc.aadder_L4_nomodern.pca$explained_variance * 100, 2), "%")

# Extract loadings (eigenvectors) from PC analysis
pca_loadings_df <- as.data.frame(decontam.outrem.calc.aadder_L4_nomodern.pca$loadings$X) %>%
  rownames_to_column(var = "seed_id")

# Identify the 10 pathways that have highest positive and negative loadings in PC1
pc1_pws_nomod <- pca_loadings_df %>%
  top_n(10, PC1) %>%
  mutate(Protein = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           }))

pc1_neg_nomod <- pca_loadings_df %>%
  top_n(-10, PC1) %>%
  mutate(Protein = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           }))

# write.table(pc1_pws_nomod, "../pc1_pws_nomod.tsv", sep = "\t", quote = F, row.names = F)
# write.table(pc1_neg_nomod, "../pc1_neg_nomod.tsv", sep = "\t", quote = F, row.names = F)

# Calculate correction factor for coordinate systems of principal coordinates
# so that they have the same units as the rotations of PC1 and PC2; from
# ggfortify::autoplot
corr_lam <- decontam.outrem.calc.aadder_L4_nomodern.pca$sdev[c("PC1", "PC2")] * sqrt(nrow(decontam.outrem.calc.aadder_L4_nomodern.pca_df))

# make Host Genus a factor for plotting
decontam.outrem.calc.aadder_L4_nomodern.pca_df$Host_Common <- factor(decontam.outrem.calc.aadder_L4_nomodern.pca_df$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Neanderthal)", "Homo (Modern Human)", "Plaque"))


# Plot the top 10 negative and positive paths that are strongest loading in PC1 
pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc1bi <- decontam.outrem.calc.aadder_L4_nomodern.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  scale_y_reverse() +
  geom_segment(data = pc1_pws_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pc1_neg_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 1, 2, 6)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(legend.position = "top")

pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc1bi

# ggsave("../pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc1bi.pdf", plot = pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc1bi, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc1bi_label <- decontam.outrem.calc.aadder_L4_nomodern.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  scale_y_reverse() +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc1_pws_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc1_pws_nomod,
                   aes(x = PC1, y = PC2, label = Protein),
                   size = 1.5, colour = "grey20") +
    geom_segment(data = pc1_neg_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc1_neg_nomod,
                   aes(x = PC1, y = PC2, label = Protein),
                   size = 1.5, colour = "grey20", force = 4) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 1, 2, 6)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(legend.position = "top")

pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc1bi_label

# ggsave("../pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc1bi_label.pdf", plot = pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc1bi_label, device = "pdf",
#        scale = 1, width = 7, height = 25, units = c("in"),
#        dpi = 300)

```


And now we need to repeat the PC2 bi-plots without the modern human 
samples to ensure the loadings aren't driven mainly by modern samples.
```{r pc2_biplots_nomodern}
# select out just the enzymes
decontam.outrem.calc.aadder_L4_nomodern <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4_nomodern <- anti_join(decontam.outrem.calc.aadder_L4_nomodern, contaminants)
rownames(decontam.outrem.calc.aadder_L4_nomodern) <- decontam.outrem.calc.aadder_L4_nomodern$Category

# remove outliers, filter for >0.05
decontam.outrem.calc.aadder_L4_nomodern <- decontam.outrem.calc.aadder_L4_nomodern %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("Category","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  adorn_totals(where = "col", name = "Sum_enzymes") %>%
  mutate(Percent = Sum_enzymes/sum(Sum_enzymes)*100) %>%
  filter(Percent > 0.05) %>%
  select(-one_of(c("Sum_enzymes","Percent"))) %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,"JAE|VLC"))


rownames(decontam.outrem.calc.aadder_L4_nomodern) <- decontam.outrem.calc.aadder_L4_nomodern$SampleID
decontam.outrem.calc.aadder_L4_nomodern <- as.matrix(decontam.outrem.calc.aadder_L4_nomodern %>% 
  select(-SampleID))

# add +1 to all values to be able to do a CLR transformation before running PCA
decontam.outrem.calc.aadder_L4_nomodern <- decontam.outrem.calc.aadder_L4_nomodern+1
decontam.outrem.calc.aadder_L4_nomodern.pca <- mixOmics::pca(decontam.outrem.calc.aadder_L4_nomodern, ncomp = 3, logratio = 'CLR')
decontam.outrem.calc.aadder_L4_nomodern.pca
plot(decontam.outrem.calc.aadder_L4_nomodern.pca)

# Borrowed from Alex's coprolite CAZymes bi-plot script
# Extract coordinates from mixOmics object and convert into long table dataframe
decontam.outrem.calc.aadder_L4_nomodern.pca_df <- as.data.frame(decontam.outrem.calc.aadder_L4_nomodern.pca$x) %>%
  rownames_to_column("SampleID") %>%
  # Add meta data
  inner_join(fullmetadata, by="SampleID") %>%
  select(SampleID,Host_Genus,Host_Common,PC1,PC2,PC3)

# Extract explained variance
pca_expl_vars <- paste0(round(decontam.outrem.calc.aadder_L4_nomodern.pca$explained_variance * 100, 2), "%")

# Extract loadings (eigenvectors) from PC analysis
pca_loadings_df <- as.data.frame(decontam.outrem.calc.aadder_L4_nomodern.pca$loadings$X) %>%
  rownames_to_column(var = "seed_id")

# Identify the 10 pathways that have highest positive and negative loadings in PC1
pc2_pws_nomod <- pca_loadings_df %>%
  top_n(10, PC2) %>%
  mutate(Protein = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           }))

pc2_neg_nomod <- pca_loadings_df %>%
  top_n(-10, PC2) %>%
  mutate(Protein = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           }))

# write.table(pc2_pws_nomod, "../pc2_pws_nomod.tsv", sep = "\t", quote = F, row.names = F)
# write.table(pc2_neg_nomod, "../pc2_neg_nomod.tsv", sep = "\t", quote = F, row.names = F)

# Calculate correction factor for coordinate systems of principal coordinates
# so that they have the same units as the rotations of PC1 and PC2; from
# ggfortify::autoplot
corr_lam <- decontam.outrem.calc.aadder_L4_nomodern.pca$sdev[c("PC1", "PC2")] * sqrt(nrow(decontam.outrem.calc.aadder_L4_nomodern.pca_df))

# make Host Genus a factor for plotting
decontam.outrem.calc.aadder_L4_nomodern.pca_df$Host_Common <- factor(decontam.outrem.calc.aadder_L4_nomodern.pca_df$Host_Common, 
                                 levels = c("Alouatta", "Gorilla", "Pan", "Homo (Neanderthal)", "Homo (Modern Human)", "Plaque"))


# Plot the top 10 negative and positive paths that are strongest loading in PC1 
pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc2bi <- decontam.outrem.calc.aadder_L4_nomodern.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  scale_y_reverse() +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc2_pws_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pc2_neg_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 1, 2, 6)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(legend.position = "top")

pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc2bi

# ggsave("../pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc2bi.pdf", plot = pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc2bi, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc2bi_label <- decontam.outrem.calc.aadder_L4_nomodern.pca_df %>%
  mutate(PC1 = PC1 / corr_lam[1],
         PC2 = PC2 / corr_lam[2]) %>%
  ggplot(.,
         aes(x = PC1, y = PC2, colour = Host_Common)) +
  scale_y_reverse() +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
  geom_segment(data = pc2_pws_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "grey50",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc2_pws_nomod,
                   aes(x = PC1, y = PC2, label = Protein),
                   size = 1.5, colour = "grey20") +
    geom_segment(data = pc2_neg_nomod,
               aes(xend = PC1, yend = PC2),
               x = 0, y = 0, colour = "black",
               size = 0.5,
               arrow = arrow(length = unit(0.03, "npc"))) +
  geom_label_repel(data = pc2_neg_nomod,
                   aes(x = PC1, y = PC2, label = Protein),
                   size = 1.5, colour = "grey20", force = 4) +
labs(x = paste("PC1 -", pca_expl_vars[1]),
       y = paste("PC2 -", pca_expl_vars[2])) +
  scale_colour_manual(values = common_colours[c(1, 2, 3, 4, 5)]) +
  scale_shape_manual(values = c(8, 0, 1, 2, 6)) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(legend.position = "top")

pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc2bi_label

# ggsave("../pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc2bi_label.pdf", plot = pca_decontam.outrem.calc.aadder_L4_nomodern_pc1pc2_pc2bi_label, device = "pdf",
#        scale = 1, width = 7, height = 25, units = c("in"),
#        dpi = 300)

```


##Species contributions to the top 10 SEED proteins
```{r seed_contribution_files}
# read in the files from MALT, modified to have codes for the protein names b/c R can't handle the names with () - etc.
protein_code_list <- fread("../04-analysis/screening/aadder/tables/protein_code_list.tsv")
protein_code_list_nomod <- fread("../04-analysis/screening/aadder/tables/protein_code_list_nomod.tsv")
alouatta.seed.biplot.species <- fread("../04-analysis/screening/aadder/tables/Alouatta-SEED.tsv")
gorilla.seed.biplot.species <- fread("../04-analysis/screening/aadder/tables/Gorilla-SEED.tsv")
pan.seed.biplot.species <- fread("../04-analysis/screening/aadder/tables/Pan-SEED.tsv")
homo.seed.biplot.species <- fread("../04-analysis/screening/aadder/tables/Homo-SEED.tsv")

host_genus.biplot.species <- bind_rows(alouatta.seed.biplot.species,gorilla.seed.biplot.species)
host_genus.biplot.species <- bind_rows(host_genus.biplot.species,pan.seed.biplot.species)
host_genus.biplot.species <- bind_rows(host_genus.biplot.species,homo.seed.biplot.species)

# Now make a separate column with the Genus for summarizing in the next step
host_genus.biplot.species <- host_genus.biplot.species %>% mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]}))



```

```{r seed_contribution_files_nomod}
# read in the files from MALT, modified to have codes for the protein names b/c R can't handle the names with () - etc.
homo_nomod.seed.biplot.species <- fread("../04-analysis/screening/aadder/tables/Homo_nomod-SEED.tsv")

host_genus_nomod.biplot.species <- bind_rows(alouatta.seed.biplot.species,gorilla.seed.biplot.species)
host_genus_nomod.biplot.species <- bind_rows(host_genus_nomod.biplot.species,pan.seed.biplot.species)
host_genus_nomod.biplot.species <- bind_rows(host_genus_nomod.biplot.species,homo_nomod.seed.biplot.species)

# Now make a separate column with the Genus for summarizing in the next step
host_genus_nomod.biplot.species <- host_genus_nomod.biplot.species %>% mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]}))



```


```{r seed_contributions, eval = F}
# Make individual files for PC1 positive, PC1 negative, PC2 positive, and PC2 negative, summarized at Genus level
host_genus.biplot.species.pc1_pws_homo <- host_genus.biplot.species %>% 
                                  mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]
                                  })) %>%
                                  filter(str_detect(Host_Genus, "Homo")) %>%
                                  filter(str_detect(PC1code, "pc1p1")) %>%
                                  filter(!str_detect(PC1code, "pc1p10")) %>%
                                  group_by(Genus) %>%
                                  summarise(Sum = sum(Count)) %>%
                                  mutate(Percent = Sum/sum(Sum)*100) %>%
                                  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_pws_homo['Protein'] <- "pc1p1"
host_genus.biplot.species.pc1_pws_homo['Host_Genus'] <- 'Homo'

host_genus.biplot.species.pc1_pws_pan <- host_genus.biplot.species %>% 
                                  mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]
                                  })) %>%
                                  filter(str_detect(Host_Genus, "Pan")) %>%
                                  filter(str_detect(PC1code, "pc1p1")) %>%
                                  filter(!str_detect(PC1code, "pc1p10")) %>%
                                  group_by(Genus) %>%
                                  summarise(Sum = sum(Count)) %>%
                                  mutate(Percent = Sum/sum(Sum)*100) %>%
                                  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_pws_pan['Protein'] <- "pc1p1"
host_genus.biplot.species.pc1_pws_pan['Host_Genus'] <- 'Pan'

pc1_pws_stats <- bind_rows(host_genus.biplot.species.pc1_pws_homo,host_genus.biplot.species.pc1_pws_pan)

host_genus.biplot.species.pc1_pws_gorilla <- host_genus.biplot.species %>% 
                                  mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]
                                  })) %>%
                                  filter(str_detect(Host_Genus, "Gorilla")) %>%
                                  filter(str_detect(PC1code, "pc1p1")) %>%
                                  filter(!str_detect(PC1code, "pc1p10")) %>%
                                  group_by(Genus) %>%
                                  summarise(Sum = sum(Count)) %>%
                                  mutate(Percent = Sum/sum(Sum)*100) %>%
                                  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_pws_gorilla['Protein'] <- "pc1p1"
host_genus.biplot.species.pc1_pws_gorilla['Host_Genus'] <- 'Gorilla'

pc1_pws_stats <- bind_rows(pc1_pws_stats,host_genus.biplot.species.pc1_pws_gorilla)

host_genus.biplot.species.pc1_pws_alouatta <- host_genus.biplot.species %>% 
                                  mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]
                                  })) %>%
                                  filter(str_detect(Host_Genus, "Alouatta")) %>%
                                  filter(str_detect(PC1code, "pc1p1")) %>%
                                  filter(!str_detect(PC1code, "pc1p10")) %>%
                                  group_by(Genus) %>%
                                  summarise(Sum = sum(Count)) %>%
                                  mutate(Percent = Sum/sum(Sum)*100) %>%
                                  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_pws_alouatta['Protein'] <- "pc1p1"
host_genus.biplot.species.pc1_pws_alouatta['Host_Genus'] <- 'Alouatta'

pc1_pws_stats <- bind_rows(pc1_pws_stats,host_genus.biplot.species.pc1_pws_alouatta)

host_genus.biplot.species.pc1_neg_homo <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC1code, "pc1n1")) %>%
  filter(!str_detect(PC1code, "pc1n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_neg_homo['Protein'] <- "pc1n1"
host_genus.biplot.species.pc1_neg_homo['Host_Genus'] <- 'Homo'

host_genus.biplot.species.pc1_neg_pan <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC1code, "pc1n1")) %>%
  filter(!str_detect(PC1code, "pc1n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_neg_pan['Protein'] <- "pc1n1"
host_genus.biplot.species.pc1_neg_pan['Host_Genus'] <- 'Pan'

pc1_neg_stats <- bind_rows(host_genus.biplot.species.pc1_neg_homo,host_genus.biplot.species.pc1_neg_pan)

host_genus.biplot.species.pc1_neg_gorilla <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC1code, "pc1n1")) %>%
  filter(!str_detect(PC1code, "pc1n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_neg_gorilla['Protein'] <- "pc1n1"
host_genus.biplot.species.pc1_neg_gorilla['Host_Genus'] <- 'Gorilla'

pc1_neg_stats <- bind_rows(pc1_neg_stats,host_genus.biplot.species.pc1_neg_gorilla)

host_genus.biplot.species.pc1_neg_alouatta <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC1code, "pc1n1")) %>%
  filter(!str_detect(PC1code, "pc1n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_neg_alouatta['Protein'] <- "pc1n1"
host_genus.biplot.species.pc1_neg_alouatta['Host_Genus'] <- 'Alouatta'

pc1_neg_stats <- bind_rows(pc1_neg_stats,host_genus.biplot.species.pc1_neg_alouatta)

host_genus.biplot.species.pc2_pws_homo <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC2code, "pc2p1")) %>%
  filter(!str_detect(PC2code, "pc2p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_pws_homo['Protein'] <- "pc2p1"
host_genus.biplot.species.pc2_pws_homo['Host_Genus'] <- 'Homo'

host_genus.biplot.species.pc2_pws_pan <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC2code, "pc2p1")) %>%
  filter(!str_detect(PC2code, "pc2p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_pws_pan['Protein'] <- "pc2p1"
host_genus.biplot.species.pc2_pws_pan['Host_Genus'] <- 'Pan'

pc2_pws_stats <- bind_rows(host_genus.biplot.species.pc2_pws_homo,host_genus.biplot.species.pc2_pws_pan)

host_genus.biplot.species.pc2_pws_gorilla <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC2code, "pc2p1")) %>%
  filter(!str_detect(PC2code, "pc2p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_pws_gorilla['Protein'] <- "pc2p1"
host_genus.biplot.species.pc2_pws_gorilla['Host_Genus'] <- 'Gorilla'

pc2_pws_stats <- bind_rows(pc2_pws_stats,host_genus.biplot.species.pc2_pws_gorilla)

host_genus.biplot.species.pc2_pws_alouatta <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC2code, "pc2p1")) %>%
  filter(!str_detect(PC2code, "pc2p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_pws_alouatta['Protein'] <- "pc2p1"
host_genus.biplot.species.pc2_pws_alouatta['Host_Genus'] <- 'Alouatta'

pc2_pws_stats <- bind_rows(pc2_pws_stats,host_genus.biplot.species.pc2_pws_alouatta)

host_genus.biplot.species.pc2_neg_homo <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC2code, "pc2n1")) %>%
  filter(!str_detect(PC2code, "pc2n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_neg_homo['Protein'] <- "pc2n1"
host_genus.biplot.species.pc2_neg_homo['Host_Genus'] <- 'Homo'

host_genus.biplot.species.pc2_neg_pan <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC2code, "pc2n1")) %>%
  filter(!str_detect(PC2code, "pc2n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_neg_pan['Protein'] <- "pc2n1"
host_genus.biplot.species.pc2_neg_pan['Host_Genus'] <- 'Pan'

pc2_neg_stats <- bind_rows(host_genus.biplot.species.pc2_neg_homo,host_genus.biplot.species.pc2_neg_pan)

host_genus.biplot.species.pc2_neg_gorilla <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC2code, "pc2n1")) %>%
  filter(!str_detect(PC2code, "pc2n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_neg_gorilla['Protein'] <- "pc2n1"
host_genus.biplot.species.pc2_neg_gorilla['Host_Genus'] <- 'Gorilla'

pc2_neg_stats <- bind_rows(pc2_neg_stats,host_genus.biplot.species.pc2_neg_gorilla)

host_genus.biplot.species.pc2_neg_alouatta <- host_genus.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC2code, "pc2n1")) %>%
  filter(!str_detect(PC2code, "pc2n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_neg_alouatta['Protein'] <- "pc2n1"
host_genus.biplot.species.pc2_neg_alouatta['Host_Genus'] <- 'Alouatta'

pc2_neg_stats <- bind_rows(pc2_neg_stats,host_genus.biplot.species.pc2_neg_alouatta)

# now add orthologs 2-10 to each table by changing pc1p#/pc1n#/pc2p#/pc2n# in the sections below
host_genus.biplot.species.pc1_pws_homo <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC1code, "pc1p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_pws_homo['Protein'] <- "pc1p10"
host_genus.biplot.species.pc1_pws_homo['Host_Genus'] <- 'Homo'

pc1_pws_stats <- bind_rows(pc1_pws_stats, host_genus.biplot.species.pc1_pws_homo)

host_genus.biplot.species.pc1_pws_pan <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC1code, "pc1p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_pws_pan['Protein'] <- "pc1p10"
host_genus.biplot.species.pc1_pws_pan['Host_Genus'] <- 'Pan'

pc1_pws_stats <- bind_rows(pc1_pws_stats, host_genus.biplot.species.pc1_pws_pan)

host_genus.biplot.species.pc1_pws_gorilla <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC1code, "pc1p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_pws_gorilla['Protein'] <- "pc1p10"
host_genus.biplot.species.pc1_pws_gorilla['Host_Genus'] <- 'Gorilla'

pc1_pws_stats <- bind_rows(pc1_pws_stats, host_genus.biplot.species.pc1_pws_gorilla)

host_genus.biplot.species.pc1_pws_alouatta <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC1code, "pc1p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_pws_alouatta['Protein'] <- "pc1p10"
host_genus.biplot.species.pc1_pws_alouatta['Host_Genus'] <- 'Alouatta'

pc1_pws_stats <- bind_rows(pc1_pws_stats, host_genus.biplot.species.pc1_pws_alouatta)

host_genus.biplot.species.pc1_neg_homo <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC1code, "pc1n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_neg_homo['Protein'] <- "pc1n10"
host_genus.biplot.species.pc1_neg_homo['Host_Genus'] <- 'Homo'

pc1_neg_stats <- bind_rows(pc1_neg_stats, host_genus.biplot.species.pc1_neg_homo)

host_genus.biplot.species.pc1_neg_pan <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC1code, "pc1n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_neg_pan['Protein'] <- "pc1n10"
host_genus.biplot.species.pc1_neg_pan['Host_Genus'] <- 'Pan'

pc1_neg_stats <- bind_rows(pc1_neg_stats, host_genus.biplot.species.pc1_neg_pan)

host_genus.biplot.species.pc1_neg_gorilla <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC1code, "pc1n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_neg_gorilla['Protein'] <- "pc1n10"
host_genus.biplot.species.pc1_neg_gorilla['Host_Genus'] <- 'Gorilla'

pc1_neg_stats <- bind_rows(pc1_neg_stats, host_genus.biplot.species.pc1_neg_gorilla)

host_genus.biplot.species.pc1_neg_alouatta <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC1code, "pc1n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc1_neg_alouatta['Protein'] <- "pc1n10"
host_genus.biplot.species.pc1_neg_alouatta['Host_Genus'] <- 'Alouatta'

pc1_neg_stats <- bind_rows(pc1_neg_stats, host_genus.biplot.species.pc1_neg_alouatta)

host_genus.biplot.species.pc2_pws_homo <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC2code, "pc2p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_pws_homo['Protein'] <- "pc2p10"
host_genus.biplot.species.pc2_pws_homo['Host_Genus'] <- 'Homo'

pc2_pws_stats <- bind_rows(pc2_pws_stats, host_genus.biplot.species.pc2_pws_homo)

host_genus.biplot.species.pc2_pws_pan <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC2code, "pc2p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_pws_pan['Protein'] <- "pc2p10"
host_genus.biplot.species.pc2_pws_pan['Host_Genus'] <- 'Pan'

pc2_pws_stats <- bind_rows(pc2_pws_stats, host_genus.biplot.species.pc2_pws_pan)

host_genus.biplot.species.pc2_pws_gorilla <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC2code, "pc2p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_pws_gorilla['Protein'] <- "pc2p10"
host_genus.biplot.species.pc2_pws_gorilla['Host_Genus'] <- 'Gorilla'

pc2_pws_stats <- bind_rows(pc2_pws_stats, host_genus.biplot.species.pc2_pws_gorilla)

host_genus.biplot.species.pc2_pws_alouatta <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC2code, "pc2p10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_pws_alouatta['Protein'] <- "pc2p10"
host_genus.biplot.species.pc2_pws_alouatta['Host_Genus'] <- 'Alouatta'

pc2_pws_stats <- bind_rows(pc2_pws_stats, host_genus.biplot.species.pc2_pws_alouatta)

host_genus.biplot.species.pc2_neg_homo <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC2code, "pc2n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_neg_homo['Protein'] <- "pc2n10"
host_genus.biplot.species.pc2_neg_homo['Host_Genus'] <- 'Homo'

pc2_neg_stats <- bind_rows(pc2_neg_stats, host_genus.biplot.species.pc2_neg_homo)

host_genus.biplot.species.pc2_neg_pan <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC2code, "pc2n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_neg_pan['Protein'] <- "pc2n10"
host_genus.biplot.species.pc2_neg_pan['Host_Genus'] <- 'Pan'

pc2_neg_stats <- bind_rows(pc2_neg_stats, host_genus.biplot.species.pc2_neg_pan)

host_genus.biplot.species.pc2_neg_gorilla <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC2code, "pc2n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_neg_gorilla['Protein'] <- "pc2n10"
host_genus.biplot.species.pc2_neg_gorilla['Host_Genus'] <- 'Gorilla'

pc2_neg_stats <- bind_rows(pc2_neg_stats, host_genus.biplot.species.pc2_neg_gorilla)

host_genus.biplot.species.pc2_neg_alouatta <- host_genus.biplot.species %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC2code, "pc2n10")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus.biplot.species.pc2_neg_alouatta['Protein'] <- "pc2n10"
host_genus.biplot.species.pc2_neg_alouatta['Host_Genus'] <- 'Alouatta'

pc2_neg_stats <- bind_rows(pc2_neg_stats, host_genus.biplot.species.pc2_neg_alouatta)

# save the table to load again next time so you don't have to run this part again
# write.table(pc1_pws_stats, "../pc1_pws_stats_seed.tsv", quote=F, sep="\t", row.names = F)
# write.table(pc1_neg_stats, "../pc1_neg_stats_seed.tsv", quote=F, sep="\t", row.names = F)
# write.table(pc2_pws_stats, "../pc2_pws_stats_seed.tsv", quote=F, sep="\t", row.names = F)
# write.table(pc2_neg_stats, "../pc2_neg_stats_seed.tsv", quote=F, sep="\t", row.names = F)

```


And plot with bar charts
```{r seed_contributions_graph_prep}
# first join the protein codes with the protein names
pc1_pws <- inner_join(pc1_pws, protein_code_list, by = "Protein")
pc1_pws_top10 <- pc1_pws$PC1code
pc1_neg <- inner_join(pc1_neg, protein_code_list, by = "Protein")
pc1_neg_top10 <- pc1_neg$PC1code
pc2_pws <- inner_join(pc2_pws, protein_code_list, by = "Protein")
pc2_pws_top10 <- pc2_pws$PC2code
pc2_neg <- inner_join(pc2_neg, protein_code_list, by = "Protein")
pc2_neg_top10 <- pc2_neg$PC2code

# read in the saved files
pc1_pws_stats <- fread("../04-analysis/screening/aadder/tables/pc1_pws_stats_seed.tsv")
pc1_neg_stats <- fread("../04-analysis/screening/aadder/tables/pc1_neg_stats_seed.tsv")
pc2_pws_stats <- fread("../04-analysis/screening/aadder/tables/pc2_pws_stats_seed.tsv")
pc2_neg_stats <- fread("../04-analysis/screening/aadder/tables/pc2_neg_stats_seed.tsv")

# make the Host Genus a factor for plotting
pc1_pws_stats <- pc1_pws_stats %>% select(Protein,Host_Genus,Genus,Sum,Percent)
pc1_pws_stats$Host_Genus <- factor(pc1_pws_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc1_neg_stats <- pc1_neg_stats %>% select(Protein,Host_Genus,Genus,Sum,Percent)
pc1_neg_stats$Host_Genus <- factor(pc1_neg_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc2_pws_stats <- pc2_pws_stats %>% select(Protein,Host_Genus,Genus,Sum,Percent)
pc2_pws_stats$Host_Genus <- factor(pc2_pws_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc2_neg_stats <- pc2_neg_stats %>% select(Protein,Host_Genus,Genus,Sum,Percent)
pc2_neg_stats$Host_Genus <- factor(pc2_neg_stats$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))

```

Now graph
```{r seed_contributions_graphs}
# remove all species with <15% contribution and fill in that % with Other 
pc1_pws_stats_extra <- lapply(pc1_pws_top10, function(eclass) {
 statssum <- pc1_pws_stats %>%
   group_by(Host_Genus) %>%
   filter(Protein == eclass) %>%
   filter(Percent <=15) %>%
#   filter(Percent >15) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc1_pws_stats_extra['Protein'] <- c("pc1p10","pc1p10","pc1p10","pc1p10",
                                    "pc1p2","pc1p2","pc1p2","pc1p2",
                                    "pc1p9","pc1p9","pc1p9","pc1p9",
                                    "pc1p5","pc1p5","pc1p5","pc1p5",
                                    "pc1p6","pc1p6","pc1p6","pc1p6",
                                    "pc1p1","pc1p1","pc1p1","pc1p1",
                                    "pc1p3","pc1p3","pc1p3","pc1p3",
                                    "pc1p8","pc1p8","pc1p8","pc1p8",
                                    "pc1p7","pc1p7","pc1p7","pc1p7",
                                    "pc1p4","pc1p4","pc1p4","pc1p4")
pc1_pws_stats_extra['Sum'] <- NA
pc1_pws_stats_extra['Genus'] <- 'Other'
pc1_pws_stats_extra <- pc1_pws_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(Protein,Host_Genus,Genus,Sum,Percent)

pc1_pws_stats_extra <- pc1_pws_stats_extra %>%
  bind_rows(., pc1_pws_stats %>%
              filter(Percent >15)) %>%
  filter(!str_detect(Genus, "Total"))

pc1_pws_stats_extra$Protein <- factor(pc1_pws_stats_extra$Protein, levels = c("pc1p1","pc1p2","pc1p3","pc1p4",
                                      "pc1p5","pc1p6","pc1p7","pc1p8","pc1p9","pc1p10"))
pc1_pws_stats_extra$Genus <- factor(pc1_pws_stats_extra$Genus, levels = c("Other",
"[Eubacterium]",
"Actinobaculum",
"Actinomyces",
"Aggregatibacter",
"Capnocytophaga",
"Catonella",
"Eggerthia",
"Fusobacterium",
"Macrococcus",
"Olsenella",
"Peptostreptococcus",
"Porphyromonas",
"Prevotella",
"Pseudopropionibacterium",
"Pseudoramibacter",
"Slackia",
"Streptococcus",
"Tannerella",
"Treponema"))

pc1_pws_stats_extra_protein_plot <- pc1_pws_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                               "#3F4788FF","#238A8DFF","#56C667FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protein, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC1 positive") +
    theme(title = element_text(size=10))

pc1_pws_stats_extra_protein_plot

# ggsave("../pc1_pws_stats_extra_protein_plot.pdf", plot = pc1_pws_stats_extra_protein_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)


pc1_pws_stats_extra_hostgenus_plot <- pc1_pws_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Protein, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                               "#3F4788FF","#238A8DFF","#56C667FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC1 positive") +
    theme(title = element_text(size=10))

pc1_pws_stats_extra_hostgenus_plot

ggsave("../pc1_pws_stats_extra_hostgenus_plot.pdf", plot = pc1_pws_stats_extra_hostgenus_plot, device = "pdf",
       scale = 1, width = 8, height = 5, units = c("in"),
       dpi = 300)

# Now PC1 negative
pc1_neg_stats_extra <- lapply(pc1_neg_top10, function(eclass) {
 statssum <- pc1_neg_stats %>%
   group_by(Host_Genus) %>%
   filter(Protein == eclass) %>%
   filter(Percent <=15) %>%
#   filter(Percent >15) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc1_neg_stats_extra['Protein'] <- c("pc1n1","pc1n1","pc1n1","pc1n1",
                                    "pc1n6","pc1n6","pc1n6","pc1n6",
                                    "pc1n5","pc1n5","pc1n5","pc1n5",
                                    "pc1n9","pc1n9","pc1n9","pc1n9",
                                    "pc1n3","pc1n3","pc1n3","pc1n3",
                                    "pc1n2","pc1n2","pc1n2","pc1n2",
                                    "pc1n10","pc1n10","pc1n10","pc1n10",
                                    "pc1n7","pc1n7","pc1n7","pc1n7",
                                    "pc1n4","pc1n4","pc1n4","pc1n4",
                                    "pc1n8")
pc1_neg_stats_extra['Sum'] <- NA
pc1_neg_stats_extra['Genus'] <- 'Other'
pc1_neg_stats_extra <- pc1_neg_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(Protein,Host_Genus,Genus,Sum,Percent)

pc1_neg_stats_extra <- pc1_neg_stats_extra %>%
  bind_rows(., pc1_neg_stats %>%
              filter(Percent >15)) %>%
  filter(!str_detect(Genus, "Total"))

pc1_neg_stats_extra$Protein <- factor(pc1_neg_stats_extra$Protein, levels = c("pc1n1","pc1n2","pc1n3","pc1n4",
                                      "pc1n5","pc1n6","pc1n7","pc1n8","pc1n9","pc1n10"))

pc1_neg_stats_extra$Genus <- factor(pc1_neg_stats_extra$Genus, levels = c("Other",
"Acinetobacter",
"Aggregatibacter",
"Campylobacter",
"Cardiobacterium",
"Fusobacterium",
"Lautropia",
"Neisseria",
"Ottowia",
"Stenotrophomonas",
"Streptococcus",
"Streptomyces"))

pc1_neg_stats_extra_protein_plot <- pc1_neg_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#400F73FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protein, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC1 negative") +
    theme(title = element_text(size=10))

pc1_neg_stats_extra_protein_plot

# ggsave("../pc1_neg_stats_extra_protein_plot.pdf", plot = pc1_neg_stats_extra_protein_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc1_neg_stats_extra_hostgenus_plot <- pc1_neg_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Protein, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#400F73FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC1 negative") +
    theme(title = element_text(size=10))

pc1_neg_stats_extra_hostgenus_plot

# ggsave("../pc1_neg_stats_extra_hostgenus_plot.pdf", plot = pc1_neg_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

# Now PC2 positive
pc2_pws_stats_extra <- lapply(pc2_pws_top10, function(eclass) {
 statssum <- pc2_pws_stats %>%
   group_by(Host_Genus) %>%
   filter(Protein == eclass) %>%
   filter(Percent <=15) %>%
#   filter(Percent >15) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc2_pws_stats_extra['Protein'] <- c("pc2p10","pc2p10","pc2p10","pc2p10",
                                    "pc2p4","pc2p4","pc2p4","pc2p4",
                                    "pc2p8","pc2p8","pc2p8","pc2p8",
                                    "pc2p5","pc2p5","pc2p5","pc2p5",
                                    "pc2p2","pc2p2","pc2p2","pc2p2",
                                    "pc2p9","pc2p9","pc2p9","pc2p9",
                                    "pc2p3","pc2p3","pc2p3","pc2p3",
                                    "pc2p7","pc2p7","pc2p7","pc2p7",
                                    "pc2p6","pc2p6","pc2p6","pc2p6",
                                    "pc2p1","pc2p1","pc2p1","pc2p1")
pc2_pws_stats_extra['Sum'] <- NA
pc2_pws_stats_extra['Genus'] <- 'Other'
pc2_pws_stats_extra <- pc2_pws_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(Protein,Host_Genus,Genus,Sum,Percent)

pc2_pws_stats_extra <- pc2_pws_stats_extra %>%
  bind_rows(., pc2_pws_stats %>%
              filter(Percent >15)) %>%
  filter(!str_detect(Genus, "Total"))

pc2_pws_stats_extra$Protein <- factor(pc2_pws_stats_extra$Protein, levels = c("pc2p1","pc2p2","pc2p3","pc2p4",
                                      "pc2p5","pc2p6","pc2p7","pc2p8","pc2p9","pc2p10"))

pc2_pws_stats_extra$Genus <- factor(pc2_pws_stats_extra$Genus, levels = c("Other",
"Acinetobacter",
"Aggregatibacter",
"Atopobium",
"Bacteroidetes",
"Campylobacter",
"Capnocytophaga",
"Cardiobacterium",
"Corynebacterium",
"Desulfobulbus",
"Dialister",
"Fusobacterium",
"Lautropia",
"Macrococcus",
"Neisseria",
"Olsenella",
"Phocaeicola",
"Slackia",
"Streptomyces",
"Treponema"))

pc2_pws_stats_extra_protein_plot <- pc2_pws_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
   scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                               "#3F4788FF","#238A8DFF","#56C667FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protein, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC2 positive") +
    theme(title = element_text(size=10))

pc2_pws_stats_extra_protein_plot

# ggsave("../pc2_pws_stats_extra_protein_plot.pdf", plot = pc2_pws_stats_extra_protein_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_pws_stats_extra_hostgenus_plot <- pc2_pws_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Protein, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
   scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                               "#3F4788FF","#238A8DFF","#56C667FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC2 positive") +
    theme(title = element_text(size=10))

pc2_pws_stats_extra_hostgenus_plot

# ggsave("../pc2_pws_stats_extra_hostgenus_plot.pdf", plot = pc2_pws_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

# Now PC2 negative
pc2_neg_stats_extra <- lapply(pc2_neg_top10, function(eclass) {
 statssum <- pc2_neg_stats %>%
   group_by(Host_Genus) %>%
   filter(Protein == eclass) %>%
   filter(Percent <=15) %>%
#   filter(Percent >15) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc2_neg_stats_extra['Protein'] <- c("pc2n8","pc2n8","pc2n8","pc2n8",
                                    "pc2n9","pc2n9","pc2n9","pc2n9",
                                    "pc2n1","pc2n1","pc2n1","pc2n1",
                                    "pc2n3","pc2n3","pc2n3","pc2n3",
                                    "pc2n10","pc2n10","pc2n10","pc2n10",
                                    "pc2n5","pc2n5","pc2n5","pc2n5",
                                    "pc2n7","pc2n7","pc2n7","pc2n7",
                                    "pc2n6","pc2n6","pc2n6","pc2n6",
                                    "pc2n4","pc2n4","pc2n4","pc2n4",
                                    "pc2n2")
pc2_neg_stats_extra['Sum'] <- NA
pc2_neg_stats_extra['Genus'] <- 'Other'
pc2_neg_stats_extra <- pc2_neg_stats_extra %>%
  rename(Percent = Remaining) %>%
  select(Protein,Host_Genus,Genus,Sum,Percent)

pc2_neg_stats_extra <- pc2_neg_stats_extra %>%
  bind_rows(., pc2_neg_stats %>%
              filter(Percent >15)) %>%
  filter(!str_detect(Genus, "Total"))

pc2_neg_stats_extra$Protein <- factor(pc2_neg_stats_extra$Protein, levels = c("pc2n1","pc2n2","pc2n3","pc2n4",
                                      "pc2n5","pc2n6","pc2n7","pc2n8","pc2n9","pc2n10"))

pc2_neg_stats_extra$Genus <- factor(pc2_neg_stats_extra$Genus, levels = c("Other",
"Abiotrophia",
"Actinomyces",
"Aggregatibacter",
"Atopobium",
"Capnocytophaga",
"Catonella",
"Filifactor",
"Macrococcus",
"Oribacterium",
"Parabacteroides",
"Parvimonas",
"Porphyromonas",
"Schaalia",
"Stenotrophomonas",
"Streptococcus",
"Streptomyces",
"Treponema"))

pc2_neg_stats_extra_protein_plot <- pc2_neg_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF","#3F4788FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protein, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC2 negative") +
    theme(title = element_text(size=10))

pc2_neg_stats_extra_protein_plot

# ggsave("../pc2_neg_stats_extra_protein_plot.pdf", plot = pc2_neg_stats_extra_protein_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_neg_stats_extra_hostgenus_plot <- pc2_neg_stats_extra %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Protein, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF","#3F4788FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC2 negative") +
    theme(title = element_text(size=10))

pc2_neg_stats_extra_hostgenus_plot

# ggsave("../pc2_neg_stats_extra_hostgenus_plot.pdf", plot = pc2_neg_stats_extra_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

```

Repeat for the data without modern humans
```{r seed_contributions_nomod, eval = F}
pc1_pws_nomod <- inner_join(pc1_pws_nomod, protein_code_list_nomod, by = "Protein")
pc1_neg_nomod <- inner_join(pc1_neg_nomod, protein_code_list_nomod, by = "Protein")
pc2_pws_nomod <- inner_join(pc2_pws_nomod, protein_code_list_nomod, by = "Protein")
pc2_neg_nomod <- inner_join(pc2_neg_nomod, protein_code_list_nomod, by = "Protein")


# Make individual files for PC1 positive, PC1 negative, PC2 positive, and PC2 negative, summarized at Genus level
host_genus_nomod.biplot.species.pc1_pws_homo <- host_genus_nomod.biplot.species %>% 
                                  mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]
                                  })) %>%
                                  filter(str_detect(Host_Genus, "Homo")) %>%
                                  filter(str_detect(PC1nomodcode, "pc1p1nomod")) %>%
                                  filter(!str_detect(PC1nomodcode, "pc1p10nomod")) %>%
                                  group_by(Genus) %>%
                                  summarise(Sum = sum(Count)) %>%
                                  mutate(Percent = Sum/sum(Sum)*100) %>%
                                  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_pws_homo['Protein'] <- "pc1p1nomod"
host_genus_nomod.biplot.species.pc1_pws_homo['Host_Genus'] <- 'Homo'

host_genus_nomod.biplot.species.pc1_pws_pan <- host_genus_nomod.biplot.species %>% 
                                  mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]
                                  })) %>%
                                  filter(str_detect(Host_Genus, "Pan")) %>%
                                  filter(str_detect(PC1nomodcode, "pc1p1nomod")) %>%
                                  filter(!str_detect(PC1nomodcode, "pc1p10nomod")) %>%
                                  group_by(Genus) %>%
                                  summarise(Sum = sum(Count)) %>%
                                  mutate(Percent = Sum/sum(Sum)*100) %>%
                                  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_pws_pan['Protein'] <- "pc1p1nomod"
host_genus_nomod.biplot.species.pc1_pws_pan['Host_Genus'] <- 'Pan'

pc1_pws_nomod_stats <- bind_rows(host_genus_nomod.biplot.species.pc1_pws_homo,host_genus_nomod.biplot.species.pc1_pws_pan)

host_genus_nomod.biplot.species.pc1_pws_gorilla <- host_genus_nomod.biplot.species %>% 
                                  mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]
                                  })) %>%
                                  filter(str_detect(Host_Genus, "Gorilla")) %>%
                                  filter(str_detect(PC1nomodcode, "pc1p1nomod")) %>%
                                  filter(!str_detect(PC1nomodcode, "pc1p10nomod")) %>%
                                  group_by(Genus) %>%
                                  summarise(Sum = sum(Count)) %>%
                                  mutate(Percent = Sum/sum(Sum)*100) %>%
                                  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_pws_gorilla['Protein'] <- "pc1p1nomod"
host_genus_nomod.biplot.species.pc1_pws_gorilla['Host_Genus'] <- 'Gorilla'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats,host_genus_nomod.biplot.species.pc1_pws_gorilla)

host_genus_nomod.biplot.species.pc1_pws_alouatta <- host_genus_nomod.biplot.species %>% 
                                  mutate(Genus = sapply(Species, function(f) {
                                  unlist(str_split(f, " "))[1]
                                  })) %>%
                                  filter(str_detect(Host_Genus, "Alouatta")) %>%
                                  filter(str_detect(PC1nomodcode, "pc1p1nomod")) %>%
                                  filter(!str_detect(PC1nomodcode, "pc1p10nomod")) %>%
                                  group_by(Genus) %>%
                                  summarise(Sum = sum(Count)) %>%
                                  mutate(Percent = Sum/sum(Sum)*100) %>%
                                  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_pws_alouatta['Protein'] <- "pc1p1nomod"
host_genus_nomod.biplot.species.pc1_pws_alouatta['Host_Genus'] <- 'Alouatta'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats,host_genus_nomod.biplot.species.pc1_pws_alouatta)

host_genus_nomod.biplot.species.pc1_neg_homo <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC1nomodcode, "pc1n1nomod")) %>%
  filter(!str_detect(PC1nomodcode, "pc1n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_neg_homo['Protein'] <- "pc1n1nomod"
host_genus_nomod.biplot.species.pc1_neg_homo['Host_Genus'] <- 'Homo'

host_genus_nomod.biplot.species.pc1_neg_pan <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC1nomodcode, "pc1n1nomod")) %>%
  filter(!str_detect(PC1nomodcode, "pc1n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_neg_pan['Protein'] <- "pc1n1nomod"
host_genus_nomod.biplot.species.pc1_neg_pan['Host_Genus'] <- 'Pan'

pc1_neg_nomod_stats <- bind_rows(host_genus_nomod.biplot.species.pc1_neg_homo,host_genus_nomod.biplot.species.pc1_neg_pan)

host_genus_nomod.biplot.species.pc1_neg_gorilla <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC1nomodcode, "pc1n1nomod")) %>%
  filter(!str_detect(PC1nomodcode, "pc1n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_neg_gorilla['Protein'] <- "pc1n1nomod"
host_genus_nomod.biplot.species.pc1_neg_gorilla['Host_Genus'] <- 'Gorilla'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats,host_genus_nomod.biplot.species.pc1_neg_gorilla)

host_genus_nomod.biplot.species.pc1_neg_alouatta <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC1nomodcode, "pc1n1nomod")) %>%
  filter(!str_detect(PC1nomodcode, "pc1n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_neg_alouatta['Protein'] <- "pc1n1nomod"
host_genus_nomod.biplot.species.pc1_neg_alouatta['Host_Genus'] <- 'Alouatta'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats,host_genus_nomod.biplot.species.pc1_neg_alouatta)

host_genus_nomod.biplot.species.pc2_pws_homo <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC2nomodcode, "pc2p1nomod")) %>%
  filter(!str_detect(PC2nomodcode, "pc2p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_pws_homo['Protein'] <- "pc2p1nomod"
host_genus_nomod.biplot.species.pc2_pws_homo['Host_Genus'] <- 'Homo'

host_genus_nomod.biplot.species.pc2_pws_pan <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC2nomodcode, "pc2p1nomod")) %>%
  filter(!str_detect(PC2nomodcode, "pc2p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_pws_pan['Protein'] <- "pc2p1nomod"
host_genus_nomod.biplot.species.pc2_pws_pan['Host_Genus'] <- 'Pan'

pc2_pws_nomod_stats <- bind_rows(host_genus_nomod.biplot.species.pc2_pws_homo,host_genus_nomod.biplot.species.pc2_pws_pan)

host_genus_nomod.biplot.species.pc2_pws_gorilla <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC2nomodcode, "pc2p1nomod")) %>%
  filter(!str_detect(PC2nomodcode, "pc2p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_pws_gorilla['Protein'] <- "pc2p1nomod"
host_genus_nomod.biplot.species.pc2_pws_gorilla['Host_Genus'] <- 'Gorilla'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats,host_genus_nomod.biplot.species.pc2_pws_gorilla)

host_genus_nomod.biplot.species.pc2_pws_alouatta <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC2nomodcode, "pc2p1nomod")) %>%
  filter(!str_detect(PC2nomodcode, "pc2p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_pws_alouatta['Protein'] <- "pc2p1nomod"
host_genus_nomod.biplot.species.pc2_pws_alouatta['Host_Genus'] <- 'Alouatta'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats,host_genus_nomod.biplot.species.pc2_pws_alouatta)

host_genus_nomod.biplot.species.pc2_neg_homo <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC2nomodcode, "pc2n1nomod")) %>%
  filter(!str_detect(PC2nomodcode, "pc2n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_neg_homo['Protein'] <- "pc2n1nomod"
host_genus_nomod.biplot.species.pc2_neg_homo['Host_Genus'] <- 'Homo'

host_genus_nomod.biplot.species.pc2_neg_pan <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC2nomodcode, "pc2n1nomod")) %>%
  filter(!str_detect(PC2nomodcode, "pc2n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_neg_pan['Protein'] <- "pc2n1nomod"
host_genus_nomod.biplot.species.pc2_neg_pan['Host_Genus'] <- 'Pan'

pc2_neg_nomod_stats <- bind_rows(host_genus_nomod.biplot.species.pc2_neg_homo,host_genus_nomod.biplot.species.pc2_neg_pan)

host_genus_nomod.biplot.species.pc2_neg_gorilla <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC2nomodcode, "pc2n1nomod")) %>%
  filter(!str_detect(PC2nomodcode, "pc2n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_neg_gorilla['Protein'] <- "pc2n1nomod"
host_genus_nomod.biplot.species.pc2_neg_gorilla['Host_Genus'] <- 'Gorilla'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats,host_genus_nomod.biplot.species.pc2_neg_gorilla)

host_genus_nomod.biplot.species.pc2_neg_alouatta <- host_genus_nomod.biplot.species %>% 
  mutate(Genus = sapply(Species, function(f) {
    unlist(str_split(f, " "))[1]
  })) %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC2nomodcode, "pc2n1nomod")) %>%
  filter(!str_detect(PC2nomodcode, "pc2n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_neg_alouatta['Protein'] <- "pc2n1nomod"
host_genus_nomod.biplot.species.pc2_neg_alouatta['Host_Genus'] <- 'Alouatta'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats,host_genus_nomod.biplot.species.pc2_neg_alouatta)

# now add orthologs 2-10 to each table by changing pc1p#/pc1n#/pc2p#/pc2n# in the sections below
host_genus_nomod.biplot.species.pc1_pws_homo <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC1nomodcode, "pc1p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_pws_homo['Protein'] <- "pc1p10nomod"
host_genus_nomod.biplot.species.pc1_pws_homo['Host_Genus'] <- 'Homo'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats, host_genus_nomod.biplot.species.pc1_pws_homo)

host_genus_nomod.biplot.species.pc1_pws_pan <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC1nomodcode, "pc1p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_pws_pan['Protein'] <- "pc1p10nomod"
host_genus_nomod.biplot.species.pc1_pws_pan['Host_Genus'] <- 'Pan'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats, host_genus_nomod.biplot.species.pc1_pws_pan)

host_genus_nomod.biplot.species.pc1_pws_gorilla <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC1nomodcode, "pc1p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_pws_gorilla['Protein'] <- "pc1p10nomod"
host_genus_nomod.biplot.species.pc1_pws_gorilla['Host_Genus'] <- 'Gorilla'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats, host_genus_nomod.biplot.species.pc1_pws_gorilla)

host_genus_nomod.biplot.species.pc1_pws_alouatta <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC1nomodcode, "pc1p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_pws_alouatta['Protein'] <- "pc1p10nomod"
host_genus_nomod.biplot.species.pc1_pws_alouatta['Host_Genus'] <- 'Alouatta'

pc1_pws_nomod_stats <- bind_rows(pc1_pws_nomod_stats, host_genus_nomod.biplot.species.pc1_pws_alouatta)

host_genus_nomod.biplot.species.pc1_neg_homo <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC1nomodcode, "pc1n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_neg_homo['Protein'] <- "pc1n10nomod"
host_genus_nomod.biplot.species.pc1_neg_homo['Host_Genus'] <- 'Homo'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats, host_genus_nomod.biplot.species.pc1_neg_homo)

host_genus_nomod.biplot.species.pc1_neg_pan <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC1nomodcode, "pc1n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_neg_pan['Protein'] <- "pc1n10nomod"
host_genus_nomod.biplot.species.pc1_neg_pan['Host_Genus'] <- 'Pan'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats, host_genus_nomod.biplot.species.pc1_neg_pan)

host_genus_nomod.biplot.species.pc1_neg_gorilla <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC1nomodcode, "pc1n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_neg_gorilla['Protein'] <- "pc1n10nomod"
host_genus_nomod.biplot.species.pc1_neg_gorilla['Host_Genus'] <- 'Gorilla'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats, host_genus_nomod.biplot.species.pc1_neg_gorilla)

host_genus_nomod.biplot.species.pc1_neg_alouatta <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC1nomodcode, "pc1n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc1_neg_alouatta['Protein'] <- "pc1n10nomod"
host_genus_nomod.biplot.species.pc1_neg_alouatta['Host_Genus'] <- 'Alouatta'

pc1_neg_nomod_stats <- bind_rows(pc1_neg_nomod_stats, host_genus_nomod.biplot.species.pc1_neg_alouatta)

host_genus_nomod.biplot.species.pc2_pws_homo <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC2nomodcode, "pc2p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_pws_homo['Protein'] <- "pc2p10nomod"
host_genus_nomod.biplot.species.pc2_pws_homo['Host_Genus'] <- 'Homo'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats, host_genus_nomod.biplot.species.pc2_pws_homo)

host_genus_nomod.biplot.species.pc2_pws_pan <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC2nomodcode, "pc2p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_pws_pan['Protein'] <- "pc2p10nomod"
host_genus_nomod.biplot.species.pc2_pws_pan['Host_Genus'] <- 'Pan'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats, host_genus_nomod.biplot.species.pc2_pws_pan)

host_genus_nomod.biplot.species.pc2_pws_gorilla <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC2nomodcode, "pc2p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_pws_gorilla['Protein'] <- "pc2p10nomod"
host_genus_nomod.biplot.species.pc2_pws_gorilla['Host_Genus'] <- 'Gorilla'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats, host_genus_nomod.biplot.species.pc2_pws_gorilla)

host_genus_nomod.biplot.species.pc2_pws_alouatta <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC2nomodcode, "pc2p10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_pws_alouatta['Protein'] <- "pc2p10nomod"
host_genus_nomod.biplot.species.pc2_pws_alouatta['Host_Genus'] <- 'Alouatta'

pc2_pws_nomod_stats <- bind_rows(pc2_pws_nomod_stats, host_genus_nomod.biplot.species.pc2_pws_alouatta)

host_genus_nomod.biplot.species.pc2_neg_homo <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Homo")) %>%
  filter(str_detect(PC2nomodcode, "pc2n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_neg_homo['Protein'] <- "pc2n10nomod"
host_genus_nomod.biplot.species.pc2_neg_homo['Host_Genus'] <- 'Homo'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats, host_genus_nomod.biplot.species.pc2_neg_homo)

host_genus_nomod.biplot.species.pc2_neg_pan <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Pan")) %>%
  filter(str_detect(PC2nomodcode, "pc2n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_neg_pan['Protein'] <- "pc2n10nomod"
host_genus_nomod.biplot.species.pc2_neg_pan['Host_Genus'] <- 'Pan'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats, host_genus_nomod.biplot.species.pc2_neg_pan)

host_genus_nomod.biplot.species.pc2_neg_gorilla <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Gorilla")) %>%
  filter(str_detect(PC2nomodcode, "pc2n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_neg_gorilla['Protein'] <- "pc2n10nomod"
host_genus_nomod.biplot.species.pc2_neg_gorilla['Host_Genus'] <- 'Gorilla'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats, host_genus_nomod.biplot.species.pc2_neg_gorilla)

host_genus_nomod.biplot.species.pc2_neg_alouatta <- host_genus_nomod.biplot.species %>%
  filter(str_detect(Host_Genus, "Alouatta")) %>%
  filter(str_detect(PC2nomodcode, "pc2n10nomod")) %>%
  group_by(Genus) %>%
  summarise(Sum = sum(Count)) %>%
  drop_na() %>%
  mutate(Percent = Sum/sum(Sum)*100) %>%
  adorn_totals(where = "row", name = "Total")
host_genus_nomod.biplot.species.pc2_neg_alouatta['Protein'] <- "pc2n10nomod"
host_genus_nomod.biplot.species.pc2_neg_alouatta['Host_Genus'] <- 'Alouatta'

pc2_neg_nomod_stats <- bind_rows(pc2_neg_nomod_stats, host_genus_nomod.biplot.species.pc2_neg_alouatta)

# save the table to load again next time so you don't have to run this part again
# write.table(pc1_pws_nomod_stats, "../pc1_pws_stats_seed_nomod.tsv", quote=F, sep="\t", row.names = F)
# write.table(pc1_neg_nomod_stats, "../pc1_neg_stats_seed_nomod.tsv", quote=F, sep="\t", row.names = F)
# write.table(pc2_pws_nomod_stats, "../pc2_pws_stats_seed_nomod.tsv", quote=F, sep="\t", row.names = F)
# write.table(pc2_neg_nomod_stats, "../pc2_neg_stats_seed_nomod.tsv", quote=F, sep="\t", row.names = F)

```

```{r seed_contributions_graph_prep_nomod}
# first join the protein codes with the protein names
pc1_pws_top10_nomod <- protein_code_list_nomod %>% filter(str_detect(PC1nomodcode, "pc1p")) %>% select(PC1nomodcode)
pc1_pws_top10_nomod <- pc1_pws_top10_nomod$PC1nomodcode
pc1_neg_top10_nomod <- protein_code_list_nomod %>% filter(str_detect(PC1nomodcode, "pc1n")) %>% select(PC1nomodcode)
pc1_neg_top10_nomod <- pc1_neg_top10_nomod$PC1nomodcode
pc2_pws_top10_nomod <- protein_code_list_nomod %>% filter(str_detect(PC2nomodcode, "pc2p")) %>% select(PC2nomodcode)
pc2_pws_top10_nomod <- pc2_pws_top10_nomod$PC2nomodcode
pc2_neg_top10_nomod <- protein_code_list_nomod %>% filter(str_detect(PC2nomodcode, "pc2n")) %>% select(PC2nomodcode)
pc2_neg_top10_nomod <- pc2_neg_top10_nomod$PC2nomodcode

# read in the saved files
pc1_pws_stats_nomod <- fread("../04-analysis/screening/aadder/tables/pc1_pws_stats_seed_nomod.tsv")
pc1_neg_stats_nomod <- fread("../04-analysis/screening/aadder/tables/pc1_neg_stats_seed_nomod.tsv")
pc2_pws_stats_nomod <- fread("../04-analysis/screening/aadder/tables/pc2_pws_stats_seed_nomod.tsv")
pc2_neg_stats_nomod <- fread("../04-analysis/screening/aadder/tables/pc2_neg_stats_seed_nomod.tsv")

# make the Host Genus a factor for plotting
pc1_pws_stat_nomods <- pc1_pws_stats_nomod %>% select(Protein,Host_Genus,Genus,Sum,Percent)
pc1_pws_stats_nomod$Host_Genus <- factor(pc1_pws_stats_nomod$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc1_neg_stats_nomod <- pc1_neg_stats_nomod %>% select(Protein,Host_Genus,Genus,Sum,Percent)
pc1_neg_stats_nomod$Host_Genus <- factor(pc1_neg_stats_nomod$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc2_pws_stats_nomod <- pc2_pws_stats_nomod %>% select(Protein,Host_Genus,Genus,Sum,Percent)
pc2_pws_stats_nomod$Host_Genus <- factor(pc2_pws_stats_nomod$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))
pc2_neg_stats_nomod <- pc2_neg_stats_nomod %>% select(Protein,Host_Genus,Genus,Sum,Percent)
pc2_neg_stats_nomod$Host_Genus <- factor(pc2_neg_stats_nomod$Host_Genus, levels = c("Homo","Pan","Gorilla","Alouatta"))

```

```{r seed_contributions_graphs_nomod}

# remove all species with <15% contribution and fill in that % with Other 
pc1_pws_stats_extra_nomod <- lapply(pc1_pws_top10_nomod, function(eclass) {
 statssum <- pc1_pws_stats_nomod %>%
   group_by(Host_Genus) %>%
   filter(Protein == eclass) %>%
   filter(Percent <=15) %>%
#   filter(Percent >15) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc1_pws_stats_extra_nomod['Protein'] <- c("pc1p10","pc1p10","pc1p10","pc1p10",
                                    "pc1p1","pc1p1","pc1p1","pc1p1",
                                    "pc1p2","pc1p2","pc1p2","pc1p2",
                                    "pc1p3","pc1p3","pc1p3","pc1p3",
                                    "pc1p4","pc1p4","pc1p4","pc1p4",
                                    "pc1p5","pc1p5","pc1p5","pc1p5",
                                    "pc1p6","pc1p6","pc1p6","pc1p6",
                                    "pc1p7","pc1p7","pc1p7","pc1p7",
                                    "pc1p8","pc1p8","pc1p8","pc1p8",
                                    "pc1p9","pc1p9","pc1p9","pc1p9")
pc1_pws_stats_extra_nomod['Sum'] <- NA
pc1_pws_stats_extra_nomod['Genus'] <- 'Other'
pc1_pws_stats_extra_nomod <- pc1_pws_stats_extra_nomod %>%
  rename(Percent = Remaining) %>%
  select(Protein,Host_Genus,Genus,Sum,Percent)

# remove 'nomod' from the protein names to be able to merge with the columns just made. We don't need 'nomod' in the plots, it looks cleaner without. 
pc1_pws_stats_nomod$Protein <- gsub("nomod", "", pc1_pws_stats_nomod$Protein)

pc1_pws_stats_extra_nomod <- pc1_pws_stats_extra_nomod %>%
  bind_rows(., pc1_pws_stats_nomod %>%
              filter(Percent >15)) %>%
  filter(!str_detect(Genus, "Total"))

pc1_pws_stats_extra_nomod$Protein <- factor(pc1_pws_stats_extra_nomod$Protein, levels = c("pc1p1","pc1p2","pc1p3","pc1p4",
                                      "pc1p5","pc1p6","pc1p7","pc1p8","pc1p9","pc1p10"))
pc1_pws_stats_extra_nomod$Genus <- factor(pc1_pws_stats_extra_nomod$Genus, levels = c("Other",
"[Eubacterium]",
"Aggregatibacter",
"Capnocytophaga",
"Catonella",
"Desulfobulbus",
"Eggerthia",
"Fusobacterium",
"Leptotrichia",
"Macrococcus",
"Methanobrevibacter",
"Olsenella",
"Parvimonas",
"Peptoniphilus",
"Peptostreptococcus",
"Porphyromonas",
"Prevotella",
"Pseudopropionibacterium",
"Pseudoramibacter",
"Slackia",
"Streptococcus",
"Tannerella",
"Treponema"))

pc1_pws_stats_extra_nomod_protein_plot <- pc1_pws_stats_extra_nomod %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                               "#3F4788FF","#238A8DFF","#56C667FF","#FDE725FF",
                               "#74D055FF","#952C80FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protein, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC1 positive, no modern Homo") +
    theme(title = element_text(size=10))

pc1_pws_stats_extra_nomod_protein_plot

# ggsave("../pc1_pws_stats_extra_nomod_protein_plot.pdf", plot = pc1_pws_stats_extra_nomod_protein_plot, device = "pdf",
#        scale = 1, width = 10, height = 5, units = c("in"),
#        dpi = 300)

pc1_pws_stats_extra_nomod_hostgenus_plot <- pc1_pws_stats_extra_nomod %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Protein, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF","#FCFDBFFF",
                               "#3F4788FF","#238A8DFF","#56C667FF","#FDE725FF",
                               "#74D055FF","#DCE318FF","#FDE4A6FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC1 positive, no modern Homo") +
    theme(title = element_text(size=10))

pc1_pws_stats_extra_nomod_hostgenus_plot

# ggsave("../pc1_pws_stats_extra_nomod_hostgenus_plot.pdf", plot = pc1_pws_stats_extra_nomod_hostgenus_plot, device = "pdf",
#        scale = 1, width = 10, height = 5, units = c("in"),
#        dpi = 300)

# Now PC1 negative
pc1_neg_stats_extra_nomod <- lapply(pc1_neg_top10_nomod, function(eclass) {
 statssum <- pc1_neg_stats_nomod %>%
   group_by(Host_Genus) %>%
   filter(Protein == eclass) %>%
   filter(Percent <=15) %>%
#   filter(Percent >15) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc1_neg_stats_extra_nomod['Protein'] <- c("pc1n10","pc1n10","pc1n10","pc1n10",
                                    "pc1n1","pc1n1","pc1n1","pc1n1",
                                    "pc1n2","pc1n2","pc1n2","pc1n2",
                                    "pc1n3","pc1n3","pc1n3","pc1n3",
                                    "pc1n4","pc1n4","pc1n4","pc1n4",
                                    "pc1n5","pc1n5","pc1n5","pc1n5",
                                    "pc1n6","pc1n6","pc1n6","pc1n6",
                                    "pc1n7","pc1n7","pc1n7","pc1n7",
                                    "pc1n8","pc1n8","pc1n8","pc1n8",
                                    "pc1n9","pc1n9","pc1n9","pc1n9")
pc1_neg_stats_extra_nomod['Sum'] <- NA
pc1_neg_stats_extra_nomod['Genus'] <- 'Other'
pc1_neg_stats_extra_nomod <- pc1_neg_stats_extra_nomod %>%
  rename(Percent = Remaining) %>%
  select(Protein,Host_Genus,Genus,Sum,Percent)

# remove 'nomod' from the protein names to be able to merge with the columns just made. We don't need 'nomod' in the plots, it looks cleaner without. 
pc1_neg_stats_nomod$Protein <- gsub("nomod", "", pc1_neg_stats_nomod$Protein)

pc1_neg_stats_extra_nomod <- pc1_neg_stats_extra_nomod %>%
  bind_rows(., pc1_neg_stats_nomod %>%
              filter(Percent >15)) %>%
  filter(!str_detect(Genus, "Total"))

pc1_neg_stats_extra_nomod$Protein <- factor(pc1_neg_stats_extra_nomod$Protein, levels = c("pc1n1","pc1n2","pc1n3","pc1n4",
                                      "pc1n5","pc1n6","pc1n7","pc1n8","pc1n9","pc1n10"))

pc1_neg_stats_extra_nomod$Genus <- factor(pc1_neg_stats_extra_nomod$Genus, levels = c("Other",
"Acinetobacter",
"Aggregatibacter",
"Brachymonas",
"Campylobacter",
"Cardiobacterium",
"Fusobacterium",
"Lautropia",
"Neisseria",
"Ottowia",
"Pseudopropionibacterium",
"Streptococcus"))

pc1_neg_stats_extra_nomod_protein_plot <- pc1_neg_stats_extra_nomod %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#FDE725FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protein, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC1 negative, no modern Homo") +
    theme(title = element_text(size=10))

pc1_neg_stats_extra_nomod_protein_plot

# ggsave("../pc1_neg_stats_extra_nomod_protein_plot.pdf", plot = pc1_neg_stats_extra_nomod_protein_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc1_neg_stats_extra_nomod_hostgenus_plot <- pc1_neg_stats_extra_nomod %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Protein, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#FDE725FF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC1 negative, no modern Homo") +
    theme(title = element_text(size=10))

pc1_neg_stats_extra_nomod_hostgenus_plot

# ggsave("../pc1_neg_stats_extra_nomod_hostgenus_plot.pdf", plot = pc1_neg_stats_extra_nomod_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

# Now PC2 positive
pc2_pws_stats_extra_nomod <- lapply(pc2_pws_top10_nomod, function(eclass) {
 statssum <- pc2_pws_stats_nomod %>%
   group_by(Host_Genus) %>%
   filter(Protein == eclass) %>%
   filter(Percent <=15) %>%
#   filter(Percent >15) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc2_pws_stats_extra_nomod['Protein'] <- c("pc2p1","pc2p1","pc2p1","pc2p1",
                                          "pc2p2","pc2p2","pc2p2","pc2p2",
                                          "pc2p10","pc2p10","pc2p10","pc2p10",
                                          "pc2p3","pc2p3","pc2p3","pc2p3",
                                          "pc2p4","pc2p4","pc2p4","pc2p4",
                                          "pc2p5","pc2p5","pc2p5","pc2p5",
                                          "pc2p6","pc2p6","pc2p6","pc2p6",
                                          "pc2p7","pc2p7","pc2p7","pc2p7",
                                          "pc2p8","pc2p8","pc2p8","pc2p8",
                                          "pc2p9","pc2p9","pc2p9","pc2p9")
pc2_pws_stats_extra_nomod['Sum'] <- NA
pc2_pws_stats_extra_nomod['Genus'] <- 'Other'
pc2_pws_stats_extra_nomod <- pc2_pws_stats_extra_nomod %>%
  rename(Percent = Remaining) %>%
  select(Protein,Host_Genus,Genus,Sum,Percent)

# remove 'nomod' from the protein names to be able to merge with the columns just made. We don't need 'nomod' in the plots, it looks cleaner without. 
pc2_pws_stats_nomod$Protein <- gsub("nomod", "", pc2_pws_stats_nomod$Protein)

pc2_pws_stats_extra_nomod <- pc2_pws_stats_extra_nomod %>%
  bind_rows(., pc2_pws_stats_nomod %>%
              filter(Percent >15)) %>%
  filter(!str_detect(Genus, "Total"))

pc2_pws_stats_extra_nomod$Protein <- factor(pc2_pws_stats_extra_nomod$Protein, levels = c("pc2p1","pc2p2","pc2p3","pc2p4",
                                      "pc2p5","pc2p6","pc2p7","pc2p8","pc2p9","pc2p10"))

pc2_pws_stats_extra_nomod$Genus <- factor(pc2_pws_stats_extra_nomod$Genus, levels = c("Other",
"Aggregatibacter",
"Bacteroidetes",
"Campylobacter",
"Capnocytophaga",
"Cardiobacterium",
"Desulfobulbus",
"Dialister",
"Fusobacterium",
"Lautropia",
"Neisseria",
"Phocaeicola",
"Porphyromonas",
"Prevotella",
"Tannerella",
"Treponema"))

pc2_pws_stats_extra_nomod_protein_plot <- pc2_pws_stats_extra_nomod %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protein, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC2 positive, no modern Homo") +
    theme(title = element_text(size=10))

pc2_pws_stats_extra_nomod_protein_plot

# ggsave("../pc2_pws_stats_extra_nomod_protein_plot.pdf", plot = pc2_pws_stats_extra_nomod_protein_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_pws_stats_extra_nomod_hostgenus_plot <- pc2_pws_stats_extra_nomod %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Protein, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC2 positive, no modern Homo") +
    theme(title = element_text(size=10))

pc2_pws_stats_extra_nomod_hostgenus_plot

# ggsave("../pc2_pws_stats_extra_nomod_hostgenus_plot.pdf", plot = pc2_pws_stats_extra_nomod_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

# Now PC2 negative
pc2_neg_stats_extra_nomod <- lapply(pc2_neg_top10_nomod, function(eclass) {
 statssum <- pc2_neg_stats_nomod %>%
   group_by(Host_Genus) %>%
   filter(Protein == eclass) %>%
   filter(Percent <=15) %>%
#   filter(Percent >15) %>%
   summarise(Remaining = sum(Percent))
}) %>%
 bind_rows(.)

pc2_neg_stats_extra_nomod['Protein'] <- c("pc2n7","pc2n7","pc2n7","pc2n7",
                                    "pc2n9","pc2n9","pc2n9","pc2n9",
                                    "pc2n10","pc2n10","pc2n10","pc2n10",
                                    "pc2n1","pc2n1","pc2n1","pc2n1",
                                    #"pc2n2",
                                    "pc2n3","pc2n3","pc2n3","pc2n3",
                                    "pc2n4","pc2n4","pc2n4","pc2n4",
                                    "pc2n5","pc2n5","pc2n5","pc2n5",
                                    "pc2n6","pc2n6","pc2n6","pc2n6",
                                    "pc2n8","pc2n8","pc2n8","pc2n8")
pc2_neg_stats_extra_nomod['Sum'] <- NA
pc2_neg_stats_extra_nomod['Genus'] <- 'Other'
pc2_neg_stats_extra_nomod <- pc2_neg_stats_extra_nomod %>%
  rename(Percent = Remaining) %>%
  select(Protein,Host_Genus,Genus,Sum,Percent)

# remove 'nomod' from the protein names to be able to merge with the columns just made. We don't need 'nomod' in the plots, it looks cleaner without. 
pc2_neg_stats_nomod$Protein <- gsub("nomod", "", pc2_neg_stats_nomod$Protein)

pc2_neg_stats_extra_nomod <- pc2_neg_stats_extra_nomod %>%
  bind_rows(., pc2_neg_stats_nomod %>%
              filter(Percent >15)) %>%
  filter(!str_detect(Genus, "Total"))

pc2_neg_stats_extra_nomod$Protein <- factor(pc2_neg_stats_extra_nomod$Protein, levels = c("pc2n1","pc2n2","pc2n3","pc2n4",
                                      "pc2n5","pc2n6","pc2n7","pc2n8","pc2n9","pc2n10"))

pc2_neg_stats_extra_nomod$Genus <- factor(pc2_neg_stats_extra_nomod$Genus, levels = c("Other",
"Abiotrophia",
"Actinomyces",
"Aggregatibacter",
"Atopobium",
"Capnocytophaga",
"Catonella",
"Filifactor",
"Macrococcus",
"Oribacterium",
"Parabacteroides",
"Parvimonas",
"Porphyromonas",
"Schaalia",
"Stenotrophomonas",
"Streptococcus",
"Treponema"))

pc2_neg_stats_extra_nomod_protein_plot <- pc2_neg_stats_extra_nomod %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Host_Genus, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protein, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC2 negative, no modern Homo") +
    theme(title = element_text(size=10))

pc2_neg_stats_extra_nomod_protein_plot

# ggsave("../pc2_neg_stats_extra_nomod_protein_plot.pdf", plot = pc2_neg_stats_extra_nomod_protein_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

pc2_neg_stats_extra_nomod_hostgenus_plot <- pc2_neg_stats_extra_nomod %>%
  filter(!str_detect(Genus, "Total")) %>%
ggplot(., aes(x=Protein, y=Percent, fill = Genus)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  scale_fill_manual(values = c("#bdbdbd","#440154FF","#39558CFF","#1F968BFF","#74D055FF",
                               "#07071DFF","#6B1D81FF","#D6456CFF","#FEB37BFF",
                               "#453781FF","#287D8EFF","#3CBC75FF","#DCE318FF",
                               "#400F73FF","#AB337CFF","#FA815FFF")) +
  theme(text = element_text(size=18),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Host_Genus, nrow=2) +
  ylab("Percent") +
  ggtitle("Species-level assignents to SEED proteins, PC2 negative, no modern Homo") +
    theme(title = element_text(size=10))

pc2_neg_stats_extra_nomod_hostgenus_plot

# ggsave("../pc2_neg_stats_extra_nomod_hostgenus_plot.pdf", plot = pc2_neg_stats_extra_nomod_hostgenus_plot, device = "pdf",
#        scale = 1, width = 8, height = 5, units = c("in"),
#        dpi = 300)

```

# SEED statistics
Let's calculate some summary statistics for each genus to be able to look at the
differenes in SEED category assignments across hosts. Make individual tables of
stats for each SEED category for each host genus, then merge them into one big
table for comparisons.

Do this for L2, and L4 
*Note that L4 is not the decontaminated, outliers removed, & filtered set*
```{r seed_stats_tables}
# Start with the Level 2 categories

fullset.aadder_L2 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 2, ]
rownames(fullset.aadder_L2) <- fullset.aadder_L2$Category
fullset.aadder_L2 <- fullset.aadder_L2 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  inner_join(., fullmetadata %>% select(SampleID,Host_Genus,Host_Common,Host_General), by = "SampleID") %>%
  select(SampleID, Host_Genus, Host_Common, Host_General, everything(.)) %>%
  gather("Category","Counts",5:ncol(.))

fullset.aadder_L2$Counts <- as.numeric(fullset.aadder_L2$Counts)

# compute summary statistics by pathway for Homo
Homo_stats_pathway <- fullset.aadder_L2 %>%
  dplyr::filter(str_detect(Host_Genus,"Homo")) %>%
  group_by(Category) %>%
  summarise(
    count = n(),
    prevalence = sum(Counts >0),
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Homo_stats_pathway['genus'] <- 'Homo'


# compute summary statistics by pathway for Pan
Pan_stats_pathway <- fullset.aadder_L2 %>%
  dplyr::filter(str_detect(Host_Genus,"Pan")) %>%
  group_by(Category) %>%
  summarise(
    count = n(),
    prevalence = sum(Counts >0),
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Pan_stats_pathway['genus'] <- 'Pan'

# compute summary statistics by pathway for Gorilla
Gorilla_stats_pathway <- fullset.aadder_L2 %>%
  dplyr::filter(str_detect(Host_Genus,"Gorilla")) %>%
  group_by(Category) %>%
  summarise(
    count = n(),
    prevalence = sum(Counts >0),
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Gorilla_stats_pathway['genus'] <- 'Gorilla'

# compute summary statistics by pathway for Alouatta
Alouatta_stats_pathway <- fullset.aadder_L2 %>%
  dplyr::filter(str_detect(Host_Genus,"Alouatta")) %>%
  group_by(Category) %>%
  summarise(
    count = n(),
    prevalence = sum(Counts >0),
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Alouatta_stats_pathway['genus'] <- 'Alouatta'

# Combine the stats columns and use this to look at the differenes between 
#host genera for pathways of interest
allSEED_stats_L2 <- bind_rows(Homo_stats_pathway, Pan_stats_pathway, Gorilla_stats_pathway, Alouatta_stats_pathway)

# For example, what are the differences in abundance of Carbohydrates 
#between host genera? (This is strongly "Homo" based on bi-plots generated below)
allSEED_stats_L2 %>% filter(str_detect(Category, "Carbohydrates|Protein"))


# Now the Level 4 categories
fullset.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
rownames(fullset.aadder_L4) <- fullset.aadder_L4$Category
fullset.aadder_L4 <- fullset.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  inner_join(., fullmetadata %>% select(SampleID,Host_Genus,Host_Common,Host_General), by = "SampleID") %>%
  select(SampleID, Host_Genus, Host_Common, Host_General, everything(.)) %>%
  gather("Category","Counts",5:ncol(.))

fullset.aadder_L4$Counts <- as.numeric(fullset.aadder_L4$Counts)

# compute summary statistics by pathway for Homo
Homo_stats_pathway <- fullset.aadder_L4 %>%
  dplyr::filter(str_detect(Host_Genus,"Homo")) %>%
  group_by(Category) %>%
  summarise(
    count = n(),
    prevalence = sum(Counts >0),
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Homo_stats_pathway['genus'] <- 'Homo'

# compute summary statistics by pathway for Pan
Pan_stats_pathway <- fullset.aadder_L4 %>%
  dplyr::filter(str_detect(Host_Genus,"Pan")) %>%
  group_by(Category) %>%
  summarise(
    count = n(),
    prevalence = sum(Counts >0),
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Pan_stats_pathway['genus'] <- 'Pan'

# compute summary statistics by pathway for Gorilla
Gorilla_stats_pathway <- fullset.aadder_L4 %>%
  dplyr::filter(str_detect(Host_Genus,"Gorilla")) %>%
  group_by(Category) %>%
  summarise(
    count = n(),
    prevalence = sum(Counts >0),
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Gorilla_stats_pathway['genus'] <- 'Gorilla'

# compute summary statistics by pathway for Alouatta
Alouatta_stats_pathway <- fullset.aadder_L4 %>%
  dplyr::filter(str_detect(Host_Genus,"Alouatta")) %>%
  group_by(Category) %>%
  summarise(
    count = n(),
    prevalence = sum(Counts >0),
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )

# add a column that specificies the genus for merging all 4 genus tables together
Alouatta_stats_pathway['genus'] <- 'Alouatta'

# Combine the stats columns and use this to look at the differenes between 
#host genera for pathways of interest
allSEED_stats_L4 <- bind_rows(Homo_stats_pathway, Pan_stats_pathway, Gorilla_stats_pathway, Alouatta_stats_pathway)

# For example, what are the differences in abundance of Serine endopeptidase ScpC 
#between host genera? (This is strongly "Homo" based on bi-plots generated below)
allSEED_stats_L4 %>% filter(str_detect(Category, "Serine\ endopeptidase\ ScpC"))
```

The last thing we need to check is the total number of reads assigned vs not
assigned in each host genus. We want to see if Chimps have a lower proportion of
assigned reads (or a higher proportion of not assigned reads) than the other
host genera. This is different from the SEED category Unassigned, which means
the reads mapped to a protein that for which the SEED category is Unassigned (it
hasn't been specifically placed in a well-defined category such as Protein
Metabolism or Carbohydrates). We want the reads that were not assigned to any
protein.
```{r read_assignment_stats}
# read in the full AADDER table and change the first column title so it doesn't start with #
assigned.aadder <- fread("../04-analysis/screening/aadder/tables/AADDER_refseq_seed_85pc_supp0.01_20190513-ex_seedPath_to_count_summarised_Not_Assigned.txt")
assigned.aadder <- dplyr::rename(assigned.aadder, Category = `#Datasets`)

# Fix the column headers
colnames(assigned.aadder) <- gsub("_S.*_L.*_R1_.*.fastq.merged.prefixed.hg19unmapped.blastn.sam.gz","", colnames(assigned.aadder))
colnames(assigned.aadder) <- gsub("_S.*_L.*_R1_.*.fastq.combined.fq.prefixed.extractunmapped.bam.blastn.sam.gz","", colnames(assigned.aadder))
colnames(assigned.aadder) <- gsub("_S.*_L.*_R1_.*.fastq.merged.blastn.sam.gz","", colnames(assigned.aadder))
colnames(assigned.aadder) <- gsub("_L000_R1_000.fastq.merged.prefixed.hg19unmapped.blastn.sam.gz","", colnames(assigned.aadder))
colnames(assigned.aadder) <- gsub(".A0101.A0101",".A0101", colnames(assigned.aadder))

# transpose the data, create a new column of the total number of assigned reads, and merge it with the metadata. Select out only the columns of interest.
assigned.aadder <- assigned.aadder %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(Category, Counts) %>%
  select(SampleID,`SEED;`,`SEED;Not assigned;`) %>%
  rename(SEED_total_reads = `SEED;`) %>%
  rename(SEED_not_assigned_reads = `SEED;Not assigned;`) %>%
  mutate(SEED_assigned_reads = SEED_total_reads - SEED_not_assigned_reads) %>%
  mutate(SEED_proportion_assigned = (SEED_assigned_reads/SEED_total_reads)*100) %>%
  inner_join(., fullmetadata) %>%
  select(SampleID,Host_Common,Host_Genus,Host_General,SEED_total_reads,SEED_assigned_reads,SEED_not_assigned_reads,SEED_proportion_assigned)

# Now remove the outlier and control samples
assigned.outrem.aadder <- assigned.aadder %>%
  filter(!str_detect(SampleID, outliers_L4))

# Set the host genus as a factor for plotting
assigned.aadder$Host_Common <- factor(assigned.aadder$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))

assigned.aadder$Host_Genus <- factor(assigned.aadder$Host_Genus, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo",
                                          "Plaque", "Gut", "Skin", "Sediment","Control"))

assigned.outrem.aadder$Host_Common <- factor(assigned.outrem.aadder$Host_Common, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "Plaque", "Gut", "Skin", "Sediment", 
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl"))

assigned.outrem.aadder$Host_Genus <- factor(assigned.outrem.aadder$Host_Genus, 
                                levels = c("Alouatta", "Gorilla", "Pan", "Homo",
                                          "Plaque", "Gut", "Skin", "Sediment","Control"))


# graph the total number of assigned reads by host genus
assigned.aadder %>%
  ggplot(., aes(x=Host_Genus, y=SEED_assigned_reads, fill = Host_Genus)) +
  geom_violin() +
  geom_point() +
  theme_minimal() + theme(text = element_text(size = 14)) +
  scale_fill_manual(values = genus_colours) +
  theme(axis.text.x = element_text(hjust = 1)) +
  #scale_y_continuous(trans='log10') +
  ylab("Counts") +
  ggtitle("Total number of reads assigned to a SEED category")  +
  theme(plot.title = element_text(size = 10))

# graph the total number of assigned reads by host genus with outlier samples removed
aadder.assigned_counts <- assigned.outrem.aadder %>%
  ggplot(., aes(x=Host_Genus, y=SEED_assigned_reads, fill = Host_Genus)) +
  geom_violin() +
  geom_point() +
  theme_minimal() + theme(text = element_text(size = 14)) +
  scale_fill_manual(values = genus_colours) +
  theme(axis.text.x = element_text(hjust = 1)) +
  #scale_y_continuous(trans='log10') +
  ylab("Counts") +
  ggtitle("Total number of reads assigned to a SEED category, outlier samples removed") +
  theme(plot.title = element_text(size = 10))

aadder.assigned_counts

# ggsave("../aadder.assigned_counts.pdf", plot = aadder.assigned_counts, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

# graph the proportion of reads assigned for each category
aadder.assigned_proportions <- assigned.aadder %>%
  ggplot(., aes(x=Host_Common, y=SEED_proportion_assigned)) +
  geom_boxplot() +
  geom_point(size = 2, aes(colour = Host_Common, shape = Host_Common)) +
   scale_colour_manual(values = common_colours) +
   scale_shape_manual(values = common_shapes) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ylab("Percent") +
  ggtitle("Percent of reads assigned to a SEED category") +
  theme(plot.title = element_text(size = 10))

aadder.assigned_proportions
aadder.assigned_proportions <- assigned.aadder %>%
  ggplot(., aes(x=Host_Common, y=SEED_proportion_assigned)) +
  geom_boxplot() +
  geom_point(size = 2, aes(colour = Host_General, shape = Host_General)) +
   scale_colour_manual(values = general_colors) +
   scale_shape_manual(values = general_shapes) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ylab("Percent") +
  ggtitle("Percent of reads assigned to a SEED category") +
  theme(plot.title = element_text(size = 10))

aadder.assigned_proportions

# ggsave("../aadder.assigned_proportions.pdf", plot = aadder.assigned_proportions, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```

And look at the % of reads that were assigned to the Carbohydrate category, at
both level 2 and level 4. *Note this is not decontaminated/filtered*
```{r seed_percent_carbs}
# now get the % of reads that were assigned to the Carbohydrate category, at Level 2 and remove outliers
aadder.carbs_L2 <- as.data.frame(fullset.aadder) %>%
  filter(str_detect(Category, "Carbohydrates"))

# filter out just the level 4 Carbohydrate categories
aadder.carbs_L2 <- as.tibble(aadder.carbs_L2)
aadder.carbs_L2 <- aadder.carbs_L2[str_count(aadder.carbs_L2$Category, ";") == 2, ]
rownames(aadder.carbs_L2) <- aadder.carbs_L2$Category

aadder.carbs_pct <- aadder.carbs_L2 %>%
  gather("SampleID","CPM",2:ncol(.)) %>%
  spread(Category,CPM) %>%
  adorn_totals(., where = "col", name = "Carbohydrates_L2") %>%
  select(SampleID,Carbohydrates_L2) %>%
  filter(!str_detect(SampleID, outliers_L4))

aadder.carbs_pct <- aadder.carbs_pct %>%
  inner_join(., assigned.aadder, by = "SampleID") %>%
  select(SampleID,Host_Genus,Host_Common,Host_General,SEED_total_reads,SEED_assigned_reads,SEED_not_assigned_reads,SEED_proportion_assigned,Carbohydrates_L2) %>%
  mutate(Percent_carbs = Carbohydrates_L2/SEED_assigned_reads*100)

aadder.carbs_pct_L2.plot <- aadder.carbs_pct %>%
  filter(!str_detect(Host_Common,"Gut|Skin|Sediment|ExtractionControl|EnvironmentalControl|LibraryControl")) %>%
  ggplot(., aes(x=Host_Common, y=Percent_carbs)) +
  geom_boxplot() +
  geom_point(size = 3, aes(colour = Host_Common, shape = Host_Common)) +
  scale_colour_manual(values = common_colours) +
  scale_shape_manual(values = common_shapes) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  ylab("Percent") +
  ggtitle("Percent of reads assigned to the SEED Carbohydrate category L2, outliers removed") +
  theme(plot.title = element_text(size = 10))

aadder.carbs_pct_L2.plot

aadder.carbs_pct_L2.plot <- aadder.carbs_pct %>%
  filter(!str_detect(Host_Common,"Gut|Skin|Sediment|ExtractionControl|EnvironmentalControl|LibraryControl")) %>%
  ggplot(., aes(x=Host_Common, y=Percent_carbs)) +
  geom_boxplot() +
  geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
  scale_colour_manual(values = general_colors) +
  scale_shape_manual(values = general_shapes) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  ylab("Percent") +
  ggtitle("Percent of reads assigned to the SEED Carbohydrate category L2, outliers removed") +
  theme(plot.title = element_text(size = 10))

aadder.carbs_pct_L2.plot

# ggsave("../aadder.carbs_pct_L2.plot.pdf", plot = aadder.carbs_pct_L2.plot, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)


# now get the % of reads that were assigned to the Carbohydrate category, at Level 4
aadder.carbs_L4 <- as.data.frame(fullset.aadder) %>%
  filter(str_detect(Category, "Carbohydrates"))

# filter out just the level 4 Carbohydrate categories
aadder.carbs_L4 <- as.tibble(aadder.carbs_L4)
aadder.carbs_L4 <- aadder.carbs_L4[str_count(aadder.carbs_L4$Category, ";") == 4, ]
rownames(aadder.carbs_L4) <- aadder.carbs_L4$Category


aadder.carbs_pct <- aadder.carbs_L4 %>%
  gather("SampleID","CPM",2:ncol(.)) %>%
  spread(Category,CPM) %>%
  adorn_totals(., where = "col", name = "Carbohydrates_L4") %>%
  select(SampleID,Carbohydrates_L4) %>%
  filter(!str_detect(SampleID, outliers_L4))

aadder.carbs_pct <- aadder.carbs_pct %>%
  inner_join(., assigned.aadder, by = "SampleID") %>%
  select(SampleID,Host_Genus,Host_General,Host_Common,SEED_total_reads,SEED_assigned_reads,SEED_not_assigned_reads,SEED_proportion_assigned,Carbohydrates_L4) %>%
  mutate(Percent_carbs = Carbohydrates_L4/SEED_assigned_reads*100)

aadder.carbs_pct_L4.plot <- aadder.carbs_pct %>%
  filter(!str_detect(Host_Common,"Gut|Skin|Sediment|ExtractionControl|EnvironmentalControl|LibraryControl")) %>%
  ggplot(., aes(x=Host_Common, y=Percent_carbs)) +
  geom_boxplot() +
  geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
  scale_colour_manual(values = general_colors) +
  scale_shape_manual(values = general_shapes) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  ylab("Percent") +
  ggtitle("Percent of reads assigned to the SEED Carbohydrate category L4, outliers removed") +
  theme(plot.title = element_text(size = 10))

aadder.carbs_pct_L4.plot

# ggsave("../aadder.carbs_pct_L4.plot.pdf", plot = aadder.carbs_pct_L4.plot, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

```

Let's make graphs of Carbohydrate entry prevalence and abundance
```{r carbs_prev_abund}
# now get the % of reads that were assigned to the Carbohydrate category, at Level 4
aadder.carbs_L4 <- as.data.frame(fullset.aadder) %>%
  filter(str_detect(Category, "Carbohydrates"))

# filter out just the level 4 Carbohydrate categories
aadder.carbs_L4 <- as.tibble(aadder.carbs_L4)
aadder.carbs_L4 <- aadder.carbs_L4[str_count(aadder.carbs_L4$Category, ";") == 4, ]
rownames(aadder.carbs_L4) <- aadder.carbs_L4$Category


# make Relab binary to make a presence/absence table
aadder.carbs_prev <- aadder.carbs_L4 %>%
  gather("SampleID","CPM",2:ncol(.)) %>%
  rename(Present = CPM)
aadder.carbs_prev$Present[aadder.carbs_prev$Present>0] <- 1

aadder.carbs_prev <- aadder.carbs_prev %>%
  spread(Category,Present) %>%
  adorn_totals(., where = "col", name = "Prevalence") %>%
  select(SampleID,Prevalence) %>%
  inner_join(., fullmetadata) %>%
  filter(!str_detect(SampleID, outliers_L4)) %>%
  select(SampleID,Host_Genus,Host_Common,Host_General,Prevalence)

aadder.carbs_prev$Host_Common <- factor(aadder.carbs_prev$Host_Common, levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))
aadder.carbs_prev$Host_Genus <- factor(aadder.carbs_prev$Host_Genus, levels = c("Alouatta", "Gorilla", "Pan", "Homo", "Plaque"))

aadder.carbs_prev_L4.plot <- aadder.carbs_prev %>%
  filter(!str_detect(Host_Common,"Gut|Skin|Sediment|ExtractionControl|EnvironmentalControl|LibraryControl")) %>%
  ggplot(., aes(x=Host_Common, y=Prevalence)) +
  geom_boxplot() +
  geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
  scale_colour_manual(values = general_colors) +
  scale_shape_manual(values = general_shapes) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  ylab("Number of Enzymes") +
  ggtitle("Number of enzymes in the SEED Carbohydrate category L4, outliers removed") +
  theme(plot.title = element_text(size = 10))

aadder.carbs_prev_L4.plot

aadder.carbs_prev <- aadder.carbs_prev %>%
  filter(!str_detect(Host_Common,"Gut|Skin|Sediment|ExtractionControl|EnvironmentalControl|LibraryControl"))
  
pairwise.wilcox.test(aadder.carbs_prev$Prevalence, aadder.carbs_prev$Host_Genus,
                     p.adjust.method = "fdr")

pairwise.wilcox.test(aadder.carbs_prev$Prevalence, aadder.carbs_prev$Host_Common,
                     p.adjust.method = "fdr")

aadder.carbs_prev_nomod <- aadder.carbs_prev %>% filter(!str_detect(Host_General, "ModernDayHuman"))
pairwise.wilcox.test(aadder.carbs_prev_nomod$Prevalence, aadder.carbs_prev_nomod$Host_Common,
                     p.adjust.method = "fdr")

# ggsave("../aadder.carbs_prev_L4.plot.pdf", plot = aadder.carbs_prev_L4.plot, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)


aadder.carbs_abund <- aadder.carbs_L4 %>%
  gather("SampleID","CPM",2:ncol(.)) %>%
  spread(Category,CPM) %>%
  adorn_totals(., where = "col", name = "Abundance") %>%
  select(SampleID,Abundance) %>%
  inner_join(., fullmetadata) %>%
  filter(!str_detect(SampleID, outliers_L4)) %>%
  select(SampleID,Host_Genus,Host_Common,Host_General,Abundance)

aadder.carbs_abund$Host_Common <- factor(aadder.carbs_abund$Host_Common, levels = c("Alouatta", "Gorilla", "Pan", "Homo (Modern Human)", "Homo (Neanderthal)",
                                          "EnvironmentalControl","ExtractionControl", "LibraryControl", "Plaque", "Gut", 
                                          "Skin", "Sediment"))
aadder.carbs_abund$Host_Genus <- factor(aadder.carbs_abund$Host_Genus, levels = c("Alouatta", "Gorilla", "Pan", "Homo", "Plaque"))

require(scales)
aadder.carbs_abund_L4.plot <- aadder.carbs_abund %>%
  filter(!str_detect(Host_Common,"Gut|Skin|Sediment|ExtractionControl|EnvironmentalControl|LibraryControl")) %>%
  ggplot(., aes(x=Host_Common, y=Abundance)) +
  geom_boxplot() +
  geom_point(size = 3, aes(colour = Host_General, shape = Host_General)) +
  scale_colour_manual(values = general_colors) +
  scale_shape_manual(values = general_shapes) +
  theme_minimal() + theme(text = element_text(size = 14)) +
  theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  ylab("Abundance") +
  scale_y_continuous(trans='log10', labels=comma) +
  ggtitle("Abundance of reads assigned to the SEED Carbohydrate category L4, outliers removed") +
  theme(plot.title = element_text(size = 10))

aadder.carbs_abund_L4.plot

# ggsave("../aadder.carbs_abund_L4.plot.pdf", plot = aadder.carbs_abund_L4.plot, device = "pdf",
#        scale = 1, width = 7, height = 5, units = c("in"),
#        dpi = 300)

aadder.carbs_abund <- aadder.carbs_abund %>%
  filter(!str_detect(Host_Common,"Gut|Skin|Sediment|ExtractionControl|EnvironmentalControl|LibraryControl"))
  
pairwise.wilcox.test(aadder.carbs_abund$Abundance, aadder.carbs_abund$Host_Genus,
                     p.adjust.method = "fdr")

pairwise.wilcox.test(aadder.carbs_abund$Abundance, aadder.carbs_abund$Host_Common,
                     p.adjust.method = "fdr")

aadder.carbs_abund_nomod <- aadder.carbs_abund %>% filter(!str_detect(Host_General, "ModernDayHuman"))
pairwise.wilcox.test(aadder.carbs_abund_nomod$Abundance, aadder.carbs_abund_nomod$Host_Common,
                     p.adjust.method = "fdr")

```

And now finally let's make a table of summary statistics to compare the host genera
```{r seed_summary_stats}
# compute summary statistics by pathway for Homo
Homo_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Homo")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/mean*100),
    count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )  %>%
  filter(!str_detect(SEED_reads, "SEED_total_reads"))


Homo_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Homo")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/sum(Homo_SEED_assignment_stats$mean)*100),
    count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )  %>%
  filter(!str_detect(SEED_reads, "SEED_total_reads"))

# add a column that specificies the genus for merging all 4 genus tables together
Homo_SEED_assignment_stats['genus'] <- 'Homo'

# compute summary statistics by pathway for Gorilla
Gorilla_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Gorilla")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/mean*100),
    count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  ) %>%
    filter(!str_detect(SEED_reads, "SEED_total_reads"))

Gorilla_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Gorilla")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/sum(Gorilla_SEED_assignment_stats$mean)*100),
    count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  ) %>%
    filter(!str_detect(SEED_reads, "SEED_total_reads"))

# add a column that specificies the genus for merging all 4 genus tables together
Gorilla_SEED_assignment_stats['genus'] <- 'Gorilla'

# compute summary statistics by pathway for Pan
Pan_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Pan")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/mean*100),
   count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  ) %>%
  filter(!str_detect(SEED_reads, "SEED_total_reads"))

Pan_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Pan")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/sum(Pan_SEED_assignment_stats$mean)*100),
   count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  ) %>%
  filter(!str_detect(SEED_reads, "SEED_total_reads"))

# add a column that specificies the genus for merging all 4 genus tables together
Pan_SEED_assignment_stats['genus'] <- 'Pan'

# compute summary statistics by pathway for Alouatta
Alouatta_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Alouatta")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/mean*100),
    count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )  %>%
  filter(!str_detect(SEED_reads, "SEED_total_reads"))

Alouatta_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Alouatta")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/sum(Alouatta_SEED_assignment_stats$mean)*100),
    count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )  %>%
  filter(!str_detect(SEED_reads, "SEED_total_reads"))

# add a column that specificies the genus for merging all 4 genus tables together
Alouatta_SEED_assignment_stats['genus'] <- 'Alouatta'

# compute summary statistics by pathway for Alouatta
Plaque_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Plaque")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/mean*100),
    count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )  %>%
  filter(!str_detect(SEED_reads, "SEED_total_reads"))

Plaque_SEED_assignment_stats <- assigned.outrem.aadder %>%
  gather("SEED_reads","Counts", 5:ncol(.)) %>%
  filter(str_detect(Host_Genus,"Plaque")) %>%
  group_by(SEED_reads) %>%
  summarise(
    mean = mean(Counts),
    sd = sd(Counts),
    median = median(Counts),
    percent = (mean(Counts)/sum(Plaque_SEED_assignment_stats$mean)*100),
    count = n(),
    prevalence = sum(Counts >0),
    se = qnorm(.975) * sd / sqrt(count),
    IQR = IQR(Counts)
  )  %>%
  filter(!str_detect(SEED_reads, "SEED_total_reads"))

# add a column that specificies the genus for merging all 4 genus tables together
Plaque_SEED_assignment_stats['genus'] <- 'Plaque'

# Combine the stats columns and use this to look at the differenes between 
#host genera for pathways of interest
allSEED_assignment_stats <- bind_rows(Plaque_SEED_assignment_stats, Homo_SEED_assignment_stats, Pan_SEED_assignment_stats, Gorilla_SEED_assignment_stats, Alouatta_SEED_assignment_stats)
allSEED_assignment_stats <- allSEED_assignment_stats %>% select(genus, everything())
```

## Heat maps
Make heat maps of the top 10 proteins in each loading (PC1 positive, PC1 negative, PC2 positive, PC2 negative) to look at distribution in host genera. 

First we'll set the order of the samples for the heatmap, so they're based on host genus, and not clustered in the hierarchical clustering step.  
```{r, heatmap_sample_order}
# set the sample order for including modern plaque samples
sampleorder.oral = as.vector(c("OME002.A0101","OME003.A0101","OME005.A0101","OME007.A0101","OME009.A0101",
                               "ABM006.A0101","DJA002.A0101","EBO001.A0101","EBO003.A0101","FDM001.A0101","GDC002.A0101",
                               "GDC004.A0101","GDC005.A0101","IBA001.A0101","MTK001.A0101","MTM001.A0101","MTM004.A0101",
                               "MTM005.A0101","MTM008.A0101","MTM010.A0101","MTM011.A0101",
                               "MTS001.A0101","MTS002.A0101","ABM007.A0101","ABM008.A0101","CDC006.A0101",
                               "CDC008.A0101","CDC009.A0101","CDC010.A0101","CDC011.A0101","Chimp.150519",
                               "DJA007.A0101","EBO008.A0101","EBO009.A0101","KNP001.A0102",
                               "KNP003.A0102","KNP005.A0102","KNP006.A0102","KNP007.A0102","KNP009.A0102",
                               "ElSidron1","ElSidron2","SpyOld","FUM001.A0101","FUM002.A0101","GDN001.A0101",
                               "GOY005.A0101","GOY006.A0101","PES001.B0101","VLD001.A0101","ECO002.B0101",
                               "ECO004.B0101","EMN001.A0101","MOA001.A0101","OAK002.A0101",
                               "OAK003.A0101","OAK004.A0101","OAK005.A0101","OFN001.A0101","RIG001.A0101",
                               "PLV001.A0101","TAF008.B0101","TAF017.A0101","TAF018.A0101","ESA003.A0101",
                               "ESA004.A0101","ESA006.A0101","ESA007.A0101","ESA008.A0101","LIS001.A0101",
                               "LIS002.A0101","LIS003.A0101","PYK001.A0101","PYK002.A0101","PYK003.A0101",
                               "PYK004.A0101","PYK005.A0101","JAE006.A0101","JAE007.A0101","JAE008.A0101",
                               "JAE009.A0101","JAE010.A0101","JAE012.A0101","JAE013.A0101","JAE014.A0101",
                               "JAE015.A0101","JAE016.A0101","VLC001.A0101","VLC002.A0101","VLC003.A0101",
                               "VLC004.A0101","VLC005.A0101","VLC008.A0101","VLC009.A0101","VLC010.A0101",
                               "SRR061192","SRR061294","SRR061320","SRR061365","SRR061562","SRR062298",
                               "SRR062299","SRR063517","SRR1804664","SRR1804823","SRR512767",
                               "SRR513165","SRR513768","SRR513775","SRR513828","SRR514202",
                               "SRR514239","SRR514306","SRR514329"))

# set the sample order for excluding modern plaque samples
sampleorder.calc = as.vector(c("OME002.A0101","OME003.A0101","OME005.A0101","OME007.A0101","OME009.A0101",
                               "ABM006.A0101","DJA002.A0101","EBO001.A0101","EBO003.A0101","FDM001.A0101","GDC002.A0101",
                               "GDC004.A0101","GDC005.A0101","IBA001.A0101","MTK001.A0101","MTM001.A0101","MTM004.A0101",
                               "MTM005.A0101","MTM008.A0101","MTM010.A0101","MTM011.A0101",
                               "MTS001.A0101","MTS002.A0101","ABM007.A0101","ABM008.A0101","CDC006.A0101",
                               "CDC008.A0101","CDC009.A0101","CDC010.A0101","CDC011.A0101","Chimp.150519",
                               "DJA007.A0101","EBO008.A0101","EBO009.A0101","KNP001.A0102",
                               "KNP003.A0102","KNP005.A0102","KNP006.A0102","KNP007.A0102","KNP009.A0102",
                               "ElSidron1","ElSidron2","SpyOld","FUM001.A0101","FUM002.A0101","GDN001.A0101",
                               "GOY005.A0101","GOY006.A0101","PES001.B0101","VLD001.A0101","ECO002.B0101",
                               "ECO004.B0101","EMN001.A0101","MOA001.A0101","OAK002.A0101",
                               "OAK003.A0101","OAK004.A0101","OAK005.A0101","OFN001.A0101","RIG001.A0101",
                               "PLV001.A0101","TAF008.B0101","TAF017.A0101","TAF018.A0101","ESA003.A0101",
                               "ESA004.A0101","ESA006.A0101","ESA007.A0101","ESA008.A0101","LIS001.A0101",
                               "LIS002.A0101","LIS003.A0101","PYK001.A0101","PYK002.A0101","PYK003.A0101",
                               "PYK004.A0101","PYK005.A0101","JAE006.A0101","JAE007.A0101","JAE008.A0101",
                               "JAE009.A0101","JAE010.A0101","JAE012.A0101","JAE013.A0101","JAE014.A0101",
                               "JAE015.A0101","JAE016.A0101","VLC001.A0101","VLC002.A0101","VLC003.A0101",
                               "VLC004.A0101","VLC005.A0101","VLC008.A0101","VLC009.A0101","VLC010.A0101"))


sampleorder.nomod.calc = as.vector(c("OME002.A0101","OME003.A0101","OME005.A0101","OME007.A0101","OME009.A0101",
                               "ABM006.A0101","DJA002.A0101","EBO001.A0101","EBO003.A0101","FDM001.A0101","GDC002.A0101",
                               "GDC004.A0101","GDC005.A0101","IBA001.A0101","MTK001.A0101","MTM001.A0101","MTM004.A0101",
                               "MTM005.A0101","MTM008.A0101","MTM010.A0101","MTM011.A0101",
                               "MTS001.A0101","MTS002.A0101","ABM007.A0101","ABM008.A0101","CDC006.A0101",
                               "CDC008.A0101","CDC009.A0101","CDC010.A0101","CDC011.A0101","Chimp.150519",
                               "DJA007.A0101","EBO008.A0101","EBO009.A0101","KNP001.A0102",
                               "KNP003.A0102","KNP005.A0102","KNP006.A0102","KNP007.A0102","KNP009.A0102",
                               "ElSidron1","ElSidron2","SpyOld","FUM001.A0101","FUM002.A0101","GDN001.A0101",
                               "GOY005.A0101","GOY006.A0101","PES001.B0101","VLD001.A0101","ECO002.B0101",
                               "ECO004.B0101","EMN001.A0101","MOA001.A0101","OAK002.A0101",
                               "OAK003.A0101","OAK004.A0101","OAK005.A0101","OFN001.A0101","RIG001.A0101",
                               "PLV001.A0101","TAF008.B0101","TAF017.A0101","TAF018.A0101","ESA003.A0101",
                               "ESA004.A0101","ESA006.A0101","ESA007.A0101","ESA008.A0101","LIS001.A0101",
                               "LIS002.A0101","LIS003.A0101","PYK001.A0101","PYK002.A0101","PYK003.A0101",
                               "PYK004.A0101","PYK005.A0101"))

protein_code_list_all <- protein_code_list %>% full_join(.,protein_code_list_nomod) %>% select(PC1code,PC2code,PC1nomodcode,PC2nomodcode,Protein) %>% mutate_at(c('PC1code','PC2code','PC1nomodcode','PC2nomodcode'), replace_na, "none")

protein_code_list_all$PC1code <- factor(protein_code_list_all$PC1code, levels = c("pc1n1","pc1n2","pc1n3","pc1n4","pc1n5","pc1n6","pc1n7","pc1n8","pc1n9","pc1n10","pc1p1","pc1p2","pc1p3","pc1p4","pc1p5","pc1p6","pc1p7","pc1p8","pc1p9","pc1p10","none"))
protein_code_list_all$PC2code <- factor(protein_code_list_all$PC2code, levels = c("pc2n1","pc2n2","pc2n3","pc2n4","pc2n5","pc2n6","pc2n7","pc2n8","pc2n9","pc2n10","pc2p1","pc2p2","pc2p3","pc2p4","pc2p5","pc2p6","pc2p7","pc2p8","pc2p9","pc2p10","none"))

proteinorderPC1n = as.vector(c("pc1n1","pc1n2","pc1n3","pc1n4","pc1n5","pc1n6","pc1n7","pc1n8","pc1n9","pc1n10"))
proteinorderPC1p = as.vector(c("pc1p1","pc1p2","pc1p3","pc1p4","pc1p5","pc1p6","pc1p7","pc1p8","pc1p9","pc1p10"))
proteinorderPC2n = as.vector(c("pc2n1","pc2n2","pc2n3","pc2n4","pc2n5","pc2n6","pc2n7","pc2n8","pc2n9","pc2n10"))
proteinorderPC2p = as.vector(c("pc2p1","pc2p2","pc2p3","pc2p4","pc2p5","pc2p6","pc2p7","pc2p8","pc2p9","pc2p10"))

proteinorder.nomod = as.vector(c())

```


```{r heat_maps}
# make a list of the top 10 proteins with strongest loadings, unique
top10_SEED <- bind_rows(pc1_pws,pc1_neg)
top10_SEED <- bind_rows(top10_SEED,pc2_pws)
top10_SEED <- bind_rows(top10_SEED,pc2_neg)
top10_SEED <- distinct(top10_SEED %>% select(seed_id))
#top10_SEED <- str_c(top10_SEED$enzyme, collapse = "|")

# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., top10_SEED, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$seed_id
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-seed_id))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("seed_id") %>%
  mutate(enzyme = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           })) %>%
  select(enzyme, everything()) %>%
  select(-seed_id) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(enzyme,Counts)

rownames(decontam.outrem.calc.aadder_L4_clr) <- decontam.outrem.calc.aadder_L4_clr$SampleID
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- as.matrix(decontam.outrem.calc.aadder_L4_clr)
rownames(decontam.outrem.calc.aadder_L4_clr.ordered) <- sampleorder.calc

# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered, Colv = T, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins")

```

Instead let's plot individual heatmaps for PC1n, PC1p, PC2n, PC2p
```{r individual_heatmaps_pc1n}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., pc1_neg, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$PC1code
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-starts_with("PC")) %>% select(-one_of("seed_id","Protein")))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("PC1code") %>%
  select(PC1code, everything()) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(PC1code,Counts)

rownames(decontam.outrem.calc.aadder_L4_clr) <- decontam.outrem.calc.aadder_L4_clr$SampleID
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered_pc1n <- as.matrix(decontam.outrem.calc.aadder_L4_clr.ordered)
rownames(decontam.outrem.calc.aadder_L4_clr.ordered_pc1n) <- sampleorder.calc
colnames(decontam.outrem.calc.aadder_L4_clr.ordered_pc1n) <- proteinorderPC1n

# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap_pc1n <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered_pc1n, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins PC1 negative")

```

```{r individual_heatmaps_pc1p}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., pc1_pws, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$PC1code
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-starts_with("PC")) %>% select(-one_of("seed_id","Protein")))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("PC1code") %>%
  select(PC1code, everything()) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(PC1code,Counts)

rownames(decontam.outrem.calc.aadder_L4_clr) <- decontam.outrem.calc.aadder_L4_clr$SampleID
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered_pc1p <- as.matrix(decontam.outrem.calc.aadder_L4_clr.ordered)
rownames(decontam.outrem.calc.aadder_L4_clr.ordered_pc1p) <- sampleorder.calc
colnames(decontam.outrem.calc.aadder_L4_clr.ordered_pc1p) <- proteinorderPC1p

# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap_pc1p <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered_pc1p, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins PC1 negative")

```

Instead let's plot individual heatmaps for PC1n, PC1p, PC2n, PC2p
```{r individual_heatmaps_pc2n}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., pc2_neg, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$PC2code
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-starts_with("PC")) %>% select(-one_of("seed_id","Protein")))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("PC1code") %>%
  select(PC1code, everything()) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(PC1code,Counts)

rownames(decontam.outrem.calc.aadder_L4_clr) <- decontam.outrem.calc.aadder_L4_clr$SampleID
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered_pc2n <- as.matrix(decontam.outrem.calc.aadder_L4_clr.ordered)
rownames(decontam.outrem.calc.aadder_L4_clr.ordered_pc2n) <- sampleorder.calc
colnames(decontam.outrem.calc.aadder_L4_clr.ordered_pc2n) <- proteinorderPC2n

# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap_pc2n <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered_pc2n, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins PC1 negative")

```

```{r individual_heatmaps_pc2p}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., pc2_pws, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$PC2code
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-starts_with("PC")) %>% select(-one_of("seed_id","Protein")))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("PC1code") %>%
  select(PC1code, everything()) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(PC1code,Counts)

rownames(decontam.outrem.calc.aadder_L4_clr) <- decontam.outrem.calc.aadder_L4_clr$SampleID
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered_pc2p <- as.matrix(decontam.outrem.calc.aadder_L4_clr.ordered)
rownames(decontam.outrem.calc.aadder_L4_clr.ordered_pc2p) <- sampleorder.calc
colnames(decontam.outrem.calc.aadder_L4_clr.ordered_pc2p) <- proteinorderPC2p

# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap_pc2p <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered_pc2p, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins PC1 negative")

```

Make heat maps of the top 10 proteins in each loading (PC1 positive, PC1
negative, PC2 positive, PC2 negative) to look at distribution in host genera
without modern *Homo.*
```{r heat_maps_nomod}
# make a list of the top 10 proteins with strongest loadings, unique
top10_SEED <- bind_rows(pc1_pws_nomod,pc1_neg_nomod)
top10_SEED <- bind_rows(top10_SEED,pc2_pws_nomod)
top10_SEED <- bind_rows(top10_SEED,pc2_neg_nomod)
top10_SEED <- distinct(top10_SEED %>% select(seed_id))
#top10_SEED <- str_c(top10_SEED$enzyme, collapse = "|")

# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers, filter for >0.05
decontam.outrem.calc.aadder_L4_nomod <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  filter(!str_detect(SampleID, "JAE|VLC")) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., top10_SEED, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4_nomod) <- decontam.outrem.calc.aadder_L4_nomod$seed_id
decontam.outrem.calc.aadder_L4_nomod <- as.matrix(decontam.outrem.calc.aadder_L4_nomod %>% select(-seed_id))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_nomod_clr = decontam.outrem.calc.aadder_L4_nomod+1
decontam.outrem.calc.aadder_L4_nomod_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_nomod_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_nomod_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_nomod_clr)
decontam.outrem.calc.aadder_L4_nomod_clr <- decontam.outrem.calc.aadder_L4_nomod_clr %>%
  rownames_to_column("seed_id") %>%
  mutate(enzyme = sapply(seed_id, function(f) {
                                  unlist(str_split(f, ";"))[4]
                           })) %>%
  select(enzyme, everything()) %>%
  select(-seed_id) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(enzyme,Counts)

rownames(decontam.outrem.calc.aadder_L4_nomod_clr) <- decontam.outrem.calc.aadder_L4_nomod_clr$SampleID
decontam.outrem.calc.aadder_L4_nomod_clr <- decontam.outrem.calc.aadder_L4_nomod_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_nomod_clr <- decontam.outrem.calc.aadder_L4_nomod_clr[match(sampleorder.nomod.calc, rownames(decontam.outrem.calc.aadder_L4_nomod_clr)),]

decontam.outrem.calc.aadder_L4_nomod_clr.ordered <- as.matrix(decontam.outrem.calc.aadder_L4_nomod_clr)
rownames(decontam.outrem.calc.aadder_L4_nomod_clr) <- sampleorder.nomod.calc

decontam.outrem.calc.aadder_L4_nomod_clr <- as.matrix(decontam.outrem.calc.aadder_L4_nomod_clr)

# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_nomod_clr.heatmap <- heatmap.2(decontam.outrem.calc.aadder_L4_nomod_clr.ordered, Colv = T, Rowv = T, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins")

```

Now repeat the individual heat maps with the proteins from the 
PCA excluding modern humans.
```{r individual_heatmaps_pc1n_nomod}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  filter(!str_detect(SampleID,"JAE|VLC")) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., pc1_neg_nomod, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$PC1nomodcode
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-starts_with("PC")) %>% select(-one_of("seed_id","Protein")))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("PC1code") %>%
  select(PC1code, everything()) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(PC1code,Counts)

rownames(decontam.outrem.calc.aadder_L4_clr) <- decontam.outrem.calc.aadder_L4_clr$SampleID
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.nomod.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.nomod.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered_pc1n_nomod <- as.matrix(decontam.outrem.calc.aadder_L4_clr.ordered)
rownames(decontam.outrem.calc.aadder_L4_clr.ordered_pc1n_nomod) <- sampleorder.nomod.calc
colnames(decontam.outrem.calc.aadder_L4_clr.ordered_pc1n_nomod) <- proteinorderPC1n


# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap_pc1n_nomod <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered_pc1n_nomod, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins PC1 negative")

```


```{r individual_heatmaps_pc1p_nomod}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  filter(!str_detect(SampleID,"JAE|VLC")) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., pc1_pws_nomod, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$PC1nomodcode
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-starts_with("PC")) %>% select(-one_of("seed_id","Protein")))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("PC1code") %>%
  select(PC1code, everything()) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(PC1code,Counts)

rownames(decontam.outrem.calc.aadder_L4_clr) <- decontam.outrem.calc.aadder_L4_clr$SampleID
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.nomod.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.nomod.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered_pc1p_nomod <- as.matrix(decontam.outrem.calc.aadder_L4_clr.ordered)
rownames(decontam.outrem.calc.aadder_L4_clr.ordered_pc1p_nomod) <- sampleorder.nomod.calc
colnames(decontam.outrem.calc.aadder_L4_clr.ordered_pc1p_nomod) <- proteinorderPC1p


# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap_pc1p_nomod <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered_pc1p_nomod, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins PC1 positive")

```


```{r individual_heatmaps_pc2n_nomod}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  filter(!str_detect(SampleID,"JAE|VLC")) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., pc2_neg_nomod, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$PC2nomodcode
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-starts_with("PC")) %>% select(-one_of("seed_id","Protein")))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("PC2code") %>%
  select(PC2code, everything()) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(PC2code,Counts)

 
# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap_pc2n_nomod <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered_pc2n_nomod, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins PC2 negative")

```


```{r individual_heatmaps_pc2p_nomod}
# select out just the enzymes
decontam.outrem.calc.aadder_L4 <- fullset.aadder[str_count(fullset.aadder$Category, ";") == 4, ]
decontam.outrem.calc.aadder_L4 <- anti_join(decontam.outrem.calc.aadder_L4, contaminants)
rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$Category

# remove outliers
decontam.outrem.calc.aadder_L4 <- decontam.outrem.calc.aadder_L4 %>%
  gather("SampleID","Counts",2:ncol(.)) %>%
  spread(Category,Counts) %>%
  filter(!str_detect(SampleID,outliers_L4)) %>%
  filter(!str_detect(SampleID,"JAE|VLC")) %>%
  inner_join(., calcSamples, by = "SampleID") %>%
  gather("seed_id","Counts",2:ncol(.)) %>%
  spread(SampleID,Counts) %>%
  inner_join(., pc2_pws_nomod, by = "seed_id")

rownames(decontam.outrem.calc.aadder_L4) <- decontam.outrem.calc.aadder_L4$PC2nomodcode
decontam.outrem.calc.aadder_L4 <- as.matrix(decontam.outrem.calc.aadder_L4 %>% select(-starts_with("PC")) %>% select(-one_of("seed_id","Protein")))

# CLR-transform the matrix for plotting the heatmap
decontam.outrem.calc.aadder_L4_clr = decontam.outrem.calc.aadder_L4+1
decontam.outrem.calc.aadder_L4_clr <- as.matrix(logratio.transfo(decontam.outrem.calc.aadder_L4_clr, logratio = "CLR"))

decontam.outrem.calc.aadder_L4_clr <- as.data.frame.matrix(decontam.outrem.calc.aadder_L4_clr)
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>%
  rownames_to_column("PC2code") %>%
  select(PC2code, everything()) %>%
  gather("SampleID","Counts", 2:ncol(.)) %>%
  spread(PC2code,Counts)

rownames(decontam.outrem.calc.aadder_L4_clr) <- decontam.outrem.calc.aadder_L4_clr$SampleID
decontam.outrem.calc.aadder_L4_clr <- decontam.outrem.calc.aadder_L4_clr %>% select(-SampleID)

# reorder the table so the rows match as above
decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.nomod.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered <- decontam.outrem.calc.aadder_L4_clr[match(sampleorder.nomod.calc, rownames(decontam.outrem.calc.aadder_L4_clr)),]

decontam.outrem.calc.aadder_L4_clr.ordered_pc2p_nomod <- as.matrix(decontam.outrem.calc.aadder_L4_clr.ordered)
rownames(decontam.outrem.calc.aadder_L4_clr.ordered_pc2p_nomod) <- sampleorder.nomod.calc
colnames(decontam.outrem.calc.aadder_L4_clr.ordered_pc2p_nomod) <- proteinorderPC2p


# the heatmaps don't plot properly in the R markdown file and need to be plotted separately in a plain R script
# make the heatmap
# decontam.outrem.calc.aadder_L4_clr.heatmap_pc2p_nomod <- heatmap.2(decontam.outrem.calc.aadder_L4_clr.ordered_pc2p_nomod, Colv = F, Rowv = F, dendrogram = "none",
#          trace = "none", col = viridis_pal(),
#          margins = c(8,6), srtCol = 30, cexCol = 0.6, cexRow = 0.5,
#           key = T, keysize = 0.8, main = "Top loading SEED proteins PC2 positive")

```
